---
title: "Estimate Sample Size"
author: "Joaquim Aguirre Plans"
date: "8/31/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Information from: 

* https://www.youtube.com/watch?v=WVHeX5La2tg&ab_channel=ResearchTips

Install packages:

```{r}
#BiocManager::install(c("RnaSeqSampleSize"))
#BiocManager::install(c("ssizeRNA"))
#install.packages("precisely")
```

Load packages:

```{r}
require(data.table)
library(dplyr)
library(pwr)
library(PROPER)
library(RnaSeqSampleSize)
library(ssizeRNA)
library(precisely)
```

Load our dataset:

```{r}
metadata_RNA = fread("/Users/quim/Documents/DATA/BioGroup/Scipher/data/out/Meta_RNA_Cleaned.csv")
```

Methods to try in R:

* pwr
* RNASeqSampleSize
* PROPER
* ssizeRNA
* precisely

Without R:

* SSizer (http://idrblab.cn/ssizer/)
* Scotty (http://scotty.genetics.utah.edu/)


## pwr

```{r}
effect.size.large = 0.8 # Large effect size
effect.size.medium = 0.5 # Large effect size
effect.size.small = 0.2 # Large effect size
sig.level = 0.05 # Significance
power = 0.8 # Power
n.sample.large.effect <- pwr.t.test(d=effect.size.large, sig.level=sig.level, power=power, type='two.sample', alternative="two.sided")
n.sample.medium.effect <- pwr.t.test(d=effect.size.medium, sig.level=sig.level, power=power, type='two.sample', alternative="two.sided")
n.sample.small.effect <- pwr.t.test(d=effect.size.small, sig.level=sig.level, power=power, type='two.sample', alternative="two.sided")

n.sample.large.effect$n
n.sample.medium.effect$n
n.sample.small.effect$n
```


## RNASeqSampleSize

Define matrix:

```{r}
# Responders
responders <- metadata_RNA[(metadata_RNA$acr50_3m=="responder") & (metadata_RNA$acr50_6m=="responder"),]
responder.ids <- responders$Subject
responder.counts <- responders[, -c(1:52)]
responder.counts.t <- t(as.matrix(responder.counts))
colnames(responder.counts.t) <- responder.ids
rownames(responder.counts.t) <- colnames(responder.counts)

# Non-responders
nonresponders <- metadata_RNA[(metadata_RNA$acr50_3m=="nonresponder") & (metadata_RNA$acr50_6m=="nonresponder"),]
nonresponder.ids <- nonresponders$Subject
nonresponder.counts <- nonresponders[, -c(1:52)]
nonresponder.counts.t <- t(as.matrix(nonresponder.counts))
colnames(nonresponder.counts.t) <- nonresponder.ids
rownames(nonresponder.counts.t) <- colnames(nonresponder.counts)

# Merged table
RNAseq_data <- cbind(responder.counts.t, nonresponder.counts.t)
rnaseq_data_file <- '/Users/quim/Documents/DATA/BioGroup/Scipher/data/out/Meta_RNA_Cleaned_Formatted4Scotty.csv'
RNAseq_data %>% write.table(rnaseq_data_file, quote = F, sep='\t', row.names = T, col.names = T)

```

Calculate sample size using distribution from our data:

```{r}
set.seed(21)
power<-0.8
fdr<-0.05 # False discovery rate
rho<-2 # Minimal fold change between two groups

# Estimate gene read count and dispersion distribution
dataMatrixDistribution <- est_count_dispersion(RNAseq_data, group=c(rep(0, length(colnames(responder.counts.t))), rep(1,length(colnames(nonresponder.counts.t)))))

n.sample.rnaseqsamplesize.distr.fc2 <- sample_size_distribution(power=power, f=fdr, distributionObject=dataMatrixDistribution, repNumber=100, showMessage=TRUE, rho=2)
n.sample.rnaseqsamplesize.distr.fc2
n.sample.rnaseqsamplesize.distr.fc1 <- sample_size_distribution(power=power, f=fdr, distributionObject=dataMatrixDistribution, repNumber=100, showMessage=TRUE, rho=1.5)
n.sample.rnaseqsamplesize.distr.fc1
```

Calculate sample size using distribution from TCGA data:

```{r}
set.seed(21)
power<-0.8
fdr<-0.05 # False discovery rate

n.sample.rnaseqsamplesize.tcga <- sample_size_distribution(power=power,f=fdr,distributionObject="TCGA_READ", repNumber=100,showMessage=TRUE)
n.sample.rnaseqsamplesize.tcga
```

Calculate sample size without using distribution:

```{r}
power<-0.8
fdr<-0.05 # False discovery rate
rho<-2 # Minimal fold change between two groups
lambda0<-5 # Minimal average read counts
phi0<-0.5 # Maximal dispersion

n.sample.rnaseqsamplesize.nodistr <- sample_size(power=power, f=fdr, rho=rho, lambda0=lambda0, phi0=phi0)
n.sample.rnaseqsamplesize.nodistr
```


## ssizeRNA

Calculate sample size using default parameters:

```{r}
set.seed(21)
power<-0.8
fdr<-0.05
prop.nonde.genes<-0.8 # Proportion of non-differentially expressed genes
av.read.count.control <- 10 # Average read count for each gene in the control group
dispersion<-0.1 # Dispersion for each gene
fold.change <- 2

n.sample.ssize.default <- ssizeRNA_single(nGenes = length(rownames(RNAseq_data)), pi0 = prop.nonde.genes, m = 200, mu = av.read.count.control, disp = dispersion, fc = fold.change, fdr = fdr, power = power, maxN = 50)
n.sample.ssize.default$ssize
```

```{r}
set.seed(21)
power<-0.8
fdr<-0.05
prop.nonde.genes<-0.8 # Proportion of non-differentially expressed genes
av.read.count.control <- 10 # Average read count for each gene in the control group
dispersion<-0.1 # Dispersion for each gene
fold.change <- 1.5

n.sample.ssize.default <- ssizeRNA_single(nGenes = length(rownames(RNAseq_data)), pi0 = prop.nonde.genes, m = 200, mu = av.read.count.control, disp = dispersion, fc = fold.change, fdr = fdr, power = power, maxN = 50)
n.sample.ssize.default$ssize
```

Calculate sample size using the distribution parameters of our own data:

```{r}
set.seed(21)
power<-0.8
fdr<-0.05
prop.nonde.genes<-0.8 # Proportion of non-differentially expressed genes
fold.change <- 2

nonresponder.counts.t.nozero <- nonresponder.counts.t[rowSums(nonresponder.counts.t) > 0,]
responder.counts.t.nozero <- responder.counts.t[rowSums(responder.counts.t) > 0,]
RNAseq_data.nozero <- cbind(responder.counts.t.nozero, nonresponder.counts.t.nozero)

# Average read count for each gene in the control group
av.read.count.control <- as.numeric(apply(nonresponder.counts.t.nozero, 1, mean))

# dispersion for each gene
d <- DGEList(RNAseq_data.nozero)
d <- calcNormFactors(d)
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)
dispersion <- d$tagwise.dispersion

n.sample.ssize.owndata <- ssizeRNA_single(nGenes = length(rownames(RNAseq_data.nozero)), pi0 = prop.nonde.genes, mu = av.read.count.control, disp = dispersion, fc = fold.change, fdr = fdr, power = power, replace = FALSE)
n.sample.ssize.owndata$ssize
```


## Calculate sample size using PROPER

```{r}
responders <- metadata_RNA[(metadata_RNA$acr50_3m=="responder") & (metadata_RNA$acr50_6m=="responder"),]
nonresponders <- metadata_RNA[(metadata_RNA$acr50_3m=="nonresponder") & (metadata_RNA$acr50_6m=="nonresponder"),]
RNAseq_data <- rbind(responders, nonresponders)
counts.table <- RNAseq_data[, -c(1:52)]
sim.param.sample.means <- estParam(X=as.matrix(counts.table),type = 1)
#sim.param.sample.edger <- estParam(X=as.matrix(counts.table),type = 2)
sim.opts.Cheung = RNAseq.SimOptions.2grp(ngenes = length(colnames(counts.table)), p.DE=0.05, lOD="cheung", lBaselineExpr="cheung")
sim.opts.sample.means = RNAseq.SimOptions.2grp(ngenes = length(colnames(counts.table)), p.DE=0.05, lOD=sim.param.sample.means$lOD, lBaselineExpr=sim.param.sample.means$lmeans)
sim.opts.sample.edger = RNAseq.SimOptions.2grp(ngenes = length(colnames(counts.table)), p.DE=0.05, lOD=sim.param.sample.edger$lOD, lBaselineExpr=sim.param.sample.edger$lmean)
```

### Run simulation and DE detection based on Cheung parameters

```{r}
simres = runSims(Nreps = c(3, 5, 7, 10), sim.opts=sim.opts.Cheung, DEmethod="edgeR", nsims=20)
```

```{r}
powers = comparePower(simres, alpha.type="fdr", alpha.nominal=0.1, stratify.by="expr", delta=0.5)
summaryPower(powers)
```

```{r}
plotPower(powers)
```

### Run simulation and DE detection based on sample means parameters

```{r}
simres = runSims(Nreps = c(5, 10, 20, 30, 40), sim.opts=sim.opts.sample.means, DEmethod="edgeR", nsims=20)
```

```{r}
powers = comparePower(simres, alpha.type="fdr", alpha.nominal=0.1, stratify.by="expr", delta=0.5)
summaryPower(powers)
```

```{r}
plotPower(powers)
```



## precisely

Calculate sample size using default parameters:

```{r}
set.seed(21)
power<-0.8
fdr<-0.05
prop.nonde.genes<-0.8 # Proportion of non-differentially expressed genes
av.read.count.control <- 10 # Average read count for each gene in the control group
dispersion<-0.1 # Dispersion for each gene
fold.change <- 2

precisely.df <- n_risk_ratio(
  precision = 2,
  exposed = length(responder.ids)/(length(responder.ids)+length(nonresponder.ids)),
  unexposed = length(nonresponder.ids)/(length(responder.ids)+length(nonresponder.ids)),
  group_ratio = length(nonresponder.ids)/length(responder.ids),
  ci = .95
)
plot_sample_size(precisely.df, xlab = "Sample Size", ylab = "Precision", line_size = 1)
```
