---
title: "RA - Generate Data for wTO"
output:
  html_document:
    theme: united
    highlight: tang
author: Deisy Morselli Gysi & Joaquim Aguirre Plans
---
#### Northeastern University
#### Project: Scipher RA - Responder vs Non-responders (using GSE97810 data)

## Description

The first step of the whole analysis is to make sure we are dealing with coding genes and that they have a minimum count of 5 reads. Moreover, those genes cannot have the same measure in all samples, therefore, SD is also used as a measure to filter undesired effect. 

I also match the patients to their metadata, so we work only with the overlap of both datasets (RNAseq and metadata) from now on.

```{r, message=FALSE}
require(dplyr)
require(magrittr)
require(data.table)
```



```{r}
metadata = fread("/Users/quim/Documents/DATA/BioGroup/Scipher/data/RA/data/GSE97810/pData/GSE97810_pData.csv")
meta.cols <- metadata$Metadata
metadata %<>% select(-Metadata)
metadata %<>% t %<>% as.data.frame()
colnames(metadata) <- meta.cols
```



```{r}
RNAseq = fread("/Users/quim/Documents/DATA/BioGroup/Scipher/data/RA/data/GSE97810/exprs/GSE97810_exprs.csv")
genes <- RNAseq$ID_REF
samples <- colnames(RNAseq)[-1]
RNAseq %<>% select(-ID_REF) %>% t %>% as.data.frame()
colnames(RNAseq) <- genes
RNAseq$Subject <- samples
dim(RNAseq)
nonCoding_id = names(RNAseq) %>% stringr::str_detect(., ":")
nonCoding = names(RNAseq)[nonCoding_id]
head(nonCoding)
RNAseq %<>% select(-all_of(nonCoding))
dim(RNAseq)
RNA = RNAseq %>% select(-Subject)
```


```{r}
sum = colSums(RNA)
mean = colMeans(RNA)
med = apply(RNA, 2, median)
sd = apply(RNA, 2,sd)

sumG = sum[sum > 0]
meanG = mean[mean > 5]
sdG = sd[sd > 1]
medG = med[med > 5]

length(sumG)
length(meanG)
length(sdG)
length(medG)

keep = data.frame(Gene = c(names(sumG), 
                           names(meanG),
                           names(sdG), 
                           names(medG))) %>% 
        group_by(Gene) %>% 
        summarise(., n = n()) %>% 
        filter(n >= 3)
dim(RNAseq)

RNAseq %<>% select(Subject, keep$Gene)
#u = RNAseq[,-1] #%>% t %>% scale() %>% t
#RNAseq = cbind(RNAseq[,1], u) %>% as.data.frame()

dim(RNAseq)

rm(RNA)
rm(u)
rm(mean)
rm(meanG)
rm(sd)
rm(sdG)
rm(med)
rm(medG)
rm(sum)
rm(sumG)
rm(keep)
```


```{r}
names(metadata)[2]= "Subject"
metadata_RNA = dplyr::inner_join(metadata, RNAseq, by="Subject")
dim(metadata_RNA[, -c(1:729)])
```
```{r}
fwrite(metadata_RNA, "/Users/quim/Documents/DATA/BioGroup/Scipher/data/RA/data/GSE97810/exprs/GSE97810_Meta_RNA_Cleaned.csv")
```

## Session Info
```{r}
sessioninfo::session_info()
```
