---
title: "RA - Generate Data for wTO"
output:
  html_document:
    theme: united
    highlight: tang
author: Deisy Morselli Gysi & Joaquim Aguirre Plans
---
#### Northeastern University
#### Project: Scipher RA - Responder vs Non-responders to TNF drugs

## Description

The first step of the whole analysis is to make sure we are dealing with coding genes and that they have a minimum count of 5 reads. Moreover, those genes cannot have the same measure in all samples, therefore, SD is also used as a measure to filter undesired effect. 

I also match the patients to their metadata, so we work only with the overlap of both datasets (RNAseq and metadata) from now on.

```{r, message=FALSE}
require(dplyr)
require(magrittr)
require(data.table)
```


```{r}
metadata = fread("/Users/quim/Documents/DATA/BioGroup/Scipher/data/Ntk004MetaData_NEU.csv")
RNAseq = fread("/Users/quim/Documents/DATA/BioGroup/Scipher/data/Network004_hg19_ucsc_raw_counts_20200909.csv")

table(metadata %>% select(acr50_3m, acr50_6m))
```

```{r}
names(metadata)[1]= "Subject"
unfiltered_metadata_RNA = dplyr::inner_join(metadata, RNAseq)
```

```{r}
table(unfiltered_metadata_RNA %>% select(acr50_3m, acr50_6m))
genes <- colnames(RNAseq[, -c(1:7)])
print(length(genes))
```

```{r}
nonCoding_id = names(RNAseq) %>% stringr::str_detect(., ":")
nonCoding = names(RNAseq)[nonCoding_id]

head(nonCoding)
RNAseq %<>% select(-all_of(nonCoding))
RNA = RNAseq[, -c(1:7)]
```


```{r}
sum = colSums(RNA)
mean = colMeans(RNA)
med = apply(RNA, 2, median)
sd = apply(RNA, 2,sd)

sumG = sum[sum > 0]
meanG = mean[mean > 5]
sdG = sd[sd > 1]
medG = med[med > 5]


keep = data.frame(Gene = c(names(sumG), 
                           names(meanG),
                           names(sdG), 
                           names(medG))) %>% 
        group_by(Gene) %>% 
        summarise(., n = n()) %>% 
        filter(n >= 3)
dim(RNAseq)

RNAseq %<>% select(Subject, keep$Gene)
u = RNAseq[,-1] #%>% t %>% scale() %>% t
RNAseq = cbind(RNAseq[,1], u) %>% as.data.frame()

dim(RNAseq)

rm(RNA)
rm(u)
rm(mean)
rm(meanG)
rm(sd)
rm(sdG)
rm(med)
rm(medG)
rm(sum)
rm(sumG)
rm(keep)
```


```{r}
names(metadata)[1]= "Subject"
metadata_RNA = dplyr::inner_join(metadata, RNAseq)
dim(metadata_RNA[, -c(1:52)])
```
```{r}
fwrite(metadata_RNA, "../data/out/Meta_RNA_Cleaned.csv")
```

## Session Info
```{r}
sessioninfo::session_info()
```
