---
title: "Create Subsamples of gene expression data from RA patients"
author: "Joaquim Aguirre Plans"
date: "8/24/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Create subsamples of gene expression data from RA patients, with the aim of testing the effect of sample size in these subsamples.

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(igraph)
library(NetSci)
set.seed(1510)
```

## Read gene expression data

```{r read_gene_expression}
#metadata_RNA = fread("/Users/quim/Documents/DATA/BioGroup/Scipher/data/out/Meta_RNA_Cleaned.csv")
metadata_RNA = fread("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/Meta_RNA_Cleaned.csv")
patients <- metadata_RNA$Subject
```

The total number of patients is the dataset is `r length(patients)`.

We check the number of labels in these patients:

```{r define_response}
labels_R_3m = metadata_RNA$Subject[metadata_RNA$acr50_3m == "responder"]
labels_R_6m =metadata_RNA$Subject[metadata_RNA$acr50_3m == "nonresponder"]
labels_NR_3m = metadata_RNA$Subject[metadata_RNA$acr50_6m == "responder"]
labels_NR_6m = metadata_RNA$Subject[metadata_RNA$acr50_6m == "nonresponder"]
all_labels <- c(labels_R_3m, labels_R_6m, labels_NR_3m, labels_NR_6m)
labels_R = metadata_RNA$Subject[(metadata_RNA$acr50_3m == "responder")&(metadata_RNA$acr50_6m == "responder")]
labels_nR = metadata_RNA$Subject[(metadata_RNA$acr50_3m == "nonresponder")&(metadata_RNA$acr50_6m == "nonresponder")]
labels_R_nR <- c(labels_R, labels_nR)
```

* Responder (3M): `r length(labels_R_3m)`
* Responder (6M): `r length(labels_R_6m)`
* Non-responder (3M): `r length(labels_NR_3m)`
* Non-responder (6M): `r length(labels_NR_6m)`
* Total number of labels: `r length(all_labels)`
* Responder (3M & 6M): `r length(labels_R)`
* Non-responder (3M & 6M): `r length(labels_nR)`
* Total number of labels (3M & 6M): `r length(labels_R_nR)`

We check the balance of responder and non-responder patients in the dataset.

```{r table_response}
table(metadata_RNA %>% select(acr50_3m, acr50_6m))
```

## Create files of responders and non-responders data in a clean format

```{r responders_cleaning}
responders <- metadata_RNA[(metadata_RNA$acr50_3m=="responder") & (metadata_RNA$acr50_6m=="responder"),]
responder.ids <- responders$Subject
responder.counts <- responders[, -c(2:52)] # Get the counts but keep the subject ID
rnaseq_responders_data_file <- '/Users/quim/Documents/DATA/BioGroup/Scipher/data/bootstrap/responders/RNAseq_Responders_all.txt'
responder.counts %>% write.table(rnaseq_responders_data_file, quote = F, sep='\t', row.names = F, col.names = T)

```

```{r nonresponders_cleaning}
nonresponders <- metadata_RNA[(metadata_RNA$acr50_3m=="nonresponder") & (metadata_RNA$acr50_6m=="nonresponder"),]
nonresponder.ids <- nonresponders$Subject
nonresponder.counts <- nonresponders[, -c(2:52)] # Get the counts but keep the subject ID
rnaseq_nonresponders_data_file <- '/Users/quim/Documents/DATA/BioGroup/Scipher/data/bootstrap/nonresponders/RNAseq_NonResponders_all.txt'
nonresponder.counts %>% write.table(rnaseq_nonresponders_data_file, quote = F, sep='\t', row.names = F, col.names = T)

```


## Subsampling

### Randomly sampling *n* labels maintaining the original proportion (stratified bootstrapping)

We define the function to make the sampling

```{r bootstrap_function}
## Bootstrap by random sampling
create_random_subsamples <- function(rnaseq, size_list, rep, output_dir, output_file_name){
  subject.ids <- rnaseq$Subject
  seed <- 1
  for(size in size_list){
    set.seed(seed)
    seed <- seed + 1
    for(r in 1:rep){
      subsample <- sample(subject.ids, size=size, replace=FALSE)
      subsample_df = rnaseq %>% filter(Subject %in% subsample)
      output_file = paste(output_dir, paste(paste(output_file_name, 'sample', size, 'rep', r, sep='_'), '.txt', sep=''), sep='/')
      subsample_df %>% write.table(output_file, quote = F, sep='\t', row.names = F, col.names = T)
    }
  }
}
```

We randomly select labels from the non-responder dataset for each subsample. The process is repeated r times for each subsample size.

```{r subsampling_non_responders}
# Define sample size
rep = 10
step = 10
max_num_samples = length(nonresponder.ids) - 5
size_list <- seq(10, max_num_samples, 10)
size_list

# Create subsample files
output_dir = '/Users/quim/Documents/DATA/BioGroup/Scipher/data/bootstrap/nonresponders'
output_file_name = 'RNAseq_NonResponders'
create_random_subsamples(rnaseq=nonresponder.counts, size_list=size_list, rep=rep, output_dir=output_dir, output_file_name)

```

```{r subsampling_responders}
# Define sample size
rep = 10
step = 10
max_num_samples = length(responder.ids) - 5
size_list <- seq(10, max_num_samples, 10)
size_list

# Create subsample files
output_dir = '/Users/quim/Documents/DATA/BioGroup/Scipher/data/bootstrap/responders'
output_file_name = 'RNAseq_Responders'
#create_random_subsamples(rnaseq=responder.counts, size_list=size_list, rep=rep, output_dir=output_dir, output_file_name)

```

