---
title: "GEO"
author: "Joaquim Aguirre-Plans"
date: '2023-02-24'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



### Description

Analyze GEO dataset.
References:
http://genomicsclass.github.io/book/pages/GEOquery.html
https://www.bioconductor.org/packages/devel/bioc/vignettes/GEOquery/inst/doc/GEOquery.html
https://kasperdanielhansen.github.io/genbioconductor/html/GEOquery.html 
Information from the dataset:
https://pubmed.ncbi.nlm.nih.gov/36109152/

```{r, message=FALSE}
packrat::init("/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeR")
```

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(GEOquery)
library(ggplot2)
require(magrittr)
library(tidyr)
set.seed(1510)
```



### Access the GEO Series data

```{r}
geo_id = "GSE193677"
geo_dir = "/work/ccnr/j.aguirreplans/Databases/GEO"
geo_id_dir = paste(geo_dir, "/", geo_id, sep="")
raw_data_dir = paste(geo_id_dir, "/raw", sep="")
output_dir = paste(geo_id_dir, "/out", sep="")
dir.create(geo_id_dir, showWarnings = FALSE)
dir.create(raw_data_dir, showWarnings = FALSE)
dir.create(output_dir, showWarnings = FALSE)
```

`getGEO` function interprets its input to determine how to get the data from GEO and then parse the data into useful R data structures.
In the case of the dataset GSE193677, it downloads two files:
* `GPL16791.soft.gz`: Containing the phenotypic data in soft format.
* `GSE193677_series_matrix.txt.gz`: Containing the phenotypic format in series format.

```{r}
gse <- getGEO(geo_id, GSEMatrix = TRUE, destdir=raw_data_dir)
show(gse)
```

Now, `gse` contains the R data structure (of class GSE) that represents the `geo_id` entry from GEO.
If using `GSEMatrix = TRUE`, the data structure returned from this parsing is a list of ExpressionSets.

```{r}
show(pData(phenoData(gse[[1]]))[1:5,c(1,6,8)])
```

Store the metadata in a file:

```{r}
metadata_complete_df = pData(gse[[1]])
metadata_complete_file = paste(raw_data_dir, "/metadata_complete.txt", sep="")
metadata_complete_df %>% fwrite(metadata_complete_file)
metadata_complete_df = fread(metadata_complete_file)
metadata_complete_df[1:5, 1:5]
```



### Access raw files

The function `getGEOSuppFiles`, in the case of the dataset GSE193677, it downloads two files:
* `GSE193677_MSCCR_Biopsy_counts.txt.gz`: Containing the raw counts.
* `GSE193677_MSCCR_Biopsy_adjcounts.txt.gz`: Containing the adjusted counts.

```{r}
filePaths = getGEOSuppFiles(geo_id, makeDirectory = FALSE, baseDir = raw_data_dir)
filePaths
```



### Access GSE Data Tables from GEO

Sometimes GSEs are include separate data tables with the sample information. If these exist, you can use the getGSEDataTables() function. For example here is the phenoData object from a different GSE accession GSE3494 with a Data Table.

```{r}
df1 <- getGSEDataTables(geo_id)
lapply(df1, head)
```



### Analyze metadata

Number of samples by gender:

```{r}
table(metadata_complete_df$`demographics_gender:ch1`)
```

Number of samples by tissue:

```{r}
table(metadata_complete_df$`regionre:ch1`)
```

Number of samples by inflammation and disease:

```{r}
table(metadata_complete_df$`diseasetypere:ch1`)
```

Number of samples by disease:

```{r}
table(metadata_complete_df$`ibd_disease:ch1`)
```

Disease activity depending on the sample type and disease:

```{r}
table(metadata_complete_df$`diseasetypere:ch1`, metadata_complete_df$`ibd_clinicianmeasure_inactive_active:ch1`)
table(metadata_complete_df$`ibd_disease:ch1`, metadata_complete_df$`ibd_clinicianmeasure_inactive_active:ch1`)
```

Create a file with simplified metadata:

```{r}
metadata_df = metadata_complete_df %>%
  dplyr::select(title, geo_accession, `diseasetypere:ch1`, `ibd_disease:ch1`, `ibd_clinicianmeasure_inactive_active:ch1`, `demographics_gender:ch1`, `regionre:ch1`, `study_eligibility_age_at_endo:ch1`) %>%
  dplyr::rename("sample_type"=`diseasetypere:ch1`, "disease"=`ibd_disease:ch1`, "disease_activity"=`ibd_clinicianmeasure_inactive_active:ch1`, "gender"=`demographics_gender:ch1`, "tissue"=`regionre:ch1`, "age"=`study_eligibility_age_at_endo:ch1`) %>%
  tidyr::separate(title, into=c("sample_id", NA, NA), sep = ",")
metadata_simple_file = paste(output_dir, "/metadata_simple.txt", sep="")
metadata_df %>% fwrite(metadata_simple_file)
metadata_df = fread(metadata_simple_file)
head(metadata_df)
```


### Access the raw counts

The raw counts are stored in the file `GSE193677_MSCCR_Biopsy_counts.txt.gz`.

```{r}
raw_counts_original_file = paste(raw_data_dir, "/GSE193677_MSCCR_Biopsy_counts.txt.gz", sep="")
```

Read the gz file:

```{r}
raw_counts_original_df = read.table(gzfile(raw_counts_original_file))
raw_counts_original_df = cbind(data.frame(Ensembl=row.names(raw_counts_original_df)), raw_counts_original_df)
```

```{r}
dim(raw_counts_original_df)
raw_counts_original_df[1:5, 1:5]
```

Filter the samples that are not in the metadata:

```{r}
raw_counts_original_df = raw_counts_original_df[, which((names(raw_counts_original_df) %in% c("Ensembl", metadata_df$sample_id)==TRUE))]
dim(raw_counts_original_df)
```

Are there repeated genes? No:

```{r}
nrow(raw_counts_original_df)
length(unique(raw_counts_original_df$Ensembl))
```


### Gene pre-processing

In the gene expression dataframe, there are `r nrow(raw_counts_original_df)` genes in Ensembl notation. We will transform these genes into HGNC gene symbol notation.

First we read the file `multiple_hgnc_dict_set_2022-01-01.txt`, containing gene symbols and their aliases in other notations:

```{r}
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_dict_file = paste(gene_annotations_dir, 'multiple_hgnc_dict_set_2022-01-01.txt', sep='/')
gene_dict_df = fread(gene_dict_file) %>% unique()
gene_dict_df = gene_dict_df[!((gene_dict_df$HGNC_Symbol=="")|(gene_dict_df$alias=="")),]
gene_dict_df$HGNC_Symbol = toupper(gene_dict_df$HGNC_Symbol)
gene_dict_df$alias = toupper(gene_dict_df$alias)
gene_dict_df %<>% unique()
head(gene_dict_df)
```

We use this dataframe to map Ensembl IDs to HGNC symbol IDs:

```{r}
raw_counts_df = inner_join(raw_counts_original_df, gene_dict_df, by=c("Ensembl"="alias"))
raw_counts_df = cbind((raw_counts_df %>% select("Ensembl", "HGNC_Symbol")), (raw_counts_df %>% select(-"Ensembl", -"HGNC_Symbol"))) %>% unique()
head(raw_counts_df[1:5,1:5])
```

```{r}
dim(raw_counts_df)
```

Now, we have `r length(unique(raw_counts_df$HGNC_Symbol))` unique HGNC symbols mapped to `r length(unique(raw_counts_df$Ensembl))` unique Ensembl IDs from a total of `r nrow(raw_counts_df)` rows. The problem is that there are Ensembl IDs mapped to the same HGNC code and viceversa. So, we have to do the following procedures:
- If an Ensembl ID has multiple HGNC symbols, keep only one symbol
- If a HGNC symbol has multiple Ensembl IDs, calculate the median

We get the HGNC symbols with multiple Ensembl IDs:

```{r}
repeated_ensembl = unique(raw_counts_df$Ensembl[duplicated(raw_counts_df$Ensembl)])
length(repeated_ensembl)
repeated_ensembl_df = raw_counts_df[raw_counts_df$Ensembl %in% repeated_ensembl,][,c("Ensembl", "HGNC_Symbol")]
repeated_ensembl_df$HGNC_length = nchar(repeated_ensembl_df$HGNC_Symbol)
repeated_ensembl_df = repeated_ensembl_df %>% arrange(Ensembl, HGNC_length, HGNC_Symbol) %>% unique()
head(repeated_ensembl_df)
```

We keep the smallest symbol, and if not, the first (in alphabetical order):

```{r}
repeated_ensembl_distinct_df = repeated_ensembl_df %>% distinct(Ensembl, .keep_all = TRUE) %>% select(Ensembl, HGNC_Symbol) # dplyr::distinct keeps only the first distinct row in terms of Ensembl
head(repeated_ensembl_distinct_df)
hgnc_to_remove = repeated_ensembl_df$HGNC_Symbol[!(repeated_ensembl_df$HGNC_Symbol %in% repeated_ensembl_distinct_df$HGNC_Symbol)] # we get the HGNC_Symbol to remove
raw_counts_df %<>% filter(!(HGNC_Symbol %in% hgnc_to_remove)) # remove HGNC symbols selected before
```

```{r}
dim(raw_counts_df)
```

We get the Ensembl IDs with multiple HGNC symbols:

```{r}
repeated_hgnc = unique(raw_counts_df$HGNC_Symbol[duplicated(raw_counts_df$HGNC_Symbol)])
length(repeated_hgnc)
repeated_hgnc_df = (raw_counts_df[raw_counts_df$HGNC_Symbol %in% repeated_hgnc,][,c("Ensembl", "HGNC_Symbol")]) %>% arrange(HGNC_Symbol, Ensembl)
head(repeated_hgnc_df)
```

We calculate the median for the HGNC genes with multiple Ensembl IDs:

```{r}
raw_counts_df = subset(raw_counts_df, select=-Ensembl)
repeated_hgnc_aggregated = aggregate(.~HGNC_Symbol, raw_counts_df[raw_counts_df$HGNC_Symbol %in% repeated_hgnc,], median)
raw_counts_df = rbind(raw_counts_df[!(raw_counts_df$HGNC_Symbol %in% repeated_hgnc),], repeated_hgnc_aggregated)
dim(raw_counts_df)
```

```{r}
raw_counts_processed_file = paste(output_dir, "/GSE193677_MSCCR_Biopsy_counts_processed.txt", sep="")
raw_counts_df %>% fwrite(raw_counts_processed_file, sep='\t')
raw_counts_df = fread(raw_counts_processed_file)
raw_counts_df[1:5,1:5]
```


