---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Description

Create figures for the manuscript, extended abstract and poster.

```{r}
library(data.table)
library(dplyr)
library(ggplot2)
require(ggrepel)
require(magrittr)
library(patchwork)
library(tidyr)
set.seed(1510)
options(bitmapType='cairo')
`%ni%` <- Negate(`%in%`)
```

### Parse RNAseq data and calculate mean, SD, CV

Define input directories and files:

```{r}
#type_data_selection = 'pearson_aracne'
type_data_selection = 'pearson_pval_0.05'
input_dir = paste('/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data/example_', type_data_selection, sep = '')
topology_results_file = paste(input_dir, 'topology_results_pearson_pval_0.05.txt', sep='/')
results_selected_df = fread(topology_results_file)
type_datasets_selected = unique(results_selected_df$type_dataset)
gtex_dir = "/work/ccnr/j.aguirreplans/Databases/GTEx/v8"
tcga_dir = "/work/ccnr/j.aguirreplans/Databases/TCGA/2022-11-18-Dataset/TCGA/out"
scipher_rnaseq_dir = "/work/ccnr/j.aguirreplans/data/Scipher/Dec2021/00_data/rnaseq_filtered_files_by_group"
data_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data'
plots_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots'
```

Read RNA-seq files for each database and calculate SD of expression samples for each gene.

#### Parse GTEx

```{r}
gtex_sd_genes_df = data.frame()
gtex_sd_samples_df = data.frame()
gtex_meta_df = data.frame()

for (type_counts in c("reads")) {
  # Get all datasets in GTEx directory
  gtex_rnaseq_dir = paste(gtex_dir, type_counts, "rnaseq_filtered_files_by_tissue", sep="/")
  rnaseq_files = list.files(gtex_rnaseq_dir)
  for (rnaseq_file in rnaseq_files){
    rnaseq_file_path = paste(gtex_rnaseq_dir, rnaseq_file, sep="/")
    # Get sex in file name
    rnaseq_spl = strsplit(gsub(".gct", "", rnaseq_file), split="_", fixed = TRUE)[[1]]
    sex = NA
    type_dataset_selected = paste(rnaseq_spl[1], tolower(rnaseq_spl[3]), sep = ":")
    if (length(rnaseq_spl) == 4) {
      sex = rnaseq_spl[4]
      type_dataset_selected = paste(type_dataset_selected, sex, sep = "_")
    }
    if (type_dataset_selected %in% type_datasets_selected){
    
        # Read dataset
        rnaseq = fread(rnaseq_file_path) %>% as.data.frame()
        # Define data frame if it is still empty
        if (nrow(gtex_sd_genes_df) == 0){
          # SD for each gene in dataset
          gtex_sd_genes_df = data.frame(gene = rnaseq$Gene, 
                                        sd = apply(rnaseq[,-1], 1, sd), 
                                        mean = apply(rnaseq[,-1], 1, mean), 
                                        dataset = rnaseq_spl[1],
                                        type_dataset = rnaseq_spl[3], 
                                        type_counts = type_counts, 
                                        sex = sex)
          # SD for each sample in dataset
          gtex_sd_samples_df = data.frame(sample = colnames(rnaseq[,-1]), 
                                          sd = apply(rnaseq[,-1], 2, sd), 
                                          mean = apply(rnaseq[,-1], 2, mean), 
                                          dataset = rnaseq_spl[1], 
                                          type_dataset = rnaseq_spl[3], 
                                          type_counts = type_counts, 
                                          sex = sex)
          # Number of genes and samples for each dataset
          gtex_meta_df = data.frame(dataset = rnaseq_spl[1], 
                                    type_dataset = rnaseq_spl[3], 
                                    type_counts = type_counts, 
                                    sex = sex, 
                                    num_samples = ncol(rnaseq[,-1]), 
                                    num_genes = nrow(rnaseq[,-1]))
        # Append results if dataframe is already created
        } else {
          gtex_sd_genes_df = rbind(gtex_sd_genes_df, 
                                   data.frame(gene = rnaseq$Gene, 
                                              sd = apply(rnaseq[,-1], 1, sd), 
                                              mean = apply(rnaseq[,-1], 1, mean), 
                                              dataset = rnaseq_spl[1], 
                                              type_dataset = rnaseq_spl[3], 
                                              type_counts = type_counts,
                                              sex = sex)
                                   )
          gtex_sd_samples_df = rbind(gtex_sd_samples_df, 
                                     data.frame(sample = colnames(rnaseq[,-1]), 
                                                sd = apply(rnaseq[,-1], 2, sd), 
                                                mean = apply(rnaseq[,-1], 2, mean), 
                                                dataset = rnaseq_spl[1], 
                                                type_dataset = rnaseq_spl[3], 
                                                type_counts = type_counts,
                                                sex = sex)
                                     )
          gtex_meta_df = rbind(gtex_meta_df, 
                               data.frame(dataset = rnaseq_spl[1], 
                                          type_dataset = rnaseq_spl[3], 
                                          type_counts = type_counts, 
                                          sex = sex, 
                                          num_samples = ncol(rnaseq[,-1]), 
                                          num_genes = nrow(rnaseq[,-1]))
                               )
        }
      }
    }
}

# Calculate CV
gtex_sd_genes_df$cv = (gtex_sd_genes_df$sd/gtex_sd_genes_df$mean)*100
gtex_sd_samples_df$cv = (gtex_sd_samples_df$sd/gtex_sd_samples_df$mean)*100
```

```{r}
head(gtex_sd_genes_df)
head(gtex_sd_samples_df)
head(gtex_meta_df)
```


#### Parse TCGA

```{r}
tcga_sd_genes_df = data.frame()
tcga_sd_samples_df = data.frame()
tcga_meta_df = data.frame()

for (type_counts in c("reads")) {
  for (subclassification in c("tumor", "tissue", "subtype")) {
    # Get all datasets in TCGA directory
    tcga_rnaseq_dir = paste(tcga_dir, type_counts, subclassification, "filter_genes_low_counts/rnaseq_filtered_files_by_sample_group", sep="/")
    rnaseq_files = list.files(tcga_rnaseq_dir)
    for (rnaseq_file in rnaseq_files) {
      # Get sex in file name
      rnaseq_file_path = paste(tcga_rnaseq_dir, rnaseq_file, sep="/")
      rnaseq_spl = strsplit(gsub(".csv", "", rnaseq_file), split="_", fixed = TRUE)[[1]]
      sex = NA
      type_dataset_selected = paste(rnaseq_spl[1], tolower(rnaseq_spl[3]), sep = ":")
      if(length(rnaseq_spl) == 4) {
        sex = rnaseq_spl[4]
        type_dataset_selected = paste(type_dataset_selected, sex, sep = "_")
      }
      if (type_dataset_selected %in% type_datasets_selected){
        # Read dataset
        rnaseq = fread(rnaseq_file_path) %>% as.data.frame()
        # Define data frame if it is still empty
        if (nrow(tcga_meta_df) == 0) {
          tcga_sd_genes_df = data.frame(gene = rnaseq$Gene, 
                                        sd = apply(rnaseq[,-1], 1, sd), 
                                        mean = apply(rnaseq[,-1], 1, mean), 
                                        dataset = rnaseq_spl[1], 
                                        type_dataset = rnaseq_spl[3], 
                                        type_counts = type_counts, 
                                        subclassification = subclassification, 
                                        sex = sex)
          tcga_sd_samples_df = data.frame(sample = colnames(rnaseq[,-1]), 
                                          sd = apply(rnaseq[,-1], 2, sd), 
                                          mean = apply(rnaseq[,-1], 2, mean), 
                                          dataset = rnaseq_spl[1], 
                                          type_dataset = rnaseq_spl[3], 
                                          type_counts = type_counts, 
                                          subclassification = subclassification, 
                                          sex = sex)
          tcga_meta_df = data.frame(dataset = rnaseq_spl[1], 
                                    type_dataset = rnaseq_spl[3], 
                                    type_counts = type_counts, 
                                    subclassification = subclassification, 
                                    sex = sex, 
                                    num_samples = ncol(rnaseq[,-1]), 
                                    num_genes = nrow(rnaseq[,-1]))
        # Append results if dataframe is already created
        } else {
          tcga_sd_genes_df = rbind(tcga_sd_genes_df, 
                                   data.frame(gene = rnaseq$Gene, 
                                              sd = apply(rnaseq[,-1], 1, sd),
                                              mean = apply(rnaseq[,-1], 1, mean), 
                                              dataset = rnaseq_spl[1], 
                                              type_dataset = rnaseq_spl[3], 
                                              type_counts = type_counts, 
                                              subclassification = subclassification, 
                                              sex = sex)
                                   )
          tcga_sd_samples_df = rbind(tcga_sd_samples_df, 
                                     data.frame(sample = colnames(rnaseq[,-1]), 
                                                sd = apply(rnaseq[,-1], 2, sd), 
                                                mean = apply(rnaseq[,-1], 2, mean), 
                                                dataset = rnaseq_spl[1], 
                                                type_dataset = rnaseq_spl[3], 
                                                type_counts = type_counts, 
                                                subclassification = subclassification, 
                                                sex = sex)
                                     )
          tcga_meta_df = rbind(tcga_meta_df, 
                               data.frame(dataset = rnaseq_spl[1], 
                                          type_dataset = rnaseq_spl[3], 
                                          type_counts = type_counts, 
                                          subclassification = subclassification, 
                                          sex = sex, 
                                          num_samples = ncol(rnaseq[,-1]), 
                                          num_genes = nrow(rnaseq[,-1]))
                               )
        }
      }
    }
  }
}

# Calculate CV
tcga_sd_genes_df$cv = (tcga_sd_genes_df$sd/tcga_sd_genes_df$mean)*100
tcga_sd_samples_df$cv = (tcga_sd_samples_df$sd/tcga_sd_samples_df$mean)*100

# Save results
tcga_sd_genes_file = paste(data_dir, "tcga_variation_by_gene.txt", sep="/")
tcga_sd_genes_df %>% fwrite(tcga_sd_genes_file)
```

```{r}
head(tcga_sd_genes_df)
head(tcga_sd_samples_df)
head(tcga_meta_df)
```


#### Parse Scipher

```{r}
scipher_sd_genes_df = data.frame()
scipher_sd_samples_df = data.frame()
scipher_meta_df = data.frame()
type_counts="reads"

# Get all datasets in Scipher directory
rnaseq_files = list.files(scipher_rnaseq_dir)

for (rnaseq_file in rnaseq_files){
  rnaseq_file_path = paste(scipher_rnaseq_dir, rnaseq_file, sep="/")
  # Get sex in file name
  rnaseq_spl = strsplit(gsub(".csv", "", rnaseq_file), split="_", fixed=T)[[1]]
  sex = NA
  if(length(rnaseq_spl) == 4){
    sex = rnaseq_spl[4]
  }
  # Read dataset
  rnaseq = fread(rnaseq_file_path) %>% as.data.frame()
  # Define data frame if it is still empty
  if (nrow(scipher_meta_df) == 0){
    scipher_sd_genes_df = data.frame(gene = rnaseq$Gene, 
                                     sd = apply(rnaseq[,-1], 1, sd), 
                                     mean = apply(rnaseq[,-1], 1, mean), 
                                     dataset = rnaseq_spl[1], 
                                     type_dataset = rnaseq_spl[3], 
                                     type_counts = type_counts,
                                     sex = sex)
    scipher_sd_samples_df = data.frame(sample = colnames(rnaseq[,-1]), 
                                       sd = apply(rnaseq[,-1], 2, sd), 
                                       mean = apply(rnaseq[,-1], 2, mean), 
                                       dataset = rnaseq_spl[1], 
                                       type_dataset = rnaseq_spl[3], 
                                       type_counts = type_counts,
                                       sex = sex)
    scipher_meta_df = data.frame(dataset = rnaseq_spl[1], 
                                 type_dataset = rnaseq_spl[3], 
                                 type_counts = type_counts, 
                                 sex = sex, 
                                 num_samples = ncol(rnaseq[,-1]), 
                                 num_genes = nrow(rnaseq[,-1]))
  # Append results if dataframe is already created
  } else {
    scipher_sd_genes_df = rbind(scipher_sd_genes_df, 
                                data.frame(gene = rnaseq$Gene, 
                                           sd = apply(rnaseq[,-1], 1, sd), 
                                           mean = apply(rnaseq[,-1], 1, mean), 
                                           dataset = rnaseq_spl[1], 
                                           type_dataset = rnaseq_spl[3], 
                                           type_counts = type_counts,
                                           sex = sex)
                                )
    scipher_sd_samples_df = rbind(scipher_sd_samples_df, 
                                  data.frame(sample = colnames(rnaseq[,-1]), 
                                             sd = apply(rnaseq[,-1], 2, sd), 
                                             mean = apply(rnaseq[,-1], 2, mean), 
                                             dataset = rnaseq_spl[1], 
                                             type_dataset = rnaseq_spl[3], 
                                             type_counts = type_counts,
                                             sex = sex)
                                  )
    scipher_meta_df = rbind(scipher_meta_df, 
                            data.frame(dataset = rnaseq_spl[1], 
                                       type_dataset = rnaseq_spl[3], 
                                       type_counts = type_counts, 
                                       sex = sex, 
                                       num_samples = ncol(rnaseq[,-1]), 
                                       num_genes = nrow(rnaseq[,-1]))
                            )
  }
}

# Calculate CV
scipher_sd_genes_df$cv = (scipher_sd_genes_df$sd/scipher_sd_genes_df$mean)*100
scipher_sd_samples_df$cv = (scipher_sd_samples_df$sd/scipher_sd_samples_df$mean)*100
```

```{r}
head(scipher_sd_genes_df)
head(scipher_sd_samples_df)
scipher_meta_df
```

#### Parse GSE193677

```{r}
gse193677_sd_genes_df = data.frame()
gse193677_sd_samples_df = data.frame()
gse193677_meta_df = data.frame()

rnaseq_file_path = "/work/ccnr/j.aguirreplans/Databases/GEO/GSE193677/out/reads/filter_genes_low_counts/rnaseq_GSE193677.txt"
rnaseq = fread(rnaseq_file_path) %>% as.data.frame()
gse193677_sd_genes_df = data.frame(gene = rnaseq$Gene, 
                                   sd = apply(rnaseq[,-1], 1, sd), 
                                   mean = apply(rnaseq[,-1], 1, mean), 
                                   dataset = "gse193677", 
                                   type_dataset = "gse193677", 
                                   type_counts = "reads",
                                   sex = NA)
gse193677_sd_samples_df = data.frame(sample = colnames(rnaseq[,-1]), 
                                     sd = apply(rnaseq[,-1], 2, sd), 
                                     mean = apply(rnaseq[,-1], 2, mean), 
                                     dataset = "gse193677", 
                                     type_dataset = "gse193677", 
                                     type_counts = "reads",
                                     sex = NA)
gse193677_meta_df = data.frame(dataset = "gse193677", 
                               type_dataset = "gse193677", 
                               type_counts = "reads", 
                               sex = NA, 
                               num_samples = ncol(rnaseq[,-1]), 
                               num_genes = nrow(rnaseq[,-1]))

# Calculate CV
gse193677_sd_genes_df$cv = (gse193677_sd_genes_df$sd/gse193677_sd_genes_df$mean)*100
gse193677_sd_samples_df$cv = (gse193677_sd_samples_df$sd/gse193677_sd_samples_df$mean)*100
```

```{r}
head(gse193677_sd_genes_df)
head(gse193677_sd_samples_df)
gse193677_meta_df
```

#### Merge datasets

Merging of SD / mean calculation by gene:

```{r}
# Full join of gene SD datasets
sd_genes_df = gtex_sd_genes_df %>% 
  full_join(tcga_sd_genes_df) %>% 
  full_join(scipher_sd_genes_df) %>%
  full_join(gse193677_sd_genes_df)

# Write full dataset name
sd_genes_df$dataset_name = paste(sd_genes_df$dataset, 
                                 sd_genes_df$type_dataset, 
                                 sep = ":")
sd_genes_df$dataset_name = tolower(sd_genes_df$dataset_name)
sd_genes_df$dataset_name = ifelse(!((sd_genes_df$subclassification == "") | (is.na(sd_genes_df$subclassification))),
                                  paste(sd_genes_df$dataset_name, 
                                        sd_genes_df$subclassification, 
                                        sep = "-"), 
                                  sd_genes_df$dataset_name)
sd_genes_df$dataset_name = ifelse(!((sd_genes_df$sex == "") | (is.na(sd_genes_df$sex))),
                                  paste(sd_genes_df$dataset_name, 
                                        sd_genes_df$sex, 
                                        sep = "_"), 
                                  sd_genes_df$dataset_name)

# Write output file
sd_genes_file = paste(data_dir, "variation_by_gene.txt", sep="/")
sd_genes_df %>% fwrite(sd_genes_file)
head(sd_genes_df)
```

Merging of SD / mean calculation by samples:

```{r}
# Full join of sample SD datasets
sd_samples_df = gtex_sd_samples_df %>% 
  full_join(tcga_sd_samples_df) %>% 
  full_join(scipher_sd_samples_df) %>%
  full_join(gse193677_sd_samples_df)

# Write full dataset name
sd_samples_df$dataset_name = paste(sd_samples_df$dataset, 
                                   sd_samples_df$type_dataset, 
                                   sep = ":")
sd_samples_df$dataset_name = tolower(sd_samples_df$dataset_name)
sd_samples_df$dataset_name = ifelse(!(sd_samples_df$subclassification == ""),
                                  paste(sd_samples_df$dataset_name, 
                                        sd_samples_df$subclassification, 
                                        sep = "_"), 
                                  sd_samples_df$dataset_name)
sd_samples_df$dataset_name = ifelse(!(sd_samples_df$sex == ""), 
                                    paste(sd_samples_df$dataset_name, 
                                          sd_samples_df$sex, 
                                          sep = "_"), 
                                    sd_samples_df$dataset_name)

# Write output file
sd_samples_file = paste(data_dir, "variation_by_sample.txt", sep="/")
sd_samples_df %>% fwrite(sd_samples_file)
head(sd_samples_df)
```

Merging of metadata:

```{r}
# Full join of metadata
meta_df = gtex_meta_df %>% 
  full_join(tcga_meta_df) %>% 
  full_join(scipher_meta_df) %>%
  full_join(gse193677_meta_df)

# Write full dataset name
meta_df$dataset_name = paste(meta_df$dataset, 
                                   meta_df$type_dataset, 
                                   sep = ":")
meta_df$dataset_name = tolower(meta_df$dataset_name)
meta_df$dataset_name = ifelse(!(meta_df$subclassification == ""),
                              paste(meta_df$dataset_name, 
                                    meta_df$subclassification,
                                    sep = "_"), 
                              meta_df$dataset_name)
meta_df$dataset_name = ifelse(!(meta_df$sex == ""), 
                              paste(meta_df$dataset_name, 
                                    meta_df$sex, 
                                    sep = "_"), 
                              meta_df$dataset_name)

# Write output file
meta_file = paste(data_dir, "metadata_summary.txt", sep="/")
meta_df %>% fwrite(meta_file)
head(meta_df)
```

### Analyze results

#### Genes variation

```{r}
#datasets_selected = c("Artery.Tibial", "Whole.Blood", "TCGA-BRCA", "TCGA-COAD", "scipher.complete.dataset")
#datasets_selected = c("Breast.Mammary.Tissue", "Thyroid", "TCGA-BRCA", "TCGA-THCA", "scipher.complete.dataset")
datasets_selected = c("Breast.Mammary.Tissue", "Thyroid", "Whole.Blood", "TCGA-BRCA", "TCGA-THCA", "scipher.complete.dataset", "gse193677")
parameters = c("sd", "mean", "cv")
parameter2label <- list("sd"="SD", "mean"="mean", "cv"="CV")
distribution_plots = list()
for (x in 1:length(parameters)){
  parameter = parameters[x]
  distribution_plot = sd_genes_df %>%
    #filter(sd < 500000) %>%
    #filter(sd < 5000) %>%
    filter(type_dataset %in% datasets_selected) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "scipher.complete.dataset", "R. arthritis")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "Thyroid", "GTEx: Thyroid")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "Breast.Mammary.Tissue", "GTEx: Breast")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "Whole.Blood", "GTEx: Whole blood")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "TCGA-THCA", "TCGA: Thyroid cancer")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "TCGA-BRCA", "TCGA: Breast cancer")) %>%
    ggplot(aes(.data[[parameter]], color=type_dataset)) +
    geom_freqpoly(size=1.5) +
    #geom_histogram(alpha=0.2) +
    #scale_color_manual(values = c("#D55E00", "#E69F00", "#44AA99", "#0072B2", "#56B4E9")) +
    scale_color_manual(values = c("#D55E00", "#E69F00", "#44AA99", "#65F3BF", "#0072B2", "#56B4E9", "purple")) +
    scale_x_log10() + 
    scale_y_log10() + 
    labs(x=paste("Gene expression ", parameter2label[[parameter]], " (Log)", sep=""), y = "Number of genes (Log)") +
    theme_linedraw() + 
    guides(col=guide_legend(title="Dataset")) +
    theme(aspect.ratio=1, plot.title =  element_text(size = 15, face="bold"), axis.title = element_text(size = 14, face="bold"), axis.text = element_text(size = 13), legend.text = element_text(size = 13), legend.title=element_text(size=14, face="bold"))
  
  # Save distribution plot without changes in the theme, because we will tune it afterwards with patchwork
  distribution_plots[[x]] = distribution_plot
  
  plot_file = paste(plots_dir, "/distribution_", parameter, "_genes.png", sep="")
  ggsave(
    plot_file,
    plot=distribution_plot,
    dpi = 1200,
    #width = 10000,
    #height = 6000,
    units = c("px")
  )
  print(distribution_plot)
}
```


```{r}
input_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables'
a_vs_fraction_corr_file = paste(input_dir, 'a_vs_fraction_sig_correlations_pearson_pval_0.05.txt', sep='/')
a_vs_fraction_corr_df = fread(a_vs_fraction_corr_file)
head(a_vs_fraction_corr_df)
```

```{r}
sd_cut=10
mean_cut=10
cv_cut=50
sd_genes_df %>% group_by(dataset, type_dataset) %>% summarize(mean_sd=mean(sd), median_sd=median(sd), mean_mean=mean(mean), median_mean=median(mean), mean_cv=mean(cv), median_cv=median(cv))
sd_genes_vs_a_df = sd_genes_df %>%
  group_by(dataset, type_dataset, dataset_name) %>%
  summarize(
    num_genes_above_sd_cut=length(sd[sd >= sd_cut]),
    num_genes_above_mean_cut=length(mean[mean >= mean_cut]),
    num_genes_above_cv_cut=length(cv[cv >= cv_cut]),
    num_genes=n()
  ) %>%
  ungroup() %>%
  mutate(
    fraction_genes_above_sd_cut = num_genes_above_sd_cut / num_genes,
    fraction_genes_above_mean_cut = num_genes_above_mean_cut / num_genes,
    fraction_genes_above_cv_cut = num_genes_above_cv_cut / num_genes
  ) %>%
  inner_join(
    a_vs_fraction_corr_df,
    by = c("dataset_name" = "dataset_name")
  ) %>%
  dplyr::select(!c("type_correlation", "correlation", "sample_size_statistical", "sample_size_statistical_corrected", "num_edges_from_statistical_corrected", "num_edges_from_statistical_corrected_norm")) %>%
  unique()

head(sd_genes_vs_a_df)
```

```{r}
parameters = c("sd", "mean", "cv")
parameter2label <- list("sd"="SD", "mean"="mean", "cv"="CV")
scatter_plots = list()
for (x in 1:length(parameters)){
  parameter = parameters[x]
  if (parameter == "cv"){
    cut = cv_cut
  } else if (parameter == "sd"){
    cut = sd_cut
  } else if (parameter == "mean"){
    cut = mean_cut
  }
  param_vs_a_plot = sd_genes_vs_a_df %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "scipher:scipher.complete.dataset", "R. arthritis")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:thyroid", "GTEx: Thyroid")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:breast.mammary.tissue", "GTEx: Breast")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-thca", "TCGA: Thyroid cancer")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-brca", "TCGA: Breast cancer")) %>%
    ggplot(aes(x=a, y=.data[[paste("fraction_genes_above_", parameter, "_cut", sep="")]])) + 
    geom_point() +
    geom_point(data = . %>% filter(dataset_name == "TCGA: Thyroid cancer"), color = "#56B4E9") +
    geom_point(data = . %>% filter(dataset_name == "TCGA: Breast cancer"), color = "#0072B2") +
    geom_point(data = . %>% filter(dataset_name == "R. arthritis"), color = "#44AA99") +
    geom_point(data = . %>% filter(dataset_name == "GTEx: Thyroid"), color = "#E69F00") +
    geom_point(data = . %>% filter(dataset_name == "GTEx: Breast"), color = "#D55E00") +
    xlab(expression(alpha)) +
    ylab(paste("Fraction of genes with ", parameter2label[[parameter]], " above ", cut, sep="")) +
    theme_linedraw() +
    theme(aspect.ratio=1, plot.title =  element_text(size = 15, face="bold"), axis.title = element_text(size = 14, face="bold"), axis.text = element_text(size = 13), legend.text = element_text(size = 13), legend.title=element_text(size=14, face="bold")) +
    geom_text_repel(
      data = subset((sd_genes_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:thyroid", "GTEx: Thyroid"))), dataset_name == "GTEx: Thyroid"),
      aes(label = dataset_name), col = "#E69F00",
      size = 4,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_genes_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:breast.mammary.tissue", "GTEx: Breast"))), dataset_name == "GTEx: Breast"),
      aes(label = dataset_name), col = "#D55E00",
      size = 4,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_genes_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "scipher:scipher.complete.dataset", "R. arthritis"))), dataset_name == "R. arthritis"),
      aes(label = dataset_name), col = "#44AA99",
      size = 4,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_genes_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-brca", "TCGA: Breast cancer"))), dataset_name == "TCGA: Breast cancer"),
      aes(label = dataset_name), col = "#0072B2",
      size = 4,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_genes_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-thca", "TCGA: Thyroid cancer"))), dataset_name == "TCGA: Thyroid cancer"),
      aes(label = dataset_name), col = "#56B4E9",
      size = 4,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    )
  plot_file = paste(plots_dir, "/a_vs_fraction_genes_above_", parameter, "_cut.png", sep="")
  ggsave(
    plot_file,
    dpi = 1200,
    width = 6000,
    #height = 6000,
    units = c("px")
  )
  print(param_vs_a_plot)
  scatter_plots[[x]] = param_vs_a_plot
}
```

Create a panel with the CV results:

```{r}
# Use patchwork to concatenate plots
distribution_plots[[3]] / scatter_plots[[3]] + 
  plot_layout(guides = 'collect') + # Center legend
  plot_annotation(tag_levels = 'A') & # Include letters 
    theme(plot.tag = element_text(face = 'bold')) # that are bold

plot_file = paste(plots_dir, "/cv_results.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  #width = 9000,
  height = 10000,
  units = c("px")
)

```

#### Samples variation

```{r}
datasets_selected = c("Breast.Mammary.Tissue", "Thyroid", "TCGA-BRCA", "TCGA-THCA", "scipher.complete.dataset")
for (parameter in c("sd", "mean", "cv")){
  distribution_plot = sd_samples_df %>%
    #filter(sd < 500000) %>%
    #filter(sd < 5000) %>%
    filter(type_dataset %in% datasets_selected) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "scipher.complete.dataset", "R. arthritis")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "Thyroid", "GTEx: Thyroid")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "Breast.Mammary.Tissue", "GTEx: Breast")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "TCGA-THCA", "TCGA: Thyroid cancer")) %>%
    mutate(type_dataset = replace(type_dataset, type_dataset == "TCGA-BRCA", "TCGA: Breast cancer")) %>%
    ggplot(aes(.data[[parameter]], color=type_dataset)) +
    geom_freqpoly(size=1.5) +
    #geom_histogram(alpha=0.2) +
    scale_color_manual(values = c("#D55E00", "#E69F00", "#44AA99", "#0072B2", "#56B4E9")) +
    scale_x_log10() + 
    scale_y_log10() + 
    labs(x=paste("Gene expression ", parameter, " (Log)", sep=""), y = "Number of samples (Log)") +
    theme_linedraw() + 
    guides(col=guide_legend(title="Dataset")) +
    theme(plot.title =  element_text(size = 17, face="bold"), axis.title = element_text(size = 16, face="bold"), axis.text = element_text(size = 15), legend.text = element_text(size = 14), legend.title=element_text(size=15, face="bold"))
  plot_file = paste(plots_dir, "/distribution_", parameter, "_samples.png", sep="")
  ggsave(
    plot_file,
    plot=distribution_plot,
    dpi = 1200,
    width = 10000,
    height = 6000,
    units = c("px")
  )
  print(distribution_plot)
}
```

```{r}
sd_cut=1000
mean_cut=1000
cv_cut=1000
sd_samples_df %>% group_by(dataset, type_dataset) %>% summarize(mean_sd=mean(sd), median_sd=median(sd), mean_mean=mean(mean), median_mean=median(mean), mean_cv=mean(cv), median_cv=median(cv))
sd_samples_vs_a_df = sd_samples_df %>% group_by(dataset, type_dataset, dataset_name) %>% summarize(num_samples_above_sd_cut=length(sd[sd >= sd_cut]), num_samples_above_mean_cut=length(mean[mean >= mean_cut]), num_samples_above_cv_cut=length(cv[cv >= cv_cut]), num_samples=n()) %>% ungroup() %>% mutate(fraction_samples_above_sd_cut = num_samples_above_sd_cut / num_samples, fraction_samples_above_mean_cut = num_samples_above_mean_cut / num_samples, fraction_samples_above_cv_cut = num_samples_above_cv_cut / num_samples) %>% inner_join(a_vs_fraction_corr_df, by=c("dataset_name"="dataset_name")) %>% select(!c("type_correlation", "correlation", "sample_size_statistical", "sample_size_statistical_corrected", "num_edges_from_statistical_corrected", "num_edges_from_statistical_corrected_norm")) %>% unique()

sd_samples_vs_a_df
```

```{r}
for (parameter in c("sd", "mean", "cv")){
  if (parameter == "cv"){
    cut = cv_cut
  } else if (parameter == "sd"){
    cut = sd_cut
  } else if (parameter == "mean"){
    cut = mean_cut
  }
  param_vs_a_plot = sd_samples_vs_a_df %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "scipher:scipher.complete.dataset", "R. arthritis")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:thyroid", "GTEx: Thyroid")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:breast.mammary.tissue", "GTEx: Breast")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-thca", "TCGA: Thyroid cancer")) %>%
    mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-brca", "TCGA: Breast cancer")) %>%
    ggplot(aes(x=a, y=.data[[paste("fraction_samples_above_", parameter, "_cut", sep="")]])) + 
    geom_point() +
    geom_point(data = . %>% filter(dataset_name == "TCGA: Thyroid cancer"), color = "#56B4E9") +
    geom_point(data = . %>% filter(dataset_name == "TCGA: Breast cancer"), color = "#0072B2") +
    geom_point(data = . %>% filter(dataset_name == "R. arthritis"), color = "#44AA99") +
    geom_point(data = . %>% filter(dataset_name == "GTEx: Thyroid"), color = "#E69F00") +
    geom_point(data = . %>% filter(dataset_name == "GTEx: Breast"), color = "#D55E00") +
    xlab(expression(alpha)) +
    ylab(paste("Fraction of samples with ", parameter, " above ", cut, sep="")) +
    theme_linedraw() +
    theme(plot.title =  element_text(size = 17, face="bold"), axis.title = element_text(size = 16, face="bold"), axis.text = element_text(size = 15), legend.text = element_text(size = 14), legend.title=element_text(size=15, face="bold")) +
    geom_text_repel(
      data = subset((sd_samples_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:thyroid", "GTEx: Thyroid"))), dataset_name == "GTEx: Thyroid"),
      aes(label = dataset_name), col = "#E69F00",
      size = 5,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_samples_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:breast.mammary.tissue", "GTEx: Breast"))), dataset_name == "GTEx: Breast"),
      aes(label = dataset_name), col = "#D55E00",
      size = 5,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_samples_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "scipher:scipher.complete.dataset", "R. arthritis"))), dataset_name == "R. arthritis"),
      aes(label = dataset_name), col = "#44AA99",
      size = 5,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_samples_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-brca", "TCGA: Breast cancer"))), dataset_name == "TCGA: Breast cancer"),
      aes(label = dataset_name), col = "#0072B2",
      size = 5,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    geom_text_repel(
      data = subset((sd_samples_vs_a_df %>% mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-thca", "TCGA: Thyroid cancer"))), dataset_name == "TCGA: Thyroid cancer"),
      aes(label = dataset_name), col = "#56B4E9",
      size = 5,
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.3, "lines")
    )
  plot_file = paste(plots_dir, "/a_vs_fraction_samples_above_", parameter, "_cut.png", sep="")
  ggsave(
    plot_file,
    dpi = 1200,
    width = 6400,
    height = 6000,
    units = c("px")
  )
  print(param_vs_a_plot)
}
```

#### Wilcoxon test between distribution of SDs

```{r}
#pairs = t(combn(c("Artery.Tibial", "Whole.Blood", "TCGA-BRCA", "TCGA-COAD", "scipher.complete.dataset"),2))
pairs = t(combn(c("Breast.Mammary.Tissue", "Thyroid", "TCGA-BRCA", "TCGA-THCA", "scipher.complete.dataset"),2))
cols = c("dataset1", "dataset2", "wilcox", "pval")
wilcox_result_df = data.frame(matrix(ncol=length(cols),nrow=0, dimnames=list(NULL, cols)))
for (x in 1:nrow(pairs)){
  pair=pairs[x,]
  comp_df = inner_join((sd_df %>% filter(type_dataset == pair[1])), (sd_df %>% filter(type_dataset == pair[2])), by="gene")
  wilcox_res = wilcox.test(comp_df$sd.x, comp_df$sd.y, paired = TRUE, alternative = "two.sided")
  wilcox_result_df = rbind(wilcox_result_df, data.frame(dataset1=pair[1], dataset2=pair[2], wilcox=wilcox_res$statistic, pval=wilcox_res$p.value))
}
wilcox_result_df
```
