---
title: "Parse co-expression analysis results"
author: "Joaquim Aguirre-Plans"
date: "1/18/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Parse the results of the analysis of co-expression networks.

```{r, message=FALSE, echo=FALSE, results='hide'}
library(data.table)
library(dplyr)
set.seed(1510)
options(bitmapType='cairo')
```

## Parse results

```{r parse_results, echo=FALSE, results='hide'}
# Input data
input_dir <- '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out'
type_counts = "reads"

# Define result dataframes
cols_topology = c("method", "dataset", "type_dataset", "subclassification", "size", "rep", "threshold", "num_nodes", "num_edges", "av_degree", "av_path_length", "av_clustering_coef", "num_components", "num_lcc_nodes", "num_lcc_edges", "lcc_z", "lcc_pvalue", "max_k", "num_main_core_nodes", "num_main_core_edges")
topology_df = data.frame(matrix(ncol=length(cols_topology),nrow=0, dimnames=list(NULL, cols_topology)))
cols_ppi = c("method", "dataset", "type_dataset", "subclassification", "size", "rep", "threshold", "num_ppi_nodes", "num_ppi_edges", "fraction_ppi_nodes", "fraction_ppi_edges", "num_ppi_main_core_nodes", "num_ppi_main_core_edges", "fraction_ppi_main_core_nodes", "fraction_ppi_main_core_edges")
ppi_results_df = data.frame(matrix(ncol=length(cols_ppi),nrow=0, dimnames=list(NULL, cols_ppi)))
cols_essential = c("method", "dataset", "type_dataset", "subclassification", "size", "rep", "threshold", "num_essential_genes", "fraction_essential_genes", "num_components", "num_lcc_nodes", "fraction_essential_lcc_nodes", "num_lcc_edges", "lcc_z", "lcc_pvalue")
essential_gene_results_df = data.frame(matrix(ncol=length(cols_essential),nrow=0, dimnames=list(NULL, cols_essential)))
cols_disease = c("method", "dataset", "type_dataset", "subclassification", "size", "rep", 'threshold', 'disease', 'disease_class', 'num_disease_genes', 'fraction_disease_genes', 'num_disease_components', 'num_disease_lcc_nodes', 'fraction_disease_lcc_nodes', 'num_disease_lcc_edges', "disease_lcc_z", "disease_lcc_pvalue")
disease_gene_results_df <- setNames(data.frame(matrix(ncol = length(cols_disease), nrow = 0)), cols_disease)

for(dataset in c('scipher', 'gtex', 'tcga', 'geo')){
  if (dataset == 'geo'){
    geo_data <- TRUE
    datasets <- list.files(paste(input_dir, "analysis_geo", sep = "/"))
  } else {
    geo_data <- FALSE
    datasets = c(dataset)
  }
  for (dataset in datasets) {
    if (dataset == 'tcga'){
      subclassifications = c('tumor', 'tissue', 'subtype')
    } else if (!(dataset == 'tcga') & isTRUE(geo_data)) {
      subclassifications = c('disease_inflamed_vs_control')
    } else {
      subclassifications = c(NA)
    }
    for (subclassification in subclassifications){
      if (isTRUE(geo_data)){
          subdatasets_dir = paste(input_dir, "analysis_geo", dataset, type_counts, subclassification, sep="/")
      } else {
        if (is.na(subclassification)){
          subdatasets_dir = paste(input_dir, paste("analysis_", dataset, sep=""), type_counts, sep="/")
          
        } else {
          subdatasets_dir = paste(input_dir, paste("analysis_", dataset, sep=""), type_counts, subclassification, sep="/")
        }
      }
      types_dataset = list.files(subdatasets_dir)
      for(type_dataset in types_dataset){
        result_files = list.files(paste(subdatasets_dir, type_dataset, sep="/"))
        for(result_file in result_files){
          input_file = paste(subdatasets_dir, type_dataset, result_file, sep="/")
          info_file = gsub("_analysis_topology.txt|_analysis_essential_genes.txt|_analysis_disease_genes.txt|_analysis_ppi.txt","",result_file)
          file_split = strsplit(info_file, split="_")[[1]]
          if (("threshold" %in% file_split) | ("edges" %in% file_split)){
            method = file_split[1]
            threshold = as.numeric(file_split[(length(file_split))])
            size = file_split[(length(file_split)-4)]
            r = file_split[(length(file_split))-2]
          } else {
            method = file_split[1]
            size = file_split[(length(file_split)-2)]
            r = file_split[(length(file_split))]
          }
          if(r=="samples"){
            size = "all"
            r = NA
          } else {
            size = as.numeric(size)
            r = as.numeric(r)
          }
          individual_df <- fread(input_file)
          if(grepl("analysis_topology", result_file, fixed = TRUE)){
            topology_df <- rbind(topology_df, cbind(data.frame(method=rep(method, nrow(individual_df)), dataset=rep(dataset, nrow(individual_df)), type_dataset=rep(type_dataset, nrow(individual_df)), subclassification=rep(subclassification, nrow(individual_df)), size=rep(size, nrow(individual_df)), rep=rep(r, nrow(individual_df))), individual_df))
          } else if(grepl("analysis_ppi", result_file, fixed = TRUE)){
            ppi_results_df <- rbind(ppi_results_df, cbind(data.frame(method=rep(method, nrow(individual_df)), dataset=rep(dataset, nrow(individual_df)), type_dataset=rep(type_dataset, nrow(individual_df)), subclassification=rep(subclassification, nrow(individual_df)), size=rep(size, nrow(individual_df)), rep=rep(r, nrow(individual_df))), individual_df))
          } else if(grepl("analysis_essential_genes", result_file, fixed = TRUE)){
            essential_gene_results_df <- rbind(essential_gene_results_df, cbind(data.frame(method=rep(method, nrow(individual_df)), dataset=rep(dataset, nrow(individual_df)), type_dataset=rep(type_dataset, nrow(individual_df)), subclassification=rep(subclassification, nrow(individual_df)), size=rep(size, nrow(individual_df)), rep=rep(r, nrow(individual_df))), individual_df))
          } else if(grepl("analysis_disease_genes", result_file, fixed = TRUE)){
            disease_gene_results_df <- rbind(disease_gene_results_df, cbind(data.frame(method=rep(method, nrow(individual_df)), dataset=rep(dataset, nrow(individual_df)), type_dataset=rep(type_dataset, nrow(individual_df)), subclassification=rep(subclassification, nrow(individual_df)), size=rep(size, nrow(individual_df)), rep=rep(r, nrow(individual_df))), individual_df))
          }
        }
      }
    } 
  }
}
```


```{r output_results, echo=FALSE, results='hide'}
# Define output files
output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data'
topology_file = paste(output_dir, 'analysis_topology.csv', sep='/')
ppi_results_file = paste(output_dir, 'analysis_ppi.csv', sep='/')
disease_genes_file = paste(output_dir, 'analysis_disease_genes.csv', sep='/')
essential_genes_file = paste(output_dir, 'analysis_essential_genes.csv', sep='/')

# Write results
topology_df %>% fwrite(topology_file)
ppi_results_df %>% fwrite(ppi_results_file)
essential_gene_results_df %>% fwrite(essential_genes_file)
disease_gene_results_df %>% fwrite(disease_genes_file)
```


```{r output_num_edges, echo=FALSE, results='hide'}
# Write the clean dataset of number of significant edges to analyze
# Filter by results of the Scipher dataset obtained by Pearson correlation and threshold 0.05
edges_pearson_df = topology_df %>% filter(((method == "pearson") & (type_correlation == "all") & (threshold == 0.05)))
edges_pearson_df$dataset = paste(edges_pearson_df$dataset, edges_pearson_df$type_dataset, sep=":")
edges_pearson_df = edges_pearson_df %>% select(dataset, size, rep, num_edges) %>% arrange(dataset, size, rep)
head(edges_pearson_df)

# Write results
#output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables'
output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data'
edges_file = paste(output_dir, 'example_pearson_pval_0.05', 'pearson_pval_0.05_num_edges_vs_num_samples.txt', sep='/')
edges_pearson_df %>% fwrite(edges_file)
```

```{r output_numbers_datasets, echo=FALSE, results='hide'}
# Check number of nodes and edges in complete graph for each dataset
cols = c("dataset", "total_num_nodes", "total_num_edges")
numbers_df = data.frame(matrix(ncol=length(cols),nrow=0, dimnames=list(NULL, cols)))

# For scipher
scipher_data_dir = '/work/ccnr/j.aguirreplans/data/Scipher/Dec2021'
genes_filt_dir = paste(scipher_data_dir, '00_data/genes_from_rnaseq_filtered_files_by_group', sep='/')
groups = unique((topology_df %>% filter(dataset == "scipher"))$type_dataset)
for (group in groups){
  genes_dataset_file = paste(genes_filt_dir, paste("scipher_rnaseq_gene_info_", group, ".csv", sep=''), sep='/')
  genes_dataset_df = fread(genes_dataset_file)
  num_genes = length(unique((genes_dataset_df %>% filter(enough_counts == TRUE) %>% unique())$HGNC_Symbol)) # Total number of genes
  num_edges = num_genes*(num_genes-1)/2 # Total number of edges
  numbers_df <- rbind(numbers_df, data.frame(dataset=paste("scipher", group, sep=":"), total_num_nodes=num_genes, total_num_edges=num_edges))
}

# For tcga
gdc_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-11-18-Dataset/TCGA'
output_dir = paste(gdc_dir, 'out', sep='/')
projects = unique((topology_df %>% filter(dataset == "tcga"))$type_dataset)
subclassifications = unique((topology_df %>% filter(dataset == "tcga"))$subclassification)
#projects = c("TCGA", "TCGA-BRCA", "TCGA-UCEC")
for (project_id in projects){
  for (subclassification in subclassifications){
    genes_filt_dir = paste(output_dir, type_counts, subclassification, 'filter_genes_low_counts/genes_filtered_files_by_sample_group', sep='/')
    genes_project_file = paste(genes_filt_dir, paste("rnaseq_gene_info_", project_id, ".csv", sep=''), sep='/')
    if(file.exists(genes_project_file)){
      genes_dataset_df = fread(genes_project_file)
      num_genes = length(unique((genes_dataset_df %>% filter(enough_counts == TRUE) %>% unique())$HGNC_Symbol)) # Total number of genes
      num_edges = num_genes*(num_genes-1)/2 # Total number of edges
      dataset = paste("tcga", project_id, sep=":")
      #dataset = paste("tcga:", project_id, "-", subclassification, sep="")
      numbers_df <- rbind(numbers_df, data.frame(dataset=dataset, total_num_nodes=num_genes, total_num_edges=num_edges))
    }
  }
}

# For gtex
gtex_dir = '/work/ccnr/j.aguirreplans/Databases/GTEx/v8'
genes_filt_dir = paste(gtex_dir, type_counts, 'genes_from_rnaseq_filtered_files_by_tissue', sep='/')
#genes_dataset_files = list.files(genes_filt_dir)
tissues = unique((topology_df %>% filter(dataset == "gtex"))$type_dataset)
#tissues = c("Whole.Blood")
for (tissue in tissues){
  genes_dataset_filt_file = paste(genes_filt_dir, paste("gtex_rnaseq_gene_info_", tissue, ".csv", sep=''), sep='/')
  genes_dataset_df = fread(genes_dataset_filt_file)
  num_genes = length(unique((genes_dataset_df %>% filter(enough_counts == TRUE) %>% unique())$HGNC_Symbol)) # Total number of genes
  num_edges = num_genes*(num_genes-1)/2 # Total number of edges
  numbers_df <- rbind(numbers_df, data.frame(dataset=paste("gtex", tissue, sep=":"), total_num_nodes=num_genes, total_num_edges=num_edges))
}

# For geo
geo_dir = '/work/ccnr/j.aguirreplans/Databases/GEO/GSE193677'
output_dir = paste(geo_dir, 'out', sep='/')
projects = unique((topology_df %>% filter(dataset == "GSE193677"))$type_dataset)
subclassifications = unique((topology_df %>% filter(dataset == "GSE193677"))$subclassification)
#projects = c("TCGA", "TCGA-BRCA", "TCGA-UCEC")
for (project_id in projects){
  for (subclassification in subclassifications){
    genes_filt_dir = paste(output_dir, type_counts, 'filter_genes_low_counts', sep='/')
    genes_project_file = paste(genes_filt_dir, paste("rnaseq_gene_info_GSE193677.csv", sep=''), sep='/')
    if(file.exists(genes_project_file)){
      genes_dataset_df = fread(genes_project_file)
      num_genes = length(unique((genes_dataset_df %>% filter(enough_counts == TRUE) %>% unique())$HGNC_Symbol)) # Total number of genes
      num_edges = num_genes*(num_genes-1)/2 # Total number of edges
      dataset = paste("gse193677", project_id, sep=":")
      numbers_df <- rbind(numbers_df, data.frame(dataset=dataset, total_num_nodes=num_genes, total_num_edges=num_edges))
    }
  }
}

# Write table
#output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables'
output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data'
numbers_file = paste(output_dir, 'dataset_numbers_complete_graph.txt', sep='/')
numbers_df %>% fwrite(numbers_file)

numbers_df
```


