---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Description

Create figures for the manuscript, extended abstract and poster.

```{r}
library(data.table)
library(dplyr)
library(ggplot2)
require(magrittr)
library(tidyr)
set.seed(1510)
options(bitmapType='cairo')
`%ni%` <- Negate(`%in%`)
```

### Analyze RNAseq data

Define input directories and files:

```{r}
gtex_dir = "/work/ccnr/j.aguirreplans/Databases/GTEx/v8"
tcga_dir = "/work/ccnr/j.aguirreplans/Databases/TCGA/2022-11-18-Dataset/TCGA/out"
scipher_rnaseq_dir = "/work/ccnr/j.aguirreplans/data/Scipher/Dec2021/00_data/rnaseq_filtered_files_by_group"
geo_dir = "/work/ccnr/j.aguirreplans/Databases/GEO/GSE193677/out"
plots_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots'
```

Read RNA-seq files for each database and keep the number of samples and genes for each dataset:

```{r}
gtex_meta_df = data.frame()
for (type_counts in c("reads", "tpm")){
  gtex_rnaseq_dir = paste(gtex_dir, type_counts, "rnaseq_filtered_files_by_tissue", sep="/")
  rnaseq_files = list.files(gtex_rnaseq_dir)
  for (rnaseq_file in rnaseq_files){
    rnaseq_file_path = paste(gtex_rnaseq_dir, rnaseq_file, sep="/")
    rnaseq_spl = strsplit(gsub(".gct", "", rnaseq_file), split="_", fixed=T)[[1]]
    sex=NA
    if(length(rnaseq_spl)==4){
      sex=rnaseq_spl[4]
    }
    rnaseq = fread(rnaseq_file_path) %>% as.data.frame()
    if (nrow(gtex_meta_df) == 0){
      gtex_meta_df = data.frame(dataset=rnaseq_spl[1], type_dataset=rnaseq_spl[3], type_counts=type_counts, sex=sex, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1]))
    } else {
      gtex_meta_df = rbind(gtex_meta_df, data.frame(dataset=rnaseq_spl[1], type_dataset=rnaseq_spl[3], type_counts=type_counts, sex=sex, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1])))
    }
  }
}
```

```{r}
head(gtex_meta_df)
```

```{r}
tcga_meta_df = data.frame()
for (type_counts in c("reads", "tpm")){
  for (subclassification in c("tumor", "tissue", "subtype")){
    tcga_rnaseq_dir = paste(tcga_dir, type_counts, subclassification, "filter_genes_low_counts/rnaseq_filtered_files_by_sample_group", sep="/")
    rnaseq_files = list.files(tcga_rnaseq_dir)
    for (rnaseq_file in rnaseq_files){
      rnaseq_file_path = paste(tcga_rnaseq_dir, rnaseq_file, sep="/")
      rnaseq_spl = strsplit(gsub(".csv", "", rnaseq_file), split="_", fixed=T)[[1]]
      sex=NA
      if(length(rnaseq_spl)==4){
        sex=rnaseq_spl[4]
      }
      rnaseq = fread(rnaseq_file_path) %>% as.data.frame()
      if (nrow(tcga_meta_df) == 0){
        tcga_meta_df = data.frame(dataset=rnaseq_spl[1], type_dataset=rnaseq_spl[3], type_counts=type_counts, subclassification=subclassification, sex=sex, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1]))
      } else {
        tcga_meta_df = rbind(tcga_meta_df, data.frame(dataset=rnaseq_spl[1], type_dataset=rnaseq_spl[3], type_counts=type_counts, subclassification=subclassification, sex=sex, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1])))
      }
    }
  }
}
```

```{r}
head(tcga_meta_df)
```

```{r}
geo_meta_df = data.frame()
geo_rnaseq_dir = "/work/ccnr/j.aguirreplans/Databases/GEO/GSE193677/out/reads/filter_genes_low_counts/rnaseq_GSE193677.txt"
rnaseq = fread(geo_rnaseq_dir) %>% as.data.frame()
for (type_dataset in c("Rectum_CD_inflamed", "Rectum_UC_inflamed", "Rectum_Control_noninflamed")) {
  if (nrow(tcga_meta_df) == 0){
    geo_meta_df = data.frame(dataset="GSE193677", type_dataset=type_dataset, type_counts="reads", subclassification="disease_inflamed_vs_control", sex=NA, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1]))
  } else {
    geo_meta_df = rbind(geo_meta_df, data.frame(dataset="GSE193677", type_dataset=type_dataset, type_counts="reads", subclassification="disease_inflamed_vs_control", sex=NA, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1])))
  }
}
```

```{r}
head(geo_meta_df)
```

```{r}
scipher_meta_df = data.frame()
rnaseq_files = list.files(scipher_rnaseq_dir)
type_counts="reads"
for (rnaseq_file in rnaseq_files){
  rnaseq_file_path = paste(scipher_rnaseq_dir, rnaseq_file, sep="/")
  rnaseq_spl = strsplit(gsub(".csv", "", rnaseq_file), split="_", fixed=T)[[1]]
  sex=NA
  if(length(rnaseq_spl)==4){
    sex=rnaseq_spl[4]
  }
  rnaseq = fread(rnaseq_file_path) %>% as.data.frame()
  if (nrow(scipher_meta_df) == 0){
    scipher_meta_df = data.frame(dataset=rnaseq_spl[1], type_dataset=rnaseq_spl[3], type_counts=type_counts, sex=sex, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1]))
  } else {
    scipher_meta_df = rbind(scipher_meta_df, data.frame(dataset=rnaseq_spl[1], type_dataset=rnaseq_spl[3], type_counts=type_counts, sex=sex, num_samples=ncol(rnaseq[,-1]), num_genes=nrow(rnaseq[,-1])))
  }
}
```

```{r}
head(scipher_meta_df)
```

```{r}
gtex_meta_df$subclassification = NA
scipher_meta_df$subclassification = NA
num_df = rbind(gtex_meta_df, tcga_meta_df, geo_meta_df, scipher_meta_df)
head(num_df)
```

```{r}
output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data'
numbers_file = paste(output_dir, 'dataset_numbers_samples_genes.txt', sep='/')
num_df %>% fwrite(numbers_file)
```

```{r}
pairs1 = c("Adrenal.Gland", "Bladder",   "Breast.Mammary.Tissue", "Uterus",   "Colon.Sigmoid", "Colon.Transverse", "Liver",     "Minor.Salivary.Gland", 
           "Esophagus.Mucosa", "Esophagus.Gastroesophageal.Junction", "Prostate",  "Stomach",   "Thyroid",   "Lung",      "Lung",      "Kidney.Cortex", 
           "Kidney.Cortex", "Kidney.Cortex", "Brain.Cortex", "Brain.Cortex", "Ovary",   "Pancreas",  "Skin.Sun.Exposed", "Testis",    "Uterus")
pairs2 = c("TCGA-ACC",      "TCGA-BLCA", "TCGA-BRCA",             "TCGA-UCS", "TCGA-READ",     "TCGA-COAD",        "TCGA-LIHC", "TCGA-HNSC",    
           "TCGA-ESCA",        "TCGA-ESCA",                           "TCGA-PRAD", "TCGA-STAD", "TCGA-THCA", "TCGA-LUAD", "TCGA-LUSC", "TCGA-KIRC", 
           "TCGA-KIRP",     "TCGA-KICH",     "TCGA-GBM",     "TCGA-LGG",     "TCGA-OV", "TCGA-PAAD", "TCGA-SKCM",        "TCGA-TGCT", "TCGA-UCEC")
pairs_df = data.frame(gtex_tissue=pairs1, tcga_tissue=pairs2) %>% arrange(gtex_tissue) %>% 
  inner_join((gtex_meta_df %>% select(-dataset, -subclassification) %>% rename(c("num_samples_gtex"="num_samples", "num_genes_gtex"="num_genes"))), by=c("gtex_tissue"="type_dataset")) %>% 
  inner_join((tcga_meta_df %>% select(-dataset) %>% rename(c("num_samples_tcga"="num_samples", "num_genes_tcga"="num_genes"))), by=c("tcga_tissue"="type_dataset", "type_counts"="type_counts", "sex"="sex"))
pairs_df$sum_num_samples = pairs_df$num_samples_gtex + pairs_df$num_samples_tcga
pairs_df$sum_num_genes = pairs_df$num_genes_gtex + pairs_df$num_genes_tcga
pairs_df = pairs_df %>% arrange(desc(sum_num_samples)) %>% relocate(subclassification, .after = sex) %>% relocate(num_genes_gtex, .after = subclassification) %>% relocate(num_genes_tcga, .after = num_genes_gtex) %>% relocate(sum_num_genes, .after = num_genes_tcga)
head(pairs_df, 10)
```

```{r}
output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data'
pairs_file = paste(output_dir, 'dataset_numbers_samples_genes_pairs_gtex_tcga.txt', sep='/')
pairs_df %>% fwrite(pairs_file)
```
