{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "40eb33ac-5838-4926-8ef3-15091938e7cd",
   "metadata": {},
   "source": [
    "# Compile TCGA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1784305d-5cd0-44a1-8784-4424c7a9b9cf",
   "metadata": {},
   "source": [
    "We first define the needed modules, functions and directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3d630aa6-10c3-4a68-b857-59dd61acac70",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Markdown as md\n",
    "import sys, os\n",
    "import json\n",
    "import gzip\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "#import rpy2.robjects as ro\n",
    "#from rpy2.robjects import pandas2ri, numpy2ri\n",
    "#from rpy2.robjects.packages import importr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6996fd15-f847-405d-87e7-8f4e4e740b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fileExist(file):\n",
    "    \"\"\"\n",
    "    Checks if a file exists AND is a file\n",
    "    \"\"\"\n",
    "    return os.path.exists(file) and os.path.isfile(file)\n",
    "\n",
    "def create_directory(directory):\n",
    "    \"\"\"\n",
    "    Checks if a directory exists and if not, creates it\n",
    "    \"\"\"\n",
    "    try:\n",
    "        os.stat(directory)\n",
    "    except:\n",
    "        os.mkdir(directory)\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a4c340af-e280-4b6a-afbe-95bbb1bf74a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define main directories\n",
    "#gdc_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-03-28-Dataset/dummy'\n",
    "#gdc_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-03-28-Dataset/TCGA'\n",
    "gdc_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-07-27-Dataset/dummy'\n",
    "#gdc_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-07-27-Dataset/TCGA'\n",
    "input_dir = os.path.join(gdc_dir, 'raw')\n",
    "data_dir = os.path.join(input_dir, 'data')\n",
    "output_dir = os.path.join(gdc_dir, 'out')\n",
    "folder_list = os.listdir(data_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "87118bda-e76b-4f33-946a-35f2bc2eb779",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dummy directory to run tests\n",
    "dummy_data_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-07-27-Dataset/dummy/raw/data'\n",
    "tcga_data_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-07-27-Dataset/TCGA/raw/data'\n",
    "dummy_files = os.listdir(dummy_data_dir)\n",
    "n_max = 400\n",
    "n = 0\n",
    "if len(dummy_files) == 0:\n",
    "    for file_id in os.listdir(tcga_data_dir):\n",
    "        if n >= n_max:\n",
    "            break\n",
    "        command = \"cp -r {} {}\".format(os.path.join(tcga_data_dir, file_id), dummy_data_dir)\n",
    "        #print(command)\n",
    "        os.system(command)\n",
    "        n+=1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22c07ce7-cdb3-494d-aa0b-04acd310922d",
   "metadata": {},
   "source": [
    "## Parse mappings of files and cases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "166d8370-4280-42e7-8be3-df1c056d5efd",
   "metadata": {},
   "source": [
    "In GDC data, the RNA-seq files are named using **file IDs** (e.g., a82ab02e-4894-43f3-adf3-8a9faebee4e3.htseq.counts.gz), and the patient metadata is named using **case IDs** (e.g., a614255d-9f1b-4e9a-8a50-435d4d123b3f). The mapping between file IDs and case IDs is found in the file `metadata.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b6a298bd-bcd9-446b-8f91-163041add34c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For data from 2022-03-28\n",
    "#mappings_file = os.path.join(input_dir, 'additional/files.2022-03-28.json')\n",
    "#with open(mappings_file) as f:\n",
    "#    mappings = json.load(f)\n",
    "\n",
    "#names_corresp = {}\n",
    "#for file in mappings:\n",
    "#    file_id = file[\"file_name\"]\n",
    "#    entity = file[\"cases\"][0] # All cases are composed of only 1 case id\n",
    "#    case_id = entity[\"case_id\"]\n",
    "#    names_corresp[file_id] = case_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e9dd9619-6979-4acc-8393-93bcd2ce0367",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For data from 2022-07-27\n",
    "metadata_file = os.path.join(input_dir, 'metadata/metadata.txt')\n",
    "metadata_df = pd.read_csv(metadata_file,sep=',',index_col=False,header=0)\n",
    "names_corresp = dict(zip(metadata_df.file_name, metadata_df.file_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2005c4c4-79c9-4ea9-8477-a08d5b5b4019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11123\n",
      "11123\n",
      "400\n"
     ]
    }
   ],
   "source": [
    "print(len(set(names_corresp.keys())))\n",
    "print(len(set(names_corresp.values())))\n",
    "print(len([file_id for file_id in folder_list if file_id in names_corresp.values()]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c66bf11-c1c8-45c3-b39f-8ebbfc715861",
   "metadata": {},
   "source": [
    "## Parse RNA-seq data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddcd32ab-7ea3-43a9-8da0-609a5952884a",
   "metadata": {},
   "source": [
    "We parse the folders inside the input directory, which store the files containing the expression."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19db038f-4ba0-4c78-8541-85d6ba09a3db",
   "metadata": {},
   "source": [
    "We parse the expression files inside the folders, and keep them into a unique dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b430ba8e-d625-44f0-962e-096854ce7f00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "The final dataframe contains **60660 rows (genes)** and **400 columns (samples)**"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_file = os.path.join(output_dir, 'dummy.csv')\n",
    "#output_file = os.path.join(output_dir, 'TCGA.csv')\n",
    "type_gene_expression = \"tpm_unstranded\"\n",
    "i=1\n",
    "if not fileExist(output_file):\n",
    "    # Parse patients data\n",
    "    main_data = None\n",
    "    for patient_folder in folder_list:\n",
    "        patient_folder_path = os.path.join(data_dir, os.fsdecode(patient_folder))\n",
    "        if(os.path.isdir(patient_folder_path)):\n",
    "            for file_id in os.listdir(patient_folder_path):\n",
    "                #if file_id in names_corresp:\n",
    "                    #case_id = names_corresp[file_id] # Get case ID\n",
    "                    file_path = os.path.join(patient_folder_path, os.fsdecode(file_id))\n",
    "                    if(len(file_id.split(\".\")) > 1 and (file_id.split(\".\")[-1] == \"gz\" or file_id.split(\".\")[-1] == \"counts\" or file_id.split(\".\")[-2] == \"augmented_star_gene_counts\")):\n",
    "                        # Uncompress (if necessary) and read file\n",
    "                        if(file_id.split(\".\")[-1] == \"gz\" and len(file_id.split(\".\")) > 2 and file_id.split(\".\")[-3] == \"htseq\"):\n",
    "                            f = gzip.open(file_path)\n",
    "                            aux_data = pd.read_csv(f,sep='\\t',index_col=0,header=None, names=[file_id])\n",
    "                        elif(file_id.split(\".\")[-1] == \"counts\" and file_id.split(\".\")[-2] == \"htseq\"):\n",
    "                            aux_data = pd.read_csv(file_path,sep='\\t',index_col=0,header=None, names=[file_id])\n",
    "                        elif(file_id.split(\".\")[-2] == \"augmented_star_gene_counts\"):\n",
    "                            aux_data = pd.read_csv(file_path,sep='\\t',index_col=0, skiprows=1, header=0)\n",
    "                            aux_data = aux_data[~aux_data.index.isin([\"N_unmapped\", \"N_multimapping\", \"N_noFeature\", \"N_ambiguous\"])]\n",
    "                            aux_data = aux_data[[type_gene_expression]].rename({type_gene_expression: file_id}, axis='columns')\n",
    "                        else:\n",
    "                            print(\"Unknown file format: {}\".format(file_id))\n",
    "                            sys.exit(10)\n",
    "                        # Concatenate previous results\n",
    "                        if main_data is not None:\n",
    "                            #main_data = pd.concat([main_data, aux_data],axis=1,copy=False)\n",
    "                            main_data = pd.merge(main_data, aux_data, left_index=True, right_index=True)\n",
    "                        else:\n",
    "                            main_data = aux_data\n",
    "                        if i%1000 == 0:\n",
    "                            print('{} files processed'.format(i))\n",
    "                        i+=1\n",
    "    # Write output file\n",
    "    main_data.to_csv(output_file, index=True, sep=',')\n",
    "else:\n",
    "    main_data = pd.read_csv(output_file, sep=',', index_col=0)\n",
    "\n",
    "md(\"The final dataframe contains **{} rows (genes)** and **{} columns (samples)**\".format(len(main_data), len(main_data.columns)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "5ed205d8-80e2-47e3-b12e-e33435c20477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>88215dd0-5841-44f1-9393-eefd8238cbb3.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>e74f321c-217f-4bdc-ad17-f132501b5157.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>09755ce8-ed89-411a-a42f-b3edc4e41eeb.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>1e100c4d-13eb-4e5a-9117-ead385902710.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>84e40c7e-2f60-4461-b14c-0859697295a4.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>ba295155-272e-43eb-9d6a-e4c9c392e68b.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>8d1641ea-7552-4d23-9298-094e0056386a.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>523c76c5-ac4d-4a69-88b8-0857af45de7b.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>5abd235a-829b-4457-8f47-ecd1adab30ca.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>49a0c23d-e0e7-4b67-b9f0-527c14105345.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>...</th>\n",
       "      <th>e2ae0ec8-6198-4a6f-8a06-197a8ec81384.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>087c2c93-d071-41e1-a2e8-a1961ba54a86.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>12e291b4-8090-40f0-975c-4ffebb4f695a.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>3a770603-105c-45ef-9b26-49918dd7bc8e.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>d3a67b19-5e2e-48ec-a887-a2d605861717.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>e121b871-b177-47ef-a4f1-de73cb338519.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>868bc289-3c43-4a0f-bbb1-703f0e9fa33a.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>26fdaa75-cea5-4355-b3b1-6b5c2c75e542.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>c8d14d90-63a8-4576-ae80-ef8582792d16.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <th>ec79ad9f-dac4-49c6-8794-5168fb8fabf8.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <td>54.5637</td>\n",
       "      <td>27.3874</td>\n",
       "      <td>50.3093</td>\n",
       "      <td>27.8715</td>\n",
       "      <td>37.8239</td>\n",
       "      <td>56.2216</td>\n",
       "      <td>28.5350</td>\n",
       "      <td>50.5159</td>\n",
       "      <td>138.9149</td>\n",
       "      <td>59.9764</td>\n",
       "      <td>...</td>\n",
       "      <td>3.3343</td>\n",
       "      <td>38.9980</td>\n",
       "      <td>47.3929</td>\n",
       "      <td>28.3164</td>\n",
       "      <td>25.3789</td>\n",
       "      <td>36.9583</td>\n",
       "      <td>24.4795</td>\n",
       "      <td>56.5504</td>\n",
       "      <td>41.6066</td>\n",
       "      <td>0.2762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>0.0966</td>\n",
       "      <td>0.0482</td>\n",
       "      <td>1.8636</td>\n",
       "      <td>0.0333</td>\n",
       "      <td>0.3392</td>\n",
       "      <td>0.2768</td>\n",
       "      <td>5.1690</td>\n",
       "      <td>0.0318</td>\n",
       "      <td>1.2867</td>\n",
       "      <td>0.0520</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0914</td>\n",
       "      <td>0.0453</td>\n",
       "      <td>0.8806</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>0.0319</td>\n",
       "      <td>3.9432</td>\n",
       "      <td>5.5485</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.13</th>\n",
       "      <td>93.9366</td>\n",
       "      <td>113.4229</td>\n",
       "      <td>85.7241</td>\n",
       "      <td>49.5446</td>\n",
       "      <td>51.3701</td>\n",
       "      <td>126.9161</td>\n",
       "      <td>101.9253</td>\n",
       "      <td>93.4030</td>\n",
       "      <td>77.2900</td>\n",
       "      <td>138.3704</td>\n",
       "      <td>...</td>\n",
       "      <td>66.3315</td>\n",
       "      <td>125.9330</td>\n",
       "      <td>168.6618</td>\n",
       "      <td>43.8181</td>\n",
       "      <td>71.2627</td>\n",
       "      <td>81.3964</td>\n",
       "      <td>70.7812</td>\n",
       "      <td>72.9181</td>\n",
       "      <td>46.3479</td>\n",
       "      <td>33.9519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>13.0767</td>\n",
       "      <td>12.6289</td>\n",
       "      <td>9.2839</td>\n",
       "      <td>1.8261</td>\n",
       "      <td>4.6496</td>\n",
       "      <td>25.4778</td>\n",
       "      <td>11.2845</td>\n",
       "      <td>8.1861</td>\n",
       "      <td>3.8629</td>\n",
       "      <td>6.4018</td>\n",
       "      <td>...</td>\n",
       "      <td>4.2738</td>\n",
       "      <td>7.5792</td>\n",
       "      <td>8.5848</td>\n",
       "      <td>2.6325</td>\n",
       "      <td>4.0725</td>\n",
       "      <td>8.4650</td>\n",
       "      <td>5.6966</td>\n",
       "      <td>28.2136</td>\n",
       "      <td>3.7234</td>\n",
       "      <td>9.2297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>2.4590</td>\n",
       "      <td>12.1395</td>\n",
       "      <td>3.2253</td>\n",
       "      <td>0.4112</td>\n",
       "      <td>1.7095</td>\n",
       "      <td>15.4251</td>\n",
       "      <td>3.6297</td>\n",
       "      <td>5.0929</td>\n",
       "      <td>1.7255</td>\n",
       "      <td>2.3896</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9529</td>\n",
       "      <td>5.1320</td>\n",
       "      <td>6.0787</td>\n",
       "      <td>0.4368</td>\n",
       "      <td>1.3253</td>\n",
       "      <td>3.4165</td>\n",
       "      <td>2.5940</td>\n",
       "      <td>16.5258</td>\n",
       "      <td>0.5460</td>\n",
       "      <td>9.1274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288669.1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288670.1</th>\n",
       "      <td>6.0340</td>\n",
       "      <td>10.3222</td>\n",
       "      <td>23.9462</td>\n",
       "      <td>4.6733</td>\n",
       "      <td>15.1532</td>\n",
       "      <td>24.1890</td>\n",
       "      <td>6.3918</td>\n",
       "      <td>8.1920</td>\n",
       "      <td>13.6630</td>\n",
       "      <td>5.1359</td>\n",
       "      <td>...</td>\n",
       "      <td>12.1709</td>\n",
       "      <td>12.0578</td>\n",
       "      <td>8.2870</td>\n",
       "      <td>9.1025</td>\n",
       "      <td>37.7176</td>\n",
       "      <td>10.4781</td>\n",
       "      <td>5.7829</td>\n",
       "      <td>30.7163</td>\n",
       "      <td>18.7374</td>\n",
       "      <td>13.0228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288671.1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288674.1</th>\n",
       "      <td>0.0519</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.0550</td>\n",
       "      <td>0.0102</td>\n",
       "      <td>0.0360</td>\n",
       "      <td>0.0910</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0293</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>0.0239</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0433</td>\n",
       "      <td>0.0484</td>\n",
       "      <td>0.0421</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.0529</td>\n",
       "      <td>0.0169</td>\n",
       "      <td>0.0440</td>\n",
       "      <td>0.1180</td>\n",
       "      <td>0.0212</td>\n",
       "      <td>0.2232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288675.1</th>\n",
       "      <td>2.7573</td>\n",
       "      <td>2.3731</td>\n",
       "      <td>0.5038</td>\n",
       "      <td>0.6137</td>\n",
       "      <td>0.3439</td>\n",
       "      <td>0.9379</td>\n",
       "      <td>1.1669</td>\n",
       "      <td>0.5035</td>\n",
       "      <td>0.2740</td>\n",
       "      <td>1.3240</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8082</td>\n",
       "      <td>0.6469</td>\n",
       "      <td>0.4818</td>\n",
       "      <td>2.5472</td>\n",
       "      <td>1.5138</td>\n",
       "      <td>0.6288</td>\n",
       "      <td>1.0647</td>\n",
       "      <td>0.4225</td>\n",
       "      <td>0.0970</td>\n",
       "      <td>0.8522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60660 rows × 400 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    88215dd0-5841-44f1-9393-eefd8238cbb3.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            54.5637                             \n",
       "ENSG00000000005.6                                              0.0966                             \n",
       "ENSG00000000419.13                                            93.9366                             \n",
       "ENSG00000000457.14                                            13.0767                             \n",
       "ENSG00000000460.17                                             2.4590                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                              6.0340                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0519                             \n",
       "ENSG00000288675.1                                              2.7573                             \n",
       "\n",
       "                    e74f321c-217f-4bdc-ad17-f132501b5157.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            27.3874                             \n",
       "ENSG00000000005.6                                              0.0482                             \n",
       "ENSG00000000419.13                                           113.4229                             \n",
       "ENSG00000000457.14                                            12.6289                             \n",
       "ENSG00000000460.17                                            12.1395                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             10.3222                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0888                             \n",
       "ENSG00000288675.1                                              2.3731                             \n",
       "\n",
       "                    09755ce8-ed89-411a-a42f-b3edc4e41eeb.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            50.3093                             \n",
       "ENSG00000000005.6                                              1.8636                             \n",
       "ENSG00000000419.13                                            85.7241                             \n",
       "ENSG00000000457.14                                             9.2839                             \n",
       "ENSG00000000460.17                                             3.2253                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             23.9462                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0550                             \n",
       "ENSG00000288675.1                                              0.5038                             \n",
       "\n",
       "                    1e100c4d-13eb-4e5a-9117-ead385902710.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            27.8715                             \n",
       "ENSG00000000005.6                                              0.0333                             \n",
       "ENSG00000000419.13                                            49.5446                             \n",
       "ENSG00000000457.14                                             1.8261                             \n",
       "ENSG00000000460.17                                             0.4112                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                              4.6733                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0102                             \n",
       "ENSG00000288675.1                                              0.6137                             \n",
       "\n",
       "                    84e40c7e-2f60-4461-b14c-0859697295a4.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            37.8239                             \n",
       "ENSG00000000005.6                                              0.3392                             \n",
       "ENSG00000000419.13                                            51.3701                             \n",
       "ENSG00000000457.14                                             4.6496                             \n",
       "ENSG00000000460.17                                             1.7095                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             15.1532                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0360                             \n",
       "ENSG00000288675.1                                              0.3439                             \n",
       "\n",
       "                    ba295155-272e-43eb-9d6a-e4c9c392e68b.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            56.2216                             \n",
       "ENSG00000000005.6                                              0.2768                             \n",
       "ENSG00000000419.13                                           126.9161                             \n",
       "ENSG00000000457.14                                            25.4778                             \n",
       "ENSG00000000460.17                                            15.4251                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             24.1890                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0910                             \n",
       "ENSG00000288675.1                                              0.9379                             \n",
       "\n",
       "                    8d1641ea-7552-4d23-9298-094e0056386a.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            28.5350                             \n",
       "ENSG00000000005.6                                              5.1690                             \n",
       "ENSG00000000419.13                                           101.9253                             \n",
       "ENSG00000000457.14                                            11.2845                             \n",
       "ENSG00000000460.17                                             3.6297                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                              6.3918                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0000                             \n",
       "ENSG00000288675.1                                              1.1669                             \n",
       "\n",
       "                    523c76c5-ac4d-4a69-88b8-0857af45de7b.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            50.5159                             \n",
       "ENSG00000000005.6                                              0.0318                             \n",
       "ENSG00000000419.13                                            93.4030                             \n",
       "ENSG00000000457.14                                             8.1861                             \n",
       "ENSG00000000460.17                                             5.0929                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                              8.1920                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0293                             \n",
       "ENSG00000288675.1                                              0.5035                             \n",
       "\n",
       "                    5abd235a-829b-4457-8f47-ecd1adab30ca.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                           138.9149                             \n",
       "ENSG00000000005.6                                              1.2867                             \n",
       "ENSG00000000419.13                                            77.2900                             \n",
       "ENSG00000000457.14                                             3.8629                             \n",
       "ENSG00000000460.17                                             1.7255                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             13.6630                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0120                             \n",
       "ENSG00000288675.1                                              0.2740                             \n",
       "\n",
       "                    49a0c23d-e0e7-4b67-b9f0-527c14105345.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            59.9764                             \n",
       "ENSG00000000005.6                                              0.0520                             \n",
       "ENSG00000000419.13                                           138.3704                             \n",
       "ENSG00000000457.14                                             6.4018                             \n",
       "ENSG00000000460.17                                             2.3896                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                              5.1359                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0239                             \n",
       "ENSG00000288675.1                                              1.3240                             \n",
       "\n",
       "                    ...  \\\n",
       "gene_id             ...   \n",
       "ENSG00000000003.15  ...   \n",
       "ENSG00000000005.6   ...   \n",
       "ENSG00000000419.13  ...   \n",
       "ENSG00000000457.14  ...   \n",
       "ENSG00000000460.17  ...   \n",
       "...                 ...   \n",
       "ENSG00000288669.1   ...   \n",
       "ENSG00000288670.1   ...   \n",
       "ENSG00000288671.1   ...   \n",
       "ENSG00000288674.1   ...   \n",
       "ENSG00000288675.1   ...   \n",
       "\n",
       "                    e2ae0ec8-6198-4a6f-8a06-197a8ec81384.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                             3.3343                             \n",
       "ENSG00000000005.6                                              0.0000                             \n",
       "ENSG00000000419.13                                            66.3315                             \n",
       "ENSG00000000457.14                                             4.2738                             \n",
       "ENSG00000000460.17                                             0.9529                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             12.1709                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0433                             \n",
       "ENSG00000288675.1                                              2.8082                             \n",
       "\n",
       "                    087c2c93-d071-41e1-a2e8-a1961ba54a86.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            38.9980                             \n",
       "ENSG00000000005.6                                              0.0000                             \n",
       "ENSG00000000419.13                                           125.9330                             \n",
       "ENSG00000000457.14                                             7.5792                             \n",
       "ENSG00000000460.17                                             5.1320                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             12.0578                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0484                             \n",
       "ENSG00000288675.1                                              0.6469                             \n",
       "\n",
       "                    12e291b4-8090-40f0-975c-4ffebb4f695a.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            47.3929                             \n",
       "ENSG00000000005.6                                              0.0914                             \n",
       "ENSG00000000419.13                                           168.6618                             \n",
       "ENSG00000000457.14                                             8.5848                             \n",
       "ENSG00000000460.17                                             6.0787                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                              8.2870                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0421                             \n",
       "ENSG00000288675.1                                              0.4818                             \n",
       "\n",
       "                    3a770603-105c-45ef-9b26-49918dd7bc8e.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            28.3164                             \n",
       "ENSG00000000005.6                                              0.0453                             \n",
       "ENSG00000000419.13                                            43.8181                             \n",
       "ENSG00000000457.14                                             2.6325                             \n",
       "ENSG00000000460.17                                             0.4368                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                              9.1025                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0139                             \n",
       "ENSG00000288675.1                                              2.5472                             \n",
       "\n",
       "                    d3a67b19-5e2e-48ec-a887-a2d605861717.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            25.3789                             \n",
       "ENSG00000000005.6                                              0.8806                             \n",
       "ENSG00000000419.13                                            71.2627                             \n",
       "ENSG00000000457.14                                             4.0725                             \n",
       "ENSG00000000460.17                                             1.3253                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             37.7176                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0529                             \n",
       "ENSG00000288675.1                                              1.5138                             \n",
       "\n",
       "                    e121b871-b177-47ef-a4f1-de73cb338519.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            36.9583                             \n",
       "ENSG00000000005.6                                              0.0551                             \n",
       "ENSG00000000419.13                                            81.3964                             \n",
       "ENSG00000000457.14                                             8.4650                             \n",
       "ENSG00000000460.17                                             3.4165                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             10.4781                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0169                             \n",
       "ENSG00000288675.1                                              0.6288                             \n",
       "\n",
       "                    868bc289-3c43-4a0f-bbb1-703f0e9fa33a.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            24.4795                             \n",
       "ENSG00000000005.6                                              0.0319                             \n",
       "ENSG00000000419.13                                            70.7812                             \n",
       "ENSG00000000457.14                                             5.6966                             \n",
       "ENSG00000000460.17                                             2.5940                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0118                             \n",
       "ENSG00000288670.1                                              5.7829                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0440                             \n",
       "ENSG00000288675.1                                              1.0647                             \n",
       "\n",
       "                    26fdaa75-cea5-4355-b3b1-6b5c2c75e542.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            56.5504                             \n",
       "ENSG00000000005.6                                              3.9432                             \n",
       "ENSG00000000419.13                                            72.9181                             \n",
       "ENSG00000000457.14                                            28.2136                             \n",
       "ENSG00000000460.17                                            16.5258                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             30.7163                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.1180                             \n",
       "ENSG00000288675.1                                              0.4225                             \n",
       "\n",
       "                    c8d14d90-63a8-4576-ae80-ef8582792d16.rna_seq.augmented_star_gene_counts.tsv  \\\n",
       "gene_id                                                                                           \n",
       "ENSG00000000003.15                                            41.6066                             \n",
       "ENSG00000000005.6                                              5.5485                             \n",
       "ENSG00000000419.13                                            46.3479                             \n",
       "ENSG00000000457.14                                             3.7234                             \n",
       "ENSG00000000460.17                                             0.5460                             \n",
       "...                                                               ...                             \n",
       "ENSG00000288669.1                                              0.0000                             \n",
       "ENSG00000288670.1                                             18.7374                             \n",
       "ENSG00000288671.1                                              0.0000                             \n",
       "ENSG00000288674.1                                              0.0212                             \n",
       "ENSG00000288675.1                                              0.0970                             \n",
       "\n",
       "                    ec79ad9f-dac4-49c6-8794-5168fb8fabf8.rna_seq.augmented_star_gene_counts.tsv  \n",
       "gene_id                                                                                          \n",
       "ENSG00000000003.15                                             0.2762                            \n",
       "ENSG00000000005.6                                              0.0000                            \n",
       "ENSG00000000419.13                                            33.9519                            \n",
       "ENSG00000000457.14                                             9.2297                            \n",
       "ENSG00000000460.17                                             9.1274                            \n",
       "...                                                               ...                            \n",
       "ENSG00000288669.1                                              0.0449                            \n",
       "ENSG00000288670.1                                             13.0228                            \n",
       "ENSG00000288671.1                                              0.0000                            \n",
       "ENSG00000288674.1                                              0.2232                            \n",
       "ENSG00000288675.1                                              0.8522                            \n",
       "\n",
       "[60660 rows x 400 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddd0ccb6-a3c4-42e1-942a-cfa543739c88",
   "metadata": {},
   "source": [
    "## Process metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa9dbb32-387e-468d-95fd-3413194a0521",
   "metadata": {},
   "source": [
    "There are two folders of metadata in GDC:\n",
    "- Biospecimen: Information about the samples, and how they were extracted.\n",
    "    - `aliquot.tsv`\n",
    "    - `analyte.tsv`\n",
    "    - `portion.tsv`\n",
    "    - `sample.tsv`\n",
    "    - `slide.tsv`\n",
    "- Clinical\n",
    "    - `clinical.tsv`: Age, if there is treatment, and which type of treatment.\n",
    "    - `exposure.tsv`: Exposure to substances such as alcohol or tobacco.\n",
    "    - `family_history.tsv`: Empty.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ae5695b-1bc1-41ca-b2e0-4055cf953add",
   "metadata": {},
   "source": [
    "Let's analyze the files in the **Biospecimen** folder:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "56b2a087-d503-4d9e-b070-3c04a4aaec12",
   "metadata": {},
   "outputs": [],
   "source": [
    "biospecimen_dir = os.path.join(input_dir, 'biospecimen')\n",
    "aliquot_file = os.path.join(biospecimen_dir, 'aliquot.tsv')\n",
    "analyte_file = os.path.join(biospecimen_dir, 'analyte.tsv')\n",
    "portion_file = os.path.join(biospecimen_dir, 'portion.tsv')\n",
    "sample_file = os.path.join(biospecimen_dir, 'sample.tsv')\n",
    "slide_file = os.path.join(biospecimen_dir, 'slide.tsv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f8067dfd-ff21-4073-bd78-4a851b2cabca",
   "metadata": {},
   "outputs": [],
   "source": [
    "aliquot_df = pd.read_csv(aliquot_file,sep='\\t')\n",
    "analyte_df = pd.read_csv(analyte_file,sep='\\t')\n",
    "portion_df = pd.read_csv(portion_file,sep='\\t')\n",
    "sample_df = pd.read_csv(sample_file,sep='\\t')\n",
    "slide_df = pd.read_csv(slide_file,sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b322eea9-6f4a-4383-af31-8834256ff954",
   "metadata": {},
   "source": [
    "Let's analyze the files in the **Clinical** folder:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fb716428-b2e9-4ea1-889b-89176be48b87",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_dir = os.path.join(input_dir, 'clinical')\n",
    "clinical_file = os.path.join(clinical_dir, 'clinical.tsv')\n",
    "exposure_file = os.path.join(clinical_dir, 'exposure.tsv')\n",
    "family_history_file = os.path.join(clinical_dir, 'family_history.tsv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "557036cf-f2df-47bd-b106-883f06af3261",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_df = pd.read_csv(clinical_file,sep='\\t')\n",
    "exposure_df = pd.read_csv(exposure_file,sep='\\t')\n",
    "family_history_df = pd.read_csv(family_history_file,sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b2c7d5bc-0600-4053-911f-99176fe6c4d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>project_id</th>\n",
       "      <th>age_at_index</th>\n",
       "      <th>age_is_obfuscated</th>\n",
       "      <th>cause_of_death</th>\n",
       "      <th>cause_of_death_source</th>\n",
       "      <th>country_of_residence_at_enrollment</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment_arm</th>\n",
       "      <th>treatment_dose</th>\n",
       "      <th>treatment_dose_units</th>\n",
       "      <th>treatment_effect</th>\n",
       "      <th>treatment_effect_indicator</th>\n",
       "      <th>treatment_frequency</th>\n",
       "      <th>treatment_intent_type</th>\n",
       "      <th>treatment_or_therapy</th>\n",
       "      <th>treatment_outcome</th>\n",
       "      <th>treatment_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>febe2ce5-737b-43b8-bc70-4194fe3ed5fb</td>\n",
       "      <td>TCGA-FG-A4MY</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>44</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-16310</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>febe2ce5-737b-43b8-bc70-4194fe3ed5fb</td>\n",
       "      <td>TCGA-FG-A4MY</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>44</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-16310</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>yes</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b2bcc97f-03a6-409a-8b91-78221cb10d7f</td>\n",
       "      <td>TCGA-YL-A9WK</td>\n",
       "      <td>TCGA-PRAD</td>\n",
       "      <td>63</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-23272</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>yes</td>\n",
       "      <td>'--</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b2bcc97f-03a6-409a-8b91-78221cb10d7f</td>\n",
       "      <td>TCGA-YL-A9WK</td>\n",
       "      <td>TCGA-PRAD</td>\n",
       "      <td>63</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-23272</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>03c3ae62-d0aa-412e-bd3c-4577fc9f919c</td>\n",
       "      <td>TCGA-BB-A6UM</td>\n",
       "      <td>TCGA-HNSC</td>\n",
       "      <td>52</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19194</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>yes</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 154 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id case_submitter_id project_id  \\\n",
       "0  febe2ce5-737b-43b8-bc70-4194fe3ed5fb      TCGA-FG-A4MY   TCGA-LGG   \n",
       "1  febe2ce5-737b-43b8-bc70-4194fe3ed5fb      TCGA-FG-A4MY   TCGA-LGG   \n",
       "2  b2bcc97f-03a6-409a-8b91-78221cb10d7f      TCGA-YL-A9WK  TCGA-PRAD   \n",
       "3  b2bcc97f-03a6-409a-8b91-78221cb10d7f      TCGA-YL-A9WK  TCGA-PRAD   \n",
       "4  03c3ae62-d0aa-412e-bd3c-4577fc9f919c      TCGA-BB-A6UM  TCGA-HNSC   \n",
       "\n",
       "  age_at_index age_is_obfuscated cause_of_death cause_of_death_source  \\\n",
       "0           44               '--            '--                   '--   \n",
       "1           44               '--            '--                   '--   \n",
       "2           63               '--            '--                   '--   \n",
       "3           63               '--            '--                   '--   \n",
       "4           52               '--            '--                   '--   \n",
       "\n",
       "  country_of_residence_at_enrollment days_to_birth days_to_death  ...  \\\n",
       "0                                '--        -16310           '--  ...   \n",
       "1                                '--        -16310           '--  ...   \n",
       "2                                '--        -23272           '--  ...   \n",
       "3                                '--        -23272           '--  ...   \n",
       "4                                '--        -19194           '--  ...   \n",
       "\n",
       "  treatment_arm treatment_dose treatment_dose_units treatment_effect  \\\n",
       "0           '--            '--                  '--              '--   \n",
       "1           '--            '--                  '--              '--   \n",
       "2           '--            '--                  '--              '--   \n",
       "3           '--            '--                  '--              '--   \n",
       "4           '--            '--                  '--              '--   \n",
       "\n",
       "  treatment_effect_indicator treatment_frequency treatment_intent_type  \\\n",
       "0                        '--                 '--                   '--   \n",
       "1                        '--                 '--                   '--   \n",
       "2                        '--                 '--                   '--   \n",
       "3                        '--                 '--                   '--   \n",
       "4                        '--                 '--                   '--   \n",
       "\n",
       "  treatment_or_therapy treatment_outcome               treatment_type  \n",
       "0                   no               '--  Pharmaceutical Therapy, NOS  \n",
       "1                  yes               '--       Radiation Therapy, NOS  \n",
       "2                  yes               '--  Pharmaceutical Therapy, NOS  \n",
       "3                   no               '--       Radiation Therapy, NOS  \n",
       "4                  yes               '--       Radiation Therapy, NOS  \n",
       "\n",
       "[5 rows x 154 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b6cf10e7-45f5-43ed-b113-13ef859ee09e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20396, 154)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3b0bd156-e258-4bda-883b-dd8f7547eff6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>project_id</th>\n",
       "      <th>age_at_onset</th>\n",
       "      <th>alcohol_days_per_week</th>\n",
       "      <th>alcohol_drinks_per_day</th>\n",
       "      <th>alcohol_history</th>\n",
       "      <th>alcohol_intensity</th>\n",
       "      <th>alcohol_type</th>\n",
       "      <th>asbestos_exposure</th>\n",
       "      <th>...</th>\n",
       "      <th>smoking_frequency</th>\n",
       "      <th>time_between_waking_and_first_smoke</th>\n",
       "      <th>tobacco_smoking_onset_year</th>\n",
       "      <th>tobacco_smoking_quit_year</th>\n",
       "      <th>tobacco_smoking_status</th>\n",
       "      <th>tobacco_use_per_day</th>\n",
       "      <th>type_of_smoke_exposure</th>\n",
       "      <th>type_of_tobacco_used</th>\n",
       "      <th>weight</th>\n",
       "      <th>years_smoked</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>febe2ce5-737b-43b8-bc70-4194fe3ed5fb</td>\n",
       "      <td>TCGA-FG-A4MY</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b2bcc97f-03a6-409a-8b91-78221cb10d7f</td>\n",
       "      <td>TCGA-YL-A9WK</td>\n",
       "      <td>TCGA-PRAD</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>03c3ae62-d0aa-412e-bd3c-4577fc9f919c</td>\n",
       "      <td>TCGA-BB-A6UM</td>\n",
       "      <td>TCGA-HNSC</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>No</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>588a48f7-947c-463e-8991-9a4ee82683aa</td>\n",
       "      <td>TCGA-D3-A8GS</td>\n",
       "      <td>TCGA-SKCM</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fd3315da-c870-4ad0-9d2a-50b1647d3e46</td>\n",
       "      <td>TCGA-J8-A4HW</td>\n",
       "      <td>TCGA-THCA</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id case_submitter_id project_id  \\\n",
       "0  febe2ce5-737b-43b8-bc70-4194fe3ed5fb      TCGA-FG-A4MY   TCGA-LGG   \n",
       "1  b2bcc97f-03a6-409a-8b91-78221cb10d7f      TCGA-YL-A9WK  TCGA-PRAD   \n",
       "2  03c3ae62-d0aa-412e-bd3c-4577fc9f919c      TCGA-BB-A6UM  TCGA-HNSC   \n",
       "3  588a48f7-947c-463e-8991-9a4ee82683aa      TCGA-D3-A8GS  TCGA-SKCM   \n",
       "4  fd3315da-c870-4ad0-9d2a-50b1647d3e46      TCGA-J8-A4HW  TCGA-THCA   \n",
       "\n",
       "  age_at_onset alcohol_days_per_week alcohol_drinks_per_day alcohol_history  \\\n",
       "0          '--                   '--                    '--    Not Reported   \n",
       "1          '--                   '--                    '--    Not Reported   \n",
       "2          '--                   '--                    '--              No   \n",
       "3          '--                   '--                    '--    Not Reported   \n",
       "4          '--                   '--                    '--    Not Reported   \n",
       "\n",
       "  alcohol_intensity alcohol_type asbestos_exposure  ... smoking_frequency  \\\n",
       "0               '--          '--               '--  ...               '--   \n",
       "1               '--          '--               '--  ...               '--   \n",
       "2               '--          '--               '--  ...               '--   \n",
       "3               '--          '--               '--  ...               '--   \n",
       "4               '--          '--               '--  ...               '--   \n",
       "\n",
       "  time_between_waking_and_first_smoke tobacco_smoking_onset_year  \\\n",
       "0                                 '--                        '--   \n",
       "1                                 '--                        '--   \n",
       "2                                 '--                        '--   \n",
       "3                                 '--                        '--   \n",
       "4                                 '--                        '--   \n",
       "\n",
       "  tobacco_smoking_quit_year tobacco_smoking_status tobacco_use_per_day  \\\n",
       "0                       '--                    '--                 '--   \n",
       "1                       '--                    '--                 '--   \n",
       "2                       '--                    '--                 '--   \n",
       "3                       '--                    '--                 '--   \n",
       "4                       '--                    '--                 '--   \n",
       "\n",
       "  type_of_smoke_exposure type_of_tobacco_used weight years_smoked  \n",
       "0                    '--                  '--    '--          '--  \n",
       "1                    '--                  '--    '--          '--  \n",
       "2                    '--                  '--    '--         35.0  \n",
       "3                    '--                  '--    '--          '--  \n",
       "4                    '--                  '--    '--          '--  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exposure_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9ca2cc95-47fd-4e32-b4cb-712cbe913152",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10198, 33)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exposure_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50bb8784-8f65-44a7-b765-f7931777b3dd",
   "metadata": {},
   "source": [
    "## Combine data and metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb892b9d-99fd-4dc8-9a83-668774f7d15a",
   "metadata": {},
   "source": [
    "Map the data and metadata can be done using the JSON file with the mappings:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c073f898-a891-4702-9f91-e0b080e55de9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['case_id',\n",
       " 'case_submitter_id',\n",
       " 'project_id',\n",
       " 'age_at_index',\n",
       " 'age_is_obfuscated',\n",
       " 'cause_of_death',\n",
       " 'cause_of_death_source',\n",
       " 'country_of_residence_at_enrollment',\n",
       " 'days_to_birth',\n",
       " 'days_to_death',\n",
       " 'ethnicity',\n",
       " 'gender',\n",
       " 'occupation_duration_years',\n",
       " 'premature_at_birth',\n",
       " 'race',\n",
       " 'vital_status',\n",
       " 'weeks_gestation_at_birth',\n",
       " 'year_of_birth',\n",
       " 'year_of_death',\n",
       " 'age_at_diagnosis',\n",
       " 'ajcc_clinical_m',\n",
       " 'ajcc_clinical_n',\n",
       " 'ajcc_clinical_stage',\n",
       " 'ajcc_clinical_t',\n",
       " 'ajcc_pathologic_m',\n",
       " 'ajcc_pathologic_n',\n",
       " 'ajcc_pathologic_stage',\n",
       " 'ajcc_pathologic_t',\n",
       " 'ajcc_staging_system_edition',\n",
       " 'anaplasia_present',\n",
       " 'anaplasia_present_type',\n",
       " 'ann_arbor_b_symptoms',\n",
       " 'ann_arbor_clinical_stage',\n",
       " 'ann_arbor_extranodal_involvement',\n",
       " 'ann_arbor_pathologic_stage',\n",
       " 'best_overall_response',\n",
       " 'breslow_thickness',\n",
       " 'burkitt_lymphoma_clinical_variant',\n",
       " 'child_pugh_classification',\n",
       " 'circumferential_resection_margin',\n",
       " 'classification_of_tumor',\n",
       " 'cog_liver_stage',\n",
       " 'cog_neuroblastoma_risk_group',\n",
       " 'cog_renal_stage',\n",
       " 'cog_rhabdomyosarcoma_risk_group',\n",
       " 'days_to_best_overall_response',\n",
       " 'days_to_diagnosis',\n",
       " 'days_to_last_follow_up',\n",
       " 'days_to_last_known_disease_status',\n",
       " 'days_to_recurrence',\n",
       " 'eln_risk_classification',\n",
       " 'enneking_msts_grade',\n",
       " 'enneking_msts_metastasis',\n",
       " 'enneking_msts_stage',\n",
       " 'enneking_msts_tumor_site',\n",
       " 'esophageal_columnar_dysplasia_degree',\n",
       " 'esophageal_columnar_metaplasia_present',\n",
       " 'figo_stage',\n",
       " 'figo_staging_edition_year',\n",
       " 'first_symptom_prior_to_diagnosis',\n",
       " 'gastric_esophageal_junction_involvement',\n",
       " 'gleason_grade_group',\n",
       " 'gleason_grade_tertiary',\n",
       " 'gleason_patterns_percent',\n",
       " 'goblet_cells_columnar_mucosa_present',\n",
       " 'greatest_tumor_dimension',\n",
       " 'gross_tumor_weight',\n",
       " 'icd_10_code',\n",
       " 'igcccg_stage',\n",
       " 'inpc_grade',\n",
       " 'inpc_histologic_group',\n",
       " 'inrg_stage',\n",
       " 'inss_stage',\n",
       " 'international_prognostic_index',\n",
       " 'irs_group',\n",
       " 'irs_stage',\n",
       " 'ishak_fibrosis_score',\n",
       " 'iss_stage',\n",
       " 'largest_extrapelvic_peritoneal_focus',\n",
       " 'last_known_disease_status',\n",
       " 'laterality',\n",
       " 'lymph_node_involved_site',\n",
       " 'lymph_nodes_positive',\n",
       " 'lymph_nodes_tested',\n",
       " 'lymphatic_invasion_present',\n",
       " 'margin_distance',\n",
       " 'margins_involved_site',\n",
       " 'masaoka_stage',\n",
       " 'medulloblastoma_molecular_classification',\n",
       " 'metastasis_at_diagnosis',\n",
       " 'metastasis_at_diagnosis_site',\n",
       " 'method_of_diagnosis',\n",
       " 'micropapillary_features',\n",
       " 'mitosis_karyorrhexis_index',\n",
       " 'mitotic_count',\n",
       " 'morphology',\n",
       " 'non_nodal_regional_disease',\n",
       " 'non_nodal_tumor_deposits',\n",
       " 'ovarian_specimen_status',\n",
       " 'ovarian_surface_involvement',\n",
       " 'papillary_renal_cell_type',\n",
       " 'percent_tumor_invasion',\n",
       " 'perineural_invasion_present',\n",
       " 'peripancreatic_lymph_nodes_positive',\n",
       " 'peripancreatic_lymph_nodes_tested',\n",
       " 'peritoneal_fluid_cytological_status',\n",
       " 'pregnant_at_diagnosis',\n",
       " 'primary_diagnosis',\n",
       " 'primary_gleason_grade',\n",
       " 'prior_malignancy',\n",
       " 'prior_treatment',\n",
       " 'progression_or_recurrence',\n",
       " 'residual_disease',\n",
       " 'satellite_nodule_present',\n",
       " 'secondary_gleason_grade',\n",
       " 'site_of_resection_or_biopsy',\n",
       " 'sites_of_involvement',\n",
       " 'supratentorial_localization',\n",
       " 'synchronous_malignancy',\n",
       " 'tissue_or_organ_of_origin',\n",
       " 'transglottic_extension',\n",
       " 'tumor_confined_to_organ_of_origin',\n",
       " 'tumor_depth',\n",
       " 'tumor_focality',\n",
       " 'tumor_grade',\n",
       " 'tumor_largest_dimension_diameter',\n",
       " 'tumor_regression_grade',\n",
       " 'tumor_stage',\n",
       " 'vascular_invasion_present',\n",
       " 'vascular_invasion_type',\n",
       " 'weiss_assessment_score',\n",
       " 'who_cns_grade',\n",
       " 'who_nte_grade',\n",
       " 'wilms_tumor_histologic_subtype',\n",
       " 'year_of_diagnosis',\n",
       " 'chemo_concurrent_to_radiation',\n",
       " 'days_to_treatment_end',\n",
       " 'days_to_treatment_start',\n",
       " 'initial_disease_status',\n",
       " 'number_of_cycles',\n",
       " 'reason_treatment_ended',\n",
       " 'regimen_or_line_of_therapy',\n",
       " 'therapeutic_agents',\n",
       " 'treatment_anatomic_site',\n",
       " 'treatment_arm',\n",
       " 'treatment_dose',\n",
       " 'treatment_dose_units',\n",
       " 'treatment_effect',\n",
       " 'treatment_effect_indicator',\n",
       " 'treatment_frequency',\n",
       " 'treatment_intent_type',\n",
       " 'treatment_or_therapy',\n",
       " 'treatment_outcome',\n",
       " 'treatment_type']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(clinical_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e53ad054-cf20-4d3f-b792-146ccb9c7cf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(112, 154)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = clinical_df[clinical_df[\"case_id\"].isin(main_data.columns)]\n",
    "metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b86203b8-dec9-42f2-ab1a-a8e0367dd250",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = metadata[[\"case_id\", \"case_submitter_id\", \"project_id\", \"cause_of_death\", \"treatment_or_therapy\", \"treatment_outcome\", \"treatment_type\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "39b0db49-a70a-4dae-9041-a3d0f33c6a9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>project_id</th>\n",
       "      <th>cause_of_death</th>\n",
       "      <th>treatment_or_therapy</th>\n",
       "      <th>treatment_outcome</th>\n",
       "      <th>treatment_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>2fabeb98-05e3-4f55-97f5-fbc675e25a3d</td>\n",
       "      <td>TCGA-62-8398</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>2fabeb98-05e3-4f55-97f5-fbc675e25a3d</td>\n",
       "      <td>TCGA-62-8398</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>'--</td>\n",
       "      <td>yes</td>\n",
       "      <td>'--</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>5d66191d-da49-4d7e-89a4-a382db338340</td>\n",
       "      <td>TCGA-D5-6931</td>\n",
       "      <td>TCGA-COAD</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>5d66191d-da49-4d7e-89a4-a382db338340</td>\n",
       "      <td>TCGA-D5-6931</td>\n",
       "      <td>TCGA-COAD</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>c92696ba-f549-480d-8d4a-5f23b27aa336</td>\n",
       "      <td>TCGA-C5-A907</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>'--</td>\n",
       "      <td>yes</td>\n",
       "      <td>'--</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19465</th>\n",
       "      <td>128fd280-9a04-4038-9269-9a945b9a4dc9</td>\n",
       "      <td>TCGA-FU-A770</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>'--</td>\n",
       "      <td>not reported</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19856</th>\n",
       "      <td>305eaef4-4644-46e3-a696-d2e4a972f691</td>\n",
       "      <td>TCGA-CZ-4865</td>\n",
       "      <td>TCGA-KIRC</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19857</th>\n",
       "      <td>305eaef4-4644-46e3-a696-d2e4a972f691</td>\n",
       "      <td>TCGA-CZ-4865</td>\n",
       "      <td>TCGA-KIRC</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19914</th>\n",
       "      <td>ef974c39-8bac-486b-a395-9d1283e08028</td>\n",
       "      <td>TCGA-ZG-A9LB</td>\n",
       "      <td>TCGA-PRAD</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19915</th>\n",
       "      <td>ef974c39-8bac-486b-a395-9d1283e08028</td>\n",
       "      <td>TCGA-ZG-A9LB</td>\n",
       "      <td>TCGA-PRAD</td>\n",
       "      <td>'--</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    case_id case_submitter_id project_id  \\\n",
       "186    2fabeb98-05e3-4f55-97f5-fbc675e25a3d      TCGA-62-8398  TCGA-LUAD   \n",
       "187    2fabeb98-05e3-4f55-97f5-fbc675e25a3d      TCGA-62-8398  TCGA-LUAD   \n",
       "544    5d66191d-da49-4d7e-89a4-a382db338340      TCGA-D5-6931  TCGA-COAD   \n",
       "545    5d66191d-da49-4d7e-89a4-a382db338340      TCGA-D5-6931  TCGA-COAD   \n",
       "666    c92696ba-f549-480d-8d4a-5f23b27aa336      TCGA-C5-A907  TCGA-CESC   \n",
       "...                                     ...               ...        ...   \n",
       "19465  128fd280-9a04-4038-9269-9a945b9a4dc9      TCGA-FU-A770  TCGA-CESC   \n",
       "19856  305eaef4-4644-46e3-a696-d2e4a972f691      TCGA-CZ-4865  TCGA-KIRC   \n",
       "19857  305eaef4-4644-46e3-a696-d2e4a972f691      TCGA-CZ-4865  TCGA-KIRC   \n",
       "19914  ef974c39-8bac-486b-a395-9d1283e08028      TCGA-ZG-A9LB  TCGA-PRAD   \n",
       "19915  ef974c39-8bac-486b-a395-9d1283e08028      TCGA-ZG-A9LB  TCGA-PRAD   \n",
       "\n",
       "      cause_of_death treatment_or_therapy treatment_outcome  \\\n",
       "186              '--                   no               '--   \n",
       "187              '--                  yes               '--   \n",
       "544              '--                   no               '--   \n",
       "545              '--                   no               '--   \n",
       "666              '--                  yes               '--   \n",
       "...              ...                  ...               ...   \n",
       "19465            '--         not reported               '--   \n",
       "19856            '--                   no               '--   \n",
       "19857            '--                   no               '--   \n",
       "19914            '--                   no               '--   \n",
       "19915            '--                   no               '--   \n",
       "\n",
       "                    treatment_type  \n",
       "186         Radiation Therapy, NOS  \n",
       "187    Pharmaceutical Therapy, NOS  \n",
       "544    Pharmaceutical Therapy, NOS  \n",
       "545         Radiation Therapy, NOS  \n",
       "666    Pharmaceutical Therapy, NOS  \n",
       "...                            ...  \n",
       "19465       Radiation Therapy, NOS  \n",
       "19856       Radiation Therapy, NOS  \n",
       "19857  Pharmaceutical Therapy, NOS  \n",
       "19914       Radiation Therapy, NOS  \n",
       "19915  Pharmaceutical Therapy, NOS  \n",
       "\n",
       "[112 rows x 7 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fce3945-3e2a-4de4-91e5-85f653935481",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0d21ebf-0017-4791-adb8-051fc000def5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c8f866f-be44-49bd-b1f5-e201b08625a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "​\n",
    "​\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects import pandas2ri, numpy2ri\n",
    "from rpy2.robjects.packages import importr\n",
    "import gzip\n",
    "import json\n",
    "​\n",
    "​\n",
    "​\n",
    "ro.pandas2ri.activate()\n",
    "​\n",
    "ro.r('library(biomaRt)')\n",
    "ro.r('human = useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\",  host=\"www.ensembl.org\")')\n",
    "​\n",
    "file_folder = os.path.abspath(os.path.dirname(__file__))\n",
    "gdc_folder = os.path.dirname(file_folder) +\"/gdc/\"\n",
    "folder = file_folder + '/results/'\n",
    "with open(gdc_folder + \"metadata.cart.2018-10-15.json\") as f:\n",
    "    metadata = json.load(f)\n",
    "​\n",
    "ids = []\n",
    "filenames = []\n",
    "for file in metadata:\n",
    "    filenames += [file[\"file_name\"]]\n",
    "    entity = file[\"associated_entities\"][0]\n",
    "    ids += [entity[\"case_id\"]]\n",
    "names_corresp = {filenames[i] : ids[i] for i in range(len(ids))}\n",
    "clinical = pd.read_csv(gdc_folder +\"clinical.tsv\", delimiter='\\t') \n",
    "case_id = clinical.loc[:,\"case_id\"]\n",
    "clinical.index = case_id\n",
    "​\n",
    "diagnosis = clinical.loc[:,\"primary_diagnosis\"]\n",
    "stage = clinical.loc[:,\"tumor_stage\"]\n",
    "grade = clinical.loc[:,\"tumor_grade\"]\n",
    "last_followup = clinical.loc[:,\"days_to_last_follow_up\"]\n",
    "clinical.loc[:,\"vital_status\"] = [1  if stat == \"dead\" else 0 for stat in clinical.loc[:,\"vital_status\"]]\n",
    "status = clinical.loc[:,\"vital_status\"]\n",
    "days_to_death = clinical.loc[:,\"days_to_death\"]\n",
    "age = clinical.loc[:,\"age_at_diagnosis\"] \n",
    "recurr = clinical.loc[:,\"days_to_recurrence\"]\n",
    "names_common = clinical.loc[:,\"primary_diagnosis\"]\n",
    "names_common.index = clinical.loc[:,\"submitter_id\"]\n",
    "submitters = clinical.loc[:,\"submitter_id\"]\n",
    "locations = clinical.loc[:,\"project_id\"]\n",
    "locations = [loc.split(\"-\")[-1] for loc in locations]\n",
    "loc_corresp = {case_id[i] : locations[i] for i in range(len(case_id))}\n",
    "submitters.index = clinical.loc[:,\"case_id\"]\n",
    "i = 0\n",
    "old_i=0\n",
    "folder_list =os.listdir(gdc_folder)\n",
    "for location in set(locations):\n",
    "    print(location)\n",
    "    folder_num =[]\n",
    "    folder_count = 0\n",
    "    data_list = []\n",
    "    filenames_ordered = []\n",
    "    for patient_folder in folder_list:\n",
    "        patient_folder_path = gdc_folder + os.fsdecode(patient_folder)\n",
    "        if(os.path.isdir(patient_folder_path)):\n",
    "            for patient in os.listdir(patient_folder_path):\n",
    "                filename = patient_folder_path + '/' +os.fsdecode(patient)\n",
    "                \n",
    "                \n",
    "                if((patient.split(\".\")[-1] == \"gz\" or patient.split(\".\")[-1] == \"counts\") and loc_corresp[names_corresp[patient]] == location):\n",
    "                    if(patient.split(\".\")[-1] == \"gz\"):\n",
    "                        f = gzip.open(filename)\n",
    "                        aux_data = pd.read_csv(f,sep='\\t',index_col=0,header=None, names=[patient])\n",
    "                    else:\n",
    "                        aux_data = pd.read_csv(filename,sep='\\t',index_col=0,header=None, names=[patient])\n",
    "                    filenames_ordered += [names_corresp[patient]]\n",
    "                    data_list.append(aux_data)\n",
    "                    i += 1\n",
    "                    folder_num += [patient_folder]\n",
    "    folder_list = [fold for fold in folder_list if fold not in folder_num]\n",
    "    print(location)\n",
    "    print(len(folder_num))\n",
    "    print(len(folder_list))\n",
    "    print(i-old_i)\n",
    "    if(i - old_i >1):\n",
    "        data_list = pd.concat(data_list,axis=1,copy=False)\n",
    "    elif i-old_i ==1:\n",
    "        data_list = data_list[0]\n",
    "    old_i = i\n",
    "    \"\"\"aux = [name for name in data_list.index if \"ENSG\" in name] \n",
    "    data_list = data_list.loc[aux,:]\n",
    "    aux = np.array([name for name in data_list.index if \"ENSG0\" in name])\n",
    "    aux2 = [\"ENSG0\" + name.split('R')[1] for name in data_list.index if \"ENSGR\" in name]\n",
    "    aux2 = np.array([name.split(\".\")[0] for name in aux2])\n",
    "    ro.globalenv[\"ind\"] =aux\n",
    "    ro.r(\"ind=as.vector(ind)\")\n",
    "    \n",
    "    ro.r('ann <- getBM(attributes=c(\"hgnc_symbol\",\"description\",\"chromosome_name\",\"band\",\"strand\",\"start_position\",\"end_position\",\"ensembl_gene_id_version\"),  filters=\"ensembl_gene_id_version\", values=ind, mart=human,uniqueRows=False)') \n",
    "    aux = [str(ind) for ind in ro.r('ann$hgnc_symbol')]\n",
    "    print(ro.r('ann'))\n",
    "    ro.globalenv[\"ind\"] =aux2\n",
    "    ro.r(\"ind=as.vector(ind)\")\n",
    "    ro.r('ann <- getBM(attributes=c(\"hgnc_symbol\",\"description\",\"chromosome_name\",\"band\",\"strand\",\"start_position\",\"end_position\",\"ensembl_gene_id_version\"),  filters=\"ensembl_gene_id_version\", values=ind, mart=human,uniqueRows=False)') \n",
    "    aux2 = [str(ind) +'_2' for ind in ro.r('ann$hgnc_symbol')]\n",
    "    print(ro.r('ann'))\n",
    "    print(len(aux+aux2))\n",
    "    data_list.index = aux + aux2\n",
    "    print(data_list.index)\n",
    "    data_list = data_list.T\"\"\"\n",
    "    survival = [] \n",
    "    for j in filenames_ordered:\n",
    "        if (status[j]==1 and days_to_death[j]!=\"--\" ):\n",
    "            survival += [int(days_to_death[j])]\n",
    "        elif last_followup[j]!=\"--\":\n",
    "            survival += [int(last_followup[j])]\n",
    "    binary = [status[k] for k in filenames_ordered]\n",
    "    not_found = [k  for k in range(len(filenames_ordered)) if days_to_death[filenames_ordered[k]]==\"--\" and last_followup[filenames_ordered[k]]==\"--\"]\n",
    "    data_list.to_pickle(folder + \"genes_counts_\" + location +\"_gdc\")\n",
    "    diagnosis[filenames_ordered].to_pickle(folder + \"diagnosis_\" + location +\"_gdc\")\n",
    "    stage[filenames_ordered].to_pickle(folder + \"stage_\" + location +\"_gdc\")\n",
    "    names_common[submitters[filenames_ordered]].to_pickle(folder + \"histo_\" + location +\"_gdc\")\n",
    "    \n",
    "    np.save(file_folder+\"/results/gdc_survival\",survival)\n",
    "    np.save(file_folder+\"/results/gdc_survival_binary\",binary)\n",
    "    \n",
    "    np.save(file_folder+\"/results/gdc_survival_not_found_samples\",not_found)    \n",
    "print(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
