---
title: "GEO"
author: "Joaquim Aguirre-Plans"
date: '2023-02-24'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



### Description

Analyze GEO dataset.
References:
http://genomicsclass.github.io/book/pages/GEOquery.html
https://www.bioconductor.org/packages/devel/bioc/vignettes/GEOquery/inst/doc/GEOquery.html
https://kasperdanielhansen.github.io/genbioconductor/html/GEOquery.html 
Information from the dataset:
https://pubmed.ncbi.nlm.nih.gov/36109152/

```{r, message=FALSE}
packrat::init("/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeR")
```

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(GEOquery)
library(ggplot2)
require(magrittr)
library(tidyr)
set.seed(1510)
```



### Access the GEO Series data

```{r}
geo_id = "GSE193677"
geo_dir = "/work/ccnr/j.aguirreplans/Databases/GEO"
geo_id_dir = paste(geo_dir, "/", geo_id, sep="")
raw_data_dir = paste(geo_id_dir, "/raw", sep="")
output_dir = paste(geo_id_dir, "/out", sep="")
dir.create(geo_id_dir, showWarnings = FALSE)
dir.create(raw_data_dir, showWarnings = FALSE)
dir.create(output_dir, showWarnings = FALSE)
```

`getGEO` function interprets its input to determine how to get the data from GEO and then parse the data into useful R data structures.
In the case of the dataset GSE193677, it downloads two files:
* `GPL16791.soft.gz`: Containing the phenotypic data in soft format.
* `GSE193677_series_matrix.txt.gz`: Containing the phenotypic format in series format.

```{r}
gse <- getGEO(geo_id, GSEMatrix = TRUE, destdir=raw_data_dir)
show(gse)
```

Now, `gse` contains the R data structure (of class GSE) that represents the `geo_id` entry from GEO.
If using `GSEMatrix = TRUE`, the data structure returned from this parsing is a list of ExpressionSets.

```{r}
show(pData(phenoData(gse[[1]]))[1:5,c(1,6,8)])
```

Store the metadata in a file:

```{r}
metadata_complete_df = pData(gse[[1]])
metadata_complete_file = paste(raw_data_dir, "/metadata_complete.txt", sep="")
metadata_complete_df %>% fwrite(metadata_complete_file)
metadata_complete_df = fread(metadata_complete_file)
metadata_complete_df[1:5, 1:5]
```



### Access raw files

The function `getGEOSuppFiles`, in the case of the dataset GSE193677, it downloads two files:
* `GSE193677_MSCCR_Biopsy_counts.txt.gz`: Containing the raw counts.
* `GSE193677_MSCCR_Biopsy_adjcounts.txt.gz`: Containing the adjusted counts.

```{r}
filePaths = getGEOSuppFiles(geo_id, makeDirectory = FALSE, baseDir = raw_data_dir)
filePaths
```



### Access GSE Data Tables from GEO

Sometimes GSEs are include separate data tables with the sample information. If these exist, you can use the getGSEDataTables() function. For example here is the phenoData object from a different GSE accession GSE3494 with a Data Table.

```{r}
df1 <- getGSEDataTables(geo_id)
lapply(df1, head)
```



### Analyze metadata

Number of samples by gender:

```{r}
table(metadata_complete_df$`demographics_gender:ch1`)
```

Number of samples by tissue:

```{r}
table(metadata_complete_df$`regionre:ch1`)
```

Number of samples by inflammation and disease:

```{r}
table(metadata_complete_df$`diseasetypere:ch1`)
```

Number of samples by disease:

```{r}
table(metadata_complete_df$`ibd_disease:ch1`)
```

Disease activity depending on the sample type and disease:

```{r}
table(metadata_complete_df$`diseasetypere:ch1`, metadata_complete_df$`ibd_clinicianmeasure_inactive_active:ch1`)
table(metadata_complete_df$`ibd_disease:ch1`, metadata_complete_df$`ibd_clinicianmeasure_inactive_active:ch1`)
```

Create a file with simplified metadata:

```{r}
metadata_df = metadata_complete_df %>%
  mutate(inflammation=if_else(`diseasetypere:ch1` %in% c("CD.I", "UC.I"), "inflamed", "noninflamed")) %>%
  dplyr::select(title, geo_accession, `diseasetypere:ch1`, inflammation, `ibd_disease:ch1`, `ibd_clinicianmeasure_inactive_active:ch1`, `demographics_gender:ch1`, `regionre:ch1`, `study_eligibility_age_at_endo:ch1`) %>%
  dplyr::rename("disease_inflammation"=`diseasetypere:ch1`, "disease"=`ibd_disease:ch1`, "disease_activity"=`ibd_clinicianmeasure_inactive_active:ch1`, "gender"=`demographics_gender:ch1`, "tissue"=`regionre:ch1`, "age"=`study_eligibility_age_at_endo:ch1`) %>%
  tidyr::separate(title, into=c("sample_id", NA, NA), sep = ",") %>%
  tidyr::separate(sample_id, into=c(NA, NA, "subject_id", NA, "biopsy_id"), sep = "_", remove=FALSE, convert=TRUE) %>%
  arrange(subject_id, sample_id, geo_accession) %>%
  as.data.frame()
metadata_simple_file = paste(output_dir, "/metadata_simple.txt", sep="")
metadata_df %>% fwrite(metadata_simple_file)
metadata_df = fread(metadata_simple_file)
head(metadata_df)
```

Number of samples by inflammation:

```{r}
table(metadata_df$inflammation)
```

We see that there are subjects with multiple biopsies, many of them extracted from different tissues.

```{r}
metadata_df %>% dplyr::select(subject_id, tissue) %>% group_by(tissue) %>% summarize(n_subjects=n())
```

What about the number of subjects per type of sample and tissue?

```{r}
metadata_df %>% dplyr::select(subject_id, tissue, disease_inflammation) %>% group_by(tissue, disease_inflammation) %>% summarize(n_subjects=n())
```
We see that the tissues that have a well-balanced number of inflamed, non-inflamed and control subjects are:
* **Rectum**
  * **CD**: 116 infl / 252 non-infl
  * **UC**: 114 infl / 167 non-infl
  * **Control**: 225 non-infl
* **Ileum**
  * **CD**: 162 infl / 201 non-infl
  * **UC**: 188 non-infl
  * **Control**: 121 non-infl

What about the number of subjects per disease and tissue?

```{r}
metadata_df %>% dplyr::select(subject_id, tissue, disease) %>% filter(!(disease %in% c("CD_Pouch", "UC_Pouch"))) %>% group_by(tissue, disease) %>% summarize(n_subjects=n())
```

What about the number of subjects per inflammation and tissue?

```{r}
metadata_df %>% dplyr::select(subject_id, tissue, inflammation) %>% group_by(tissue, inflammation) %>% summarize(n_subjects=n())
```

We see that the tissues that have a well-balanced number of inflamed, non-inflamed and control subjects are:
* **Rectum**: 260 infl / 644 non-infl
* **Ileum**: 162 infl / 510 non-infl

We observe that some of the biopsies from the same subject can be both inflamed or non-inflamed such as subjects 4 and 5, and can have multiple inflamed or non-inflamed samples:

```{r}
metadata_df %>% dplyr::select(subject_id, inflammation) %>% group_by(subject_id, inflammation) %>% summarize(n_inflammation=n()) %>% head(10)
```

We also observe that the inflamed biopsies that are from the same subject are from different tissues, and the same happens for non-inflamed biopses from the same subject. This means that there are no samples from the same subject, tissue and inflammation status:

```{r}
metadata_df %>% dplyr::select(subject_id, tissue, inflammation) %>% group_by(subject_id, tissue, inflammation) %>% summarize(n_inflammation=n()) %>% head(10)
table((metadata_df %>% dplyr::select(subject_id, tissue, inflammation) %>% group_by(subject_id, tissue, inflammation) %>% summarize(n_inflammation=n()))$n_inflammation)
```

Number of unique subjects from a specific disease group:

```{r}
metadata_df %>% dplyr::select(subject_id, disease) %>% group_by(disease) %>% summarize(n_subjects=n())
```

Are there subjects from different disease states? A subject can only have one disease state.

```{r}
table((metadata_df %>% dplyr::select(subject_id, disease) %>% unique() %>% group_by(subject_id) %>% summarize(n_diseases=n()))$n_diseases)
```



### Create groups of samples without repeated samples from the same subject

In gene co-expression networks, it is important not to have samples from the same subjects in the analysis because it affects a lot the correlations of the network. So, we will define some rules to create different datasets, containing different groups of samples.

#### Disease inflamed vs. control (Rectum / Ileum)

The first type of group will be inflamed samples from a specific disease vs. non-inflamed samples from control subjects:

```{r}
# Create first group, having only disease inflamed samples vs. control as non-inflamed samples
sample_group_df1 = metadata_df %>% 
  select(sample_id, subject_id, gender, tissue, disease, inflammation) %>%
  filter(tissue %in% c("Rectum", "Ileum")) %>%  # Keep only Rectum and Ileum tissue as they are the ones having more samples
  filter(!(disease %in% c("CD_Pouch", "UC_Pouch"))) %>% # Filter out disease Pouch
  filter(!((!(disease == "Control")) & (inflammation == "noninflamed"))) %>% # Filter out non-inflamed disease samples (we will only use controls as non-inflamed)
  unique() %>% 
  arrange(tissue, disease, inflammation, subject_id, sample_id) %>%
  unite("sample_group_id", c("tissue", "disease", "inflammation"), remove = FALSE) %>% # Create an ID for each group
  mutate(sample_group_type = "disease_inflamed_vs_control") # Define the type of group that we are creating
table(sample_group_df1$sample_group_id)
head(sample_group_df1)
```

#### Inflamed vs. control (Rectum / Ileum)

The second type of group will be inflamed samples from any disease vs. non-inflamed samples from control subjects:

```{r}
# Create second group, having any inflamed samples vs. control as non-inflamed samples
sample_group_df2 = metadata_df %>% 
  select(sample_id, subject_id, gender, tissue, disease, inflammation) %>%
  filter(tissue %in% c("Rectum", "Ileum")) %>%  # Keep only Rectum and Ileum tissue as they are the ones having more samples
  filter(!(disease %in% c("CD_Pouch", "UC_Pouch"))) %>% # Filter out disease Pouch
  filter(!((!(disease == "Control")) & (inflammation == "noninflamed"))) %>% # Filter out non-inflamed disease samples (we will only use controls as non-inflamed)
  unique() %>% 
  arrange(tissue, disease, inflammation, subject_id, sample_id) %>%
  unite("sample_group_id", c("tissue", "inflammation"), remove = FALSE) %>% # Create an ID for each group
  mutate(sample_group_type = "inflamed_vs_control") # Define the type of group that we are creating
table(sample_group_df2$sample_group_id)
head(sample_group_df2)
```

#### Inflamed vs. non-inflamed, removing samples from same subject (Rectum / Ileum)

The third type of group will be inflamed samples from any disease vs. non-inflamed samples from any subject.
We will make sure to remove samples from same subject in different groups, keeping only one sample for each subject.

First we select all inflamed and non-inflamed samples:

```{r}
sample_group_df3_raw = metadata_df %>% 
  select(sample_id, subject_id, gender, tissue, disease, inflammation) %>%
  filter(tissue %in% c("Rectum", "Ileum")) %>%  # Keep only Rectum and Ileum tissue as they are the ones having more samples
  filter(!(disease %in% c("CD_Pouch", "UC_Pouch"))) %>% # Filter out disease Pouch
  unique() %>% 
  arrange(tissue, disease, inflammation, subject_id, sample_id) %>%
  unite("sample_group_id", c("tissue", "inflammation"), remove = FALSE) %>% # Create an ID for each group
  mutate(sample_group_type = "inflamed_vs_noninflamed") # Define the type of group that we are creating
head(sample_group_df3_raw)
```

We count the subjects with more than one sample and separate them from the subjects with a single sample:

```{r}
sample_group_df3_uniquesamp = sample_group_df3_raw %>% group_by(tissue, subject_id) %>% mutate(n_samp = n()) %>% ungroup() %>% filter(n_samp == 1)
sample_group_df3_multiplesamp = sample_group_df3_raw %>% group_by(tissue, subject_id) %>% mutate(n_samp = n()) %>% ungroup() %>% filter(n_samp > 1)
table(sample_group_df3_uniquesamp$n_samp)
table(sample_group_df3_multiplesamp$n_samp)
```

From the subjects that have multiple samples, we select the sample from the smallest group and add it to the pool of samples that only have one subject:

```{r}
sample_group_df3 = data.frame(sample_group_df3_uniquesamp) %>% select(-n_samp)
for(subject_id_selected in unique(sample_group_df3_multiplesamp$subject_id)){
  groups_samples_selected = (sample_group_df3_multiplesamp %>% filter(subject_id == subject_id_selected))$sample_group_id # Get groups of the subject selected
  smallest_group = (sample_group_df3 %>% filter(sample_group_id %in% groups_samples_selected) %>% group_by(sample_group_id) %>% summarize(n_samp=n()) %>% ungroup() %>% filter(n_samp == min(n_samp)))$sample_group_id # Get the group with smallest number of samples among the groups of the subject selected
  sample_selected = (sample_group_df3_multiplesamp %>% filter((subject_id == subject_id_selected) & (sample_group_id == smallest_group)))$sample_id
  sample_group_df3 = rbind(sample_group_df3, (sample_group_df3_multiplesamp %>% filter(sample_id == sample_selected) %>% select(-n_samp)))
}
length(sample_group_df3$sample_id)
length(sample_group_df3$subject_id)
table(sample_group_df3$sample_group_id)
head(sample_group_df3)
```

Merge all groups and save the data frame as a file:

```{r}
sample_groups_df = rbind(sample_group_df1, sample_group_df2, sample_group_df3)
table(sample_groups_df$sample_group_id)
table(sample_groups_df$sample_group_type)
head(sample_groups_df)
sample_groups_file = paste(output_dir, "/sample_groups.txt", sep="")
fwrite(sample_groups_df, sample_groups_file, sep='\t')
```

### Access the raw counts

The raw counts are stored in the file `GSE193677_MSCCR_Biopsy_counts.txt.gz`.

```{r}
raw_counts_original_file = paste(raw_data_dir, "/GSE193677_MSCCR_Biopsy_counts.txt.gz", sep="")
```

Read the gz file:

```{r}
RNAseq_original = read.table(gzfile(raw_counts_original_file))
RNAseq_original = cbind(data.frame(Ensembl=row.names(RNAseq_original)), RNAseq_original)
```

```{r}
dim(RNAseq_original)
RNAseq_original[1:5, 1:5]
```

Filter the samples that are not in the metadata:

```{r}
RNAseq_original = RNAseq_original[, which((names(RNAseq_original) %in% c("Ensembl", metadata_df$sample_id)==TRUE))]
dim(RNAseq_original)
```

Are there repeated genes? No:

```{r}
nrow(RNAseq_original)
length(unique(RNAseq_original$Ensembl))
```



### Gene pre-processing

#### Map Ensembl to HGNC gene symbol

In the gene expression dataframe, there are `r nrow(RNAseq_original)` genes in Ensembl notation. We will transform these genes into HGNC gene symbol notation.

First we read the file `multiple_hgnc_dict_set_2022-01-01.txt`, containing gene symbols and their aliases in other notations:

```{r}
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_dict_file = paste(gene_annotations_dir, 'multiple_hgnc_dict_set_2022-01-01.txt', sep='/')
gene_dict_df = fread(gene_dict_file) %>% unique()
gene_dict_df = gene_dict_df[!((gene_dict_df$HGNC_Symbol=="")|(gene_dict_df$alias=="")),]
gene_dict_df$HGNC_Symbol = toupper(gene_dict_df$HGNC_Symbol)
gene_dict_df$alias = toupper(gene_dict_df$alias)
gene_dict_df %<>% unique()
head(gene_dict_df)
```

We use this dataframe to map Ensembl IDs to HGNC symbol IDs:

```{r}
RNAseq = inner_join(RNAseq_original, gene_dict_df, by=c("Ensembl"="alias"))
RNAseq = cbind((RNAseq %>% select("Ensembl", "HGNC_Symbol")), (RNAseq %>% select(-"Ensembl", -"HGNC_Symbol"))) %>% unique()
head(RNAseq[1:5,1:5])
```

```{r}
dim(RNAseq)
```

Now, we have `r length(unique(RNAseq$HGNC_Symbol))` unique HGNC symbols mapped to `r length(unique(RNAseq$Ensembl))` unique Ensembl IDs from a total of `r nrow(RNAseq)` rows. The problem is that there are Ensembl IDs mapped to the same HGNC code and viceversa. So, we have to do the following procedures:
- If an Ensembl ID has multiple HGNC symbols, keep only one symbol
- If a HGNC symbol has multiple Ensembl IDs, calculate the median

We get the HGNC symbols with multiple Ensembl IDs:

```{r}
repeated_ensembl = unique(RNAseq$Ensembl[duplicated(RNAseq$Ensembl)])
length(repeated_ensembl)
repeated_ensembl_df = RNAseq[RNAseq$Ensembl %in% repeated_ensembl,][,c("Ensembl", "HGNC_Symbol")]
repeated_ensembl_df$HGNC_length = nchar(repeated_ensembl_df$HGNC_Symbol)
repeated_ensembl_df = repeated_ensembl_df %>% arrange(Ensembl, HGNC_length, HGNC_Symbol) %>% unique()
head(repeated_ensembl_df)
```

We keep the smallest symbol, and if not, the first (in alphabetical order):

```{r}
repeated_ensembl_distinct_df = repeated_ensembl_df %>% distinct(Ensembl, .keep_all = TRUE) %>% select(Ensembl, HGNC_Symbol) # dplyr::distinct keeps only the first distinct row in terms of Ensembl
head(repeated_ensembl_distinct_df)
hgnc_to_remove = repeated_ensembl_df$HGNC_Symbol[!(repeated_ensembl_df$HGNC_Symbol %in% repeated_ensembl_distinct_df$HGNC_Symbol)] # we get the HGNC_Symbol to remove
RNAseq %<>% filter(!(HGNC_Symbol %in% hgnc_to_remove)) # remove HGNC symbols selected before
```

```{r}
dim(RNAseq)
```

We get the Ensembl IDs with multiple HGNC symbols:

```{r}
repeated_hgnc = unique(RNAseq$HGNC_Symbol[duplicated(RNAseq$HGNC_Symbol)])
length(repeated_hgnc)
repeated_hgnc_df = (RNAseq[RNAseq$HGNC_Symbol %in% repeated_hgnc,][,c("Ensembl", "HGNC_Symbol")]) %>% arrange(HGNC_Symbol, Ensembl)
head(repeated_hgnc_df)
```

We calculate the median for the HGNC genes with multiple Ensembl IDs:

```{r}
RNAseq = subset(RNAseq, select=-Ensembl)
repeated_hgnc_aggregated = aggregate(.~HGNC_Symbol, RNAseq[RNAseq$HGNC_Symbol %in% repeated_hgnc,], median)
RNAseq = rbind(RNAseq[!(RNAseq$HGNC_Symbol %in% repeated_hgnc),], repeated_hgnc_aggregated) %>% rename("Gene"="HGNC_Symbol")
dim(RNAseq)
RNAseq[1:5,1:5]
```



#### Filter genes without classification

We keep genes that have a classification in HGNC:

```{r}
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_category_file = paste(gene_annotations_dir, 'hgnc_category_set_2022-01-01.txt', sep='/')
gene_category_df = fread(gene_category_file)
gene_category_df$HGNC_Symbol = toupper(gene_category_df$HGNC_Symbol)
head(gene_category_df)
```

We use this dataframe to map gene category, and we only keep genes that are in this dataframe:

```{r}
dim(RNAseq)
genes_with_category = RNAseq$Gene[RNAseq$Gene %in% gene_category_df$HGNC_Symbol]
genes_dataset_df = gene_category_df %>% filter(HGNC_Symbol %in% genes_with_category)
RNAseq %<>% filter(Gene %in% genes_with_category)
dim(RNAseq)
```

The dataset contains `r as.integer(ncol(RNAseq)-1)` case IDs and `r nrow(RNAseq)` genes with category.

It contains the following number of genes for each category:

```{r}
table(genes_dataset_df$HGNC_category)
round(prop.table(table(genes_dataset_df$HGNC_category))*100,2)
```

Save the results: 

```{r}
raw_counts_processed_file = paste(output_dir, "/GSE193677_MSCCR_Biopsy_counts_processed.txt", sep="")
RNAseq %>% fwrite(raw_counts_processed_file, sep='\t')
RNAseq = fread(raw_counts_processed_file)
RNAseq[1:5,1:5]
```

