---
title: "Pre-process"
author: "Joaquim Aguirre-Plans"
date: '2023-02-24'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


### Description

Pre-process GSE193677 data before creating gene co-expression networks.

```{r, message=FALSE}
packrat::init("/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeR")
```

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
require(magrittr)
library(tidyr)
library(igraph)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```


### Define variables

```{r define_files}
# Define type of counts used
type_counts = "reads" # tpm, reads

# Define working directories
geo_id = "GSE193677"
geo_dir = "/work/ccnr/j.aguirreplans/Databases/GEO"
geo_id_dir = paste(geo_dir, "/", geo_id, sep="")
raw_data_dir = paste(geo_id_dir, "/raw", sep="")
output_dir = paste(geo_id_dir, "/out", sep="")

# Define the input files
raw_counts_processed_file = paste(output_dir, '/GSE193677_MSCCR_Biopsy_counts_processed.txt', sep='')
metadata_file = paste(output_dir, '/metadata_simple.txt', sep='')
```


### Read files

```{r}
RNAseq = fread(raw_counts_processed_file)
```

The GEO dataframe contains `r as.integer(ncol(RNAseq)-1)` columns (samples) and `r nrow(RNAseq)` rows (genes).

```{r}
head(RNAseq[1:10,1:5])
```

```{r}
metadata_df = fread(metadata_file)
head(metadata_df)
```


### Prepare groups of samples

We have the following groups of samples:

```{r}
# Include tumor and normal sample groups
sample_groups_df = metadata_df %>% select(participant_sample_type, project_id, sample_type_simplified, gender) %>% unique() %>% rename("sample_group_id" = "project_id", "sample_group_type" = "sample_type_simplified", "sex"="gender")

# Include normal samples matched to tissues
sample_groups_df = rbind(sample_groups_df, (metadata_df %>% select(participant_sample_type, tissue_simplified, sample_type_simplified, gender) %>% filter(sample_type_simplified == "normal") %>% select(-sample_type_simplified) %>% unique() %>% rename("sample_group_id" = "tissue_simplified", "sex"="gender") %>% mutate(sample_group_type = "tissue")))

# Include subtype samples
sample_groups_df = rbind(sample_groups_df, (metadata_df %>% select(participant_sample_type, cancer_subtype, sample_type_simplified, gender) %>% filter(sample_type_simplified == "tumor")  %>% filter(!(cancer_subtype == "")) %>% select(-sample_type_simplified) %>% unique() %>% rename("sample_group_id" = "cancer_subtype", "sex"="gender") %>% mutate(sample_group_type = "subtype"))) %>% arrange(sample_group_type, sample_group_id, participant_sample_type)

# Save file
sample_groups_file = paste(gdc_dir, paste('raw/metadata/sample_groups.txt', sep=""), sep='/')
fwrite(sample_groups_df, sample_groups_file, sep='\t')

head(sample_groups_df)
tail(sample_groups_df)
```