---
title: "GTEx subsampling"
author: "Joaquim Aguirre-Plans"
date: "29/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Create subsamples of gene expression data from GTEx, with the aim of testing the effect of sample size in these subsamples.

```{r, message=FALSE}
library(data.table)
library(dplyr)
set.seed(1510)
```

## Read files

```{r define_files}
# Define input files
samples_subjects_file = '/home/j.aguirreplans/Databases/GTEx/v8/GTEx_Annotations_SamplesSubjectsMerged.txt'
tpm_filt_file = '/home/j.aguirreplans/Databases/GTEx/v8/GTEx_RNASeq_gene_tpm_filtered.gct'

```

```{r read_files}
# Read input files
samples_df = fread(samples_subjects_file)
tpm_df = fread(tpm_filt_file)

```


## Subsampling

We define the list of tissues that we want to analyze:

```{r}
tissues <- c('Muscle - Skeletal', 'Whole Blood', 'Breast - Mammary Tissue', 'Testis', 'Ovary', 'Liver', 'Brain - Cortex', 'Brain - Cerebellum')

```

We define the sampling function:

```{r sampling_function}
## Random sampling with replacement
create_random_subsamples_with_replacement <- function(samples_list, size_list, rep, rnaseq, output_dir, output_file_name){
  subsamples_list = list()
  seed <- 1
  for(size in size_list){
    set.seed(seed)
    seed <- seed + 1
    for(r in 1:rep){
      subsample <- sample(samples_list, size=size, replace=TRUE)
      subsamples_list[[size]] = subsample
      subsample_df = rnaseq %>% filter(Subject %in% subsample)
      output_file = paste(output_dir, paste(paste(output_file_name, 'size', size, 'rep', r, sep='_'), '.txt', sep=''), sep='/')
      subsample_df %>% write.table(output_file, quote = F, sep='\t', row.names = F, col.names = T)
    }
  }
  return(subsamples_list)
}
```

For each tissue and sex, we select groups of samples from different sizes:

```{r}
sex_to_id = data.frame(row.names=c("male","female") , val=c(1,2))
for (tissue in tissues){
  for (sex in rownames(sex_to_id)){
    sex_id = sex_to_id[sex,]
    tissue_sex_samples = samples_df[(samples_df$SMTSD == tissue) & (samples_df$SEX == sex_id),]$SAMPID
    if (length(tissue_sex_samples) > 10){
      size_list <- seq(10, length(tissue_sex_samples), 10)
      print(tissue)
      print(sex)
      print(length(tissue_sex_samples))
      print(size_list)
      subsamples_list <- create_random_subsamples_with_replacement(samples_list=tissue_sex_samples, size_list=size_list, rep=10)
      print(subsamples_list)
    }
  }
}

```


### Randomly sampling *n* labels maintaining the original proportion (stratified bootstrapping)

We define the function to make the sampling

```{r bootstrap_function}
## Bootstrap by random sampling
create_random_subsamples <- function(rnaseq, size_list, rep, output_dir, output_file_name){
  subject.ids <- rnaseq$Subject
  seed <- 1
  for(size in size_list){
    set.seed(seed)
    seed <- seed + 1
    for(r in 1:rep){
      subsample <- sample(subject.ids, size=size, replace=FALSE)
      subsample_df = rnaseq %>% filter(Subject %in% subsample)
      subsample_df <- rnaseq %>% select(c("Name", "Description", subsample))
      output_file = paste(output_dir, paste(paste(output_file_name, 'sample', size, 'rep', r, sep='_'), '.txt', sep=''), sep='/')
      subsample_df %>% write.table(output_file, quote = F, sep='\t', row.names = F, col.names = T)
    }
  }
}
```

We randomly select labels from the non-responder dataset for each subsample. The process is repeated r times for each subsample size.

```{r subsampling_non_responders}
# Define sample size
rep = 10
step = 10
max_num_samples = length(nonresponder.ids) - 5
size_list <- seq(10, max_num_samples, 10)
size_list

# Create subsample files
output_dir = '/Users/quim/Documents/DATA/BioGroup/Scipher/data/bootstrap/nonresponders'
output_file_name = 'RNAseq_NonResponders'
create_random_subsamples(rnaseq=nonresponder.counts, size_list=size_list, rep=rep, output_dir=output_dir, output_file_name)

```

```{r subsampling_responders}
# Define sample size
rep = 10
step = 10
max_num_samples = length(responder.ids) - 5
size_list <- seq(10, max_num_samples, 10)
size_list

# Create subsample files
output_dir = '/Users/quim/Documents/DATA/BioGroup/Scipher/data/bootstrap/responders'
output_file_name = 'RNAseq_Responders'
#create_random_subsamples(rnaseq=responder.counts, size_list=size_list, rep=rep, output_dir=output_dir, output_file_name)

```

