---
title: "GTEx preprocessing"
author: "Joaquim Aguirre-Plans"
date: "26/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Pre-process GTEx data (version v8) before creating gene co-expression networks.

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
require(magrittr)
library(tidyr)
library(igraph)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```

## Define variables

```{r}
# Define working directories
gtex_dir = '/home/j.aguirreplans/Databases/GTEx/v8'
gene_annotations_dir = '/home/j.aguirreplans/data/gene_annotations/data/out'
gene_associations_dir = '/home/j.aguirreplans/data/gene_associations/data/out'

# Define input files
samples_subjects_file = paste(gtex_dir, 'GTEx_Annotations_SamplesSubjectsMerged.txt', sep='/')
tpm_file = paste(gtex_dir, 'GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct', sep='/')
```

## Read files

```{r}
tpm_df = fread(tpm_file)
```

The raw GTEx dataframe contains `r as.integer(ncol(tpm_df)-2)` sample IDs (as columns) and `r nrow(tpm_df)` genes (as rows) described in the first two columns as ***Name*** (Ensembl ID) and ***Description*** (HGNC symbol).

```{r}
head(tpm_df[1:10,1:5])
```

```{r}
samples_df = fread(samples_subjects_file)
```

The samples are described in the samples dataframe, which contains information about the subject and tissue of the sample:

```{r}
head(samples_df)
```

We filter the samples in the TPM file by keeping the ones that are in the samples information file:

```{r filter_tpm_file_samples}
dim(tpm_df[,-2])
tpm_df <- tpm_df %>% select(c("Name", "Description", samples_df$SAMPID))
dim(tpm_df[,-2])
```

We end up having a gene expression matrix of `r as.integer(ncol(tpm_df)-2)` sample IDs (as columns) and `r nrow(tpm_df)` genes (as rows).


## Gene pre-processing

### Merge repeated HGNC symbols

The genes in GTEx are annotated both as Ensembl IDs and HGNC symbols. We only want to keep HGNC symbol annotations. However, there are Ensembl IDs with the same HGNC symbols, probably isoforms:

```{r}
repeated_genes <- unique(tpm_df$Description[duplicated(tpm_df$Description)])
tpm_df[tpm_df$Description %in% repeated_genes[1:10],][1:10,1:5][order(Description),]
```

In total, there are `r length(repeated_genes)` repeated HGNC symbols. We merge these genes by calculating the median expression between them:

```{r}
tpm_no_duplicates_file = paste(gtex_dir, 'GTEx_RNASeq_gene_tpm_no_duplicates.gct', sep='/')
if (!(file.exists(tpm_no_duplicates_file))){
  tpm_df = subset(tpm_df, select=-Name)
  repeated_genes_aggregated = aggregate(.~Description, tpm_df[tpm_df$Description %in% repeated_genes,], median)
  tpm_df = rbind(tpm_df[!(tpm_df$Description %in% repeated_genes),], repeated_genes_aggregated) %>% rename("Gene"="Description")
  fwrite(tpm_df, tpm_no_duplicates_file)
} else {
  tpm_df = fread(tpm_no_duplicates_file)
}
```

After merging the repeated HGNC symbols, we end up having a gene expression matrix of `r as.integer(ncol(tpm_df)-1)` sample IDs (as columns) and `r nrow(tpm_df)` genes (as rows).

```{r}
tpm_df[tpm_df$Gene %in% repeated_genes[1:10],][1:10,1:5][order(Gene),]
```

### Keep genes properly annotated and classified by the HGNC database

[HGNC](https://www.genenames.org/) is a database of annotation, description and classification of genes. We use the information in this database to keep the genes that have been properly annotated using HGNC symbols and classified in different gene categories. We use the file `hgnc_category_set_2022-01-01.txt`, which is a dataframe that has been created after parsing and processing the HGNC database:

```{r}
gene_annotations_dir = '/home/j.aguirreplans/data/gene_annotations/data/out'
gene_category_file = paste(gene_annotations_dir, 'hgnc_category_set_2022-01-01.txt', sep='/')
gene_category_df = fread(gene_category_file)
gene_category_df$HGNC_Symbol = toupper(gene_category_df$HGNC_Symbol)
head(gene_category_df)
```

We use this dataframe to map gene category, and we only keep genes that are in this dataframe:

```{r}
dim(tpm_df)
genes_with_category = tpm_df$Gene[tpm_df$Gene %in% gene_category_df$HGNC_Symbol]
tpm_df %<>% filter(Gene %in% genes_with_category)
dim(tpm_df)
```

The filtered dataframe contains`r as.integer(ncol(tpm_df)-1)` sample IDs (as columns) and `r nrow(tpm_df)` genes with category (as rows).

It contains the following number of genes for each category:

```{r}
gtex_gene_category_df = gene_category_df %>% filter(HGNC_Symbol %in% genes_with_category) %>% unique
table(gtex_gene_category_df$HGNC_category)
round(prop.table(table(gtex_gene_category_df$HGNC_category))*100,2)
# Keep the information in a file
genes_dataset_file = paste(gtex_dir, "gtex_rnaseq_gene_info.csv", sep='/')
fwrite(gtex_gene_category_df, genes_dataset_file)
```

### Filter genes with low counts

We filter the genes that have low expression in most of the samples of each tissue/sex, keeping the ones that meet these rules:

* Sum of the expression of all samples above 0.
* Mean of the expression of all samples above 1.
* Median of the expression of all samples above 1.
* Standard deviation of all samples above 0.
* More than half of the genes with counts above 0.

```{r}
tpm_no_filt_dir = paste(gtex_dir, 'tpm_by_tissue', sep='/')
tpm_filt_dir = paste(gtex_dir, 'tpm_filtered_files_by_tissue', sep='/')
genes_filt_dir = paste(gtex_dir, 'genes_filtered_files_by_tissue', sep='/')
dir.create(tpm_no_filt_dir, showWarnings = FALSE)
dir.create(tpm_filt_dir, showWarnings = FALSE)
dir.create(genes_filt_dir, showWarnings = FALSE)
cols = c("tissue", "sex", "n_genes_raw", "n_genes_processed", "sum",  "mean", "median", "sd", "zero_sum_counts")
statistics_low_counts_df <- setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)
sex_to_id = data.frame(row.names=c("male","female", "") , val=c(1,2,""))
for (tissue in unique(samples_df$SMTSD.no.sp.char)){
  for (sex in rownames(sex_to_id)){
    
    # Get the samples of the selected tissue/sex
    if (sex == ""){
      samples_tissue_sex = samples_df[samples_df$SMTSD.no.sp.char == tissue,]$SAMPID
      tpm_tissue_sex_no_filt_file = paste(tpm_filt_dir, paste("gtex_rnaseq_", tissue, ".gct", sep=''), sep='/')
      tpm_tissue_sex_filt_file = paste(tpm_filt_dir, paste("gtex_rnaseq_", tissue, ".gct", sep=''), sep='/')
      genes_dataset_filt_file = paste(genes_filt_dir, paste("gtex_rnaseq_gene_info_", tissue, ".csv", sep=''), sep='/')
    } else {
      sex_id = sex_to_id[sex,]
      samples_tissue_sex = samples_df[(samples_df$SMTSD.no.sp.char == tissue) & (samples_df$SEX == sex_id),]$SAMPID
      tpm_tissue_sex_no_filt_file = paste(tpm_filt_dir, paste("gtex_rnaseq_", tissue, "_", sex, ".gct", sep=''), sep='/')
      tpm_tissue_sex_filt_file = paste(tpm_filt_dir, paste("gtex_rnaseq_", tissue, "_", sex, ".gct", sep=''), sep='/')
      genes_dataset_filt_file = paste(genes_filt_dir, paste("gtex_rnaseq_gene_info_", tissue, "_", sex, ".csv", sep=''), sep='/')
    }
    
    # Consider only tissue/sex combinations of more than 10 samples
    if (length(samples_tissue_sex) > 10){
      # Select expression from selected samples
      tpm_tissue_sex_df = tpm_df %>% select(c("Gene", all_of(samples_tissue_sex)))
      # Save file without filters
      fwrite(tpm_tissue_sex_df, tpm_tissue_sex_no_filt_file)
      
      # Calculate metrics of expression
      info_df = data.frame(Gene=tpm_tissue_sex_df$Gene, sum=rowSums(tpm_tissue_sex_df[,-1]))
      info_df$sd = apply(tpm_tissue_sex_df[,-1], 1, sd)
      info_df$mean = rowMeans(tpm_tissue_sex_df[,-1])
      info_df$median = apply(tpm_tissue_sex_df[,-1], 1, median)
      info_df$zero_sum_counts = as.numeric(apply(tpm_tissue_sex_df[,-1], 1, function(i) sum(i == 0) )) # Number of times that the counts are equal to 0
      # Filter low expressed genes
      info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 1) & (info_df$median > 1) & (info_df$sd > 0) & (info_df$zero_sum_counts < dim(tpm_tissue_sex_df[, -1])[2]/2) ,]
      tpm_tissue_sex_filt_df = tpm_tissue_sex_df[tpm_tissue_sex_df$Gene %in% info_filt_df$Gene,]
      fwrite(tpm_tissue_sex_filt_df, tpm_tissue_sex_filt_file)
      # Count statistics about the filter
      statistics_low_counts_df = rbind(statistics_low_counts_df, data.frame(tissue=tissue, sex=sex, n_genes_raw=length(tpm_tissue_sex_df$Gene), n_genes_processed=length(tpm_tissue_sex_filt_df$Gene), sum=length(info_df[(info_df$sum > 0),]$Gene), mean=length(info_df[(info_df$mean > 1),]$Gene), median=length(info_df[(info_df$median > 1),]$Gene), sd=length(info_df[(info_df$sd > 0),]$Gene), zero_sum_counts=length(info_df[(info_df$zero_sum_counts < dim(tpm_tissue_sex_df[, -1])[2]/2),]$Gene)))
      
      # Write gene info file
      genes_dataset_filt_df = data.frame(gtex_gene_category_df)
      genes_dataset_filt_df$enough_counts = ifelse(genes_dataset_filt_df$HGNC_Symbol %in% info_filt_df$Gene, TRUE, FALSE)
      fwrite(genes_dataset_filt_df, genes_dataset_filt_file)

    }
  }
}

statistics_low_counts_file = paste(gtex_dir, "gtex_rnaseq_low_counts_statistics.csv", sep='/')
fwrite(statistics_low_counts_df, statistics_low_counts_file)
```

These are the different number of genes after filtering by the criteria mentioned above:

```{r}
head(statistics_low_counts_df)
```

Here, we show an example (whole blood) of the number of genes for each gene category before/after filtering by gene counts.
Before filtering:

```{r}
genes_dataset_filt_file = paste(genes_filt_dir, "gtex_rnaseq_gene_info_Whole.Blood.csv", sep='/')
genes_dataset_filt_df = fread(genes_dataset_filt_file)
table(genes_dataset_filt_df$HGNC_category)
round(prop.table(table(genes_dataset_filt_df$HGNC_category))*100,2)
```

After filtering:

```{r}
table((genes_dataset_filt_df %>% filter(enough_counts == TRUE))$HGNC_category)
round(prop.table(table((genes_dataset_filt_df %>% filter(enough_counts == TRUE))$HGNC_category))*100,2)
```


## Prepare disease genes information dataset

Read the disease-gene associations dataset:

```{r}
gene_associations_dir = '/home/j.aguirreplans/data/gene_associations/data/out'
disease_genes_file = paste(gene_associations_dir, 'GDA_Filtered_04042022.csv', sep='/')
GDA = fread(disease_genes_file) %>% unique()
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$NewName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes)

Prepare the disease-gene associations for the GTEx dataset: 

```{r}
GDA[is.na(GDA)]<-0
GDA$NewName = tolower(GDA$NewName)
GDA$DescriptorName = tolower(GDA$DescriptorName)
GDA$MESHID = toupper(GDA$MESHID)
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
GDA %<>% 
  filter(Strong > 0 | 
           Weak > 0 |
           Incompatible > 0) %>%
  mutate(DiseaseName = NewName) %>%
  filter(DescriptorName %ni% c("behavioral disciplines and activities")) %>%
  mutate(MESHID_remove = stringr::str_detect(TreeNumber, pattern = "C22|C25|F04")) %>%
  mutate(MESHID_remove = ifelse(TreeNumber == "F01", TRUE,MESHID_remove)) %>%
  filter(!(MESHID_remove)) %>%
  select(DiseaseName, HGNC_Symbol, DescriptorName) %>% 
  unique() %>% 
  mutate(gene_in_dataset = ifelse(HGNC_Symbol %in% tpm_df$Gene, TRUE, FALSE))

# Make columns without special characters  
GDA$DiseaseName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DiseaseName))))
GDA$DescriptorName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DescriptorName))))

# Write output file
disease_genes_info_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/disease_genes/disease_genes_info_2022_gtex.csv"
GDA %>% fwrite(disease_genes_info_file)

head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$DiseaseName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes), and `r nrow((GDA %>% filter(gene_in_dataset == TRUE)))` disease-gene associations in the GTEx dataset (`r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$DiseaseName))` diseases, `r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$HGNC_Symbol))` genes).


## Prepare essential genes dataset

Prepare the essential genes of the TCGA dataset:

```{r}
# Read the essential genes file
essential_genes_dir = '/home/j.aguirreplans/Databases/OGEE'
essential_genes_original_file = paste(essential_genes_dir, 'OGEE_esential_genes_20190416.txt', sep='/')
# Filter by human genes
essential_genes_df = fread(essential_genes_original_file, header=TRUE) %>% rename("taxID"="#taxID")
head(essential_genes_df)
```

```{r}
# Read the gene annotations dictionary
gene_annotations_dir = '/home/j.aguirreplans/data/gene_annotations/data/out'
gene_dict_file = paste(gene_annotations_dir, 'multiple_hgnc_dict_set_2022-01-01.txt', sep='/')
gene_dict_df = fread(gene_dict_file) %>% unique()
gene_dict_df = gene_dict_df[!((gene_dict_df$HGNC_Symbol=="")|(gene_dict_df$alias=="")),]
gene_dict_df$HGNC_Symbol = toupper(gene_dict_df$HGNC_Symbol)
gene_dict_df$alias = toupper(gene_dict_df$alias)
gene_dict_df %<>% unique()
head(gene_dict_df)
```

```{r}
# Map ENSEMBL IDs to HGNC symbols
essential_genes_df = essential_genes_df %>% filter((taxID == 9606) & (essential=="E"))
essential_genes_df = essential_genes_df %>% inner_join(gene_dict_df, by=c("locus"="alias")) %>% select(taxID, sciName, HGNC_Symbol) %>% unique()
essential_genes_df$gene_in_dataset = ifelse(essential_genes_df$HGNC_Symbol %in% tpm_df$Gene, TRUE, FALSE)
head(essential_genes_df)
essential_genes_tcga_file = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/essential_genes/OGEE_essential_genes_gtex.csv'
essential_genes_df %>% fwrite(essential_genes_tcga_file)
```

There are `r nrow(essential_genes_df)` essential genes, `r nrow((essential_genes_df %>% filter(gene_in_dataset == TRUE)))` in the TCGA dataset.

