---
title: "GTEx preprocessing"
author: "Joaquim Aguirre-Plans"
date: "26/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Preprocess GTEx data (version v8).

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
require(magrittr)
library(tidyr)
library(igraph)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```

## Read files

```{r define_files}
# Define working directories
databases_dir = '/home/j.aguirreplans/Databases'
#databases_dir = '/Users/j.aguirreplans/Databases'
ppi_dir = '/home/j.aguirreplans/data/PPI'
deisy_dicts_dir = '/home/j.aguirreplans/data/DeisyDictionaries'

# Define input files
samples_subjects_file = paste(databases_dir, 'GTEx/v8/GTEx_Annotations_SamplesSubjectsMerged.txt', sep='/')
tpm_file = paste(databases_dir, 'GTEx/v8/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct', sep='/')
ppi_network_symbol_file = paste(ppi_dir, 'interactome_2019_merged_symbols.csv', sep='/')
gene_names_dict_file = paste(deisy_dicts_dir, 'Gene_Names_Dic.csv', sep='/')
gene_names_cat_file = paste(deisy_dicts_dir, 'Gene_Names_Cat.csv', sep='/')
gene_type_file = paste(deisy_dicts_dir, 'gene_type.csv', sep='/')
disease_genes_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/disease_genes/GDA_Data_Clean_Filtered_GeneNames.csv"
essential_genes_original_file = paste(databases_dir, 'OGEE/OGEE_esential_genes_20190416.txt', sep='/')
gene_info_file = paste(databases_dir, 'NCBIGene/Homo_sapiens.gene_info', sep='/')

# Define output files and dirs
tpm_no_duplicates_file = paste(databases_dir, 'GTEx/v8/GTEx_RNASeq_gene_tpm_no_duplicates.gct', sep='/')
tpm_only_ppi_file = paste(databases_dir, 'GTEx/v8/GTEx_RNASeq_gene_tpm_only_ppi.gct', sep='/')
tpm_ppi_filt_dir = paste(databases_dir, 'GTEx/v8/tpm_filtered_files_by_tissue_and_ppi_genes', sep='/')
tpm_filt_dir = paste(databases_dir, 'GTEx/v8/tpm_filtered_files_by_tissue', sep='/')
genes_filt_dir = paste(databases_dir, 'GTEx/v8/genes_filtered_files_by_tissue', sep='/')
ppi_filt_dir = paste(ppi_dir, 'interactome_tissue_specific', sep='/')
disease_genes_info_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/disease_genes/disease_genes_info_2022.csv"
essential_genes_file = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/essential_genes/OGEE_essential_genes.csv'

```

```{r read_files}
# Read input files
samples_df = fread(samples_subjects_file)
tpm_df = fread(tpm_file)

```


## Calculate the median expression of genes that have the same symbol

There are ENSEMBL genes that have the same symbol. They are probably isoforms:

```{r get_genes_with_same_symbol}
repeated_genes <- unique(tpm_df$Description[duplicated(tpm_df$Description)])
length(repeated_genes)
tpm_df[tpm_df$Description %in% repeated_genes[1:10],][,c("Name", "Description")][order(Description),]

```

We will merge these genes by calculating the median expression between them:

```{r merge_expression_of_genes_with_same_symbol}
dim(tpm_df)
if (!(file.exists(tpm_no_duplicates_file))){
  tpm_df = subset(tpm_df, select=-Name)
  repeated_genes_aggregated = aggregate(.~Description, tpm_df[tpm_df$Description %in% repeated_genes,], median)
  tpm_df = rbind(tpm_df[!(tpm_df$Description %in% repeated_genes),], repeated_genes_aggregated)
  fwrite(tpm_df, tpm_no_duplicates_file)
} else {
  tpm_df = fread(tpm_no_duplicates_file)
}
dim(tpm_df)

```


## Select genes with known annotations


```{r select_genes_known_annotations}
# Select genes from the initial dataset with known annotation
tpm_df$Description = toupper(tpm_df$Description)
genes_dataset_raw = unique(tpm_df$Description)
print(length(genes_dataset_raw))

# Read gene annotations and select the clean dataset of genes
gene_type_df = fread(gene_type_file) %>% unique()
gene_type_df$symbol = toupper(gene_type_df$symbol)
length(gene_type_df$symbol)
table(gene_type_df$gene_type)
round(prop.table(table(gene_type_df$gene_type))*100,2)
genes_dataset_clean = genes_dataset_raw[genes_dataset_raw %in% gene_type_df$symbol]
print(length(genes_dataset_clean))

# Merge both gene datasets with dataframe of gene categories
genes_dataset_clean_df = gene_type_df %>% filter(symbol %in% genes_dataset_clean) %>% unique()
table(genes_dataset_clean_df$gene_type)
round(prop.table(table(genes_dataset_clean_df$gene_type))*100,2)

# Find genes without symbol
genes_not_symbol = genes_dataset_raw[(!genes_dataset_raw %in% unique(genes_dataset_clean_df$symbol))]
print(genes_not_symbol[1:20])
print(length(genes_not_symbol))
#genes_not_symbol = genes_not_symbol[(!(toupper(gsub("_", "", genes_not_symbol))) %in% toupper(all_genes_and_synonyms))]
#gene_names_dict = fread(gene_names_dict_file)
#genes_not_symbol_synonyms = genes_not_symbol[genes_not_symbol %in% gene_names_dict$Alias]
#genes_not_symbol_dict = gene_names_dict %>% filter(Alias %in% genes_not_symbol_synonyms)

tpm_df = tpm_df %>% filter(Description %in% genes_dataset_clean_df$symbol)
dim(tpm_df)


```


## Remove genes with low TPM expression

We filter the samples in the TPM file by keeping the ones that are in the samples information file:

```{r filter_tpm_file_samples}
dim(tpm_df[,-1])
tpm_df <- tpm_df %>% select(c("Description", samples_df$SAMPID))
dim(tpm_df[,-1])
length(samples_df$SAMPID)

```

We filter the low expressed genes for each tissue and sex:

```{r remove_low_expressed_genes_for_all_tissues}
sex_to_id = data.frame(row.names=c("male","female", "") , val=c(1,2,""))
for (tissue in unique(samples_df$SMTSD.no.sp.char)){
  for (sex in rownames(sex_to_id)){
    print(tissue)
    print(sex)
    if (sex == ""){
      samples_tissue_sex = samples_df[samples_df$SMTSD.no.sp.char == tissue,]$SAMPID
      tpm_tissue_sex_filt_file = paste(tpm_filt_dir, paste("gtex_rnaseq_", tissue, ".gct", sep=''), sep='/')
      genes_dataset_filt_file = paste(genes_filt_dir, paste("gtex_rnaseq_gene_info_", tissue, ".csv", sep=''), sep='/')
    } else {
      sex_id = sex_to_id[sex,]
      samples_tissue_sex = samples_df[(samples_df$SMTSD.no.sp.char == tissue) & (samples_df$SEX == sex_id),]$SAMPID
      tpm_tissue_sex_filt_file = paste(tpm_filt_dir, paste("gtex_rnaseq_", tissue, "_", sex, ".gct", sep=''), sep='/')
      genes_dataset_filt_file = paste(genes_filt_dir, paste("gtex_rnaseq_gene_info_", tissue, "_", sex, ".csv", sep=''), sep='/')
    }
    if (length(samples_tissue_sex) > 10){
      tpm_tissue_sex_df = tpm_df %>% select(c("Description", all_of(samples_tissue_sex)))
      print(dim(tpm_tissue_sex_df))
      # Calculate metrics of expression
      info_df = data.frame(Description=tpm_tissue_sex_df$Description, sum=rowSums(tpm_tissue_sex_df[,-1]))
      info_df$sd = apply(tpm_tissue_sex_df[,-1], 1, sd)
      info_df$mean = rowMeans(tpm_tissue_sex_df[,-1])
      info_df$median = apply(tpm_tissue_sex_df[,-1], 1, median)
      info_df$zero_sum_counts = as.numeric(apply(tpm_tissue_sex_df[,-1], 1, function(i) sum(i == 0) )) # Number of times that the sum of the TPMs is equal to 0
      # Filter low expressed genes
      info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 0) & (info_df$median >= 1) & (info_df$sd > 0) & (info_df$zero_sum_counts < dim(tpm_tissue_sex_df[, -1])[2]/2) ,]
      tpm_tissue_sex_filt_df = tpm_tissue_sex_df[tpm_tissue_sex_df$Description %in% info_filt_df$Description,] %>% rename("Gene"="Description")
      print(dim(tpm_tissue_sex_filt_df))
      fwrite(tpm_tissue_sex_filt_df, tpm_tissue_sex_filt_file)
      
      # Get transposed matrix
      #sample.ids <- tpm_tissue_sex_filt_df[, 1][[1]]
      #tpm_tissue_sex_filt_df <- as.matrix(tpm_tissue_sex_filt_df[, -c(1)])
      #rownames(tpm_tissue_sex_filt_df) <- sample.ids
      #tpm_tissue_sex_filt_df.t <- t(as.matrix(tpm_tissue_sex_filt_df)) %>% as.data.frame() 
      #colnames(tpm_tissue_sex_filt_df.t) <- rownames(tpm_tissue_sex_filt_df)
      #tpm_tissue_sex_filt_df.t <- cbind(Sample = colnames(tpm_tissue_sex_filt_df), tpm_tissue_sex_filt_df.t)
      #tpm_tissue_sex_filt_file = paste(tpm_filt_dir, paste("GTEx_RNASeq_", tissue, "_", sex, ".gct", sep=''), sep='/')
      #fwrite(tpm_tissue_sex_filt_df.t, tpm_tissue_sex_filt_file)
      
      # Write gene info file
      genes_dataset_filt_df = genes_dataset_clean_df
      genes_dataset_filt_df$enough_counts = ifelse(genes_dataset_filt_df$symbol %in% info_filt_df$Description, TRUE, FALSE)
      fwrite(genes_dataset_filt_df, genes_dataset_filt_file)

    }
  }
}
```

```{r check_whole_blood_genes}
genes_dataset_filt_file = paste(genes_filt_dir, "gtex_rnaseq_gene_info_Whole.Blood.csv", sep='/')
genes_dataset_filt_df = fread(genes_dataset_filt_file)
print(length(unique(genes_dataset_filt_df$symbol)))
table(genes_dataset_filt_df$gene_type)
round(prop.table(table(genes_dataset_filt_df$gene_type))*100,2)
print(length(unique((genes_dataset_filt_df %>% filter(enough_counts == TRUE))$symbol)))
table((genes_dataset_filt_df %>% filter(enough_counts == TRUE))$gene_type)
round(prop.table(table((genes_dataset_filt_df %>% filter(enough_counts == TRUE))$gene_type))*100,2)

```


## Prepare disease genes information dataset

```{r}
# Create a disease gene file containing MESH information
if(!(file.exists(disease_genes_info_file))){
  
  GDA = fread(disease_genes_file)
  GDA[is.na(GDA)]<-0
  GDA$Symbol = toupper(GDA$hgnc_symbol)
  GDA$DiseaseName = tolower(GDA$DiseaseName)
  GDA$DescriptorName = tolower(GDA$DescriptorName)
  GDA %<>% 
    filter(Strong > 0 | 
             Weak > 0 |
             Incompatible > 0) %>%
    mutate(DiseaseName = NewName) %>%
    filter(DescriptorName %ni% c("Behavioral Disciplines and Activities")) %>%
    mutate(MESHID_remove = stringr::str_detect(TreeNumber, pattern = "C22|C25|F04")) %>%
    mutate(MESHID_remove = ifelse(TreeNumber == "F01", TRUE,MESHID_remove)) %>%
    filter(!(MESHID_remove)) %>%
    select(DiseaseName, hgnc_symbol, DescriptorName) %>% 
    #filter(hgnc_symbol %in% V(gPPInc)$name) %>% 
    unique()
    #mutate(gene_in_dataset = ifelse(hgnc_symbol %in% genes_keep_df$symbol, TRUE, FALSE))
    #group_by(DiseaseName) %>%
    #mutate(Total_Genes = n()) %>%
    #filter(Total_Genes > 10)

    # Make columns without special characters  
  GDA$DiseaseName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DiseaseName))))
  GDA$DescriptorName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DescriptorName))))
  
  GDA %>% fwrite(disease_genes_info_file)
  
} else {
  GDA = fread(disease_genes_info_file)
}

```


## Prepare essencial genes dataset

```{r}
essential_genes_df = fread(essential_genes_original_file, header=TRUE) %>% rename("taxID"="#taxID")
essential_genes_df = essential_genes_df %>% filter((taxID == 9606) & (essential=="E"))
gene_info_df = fread(gene_info_file)
gene_info_df = gene_info_df %>% separate_rows(dbXrefs, convert = TRUE) %>% select(Symbol, dbXrefs) %>% unique()
essential_genes_df = essential_genes_df %>% left_join(gene_info_df, by=c("locus"="dbXrefs")) %>% rename("gene"="Symbol")
#essential_genes_df$gene_in_dataset = ifelse(essential_genes_df$gene %in% genes_keep_df$symbol, TRUE, FALSE)
essential_genes_df %>% fwrite(essential_genes_file)
```







```{r}
rm(tpm_df)
rm(tpm_tissue_sex_df)
rm(tpm_tissue_sex_filt_df)
rm(info_df)
rm(info_filt_df)

```
























```{r remove_low_expressed_genes_for_all_tissues_ppi, eval=FALSE}
sex_to_id = data.frame(row.names=c("male","female", "") , val=c(1,2,""))
for (tissue in unique(samples_df$SMTSD.no.sp.char)){
  for (sex in rownames(sex_to_id)){
    print(tissue)
    print(sex)
    if (sex == ""){
      samples_tissue_sex = samples_df[samples_df$SMTSD.no.sp.char == tissue,]$SAMPID
      tpm_tissue_sex_filt_file = paste(tpm_ppi_filt_dir, paste("gtex_rnaseq_", tissue, ".gct", sep=''), sep='/')
      ppi_tissue_sex_filt_file = paste(ppi_filt_dir, paste("interactome_2019_", tissue, ".csv", sep=''), sep='/')
    } else {
      sex_id = sex_to_id[sex,]
      samples_tissue_sex = samples_df[(samples_df$SMTSD.no.sp.char == tissue) & (samples_df$SEX == sex_id),]$SAMPID
      tpm_tissue_sex_filt_file = paste(tpm_ppi_filt_dir, paste("gtex_rnaseq_", tissue, "_", sex, ".gct", sep=''), sep='/')
      ppi_tissue_sex_filt_file = paste(ppi_filt_dir, paste("interactome_2019_", tissue, "_", sex, ".csv", sep=''), sep='/')
    }
    if (length(samples_tissue_sex) > 10){
      tpm_tissue_sex_df = tpm_df %>% select(c("Description", all_of(samples_tissue_sex)))
      print(dim(tpm_tissue_sex_df))
      # Calculate metrics of expression
      info_df = data.frame(Description=tpm_tissue_sex_df$Description, sum=rowSums(tpm_tissue_sex_df[,-1]))
      info_df$sd = apply(tpm_tissue_sex_df[,-1], 1, sd)
      info_df$mean = rowMeans(tpm_tissue_sex_df[,-1])
      info_df$median = apply(tpm_tissue_sex_df[,-1], 1, median)
      info_df$zero_sum_counts = as.numeric(apply(tpm_tissue_sex_df[,-1], 1, function(i) sum(i == 0) )) # Number of times that the sum of the TPMs is equal to 0
      # Filter low expressed genes
      info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 0) & (info_df$median >= 1) & (info_df$sd > 0) & (info_df$zero_sum_counts < dim(tpm_tissue_sex_df[, -1])[2]/2) ,]
      tpm_tissue_sex_filt_df = tpm_tissue_sex_df[tpm_tissue_sex_df$Description %in% info_filt_df$Description,] %>% rename("Gene"="Description")
      print(dim(tpm_tissue_sex_filt_df))
      fwrite(tpm_tissue_sex_filt_df, tpm_tissue_sex_filt_file)
      
      # Get transposed matrix
      #sample.ids <- tpm_tissue_sex_filt_df[, 1][[1]]
      #tpm_tissue_sex_filt_df <- as.matrix(tpm_tissue_sex_filt_df[, -c(1)])
      #rownames(tpm_tissue_sex_filt_df) <- sample.ids
      #tpm_tissue_sex_filt_df.t <- t(as.matrix(tpm_tissue_sex_filt_df)) %>% as.data.frame() 
      #colnames(tpm_tissue_sex_filt_df.t) <- rownames(tpm_tissue_sex_filt_df)
      #tpm_tissue_sex_filt_df.t <- cbind(Sample = colnames(tpm_tissue_sex_filt_df), tpm_tissue_sex_filt_df.t)
      #tpm_tissue_sex_filt_file = paste(tpm_ppi_filt_dir, paste("GTEx_RNASeq_", tissue, "_", sex, ".gct", sep=''), sep='/')
      #fwrite(tpm_tissue_sex_filt_df.t, tpm_tissue_sex_filt_file)
      
      # Filter low expressed genes in PPI
      ppi_tissue_sex_filt_df = PPI_symbol[(PPI_symbol$proteinA_symbol %in% info_filt_df$Description) & (PPI_symbol$proteinB_symbol %in% info_filt_df$Description),]
      fwrite(ppi_tissue_sex_filt_df, ppi_tissue_sex_filt_file)
    }
  }
}
```
