---
title: "GTEx preprocessing"
author: "Joaquim Aguirre-Plans"
date: "26/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Preprocess GTEx data (version v8).

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
require(magrittr)
library(tidyr)
library(igraph)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```

## Read files

```{r define_files}
# Define working directories
databases_dir = '/home/j.aguirreplans/Databases'
#databases_dir = '/Users/j.aguirreplans/Databases'
ppi_dir = '/home/j.aguirreplans/data/PPI'
deisy_dicts_dir = '/home/j.aguirreplans/data/DeisyDictionaries'

# Define input files
subjects_file = paste(databases_dir, 'GTEx/v8/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt', sep='/')
samples_file = paste(databases_dir, 'GTEx/v8/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt', sep='/')
tpm_file = paste(databases_dir, 'GTEx/v8/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct', sep='/')
ppi_network_symbol_file = paste(ppi_dir, 'interactome_2019_merged_symbols.csv', sep='/')
gene_names_dict_file = paste(deisy_dicts_dir, 'Gene_Names_Dic.csv', sep='/')
gene_names_cat_file = paste(deisy_dicts_dir, 'Gene_Names_Cat.csv', sep='/')
gene_type_file = paste(deisy_dicts_dir, 'gene_type.csv', sep='/')
gene_info_file = paste(databases_dir, 'NCBIGene/Homo_sapiens.gene_info', sep='/')

# Define output files and dirs
samples_subjects_file = paste(databases_dir, 'GTEx/v8/GTEx_Annotations_SamplesSubjectsMerged.txt', sep='/')
tpm_no_duplicates_file = paste(databases_dir, 'GTEx/v8/GTEx_RNASeq_gene_tpm_no_duplicates.gct', sep='/')
tpm_filt_dir = paste(databases_dir, 'GTEx/v8/tpm_filtered_files_by_tissue', sep='/')
genes_filt_dir = paste(databases_dir, 'GTEx/v8/genes_filtered_files_by_tissue', sep='/')

```

```{r read_files}
# Read input file
tpm_df = fread(tpm_file)
subjects_df = fread(subjects_file)
samples_df = fread(samples_file)

```


## Merge sample with subject data

We filter the samples by just keeping the ones present in the TPM data:

```{r keep_only_samples_in_expression_data}
samples_df <- samples_df[samples_df$SAMPID %in% colnames(tpm_df),]
print(length(unique(samples_df$SAMPID)))
head(samples_df)

```

We add a column with subject ID by splitting the sample ID:

```{r add_subject_id_column_to_samples_df}
# Add column with subject ID 
# sample ID --> GTEX-14753-1626-SM-5NQ9L
# subject ID --> GTEX-14753
# info --> https://sites.google.com/broadinstitute.org/gtex-faqs/home
donor_ids <- list()
i = 1
for(sample_id in samples_df$SAMPID){
  split_list <- unlist(strsplit(sample_id, "-", fixed = TRUE))
  donor_id <- paste(split_list[1], split_list[2], sep='-')
  donor_ids[i] <- donor_id
  i<- i+1
}
samples_df$SUBJID <- unlist(donor_ids)

```

We merge samples and subjects dataframes:

```{r merge_samples_subjects_info}
samples_df <- merge(x = samples_df, y = subjects_df, by = "SUBJID", all = FALSE)

```

We keep only columns of interest:

```{r}
samples_df <- samples_df[,c("SAMPID", "SUBJID", "SEX", "AGE", "DTHHRDY", "SMTS", "SMTSD")]

```

We create tissue names without special characters:

```{r}
samples_df$SMTS.no.sp.char <- gsub(' ', '.', gsub(' - ', '.', gsub('[\\(\\)]', '', samples_df$SMTS)))
samples_df$SMTSD.no.sp.char <- gsub(' ', '.', gsub(' - ', '.', gsub('[\\(\\)]', '', samples_df$SMTSD)))

```

We remove samples without cause of death:

```{r}
samples_df <- samples_df[!(is.na(samples_df$DTHHRDY)),]
print(length(unique(samples_df$SAMPID)))
print(length(unique(samples_df$SUBJID)))
head(samples_df)
fwrite(samples_df, samples_subjects_file)

```


## Calculate the median expression of genes that have the same symbol

There are ENSEMBL genes that have the same symbol. They are probably isoforms:

```{r get_genes_with_same_symbol}
repeated_genes <- unique(tpm_df$Description[duplicated(tpm_df$Description)])
length(repeated_genes)
tpm_df[tpm_df$Description %in% repeated_genes[1:10],][,c("Name", "Description")][order(Description),]

```

We will merge these genes by calculating the median expression between them:

```{r merge_expression_of_genes_with_same_symbol}
dim(tpm_df)
if (!(file.exists(tpm_no_duplicates_file))){
  tpm_df = subset(tpm_df, select=-Name)
  repeated_genes_aggregated = aggregate(.~Description, tpm_df[tpm_df$Description %in% repeated_genes,], median)
  tpm_df = rbind(tpm_df[!(tpm_df$Description %in% repeated_genes),], repeated_genes_aggregated)
  fwrite(tpm_df, tpm_no_duplicates_file)
} else {
  tpm_df = fread(tpm_no_duplicates_file)
}
dim(tpm_df)

```


## Select genes with known annotations

```{r select_genes_known_annotations}
# Select genes from the initial dataset with known annotation
tpm_df$Description = toupper(tpm_df$Description)
genes_dataset_raw = unique(tpm_df$Description)
print(length(genes_dataset_raw))

# Read gene annotations and select the clean dataset of genes
gene_type_df = fread(gene_type_file) %>% unique()
gene_type_df$symbol = toupper(gene_type_df$symbol)
length(gene_type_df$symbol)
table(gene_type_df$gene_type)
round(prop.table(table(gene_type_df$gene_type))*100,2)
genes_dataset_clean = genes_dataset_raw[genes_dataset_raw %in% gene_type_df$symbol]
print(length(genes_dataset_clean))

# Merge both gene datasets with dataframe of gene categories
genes_dataset_clean_df = gene_type_df %>% filter(symbol %in% genes_dataset_clean) %>% unique()
table(genes_dataset_clean_df$gene_type)
round(prop.table(table(genes_dataset_clean_df$gene_type))*100,2)

# Find genes without symbol
genes_not_symbol = genes_dataset_raw[(!genes_dataset_raw %in% unique(genes_dataset_clean_df$symbol))]
print(genes_not_symbol[1:20])
print(length(genes_not_symbol))
#genes_not_symbol = genes_not_symbol[(!(toupper(gsub("_", "", genes_not_symbol))) %in% toupper(all_genes_and_synonyms))]
#gene_names_dict = fread(gene_names_dict_file)
#genes_not_symbol_synonyms = genes_not_symbol[genes_not_symbol %in% gene_names_dict$Alias]
#genes_not_symbol_dict = gene_names_dict %>% filter(Alias %in% genes_not_symbol_synonyms)

tpm_df = tpm_df %>% filter(Description %in% genes_dataset_clean_df$symbol)
dim(tpm_df)


```


