---
title: "Scipher Pre-processing (Unique Patients)"
author: "Joaquim Aguirre-Plans & Deisy Morselli Gysi"
date: "1/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Pre-process Scipher data (version v8) before creating gene co-expression networks.

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
library(igraph)
require(magrittr)
library(tidyr)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```

## Read files

```{r}
# Define working directories
scipher_data_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/Dec2021'

# Define input files
RNA_metadata_file = paste(scipher_data_dir, '00_data/RNA_metadata.csv', sep='/')
```

We read a file containing the gene expression and metadata together, which was created in a previous analysis:

```{r}
# Read RNA metadata processed and select the clean dataset of genes
RNA_metadata = fread(RNA_metadata_file)
genes_dataset = unique(toupper(names(RNA_metadata)[53:length(names(RNA_metadata))]))
print(length(genes_dataset))
```

```{r}
RNA_metadata[1:5,1:60]
```

The `RNA_metadata` dataframe contains `r ncol(RNA_metadata)` features (`r length(genes_dataset)` genes) and `r nrow(RNA_metadata)` samples.

> Important to note: There are many samples that have the same Scipher_id, meaning that they come from the same patient. They only differ by the visit (which can be either 0m, 3m or 6m) and the "batch" (which can be either "seq_fac1" or "seq_fac2").

## Combine samples from same patients and visit

We combine the Scipher_id and visit, so that we can have a unique ID for each sample:

```{r}
scipherid_visit = paste(RNA_metadata$Scipher_id, RNA_metadata$visit, sep=".")
RNA_metadata = cbind(data.frame(Scipher_id.visit=scipherid_visit), RNA_metadata) %>% relocate(Scipher_id.visit, Scipher_id, visit, batch, .before = age)
RNA_metadata[1:10,1:5]
```

We calculate the median expression of samples from same patients and visits:

```{r}
RNA_metadata = RNA_metadata %>% select(-batch)
RNA_metadata = RNA_metadata %>% group_by_at(colnames(RNA_metadata)[1:52]) %>% summarise_all(median) %>% as.data.frame()
RNA_metadata[1:10,1:5]
```

The new `RNA_metadata` dataframe contains `r ncol(RNA_metadata)` features (`r length(genes_dataset)` genes) and `r nrow(RNA_metadata)` samples.


## Gene pre-processing

### Filter genes without classification

We filter the genes in the matrix keeping only the ones that have been properly annotated and classified in the HGNC database:

```{r}
gene_annotations_dir = '/home/j.aguirreplans/data/gene_annotations/data/out'
gene_category_file = paste(gene_annotations_dir, 'hgnc_category_set_2022-01-01.txt', sep='/')
gene_category_df = fread(gene_category_file)
gene_category_df$HGNC_Symbol = toupper(gene_category_df$HGNC_Symbol)
head(gene_category_df)
```

```{r}
genes_with_category = genes_dataset[genes_dataset %in% gene_category_df$HGNC_Symbol]
genes_without_category = genes_dataset[!(genes_dataset %in% gene_category_df$HGNC_Symbol)]
genes_dataset_df = gene_category_df %>% filter(HGNC_Symbol %in% genes_with_category)
head(genes_dataset_df)
```

```{r}
table(genes_dataset_df$HGNC_category)
round(prop.table(table(genes_dataset_df$HGNC_category))*100,2)
```

From the initial `r length(genes_dataset)` genes, there are `r length(genes_with_category)` with category and `r length(genes_without_category)` without category.


### Filter genes with low counts

We filter the genes that have low expression in all samples, keeping the ones that meet these rules:

* Sum of the expression of all samples above 0.
* Mean of the expression of all samples above 5.
* Median of the expression of all samples above 5.
* Standard deviation of all samples above 1.
* More than half of the genes with counts above 0.

```{r}
# Get only the gene expression
RNA = RNA_metadata %>% select(genes_dataset_df$HGNC_Symbol)
genes = colnames(RNA)
sample_ids = RNA_metadata$Scipher_id.visit

# Transpose the matrix
RNA = t(as.matrix(RNA)) %>% as.data.frame()
colnames(RNA) = sample_ids
RNA = cbind(data.frame(Gene=genes), RNA)
RNA[1:5,1:5]
```

```{r}
# Calculate sum, mean, median and standard deviation
info_df = data.frame(Gene=RNA$Gene, sum=rowSums(RNA[,-1]))
info_df$sd = apply(RNA[,-1], 1, sd)
info_df$mean = rowMeans(RNA[,-1])
info_df$median = apply(RNA[,-1], 1, median)
info_df$zero_sum_counts = as.numeric(apply(RNA[,-1], 1, function(i) sum(i == 0) )) # Number of times that the counts are equal to 0

# We keep the genes following all criteria
info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 5) & (info_df$median > 5) & (info_df$sd > 1) & (info_df$zero_sum_counts < dim(RNA[, -1])[2]/2) ,]
```

There are `r length(info_filt_df$Gene)` genes accomplishing all the rules.

We count the following number of genes accomplishing each rule:

* Sum of the expression of all samples above 0: `r length(info_df[info_df$sum > 0,]$Gene)`.
* Mean of the expression of all samples above 5: `r length(info_df[info_df$mean > 5,]$Gene)`.
* Median of the expression of all samples above 5: `r length(info_df[info_df$median > 5,]$Gene)`.
* Standard deviation of all samples above 1: `r length(info_df[info_df$sd > 1,]$Gene)`.
* More than half of the genes with counts above 0: `r length(info_df[info_df$zero_sum_counts < dim(RNA[, -1])[2]/2,]$Gene)`.

We count the following number of filtered genes for each category:

```{r}
# Write the information about the selected genes of the dataset in a separate file
genes_dataset_df$enough_counts = ifelse(genes_dataset_df$HGNC_Symbol %in% info_filt_df$Gene, TRUE, FALSE)
genes_dataset_file = paste(scipher_data_dir, '00_data/scipher_unique_patients_rnaseq_gene_info.csv', sep='/')
fwrite(genes_dataset_df, genes_dataset_file)

# Check the percentage of genes in each category after filtering by counts
genes_dataset_filtered_df = genes_dataset_df %>% filter(HGNC_Symbol %in% info_filt_df$Gene) %>% unique
table(genes_dataset_filtered_df$HGNC_category)
round(prop.table(table(genes_dataset_filtered_df$HGNC_category))*100,2)
```

```{r}
RNAseq_processed = RNA_metadata %>% select(c("Scipher_id.visit", "Scipher_id", "visit", info_filt_df$Gene))
RNAseq_processed_t = t(as.matrix(RNAseq_processed %>% select(info_filt_df$Gene))) %>% as.data.frame()
colnames(RNAseq_processed_t) = RNA_metadata$Scipher_id.visit
RNAseq_processed_t = cbind(data.frame(Gene=rownames(RNAseq_processed_t)), RNAseq_processed_t)
RNAseq_processed_file = paste(scipher_data_dir, '00_data/scipher_unique_patients_rnaseq_counts_processed.csv', sep='/')
RNAseq_processed_t %>% fwrite(RNAseq_processed_file)
rm(RNA)
```

```{r}
RNA_metadata_processed = dplyr::inner_join((RNA_metadata %>% dplyr::select(names(RNA_metadata)[1:52])), RNAseq_processed)
RNA_metadata_processed_file = paste(scipher_data_dir, '00_data/scipher_unique_patients_metadata_rnaseq_counts.csv', sep='/')
RNA_metadata_processed %>% fwrite(RNA_metadata_processed_file)
```

We end up having `r length(RNAseq_processed_t$Gene)` genes and `r length(unique(RNAseq_processed$Scipher_id.visit))` samples from `r length(unique(RNAseq_processed$Scipher_id))` individuals.

## Analyze the metadata

We analyze the response of the patients at different time points:

```{r}
# Read file with metadata
metadata_raw_file = paste(scipher_data_dir, '00_data/RA-netk042021-10-25.xlsx', sep='/')
metadata = readxl::read_xlsx(metadata_raw_file)

metadata = metadata %>% 
  rename(CRP_metadata = CRP)

# Select type of response of patients
metadata_response = 
  metadata %>%
  select(Scipher_id, visit, 
         eular_binary, acr50) %>%
  pivot_wider(names_from = visit, 
              values_from = c(eular_binary, acr50))

head(metadata_response)
```

When combining the metadata with the RNAseq we have, in total, `r nrow(RNA_metadata_processed)` samples from `r length(unique(RNA_metadata_processed$Scipher_id))` individuals. Out of those, **`r nrow(RNA_metadata_processed[RNA_metadata_processed$visit == "0m", ])`** samples from `r length(unique(RNA_metadata_processed[RNA_metadata_processed$visit == "0m", ]$Scipher_id))` individuals are at baseline.

> Important to note: The key to both data sets is the combination of `Scipher_id` and `visit`. Do not join using `CRP`.

We focus here on the gene expression data that we have in the baseline (visit = 0m). 

```{r, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
# We filter gene expression by visit = 0m (at baseline)
RNAseq_processed_0m = RNAseq_processed %>% filter(visit == "0m")
RNA_metadata_processed_0m = dplyr::inner_join(metadata_response, RNAseq_processed_0m)
RNA_metadata_processed_0m_file = paste(scipher_data_dir, '00_data/scipher_unique_patients_metadata_rnaseq_counts_0m.csv', sep='/')
fwrite(RNA_metadata_processed_0m, RNA_metadata_processed_0m_file)
```

This is the response (in eular) of patients after 3 months:

```{r, echo=FALSE, results='asis'}
RNA_metadata_processed_0m %>% 
  group_by(eular_binary_3m) %>% 
  summarise(`#` = n()) %>%
  ungroup() %>%
  mutate(`%` = round((`#`/n()), 2))
```

This is the response (in eular) of patients after 6 months:

```{r, echo=FALSE, results='asis'}
RNA_metadata_processed_0m %>% 
  group_by(eular_binary_6m) %>% 
  summarise(`#` = n()) %>%
  ungroup() %>%
  mutate(`%` = round((`#`/n()), 2))
```

This is a summary of the response (in eular) after both 3 and 6 months:

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
RNA_metadata_processed_0m %>% 
  group_by(eular_binary_3m, eular_binary_6m) %>%
  summarise(`#` = n()) %>%
  ungroup() %>%
  mutate(`%` = round((`#`/n()), 2))
```


## Prepare disease genes information dataset

Read the disease-gene associations dataset:

```{r}
gene_associations_dir = '/home/j.aguirreplans/data/gene_associations/data/out'
disease_genes_file = paste(gene_associations_dir, 'GDA_Filtered_04042022.csv', sep='/')
GDA = fread(disease_genes_file) %>% unique()
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$NewName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes)

Prepare the disease-gene associations for the Scipher dataset: 

```{r}
disease_genes_info_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/disease_genes/disease_genes_info_2022_scipher_unique_patients.csv"

GDA[is.na(GDA)]<-0
GDA$NewName = tolower(GDA$NewName)
GDA$DescriptorName = tolower(GDA$DescriptorName)
GDA$MESHID = toupper(GDA$MESHID)
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
GDA %<>% 
  filter(Strong > 0 | 
           Weak > 0 |
           Incompatible > 0) %>%
  mutate(DiseaseName = NewName) %>%
  filter(DescriptorName %ni% c("behavioral disciplines and activities")) %>%
  mutate(MESHID_remove = stringr::str_detect(TreeNumber, pattern = "C22|C25|F04")) %>%
  mutate(MESHID_remove = ifelse(TreeNumber == "F01", TRUE,MESHID_remove)) %>%
  filter(!(MESHID_remove)) %>%
  select(DiseaseName, HGNC_Symbol, DescriptorName) %>% 
  unique() %>% 
  mutate(gene_in_dataset = ifelse(HGNC_Symbol %in% RNAseq_processed_t$Gene, TRUE, FALSE))

# Make columns without special characters  
GDA$DiseaseName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DiseaseName))))
GDA$DescriptorName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DescriptorName))))

# Write output file
GDA %>% fwrite(disease_genes_info_file)

head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$DiseaseName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes), and `r nrow((GDA %>% filter(gene_in_dataset == TRUE)))` disease-gene associations in the Scipher dataset (`r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$DiseaseName))` diseases, `r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$HGNC_Symbol))` genes).


## Prepare essential genes dataset

Prepare the essential genes of the Scipher dataset:

```{r}
# Read the essential genes file
essential_genes_dir = '/home/j.aguirreplans/Databases/OGEE'
essential_genes_original_file = paste(essential_genes_dir, 'OGEE_esential_genes_20190416.txt', sep='/')
# Filter by human genes
essential_genes_df = fread(essential_genes_original_file, header=TRUE) %>% rename("taxID"="#taxID")
head(essential_genes_df)
```

```{r}
# Read the gene annotations dictionary
gene_annotations_dir = '/home/j.aguirreplans/data/gene_annotations/data/out'
gene_dict_file = paste(gene_annotations_dir, 'multiple_hgnc_dict_set_2022-01-01.txt', sep='/')
gene_dict_df = fread(gene_dict_file) %>% unique()
gene_dict_df = gene_dict_df[!((gene_dict_df$HGNC_Symbol=="")|(gene_dict_df$alias=="")),]
gene_dict_df$HGNC_Symbol = toupper(gene_dict_df$HGNC_Symbol)
gene_dict_df$alias = toupper(gene_dict_df$alias)
gene_dict_df %<>% unique()
head(gene_dict_df)
```

```{r}
# Map ENSEMBL IDs to HGNC symbols
essential_genes_df = essential_genes_df %>% filter((taxID == 9606) & (essential=="E"))
essential_genes_df = essential_genes_df %>% inner_join(gene_dict_df, by=c("locus"="alias")) %>% select(taxID, sciName, HGNC_Symbol) %>% unique()
essential_genes_df$gene_in_dataset = ifelse(essential_genes_df$HGNC_Symbol %in% RNAseq_processed_t$Gene, TRUE, FALSE)
head(essential_genes_df)
essential_genes_scipher_file = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/essential_genes/OGEE_essential_genes_scipher_unique_patients.csv'
essential_genes_df %>% fwrite(essential_genes_scipher_file)
```

There are `r nrow(essential_genes_df)` essential genes, `r nrow((essential_genes_df %>% filter(gene_in_dataset == TRUE)))` in the Scipher dataset.

