---
title: "Create Subsamples of gene expression data from RA patients"
author: "Joaquim Aguirre Plans"
date: "8/24/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Create subsamples of gene expression data from RA patients, with the aim of testing the effect of sample size in these subsamples.

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(igraph)
library(NetSci)
```

## Read gene expression data

```{r read_gene_expression}
metadata_RNA = fread("/Users/quim/Documents/DATA/BioGroup/Scipher/data/out/Meta_RNA_Cleaned.csv")
patients <- metadata_RNA$Subject
```

The total number of patients is the dataset is `r length(patients)`.

We check the number of labels in these patients:

```{r define_response}
labels_R_3m = metadata_RNA$Subject[metadata_RNA$acr50_3m == "responder"]
labels_R_6m =metadata_RNA$Subject[metadata_RNA$acr50_3m == "nonresponder"]
labels_NR_3m = metadata_RNA$Subject[metadata_RNA$acr50_6m == "responder"]
labels_NR_6m = metadata_RNA$Subject[metadata_RNA$acr50_6m == "nonresponder"]
all_labels <- c(labels_R_3m, labels_R_6m, labels_NR_3m, labels_NR_6m)
labels_R = metadata_RNA$Subject[(metadata_RNA$acr50_3m == "responder")&(metadata_RNA$acr50_6m == "responder")]
labels_nR = metadata_RNA$Subject[(metadata_RNA$acr50_3m == "nonresponder")&(metadata_RNA$acr50_6m == "nonresponder")]
labels_R_nR <- c(labels_R, labels_nR)
```

* Responder (3M): `r length(labels_R_3m)`
* Responder (6M): `r length(labels_R_6m)`
* Non-responder (3M): `r length(labels_NR_3m)`
* Non-responder (6M): `r length(labels_NR_6m)`
* Total number of labels: `r length(all_labels)`
* Responder (3M & 6M): `r length(labels_R)`
* Non-responder (3M & 6M): `r length(labels_nR)`
* Total number of labels (3M & 6M): `r length(labels_R_nR)`

We check the balance of responder and non-responder patients in the dataset.

```{r table_response}
table(metadata_RNA %>% select(acr50_3m, acr50_6m))
```

## Create files of responders and non-responders data in a clean format

```{r responders_cleaning}
responders <- metadata_RNA[(metadata_RNA$acr50_3m=="responder") & (metadata_RNA$acr50_6m=="responder"),]
responder.ids <- responders$Subject
responder.counts <- responders[, -c(2:52)] # Get the counts but keep the subject ID
#responder.counts.t <- t(as.matrix(responder.counts))
#colnames(responder.counts.t) <- responder.ids
#rownames(responder.counts.t) <- colnames(responder.counts)
rnaseq_responders_data_file <- '/Users/quim/Documents/DATA/BioGroup/Scipher/data/out/Meta_RNA_Responders.txt'
responder.counts %>% write.table(rnaseq_responders_data_file, quote = F, sep='\t', row.names = F, col.names = T)

```

```{r nonresponders_cleaning}
nonresponders <- metadata_RNA[(metadata_RNA$acr50_3m=="nonresponder") & (metadata_RNA$acr50_6m=="nonresponder"),]
nonresponder.ids <- nonresponders$Subject
nonresponder.counts <- nonresponders[, -c(2:52)] # Get the counts but keep the subject ID
#nonresponder.counts.t <- t(as.matrix(nonresponder.counts))
#colnames(nonresponder.counts.t) <- nonresponder.ids
#rownames(nonresponder.counts.t) <- colnames(nonresponder.counts)
rnaseq_nonresponders_data_file <- '/Users/quim/Documents/DATA/BioGroup/Scipher/data/out/Meta_RNA_NonResponders.txt'
nonresponder.counts %>% write.table(rnaseq_nonresponders_data_file, quote = F, sep='\t', row.names = F, col.names = T)

```


## Subsampling

### Randomly sampling *n* labels maintaining the original proportion (stratified bootstrapping)

We define the function to make the sampling

```{r subsampling_random_labels_of_specific_proportion}
## Stratified bootstrapping
create_random_subsamples_of_specific_proportion <- function(patients_df, size_list, rep){
  patients <- patients_df$patient
  subsampling_cols <- c('size', 'rep', patients)
  subsampling_df <- data.frame(matrix(ncol=length(subsampling_cols), nrow=0))
  colnames(subsampling_df) <- subsampling_cols
  seed <- 1
  for(size in size_list){
    set.seed(seed)
    seed <- seed + 1
    for(r in 1:rep){
      samples_selection <- c()
      labels = unique(patients_df$response)
      for(label in labels){
        patients_label <- unique(patients_df$patient[patients_df$response == label])
        proportion <- unique(patients_df$proportion[patients_df$response == label])
        subsample <- sample(patients_label, size=size*proportion, replace=FALSE)
        samples_selection <- c(samples_selection, subsample)
        #print(c(size, r, label, length(subsample)))
      }
      subsampling_df <- rbind(subsampling_df, c(size, r, ifelse(patients %in% samples_selection, 1, 0)))
      colnames(subsampling_df) <- subsampling_cols
    }
  }
  return(subsampling_df)
}

select_size_list_by_label_proportion <- function(patients_df, max_num_samples, step){
  ### Checks if there is any case of a size (from one of the labels) which is larger than the length of patients from this label
  initial_size_list <- 1:max_num_samples
  initial_size_list <- initial_size_list[(initial_size_list %% step) == 0]
  size_list <- c()
  for(size in initial_size_list){
    size_ok = TRUE
    labels = unique(patients_df$response)
    for(label in labels){
      patients <- unique(patients_df$patient[patients_df$response == label])
      proportion <- unique(patients_df$proportion[patients_df$response == label])
      if(length(patients) < size*proportion){
        size_ok = FALSE
      }
    }
    if(!size_ok){
      break
    } else {
      size_list <- c(size_list, size)
    }
  }
  return(size_list)
}

```

We randomly select labels for each subsample. The process is repeated `r rep` times for each subsample size.

```{r size_list_selection_random_labels_of_original_proportion}
rep = 100
step = 10
proportion_R = 0.3
proportion_NR = 0.7
max_num_samples = length(labels_R_nR)
patients_df <- data.frame(patient=labels_R, response='R', proportion=proportion_R)
patients_df <- rbind(patients_df, data.frame(patient=labels_nR, response='NR', proportion=proportion_NR))
initial_size_list <- 1:(max_num_samples-1)
initial_size_list <- initial_size_list[(initial_size_list %% step) == 0]
initial_size_list
size_list <- select_size_list_by_label_proportion(patients_df, max_num_samples, step)
size_list
```

```{r subsampling_random_labels_of_original_proportion}
random_labels_subsampling_original_proportion_df <- create_random_subsamples_of_specific_proportion(patients_df, size_list, rep)
random_labels_subsampling_original_proportion_file <- '/Users/quim/Documents/DATA/BioGroup/Scipher/data/subsampling/subsampling_random_labels_original_proportion.txt'
random_labels_subsampling_original_proportion_df %>% write.table(random_labels_subsampling_original_proportion_file, quote = F, sep='\t', row.names = F, col.names = T)

```

