---
title: "TCGA subsampling"
author: "Joaquim Aguirre-Plans"
date: "4/26/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Create subsamples of gene expression data from GTEx, with the aim of testing the effect of sample size in these subsamples.

```{r, message=FALSE}
library(data.table)
library(dplyr)
set.seed(1510)
```

## Read files

```{r define_files}
# Define working directories
databases_dir = '/home/j.aguirreplans/Databases'
#databases_dir = '/Users/j.aguirreplans/Databases'
output_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data'
#output_dir = '/Users/quim/Documents/DATA/BioGroup/Scipher/data'

# Define input files and dirs
samples_subjects_file = '/home/j.aguirreplans/Databases/GTEx/v8/GTEx_Annotations_SamplesSubjectsMerged.txt'

```

```{r read_files}
# Read input files
samples_df = fread(samples_subjects_file)

```

## Subsampling

We define the list of tissues that we want to analyze:

```{r}
tissues <- unique(samples_df$SMTSD.no.sp.char)
#tissues <- c('Spleen', 'Whole.Blood')

```

We define the sampling function:

```{r sampling_function}
## Random sampling with replacement
create_random_subsample_datasets_with_replacement <- function(samples_list, size_list, rep, rnaseq, output_dir, output_file_name){
  subsamples_list = list()
  seed <- 1
  for(size in size_list){
    set.seed(seed)
    seed <- seed + 1
    for(r in 1:rep){
      subsample <- sample(samples_list, size=size, replace=TRUE)
      subsamples_list[[size]] = subsample
      subsample_df = rnaseq %>% filter(Sample %in% subsample)
      output_file = paste(output_dir, paste(paste(output_file_name, 'size', size, 'rep', r, sep='_'), '.txt', sep=''), sep='/')
      subsample_df %>% write.table(output_file, quote = F, sep='\t', row.names = F, col.names = T)
    }
  }
  return(subsamples_list)
}
# Same but without creating the gene expression files, just files with samples
create_random_subsamples_with_replacement <- function(samples_list, size_list, rep, output_dir, output_file_name){
  subsamples_df = data.frame(matrix(ncol=3,nrow=0, dimnames=list(NULL, c("size", "rep", "samples"))))
  seed <- 1
  for(size in size_list){
    set.seed(seed)
    seed <- seed + 1
    for(r in 1:rep){
      subsample <- sort(sample(samples_list, size=size, replace=TRUE))
      subsample_df <- data.frame(samples=subsample)
      output_file = paste(output_dir, paste(paste(output_file_name, 'size', size, 'rep', r, sep='_'), '.txt', sep=''), sep='/')
      subsample_df %>% write.table(output_file, quote = F, sep='\t', row.names = F, col.names = T)
      subsamples_df = rbind(subsamples_df, data.frame(size=size, rep=r, samples=paste(subsample, collapse = ';')))
    }
  }
  return(subsamples_df)
}

```

For each tissue and sex, we select groups of samples from different sizes:

```{r sampling_by_tissue_sex_with_repetition}
sampling_dir = paste(output_dir, 'sampling/GTEx/sampling_with_repetition', sep='/')
sex_to_id = data.frame(row.names=c("male","female","") , val=c(1,2,""))
for (tissue in tissues){
  for (sex in rownames(sex_to_id)){
    if (sex == ""){
      tissue_sex_samples = samples_df[samples_df$SMTSD.no.sp.char == tissue,]$SAMPID
      tissue_sampling_dir = paste(sampling_dir, tissue, sep='/')
      all_samples_output_file = paste(tissue_sampling_dir, paste('RNAseq_samples_', tissue, '_all_samples.txt', sep=''), sep='/')
      output_file_name = paste(paste(paste('RNAseq_samples', tissue, sep="_")))
      compilation_samples_output_file = paste(sampling_dir, paste(paste('RNAseq_samples', tissue, sep="_"), '.txt', sep=''), sep='/')
    } else {
      sex_id = sex_to_id[sex,]
      tissue_sex_samples = samples_df[(samples_df$SMTSD.no.sp.char == tissue) & (samples_df$SEX == sex_id),]$SAMPID
      tissue_sampling_dir = paste(sampling_dir, paste(tissue, sex, sep="_"), sep='/')
      all_samples_output_file = paste(tissue_sampling_dir, paste('RNAseq_samples_', tissue, '_', sex, '_all_samples.txt', sep=''), sep='/')
      output_file_name = paste(paste(paste('RNAseq_samples', tissue, sex, sep="_")))
      compilation_samples_output_file = paste(sampling_dir, paste(paste('RNAseq_samples', tissue, sex, sep="_"), '.txt', sep=''), sep='/')
    }
    if (length(tissue_sex_samples) > 10){
      size_list <- seq(10, length(tissue_sex_samples), 10)
      dir.create(tissue_sampling_dir, showWarnings = FALSE)
      # Create a file with a list of all samples
      all_samples_df <- data.frame(samples=tissue_sex_samples)
      all_samples_df %>% write.table(all_samples_output_file, quote = F, sep='\t', row.names = F, col.names = T)
      # Create files with lists of subsamples
      subsamples_df <- create_random_subsamples_with_replacement(samples_list=tissue_sex_samples, size_list=size_list, rep=10, output_dir=tissue_sampling_dir, output_file_name=output_file_name)
      # Create a final file with a compilation of all selected samples for the same tissue
      subsamples_df %>% write.table(compilation_samples_output_file, quote = F, sep='\t', row.names = F, col.names = T)
    }
  }
}

```

