---
title: "TCGA preparation"
author: "Joaquim Aguirre-Plans"
date: "3/9/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Description

Analyze TCGA dataset from GDC.

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(forcats)
library(ggplot2)
require(magrittr)
library(tidyr)
library(igraph)
library(rjson)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```


```{r}
# Define type of counts used
type_counts = "reads" # tpm, reads

# Define working directories
gdc_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-07-27-Dataset/TCGA'
input_dir = paste(gdc_dir, 'raw', sep='/')
data_dir = paste(input_dir, 'data', sep='/')
metadata_dir = paste(input_dir, 'metadata', sep='/')
additional_dir = paste(input_dir, 'additional', sep='/')
output_dir = paste(gdc_dir, 'out', type_counts, sep='/')
```

## Metadata

Let's analyze the `metadata.txt` file:

```{r}
metadata_file = paste(metadata_dir, 'metadata.txt', sep='/')
metadata_df = fread(metadata_file, header=TRUE)
glimpse(metadata_df)
```

There are `r length(unique(metadata_df$file_name))` unique file names, `r length(unique(metadata_df$file_id))` unique file IDs and `r length(unique(metadata_df$participant))` unique participant IDs, meaning that there are participants that are mapped to multiple files.

Here we can observe how most of the samples that belong to the same participant correspond to samples extracted from both tumor and normal tissue:

```{r}
metadata_with_distinct_counts_per_sample_type = metadata_df %>% select(file_id, participant, sample_type_simplified) %>% unique() %>% group_by(participant) %>% mutate(num_samples_tumor=sum(sample_type_simplified == "tumor")) %>% mutate(num_samples_normal=sum(sample_type_simplified == "normal")) %>% ungroup()
table(metadata_with_distinct_counts_per_sample_type$num_samples_tumor, metadata_with_distinct_counts_per_sample_type$num_samples_normal, dnn=c("num_samples_tumor", "num_samples_normal"))
```

```{r}
metadata_with_distinct_counts = metadata_df %>% select(file_id, participant, sample_type) %>% unique() %>% group_by(participant) %>% mutate(num_sample_types_per_participant=n_distinct(sample_type)) %>% mutate(num_samples_per_participant=n_distinct(file_id)) %>% ungroup()
table(metadata_with_distinct_counts$num_samples_per_participant, metadata_with_distinct_counts$num_sample_types_per_participant, dnn=c("num_samples_per_participant", "num_sample_types_per_participant"))
```

```{r}
head(metadata_df %>% group_by(participant) %>% filter(n()>1) %>% ungroup() %>% select(file_id, entity_submitter_id, participant, sample_code, portion_code, sample_type, project_id))
```

However, there is still a small number of samples that might belong to the same participant and also come from the same type of tissue. In these cases, the sample information might come from different samples, vials, portions or plates.
See this website for more information: https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/  

```{r}
head(metadata_df %>% select(file_id, entity_submitter_id, participant, sample, sample_code, portion_code, plate, center, sample_type, project_id) %>% unique() %>% group_by(participant) %>% filter(n()>1) %>% mutate(num_sample_types_per_participant=n_distinct(sample_type)) %>% mutate(num_samples_per_participant=n_distinct(file_id)) %>% filter((num_sample_types_per_participant == 1 & num_samples_per_participant > 1)) %>% ungroup(), 12)
```

This means that later, we will have to merge the expression of the samples that are from same participant and come from the same tissue type, but are from different vials, portions or plates. We will do it by using the participant_sample_type code:

```{r}
if(!("participant_sample_type" %in% colnames(metadata_df))){
  metadata_df$participant_sample_type = paste(metadata_df$participant, metadata_df$sample_type_simplified, sep="-")
  metadata_df %>% fwrite(metadata_file)
  metadata_df = fread(metadata_file, header=TRUE)
}
```

We check that we have the same number of samples and participants for each type of sample:

```{r}
metadata_df %>% select(participant, participant_sample_type, sample_type_simplified) %>% unique() %>% group_by(sample_type_simplified) %>% summarize(n_participant=n_distinct(participant), n_participant_sample_type=n_distinct(participant_sample_type))
```

TCGA contains `r length(unique(metadata_df$project_id))` project IDs. Each project is focused on a different type of cancer. Each project contains the following number of participants, depending on the sample type (tumor or normal):

```{r}
# Remove file IDs and keep only case IDs
metadata_cases_df = metadata_df %>% select(participant, project_id, project_disease_type, sample_type_simplified) %>% unique()

# Get table of project IDs
table(metadata_cases_df$sample_type_simplified, metadata_cases_df$project_id)
```

```{r}
# Plot project IDs
metadata_cases_count_tumor <- metadata_cases_df %>% filter(sample_type_simplified == "tumor") %>% count(project_id)
metadata_cases_count_normal <- metadata_cases_df %>% filter(sample_type_simplified == "normal") %>% count(project_id)
metadata_cases_count_by_sample_type = full_join(metadata_cases_count_tumor %>% rename(tumor=n), metadata_cases_count_normal %>% rename(normal=n), by="project_id") %>% replace(is.na(.), 0) %>% pivot_longer(-project_id, names_to="sample_type_simplified", values_to="n")
metadata_cases_plot_file = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/tcga_num_cases_per_project.png'
metadata_cases_plot <- metadata_cases_count_by_sample_type %>% 
  mutate(project_id = fct_reorder(project_id, n)) %>%
  ggplot( aes(x=project_id, y=n, fill=sample_type_simplified)) +
    geom_bar(position="dodge", stat="identity", alpha=.6, width=.4) +
    coord_flip() +
    xlab("") +
    guides(fill=guide_legend(title="Type of sample")) +
    #theme_bw()
    theme(plot.title =  element_text(size = 17, face="bold"), axis.title = element_text(size = 14, face="bold"), axis.text = element_text(size = 12), legend.text = element_text(size = 12), legend.title=element_text(size=14, face="bold"))
ggsave(metadata_cases_plot_file, plot=metadata_cases_plot, type='cairo', dpi = 300, height=8, width = 10)
#ggsave(metadata_cases_plot_file,plot = metadata_cases_plot,type='cairo',dpi = 1200,width = 9000,height = 6000,units = c("px"))

metadata_cases_plot
```

## Data

We read the processed file:

```{r}
# Define and read data file
if(!(type_counts %in% c("tpm", "reads"))){
  stop("Unknown type of data")
}
tcga_file = paste(output_dir, paste("TCGA_", type_counts, ".csv", sep=""), sep='/')
tcga_df = fread(tcga_file, header=TRUE)
```

The `tcga_df` dataframe contains `r ncol(tcga_df)` columns (samples) and `r nrow(tcga_df)` rows (genes).

```{r}
tcga_df %<>% rename("Ensembl"="gene_id")
head(tcga_df[1:10,1:5])
```

```{r}
dim(tcga_df)
```

The columns (samples) are actually file names. We select the ones present in the metadata:

```{r}
metadata_file_names = unique(metadata_df$file_name)
tcga_df = cbind(tcga_df$Ensembl, tcga_df %>% select(all_of(metadata_file_names)))
tcga_df %<>% rename("Ensembl"="V1")
dim(tcga_df)
```

Now it contains `r ncol(tcga_df)` columns (samples) and `r nrow(tcga_df)` rows (genes).


### Gene pre-processing

In the gene expression dataframe, there are `r nrow(tcga_df)` genes in Ensembl notation. We will transform these genes into HGNC gene symbol notation.

First we read the file `multiple_hgnc_dict_set_2022-01-01.txt`, containing gene symbols and their aliases in other notations:

```{r}
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_dict_file = paste(gene_annotations_dir, 'multiple_hgnc_dict_set_2022-01-01.txt', sep='/')
gene_dict_df = fread(gene_dict_file) %>% unique()
gene_dict_df = gene_dict_df[!((gene_dict_df$HGNC_Symbol=="")|(gene_dict_df$alias=="")),]
gene_dict_df$HGNC_Symbol = toupper(gene_dict_df$HGNC_Symbol)
gene_dict_df$alias = toupper(gene_dict_df$alias)
gene_dict_df %<>% unique()
head(gene_dict_df)
```

We remove the Ensembl ID part after the "." in the gene expression dataframe:

```{r}
tcga_df$Ensembl = sub("\\..*", "", tcga_df$Ensembl)
head(tcga_df[1:5,1:5])
```

We use this dataframe to map Ensembl IDs to HGNC symbol IDs:

```{r}
tcga_df = inner_join(tcga_df, gene_dict_df, by=c("Ensembl"="alias"))
tcga_df = cbind((tcga_df %>% select("Ensembl", "HGNC_Symbol")), (tcga_df %>% select(-"Ensembl", -"HGNC_Symbol"))) %>% unique()
head(tcga_df[1:5,1:5])
```

```{r}
dim(tcga_df)
```

Now, we have `r length(unique(tcga_df$HGNC_Symbol))` unique HGNC symbols mapped to `r length(unique(tcga_df$Ensembl))` unique Ensembl IDs from a total of `r nrow(tcga_df)` rows. The problem is that there are Ensembl IDs mapped to the same HGNC code and viceversa. So, we have to do the following procedures:
- If an Ensembl ID has multiple HGNC symbols, keep only one symbol
- If a HGNC symbol has multiple Ensembl IDs, calculate the median

We get the HGNC symbols with multiple Ensembl IDs:

```{r}
repeated_ensembl = unique(tcga_df$Ensembl[duplicated(tcga_df$Ensembl)])
length(repeated_ensembl)
repeated_ensembl_df = tcga_df[tcga_df$Ensembl %in% repeated_ensembl,][,c("Ensembl", "HGNC_Symbol")]
repeated_ensembl_df$HGNC_length = nchar(repeated_ensembl_df$HGNC_Symbol)
repeated_ensembl_df = repeated_ensembl_df[order(Ensembl, HGNC_length),]
head(repeated_ensembl_df)
```

We keep the smallest symbol, and if not, the first (in alphabetical order):

```{r}
repeated_ensembl_distinct_df = repeated_ensembl_df %>% distinct(Ensembl, .keep_all = TRUE) %>% select(Ensembl, HGNC_Symbol)
head(repeated_ensembl_distinct_df)
hgnc_to_remove = repeated_ensembl_df$HGNC_Symbol[!(repeated_ensembl_df$HGNC_Symbol %in% repeated_ensembl_distinct_df$HGNC_Symbol)]
tcga_df %<>% filter(!(HGNC_Symbol %in% hgnc_to_remove))
```

```{r}
dim(tcga_df)
```

We get the Ensembl IDs with multiple HGNC symbols:

```{r}
repeated_hgnc = unique(tcga_df$HGNC_Symbol[duplicated(tcga_df$HGNC_Symbol)])
length(repeated_hgnc)
repeated_hgnc_df = tcga_df[tcga_df$HGNC_Symbol %in% repeated_hgnc,][,c("Ensembl", "HGNC_Symbol")][order(HGNC_Symbol),]
head(repeated_hgnc_df)
```

We calculate the median for the HGNC genes with multiple Ensembl IDs:

```{r}
tcga_df = subset(tcga_df, select=-Ensembl)
repeated_hgnc_aggregated = aggregate(.~HGNC_Symbol, tcga_df[tcga_df$HGNC_Symbol %in% repeated_hgnc,], median)
tcga_df = rbind(tcga_df[!(tcga_df$HGNC_Symbol %in% repeated_hgnc),], repeated_hgnc_aggregated)
dim(tcga_df)
```

```{r}
tcga_processed_file = paste(output_dir, "TCGA_genes_processed.csv", sep='/')
fwrite(tcga_df, tcga_processed_file, sep='\t')
```


### Sample pre-processing

There are participant_sample_type IDs that are mapped to multiple files. We will merge the participant_sample_type IDs that have multiple file IDs, and substitute the file IDs with participant_sample_type IDs in the data file.

We identify the participant_sample_type IDs with multiple file IDs:

```{r}
mappings_tcga_df = metadata_df %>% filter(file_name %in% colnames(tcga_df)[-1]) %>% select(file_name, participant_sample_type) %>% unique()
mappings_tcga_duplicated_df = mappings_tcga_df %>% filter(participant_sample_type %in% subset(mappings_tcga_df,duplicated(participant_sample_type))$participant_sample_type)
head(mappings_tcga_duplicated_df)
```

There are `r length(unique(mappings_tcga_duplicated_df$participant_sample_type))` participant_sample_type IDs with multiple file IDs. We calculate the median between the expression of these file IDs.

```{r aggregate_participant_sample_type_ids}
genes_tcga_df = tcga_df$HGNC_Symbol
tcga_df %<>% select(-HGNC_Symbol) %>% t() %>% as.data.frame()
colnames(tcga_df) = genes_tcga_df
tcga_df = cbind(data.frame(file_name=row.names(tcga_df)), tcga_df)
tcga_df %<>% inner_join(mappings_tcga_df, by=c("file_name"="file_name"))
tcga_df %<>% relocate(file_name, participant_sample_type)
tcga_df = subset(tcga_df, select=-file_name)
tcga_participant_ids_duplicated = tcga_df[tcga_df$participant_sample_type %in% unique(mappings_tcga_duplicated_df$participant_sample_type),]
#participant_ids_aggregated = aggregate(.~participant_sample_type, tcga_participant_ids_duplicated, median)
participant_ids_aggregated = tcga_participant_ids_duplicated %>% group_by(participant_sample_type) %>% summarise(across(everything(), median, na.rm = TRUE)) %>% ungroup() %>% as.data.frame()
tcga_df = rbind(tcga_df[!(tcga_df$participant_sample_type %in% unique(mappings_tcga_duplicated_df$participant_sample_type)),], participant_ids_aggregated)
dim(tcga_df)
```

```{r transpose_processed_matrix}
# We transpose the matrix
genes = colnames((tcga_df %>% select(-participant_sample_type)))
participant_sample_type_ids = tcga_df$participant_sample_type
tcga_df = t(as.matrix((tcga_df %>% select(-participant_sample_type)))) %>% as.data.frame()
colnames(tcga_df) = participant_sample_type_ids
tcga_df = cbind(data.frame(Gene=genes), tcga_df)
```

```{r}
tcga_processed_file = paste(output_dir, "TCGA_genes_and_samples_processed.csv", sep='/')
fwrite(tcga_df, tcga_processed_file, sep='\t')
```

The processed `tcga_df` dataframe contains `r as.integer(ncol(tcga_df)-1)` columns (participant_sample_type IDs) and `r nrow(tcga_df)` rows (genes).

```{r}
head(tcga_df[1:10,1:5])
```


