---
title: "Compare Co-expression To PPI"
author: "Joaquim Aguirre-Plans"
date: '2022-11-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
packrat::init("/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeR")
```

```{r}
library(data.table)
library(dplyr)
library(FSA)
library(ggplot2)
library(igraph)
require(magrittr)
library(patchwork)
library(pROC)
library(ROCR)
library(tidyr)
library(wTO)
set.seed(1510)
options(bitmapType='cairo')
`%ni%` <- Negate(`%in%`)
```

Select parameters of the analysis:

```{r}
dataset = "gtex"
type_dataset = "Whole.Blood"
sample_group_type = "tumor" # if dataset = "tcga"
type_counts = "reads" # tpm, reads
sizes_selected = c(20, 100, 200, 300, 400, 500)
sizes_selected_comparison_plot = c(20, 100, 200, 300)
threshold_selected = 0.05
```

### Read PPI-related files and calculate distance matrices between protein pairs

Define a function to calculate the distance matrix in a network:

```{r}
#'  calculate_distance_matrix_ppi
#'  Function to calculate distances between pairs of nodes of the PPI network.
#'  @param ppi_net PPI network in igraph format.
#'  
calculate_distance_matrix_ppi = function(ppi_net){
  # Calculate distances
  ppi_distances = distances(ppi_net)
  # Order the nodes
  ppi_distances = ppi_distances[order(rownames(ppi_distances)), order(colnames(ppi_distances))]
  # Convert the matrix to an in-line format
  ppi_distances = ppi_distances %>% wTO.in.line() %>% rename(distances=wTO)
  # Remove infinite distances (from different components)
  ppi_distances = ppi_distances %>% filter(!(is.infinite(distances)))
  return(ppi_distances)
}
```

Calculate the distance matrices of the PPI and NCI networks (if necessary):

```{r}
# Calculate distances between pairs of nodes of the PPI and NCI networks
ppi_nci_distances_file = "/work/ccnr/j.aguirreplans/data/PPI/PPI_NCI_merged_distances.txt"
if (!(file.exists(ppi_nci_distances_file))){
  # Calculate PPI distances
  ppi_distances_file = "/work/ccnr/j.aguirreplans/data/PPI/PPI_2022_04042022_distances.txt"
  if (!(file.exists(ppi_distances_file))){
    # Read the PPI network
    ppi_file = "/work/ccnr/j.aguirreplans/data/PPI/PPI_2022_04042022.csv"
    ppi_df = fread(ppi_file) %>% dplyr::select(HGNC_Symbol.1, HGNC_Symbol.2) %>% filter(!(HGNC_Symbol.1 == HGNC_Symbol.2)) %>% unique()
    ppi_net = graph_from_data_frame(ppi_df, directed=F) %>% simplify()
    gsize(ppi_net)
    gorder(ppi_net)
    head(ppi_df)
    # Calculate distances of PPI
    ppi_distances = calculate_distance_matrix_ppi(ppi_net)
    # Save file
    ppi_distances %>% fwrite(ppi_distances_file)
  } else {
    # Read PPI distances
    ppi_distances = fread(ppi_distances_file)
  }
  # Calculate NCI distances
  nci_distances_file = "/work/ccnr/j.aguirreplans/data/PPI/ncPPI_PPI_2022_04042022_distances.txt"
  if (!(file.exists(nci_distances_file))){
    # Read NCI network
    nci_file = "/work/ccnr/j.aguirreplans/data/PPI/ncPPI_PPI_2022_04042022.csv"
    nci_df = fread(nci_file) %>% dplyr::select(HGNC_Symbol.1, HGNC_Symbol.2) %>% filter(!(HGNC_Symbol.1 == HGNC_Symbol.2)) %>% unique()
    nci_net = graph_from_data_frame(nci_df, directed=F) %>% simplify()
    gsize(nci_net)
    gorder(nci_net)
    head(nci_df)
    # Calculate distances of NCI
    nci_distances = calculate_distance_matrix_ppi(nci_net)
    # Save file
    nci_distances %>% fwrite(nci_distances_file)
  } else {
    # Read NCI distances
    nci_distances = fread(nci_distances_file)
  }
  # Merge PPI and NCI distances
  ppi_distances = ppi_distances %>% rename("distances_ppi"="distances")
  nci_distances = nci_distances %>% rename("distances_nci"="distances")
  ppi_nci_distances = ppi_distances %>% full_join(nci_distances, by=c("Node.1", "Node.2"))
  rm(ppi_distances)
  rm(nci_distances)
  ppi_nci_distances %>% fwrite(ppi_nci_distances_file)
} else {
  ppi_nci_distances = fread(ppi_nci_distances_file)
}
head(ppi_nci_distances)
```

The PPI has the following number of direct interactions:

```{r}
nrow(ppi_nci_distances %>% filter(distances_ppi == 1) %>% select(Node.1, Node.2))
```

The NCI has the following number of direct interactions:

```{r}
nrow(ppi_nci_distances %>% filter(distances_nci == 1) %>% select(Node.1, Node.2))
```

Save the sorted PPI and NCI networks if necessary:

```{r}
ppi_sorted_file = "/work/ccnr/j.aguirreplans/data/PPI/PPI_2022_04042022_sorted.csv"
if (!(file.exists(ppi_sorted_file))){
  ppi_nci_distances %>% filter(distances_ppi == 1) %>% select(Node.1, Node.2) %>% fwrite(ppi_sorted_file)
}
nci_sorted_file = "/work/ccnr/j.aguirreplans/data/PPI/ncPPI_PPI_2022_04042022_sorted.csv"
if (!(file.exists(nci_sorted_file))){
  ppi_nci_distances %>% filter(distances_nci == 1) %>% select(Node.1, Node.2) %>% fwrite(nci_sorted_file)
}
```

The maximum distance in the PPI network is:

```{r}
max((ppi_nci_distances %>% filter(!(is.na(distances_ppi))))$distances_ppi)
```

The maximum distance in the NCI network is:

```{r}
max(ppi_nci_distances$distances_nci)
```

Plot the distribution of distances between proteins in the PPI network:

```{r}
ppi_distances_plot = ppi_nci_distances %>% 
  filter(!(is.na(distances_ppi))) %>%
  ggplot(aes(x=distances_ppi)) + 
  geom_histogram( binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  scale_y_continuous(trans = scales::log10_trans()) +
  theme_linedraw() +
  xlab("Distance in the PPI network") +
  ylab("Log(Count)") +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica"))
ppi_distances_plot_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_distances.png"
ggsave(
  ppi_distances_plot_file,
  plot=ppi_distances_plot,
  dpi = 1200,
  units = c("px")
)
print(ppi_distances_plot)
```

Plot the distribution of distances between genes/proteins in the NCI network:

```{r}
nci_distances_plot = ppi_nci_distances %>% 
  ggplot(aes(x=distances_nci)) + 
  geom_histogram( binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  scale_y_continuous(trans = scales::log10_trans()) +
  theme_linedraw() +
  xlab("Distance in the NCI network") +
  ylab("Log(Count)") +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica"))
nci_distances_plot_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/nci_distances.png"
ggsave(
  nci_distances_plot_file,
  plot=nci_distances_plot,
  dpi = 1200,
  units = c("px")
)
print(nci_distances_plot)
```

### Define the protein pair sets to study

Read the list of genes in the co-expression networks that we are studying:

```{r}
if(dataset == "gtex"){
  coexpression_gene_info_file = sprintf("/work/ccnr/j.aguirreplans/Databases/GTEx/v8/%s/genes_from_rnaseq_filtered_files_by_tissue/gtex_rnaseq_gene_info_%s.csv", type_counts, type_dataset)
} else if(dataset == "tcga"){
  coexpression_gene_info_file = sprintf("/work/ccnr/j.aguirreplans/Databases/TCGA/2022-11-18-Dataset/TCGA/out/%s/%s/filter_genes_low_counts/genes_filtered_files_by_sample_group/rnaseq_gene_info_%s.csv", type_counts, sample_group_type, type_dataset)
} else if(dataset == "scipher"){
  coexpression_gene_info_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/Dec2021/00_data/genes_from_rnaseq_filtered_files_by_group/scipher_rnaseq_gene_info_%s.csv", type_dataset)
}

coexpression_gene_info_df = fread(coexpression_gene_info_file) %>% filter(enough_counts == TRUE)
length(unique(coexpression_gene_info_df$HGNC_Symbol))
```

Filter the matrix of distances between protein/gene pairs keeping the genes that are in the gene expression dataset selected.

```{r}
ppi_nci_distances_filt = ppi_nci_distances %>% filter((Node.1 %in% unique(coexpression_gene_info_df$HGNC_Symbol)) & (Node.2 %in% unique(coexpression_gene_info_df$HGNC_Symbol)))
rm(ppi_nci_distances)
```

The maximum distance in the PPI network after filtering by genes that are in the gene expression dataset is:

```{r}
max((ppi_nci_distances_filt %>% filter(!(is.na(distances_ppi))))$distances_ppi)
```

The maximum distance in the NCI network after filtering by genes that are in the gene expression dataset is:

```{r}
max(ppi_nci_distances_filt$distances_nci)
```

Select the PPIs in the gene expression dataset:

```{r}
ppi_interactions = ppi_nci_distances_filt %>% filter(distances_ppi == 1) %>% select(Node.1, Node.2)
nrow(ppi_interactions)
```

Select the NCI in the gene expression dataset:

```{r}
nci_interactions = ppi_nci_distances_filt %>% filter(distances_nci == 1) %>% select(Node.1, Node.2)
nrow(nci_interactions)
```

Select random control interactions (non-direct interactions) of the same size as the PPI:

```{r}
random_pp_pairs = ppi_nci_distances_filt %>% 
  filter(!(is.na(distances_ppi))) %>% filter((!(distances_nci == 1)) & (!(distances_ppi == 1))) %>% 
  #select(Node.1, Node.2) %>% 
  slice_sample(n=dim(ppi_interactions)[1], replace = FALSE)
nrow(random_pp_pairs)
```

Select random control interactions of the same size as the PPI that are far in the PPI.
How? From the maximum distance in the PPI (e.g. 6) we select all protein pairs. Then for the next distance (e.g. 5), and so on, until the number of selected pairs surpasses the number of direct interactions. In the case of GTEx whole blood, this would be at distance 4.
Then, we select randomly from this pool of protein pairs (from distances 4 to 6) the same number of pairs as the number of direct PPIs.

```{r}
max_ppi_filt_distance = max((ppi_nci_distances_filt %>% filter(!(is.na(distances_ppi))))$distances_ppi)
num_selected_interactions = dim(ppi_interactions)[1]
random_far_pp_pairs = ppi_nci_distances_filt[0,]
for (distance in (seq(max_ppi_filt_distance, 2))){
  random_far_pp_pairs = rbind(random_far_pp_pairs, ppi_nci_distances_filt %>% 
    filter(!(is.na(distances_ppi))) %>% 
    filter(distances_ppi == distance))
  if (dim(random_far_pp_pairs)[1] >= num_selected_interactions){
    break
  }
}
random_far_pp_pairs = random_far_pp_pairs %>% 
  slice_sample(n=num_selected_interactions, replace = FALSE)
nrow(random_far_pp_pairs)
```

Select random control interactions of the same size as the NCI that are far in the NCI following the same strategy as in PPIs described above.
In the case of GTEx whole blood, we select NCI pairs from distances 4 to 7.

```{r}
max_nci_filt_distance = max((ppi_nci_distances_filt %>% filter(!(is.na(distances_nci))))$distances_nci)
num_selected_interactions = dim(nci_interactions)[1]
random_far_nc_pairs = ppi_nci_distances_filt[0,]
for (distance in (seq(max_nci_filt_distance, 2))){
  random_far_nc_pairs = rbind(random_far_nc_pairs, ppi_nci_distances_filt %>% 
    filter(!(is.na(distances_nci))) %>% 
    filter(distances_nci == distance))
  if (dim(random_far_nc_pairs)[1] >= num_selected_interactions){
    break
  }
}
random_far_nc_pairs = random_far_nc_pairs %>% 
  slice_sample(n=num_selected_interactions, replace = FALSE)
nrow(random_far_nc_pairs)
```

Merge all the selected pairs in the same dataframe:

```{r}
pairs_file <- "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_analysis_pairs.txt"
if (file.exists(pairs_file)) {
  pairs <- data.table::fread(pairs_file)
} else {
  pairs <- (ppi_interactions %>% mutate(ppi=TRUE)) %>% 
    full_join((nci_interactions %>% mutate(nci=TRUE)), by=c("Node.1", "Node.2")) %>% 
    left_join(ppi_nci_distances_filt, by=c("Node.1", "Node.2")) %>% 
    full_join((ppi_nci_distances_filt %>% filter((distances_ppi >= 5) | (distances_nci >= 5))), by=c("Node.1", "Node.2", "distances_ppi", "distances_nci")) %>% 
    full_join((random_pp_pairs %>% mutate(random_not_ppi=TRUE)), by=c("Node.1", "Node.2", "distances_ppi", "distances_nci")) %>%
    full_join((random_far_pp_pairs %>% mutate(random_far_not_ppi=TRUE)), by=c("Node.1", "Node.2", "distances_ppi", "distances_nci")) %>%
    full_join((random_far_nc_pairs %>% mutate(random_far_not_nci=TRUE)), by=c("Node.1", "Node.2", "distances_ppi", "distances_nci")) %>%
    replace_na(list(ppi=FALSE, nci=FALSE, random_not_ppi=FALSE, random_far_not_ppi=FALSE, random_far_not_nci=FALSE))
  head(pairs)
  pairs %>% data.table::fwrite(pairs_file)
}
```

### Read the co-expression networks across different sample sizes and extract the co-expression of the protein pair sets

Fetch the files containing the gene co-expression networks of different sample sizes:

```{r}
coexpression_dir <- sprintf("/scratch/j.aguirreplans/Scipher/SampleSizeProject/networks_%s/reads/%s/consensus", dataset, type_dataset)
all_coexpression_files <- list.files(coexpression_dir)
coexpression_files_selected <- data.frame(matrix(ncol=length(c("size", "file")),nrow=0, dimnames=list(NULL, c("size", "file"))))

for (coexpression_file in all_coexpression_files){
  file_split <- strsplit(
    gsub(".net", "", coexpression_file),
    split = "_"
  )[[1]]
  size <- as.numeric(file_split[(length(file_split)-1)])
  if (size %in% sizes_selected){
    coexpression_files_selected <- rbind(
      coexpression_files_selected,
      data.frame(
        size = size,
        file = paste(coexpression_dir, coexpression_file, sep = "/")
      )
    )
  }
}
```

Read the co-expression networks and extract the co-expression of the protein pair sets to analyze.
Keep only co-expression values with significant p-value (above 0.05). If not, the value is set to 0.

```{r}
coexpression_pairs_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_analysis_pairs_pval_%s_%s_%s.txt", as.character(threshold_selected), dataset, type_dataset)
if (!(file.exists(coexpression_pairs_file))){
  coexpression_pairs_filt = data.frame(pairs)
  for (size_selected in sort(coexpression_files_selected$size)){
    coexpression_file = (coexpression_files_selected %>% filter(size == size_selected))$file
    coexpression_df = fread(coexpression_file)
    # Turn unsignificant co-expression values to 0
    coexpression_df$CN = ifelse(coexpression_df$pval.fisher >= threshold_selected, 0, coexpression_df$CN)
    # Rename CN column into size selected
    coexpression_df = coexpression_df %>% 
      select(-pval.fisher) %>% 
      rename(!!as.character(size_selected) := "CN")
    # Merge current co-expression with other co-expression sizes
    coexpression_pairs_filt = coexpression_pairs_filt %>% left_join(coexpression_df, by=c("Node.1", "Node.2"))
    rm(coexpression_df)
  }
  coexpression_pairs_filt %>% fwrite(coexpression_pairs_file)
} else {
  coexpression_pairs_filt = fread(coexpression_pairs_file)
}
```

Do the same, but this time without setting to 0 the statistically unsignificant co-expression values:

```{r}
coexpression_pairs_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_analysis_pairs_%s_%s.txt", dataset, type_dataset)
if (!(file.exists(coexpression_pairs_file))){
  coexpression_pairs = data.frame(pairs)
  for (size_selected in sort(coexpression_files_selected$size)){
    coexpression_file = (coexpression_files_selected %>% filter(size == size_selected))$file
    coexpression_df = fread(coexpression_file) %>%
      select(-pval.fisher) %>% 
      rename(!!as.character(size_selected) := "CN")
    # Merge current co-expression with other co-expression sizes
    coexpression_pairs = coexpression_pairs %>% left_join(coexpression_df, by=c("Node.1", "Node.2"))
    rm(coexpression_df)
  }
  coexpression_pairs %>% fwrite(coexpression_pairs_file)
} else {
  coexpression_pairs = fread(coexpression_pairs_file)
}
```


### Plot gene co-expression vs. sample size

```{r}
colors_defined = c("PPI" = "#DBA9CA",
                   "PPI & iNCI" = "#FAF69E",
                   "Random" = "#B8BECC",
                   "Random pairs in PPI" = "#B8BECC",
                   "Random (distance > 3)" = "#543F4D",
                   "Distance > 3" = "#543F4D",
                   "Random pairs in PPI (dist. > 3)" = "#543F4D",
                   "Random pairs in PPI & iNCI (dist. > 3)" = "#545334",
                   "1" = "#DBA9CA",
                   "2" = "#a8829b",
                   "3" = "#87677c",
                   "4" = "#5e4857",
                   "5" = "#40303b",
                   "6" = "#1c151a")
```


### Plot the co-expression of pairs of proteins grouped by their distance in the PPI network

Boxplot plotting the zeroes (corresponding to the unsignificant correlations):

```{r}
boxplot_ppi_coexpression_by_distances_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_boxplot_by_ppi_distance_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("1", "2", "3", "4", "5", "6")]
boxplot_ppi_coexpression_by_distances = coexpression_pairs_filt %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  mutate(across(c(distances_ppi), as.character)) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      y = abs(correlation),
      fill = distances_ppi,
      colour = distances_ppi) +
  geom_boxplot(alpha = 0.8,
               outlier.size = 0,
               outlier.alpha = 0) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        y = "Absolute co-expression weight",
        color = "PPI distance",
        fill = "PPI distance")
ggsave(
  boxplot_ppi_coexpression_by_distances_file,
  plot = boxplot_ppi_coexpression_by_distances,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
boxplot_ppi_coexpression_by_distances
```

Violin plot plotting the zeroes (corresponding to the unsignificant correlations):

```{r}
violinplot_ppi_coexpression_by_distances_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_ppi_distance_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("1", "2", "3", "4", "5", "6")]
violinplot_ppi_coexpression_by_distances = coexpression_pairs_filt %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  #filter(abs(correlation) > 0 ) %>% 
  mutate(across(c(distances_ppi), as.character)) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      y = abs(correlation),
      fill = distances_ppi,
      colour = distances_ppi) +
  geom_violin(alpha = 0.8, 
              scale = "width", 
              adjust = 0.5) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        y = "Absolute co-expression weight",
        color = "PPI distance",
        fill = "PPI distance")
ggsave(
  violinplot_ppi_coexpression_by_distances_file,
  plot = violinplot_ppi_coexpression_by_distances,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
violinplot_ppi_coexpression_by_distances
```

Violin plot showing only the weights of significant correlations (removing the unsignificant ones):

```{r}
violinplot_ppi_coexpression_by_distances_nozero_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_ppi_distance_nozero_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("1", "2", "3", "4", "5", "6")]
violinplot_ppi_coexpression_by_distances_nozero = coexpression_pairs_filt %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  filter(abs(correlation) > 0 ) %>% 
  mutate(across(c(distances_ppi), as.character)) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      #y = abs(correlation),
      y = correlation,
      fill = distances_ppi,
      colour = distances_ppi) +
  geom_violin(
    alpha = 0.8
    #scale = "width",
    #adjust = 0.5
  ) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        #y = "Absolute co-expression weight",
        y = "Co-expression weight",
        color = "PPI distance",
        fill = "PPI distance")
ggsave(
  violinplot_ppi_coexpression_by_distances_nozero_file,
  plot = violinplot_ppi_coexpression_by_distances_nozero,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
violinplot_ppi_coexpression_by_distances_nozero
```

Violin plot showing all co-expression weights (without setting unsignificant co-expression links to 0):

```{r}
violinplot_ppi_coexpression_by_distances_nofilt_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_ppi_distance_%s_%s.png", dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("1", "2", "3", "4", "5", "6")]
violinplot_ppi_coexpression_by_distances_nofilt = coexpression_pairs %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  mutate(across(c(distances_ppi), as.character)) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      #y = abs(correlation),
      y = correlation,
      fill = distances_ppi,
      colour = distances_ppi) +
  geom_violin(
    alpha = 0.8
    #scale = "width",
    #adjust = 0.5
  ) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        #y = "Absolute co-expression weight",
        y = "Co-expression weight",
        color = "PPI distance",
        fill = "PPI distance")
ggsave(
  violinplot_ppi_coexpression_by_distances_nofilt_file,
  plot = violinplot_ppi_coexpression_by_distances_nofilt,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
violinplot_ppi_coexpression_by_distances_nofilt
```


### Plot the co-expression of pairs of proteins/genes grouped by different categories (PPI vs. Random vs. Random (distance > 3))

Boxplot plotting the zeroes (corresponding to the unsignificant correlations):

```{r}
boxplot_ppi_coexpression_by_types_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_boxplot_by_pair_type_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
#colors_defined_filt = colors_defined[names(colors_defined) %in% c("PPI", "PPI & iNCI", "Random pairs in PPI", "Random pairs in PPI (dist. > 3)", "Random pairs in PPI & iNCI (dist. > 3)")]
colors_defined_filt = colors_defined[names(colors_defined) %in% c("PPI", "Random", "Random (distance > 3)")]
boxplot_ppi_coexpression_by_types = coexpression_pairs_filt %>%
  #select(-distances_ppi, -distances_nci) %>%
  select(-distances_ppi, -distances_nci, -nci, -random_far_not_nci) %>%
  #pivot_longer(c("ppi", "nci", "random_not_ppi", "random_far_not_ppi", "random_far_not_nci"), names_to = "category", values_to = "category_bool") %>%
  pivot_longer(c("ppi", "random_not_ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  #mutate(category = replace(category, category == "nci", "PPI+iNCI")) %>%
  mutate(category = replace(category, category == "ppi", "PPI")) %>%
  mutate(category = replace(category, category == "random_not_ppi", "Random")) %>%
  mutate(category = replace(category, category == "random_far_not_ppi", "Random (distance > 3)")) %>%
  #mutate(category = replace(category, category == "random_far_not_nci", "Random pairs in PPI+iNCI (d > 3)")) %>%
  #mutate(category = factor(category, levels = c("PPI","PPI & iNCI", "Random pairs in PPI","Random pairs in PPI (dist. > 3)","Random pairs in PPI & iNCI (dist. > 3)"))) %>%
  mutate(category = factor(category, levels = c("PPI", "Random", "Random (distance > 3)"))) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      y = abs(correlation),
      fill = category,
      colour = category) +
  geom_boxplot(alpha = 0.8,
               outlier.size = 0,
               outlier.alpha = 0) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        y = "Absolute co-expression weight",
        color = "",
        fill = "")
ggsave(
  boxplot_ppi_coexpression_by_types_file,
  plot = boxplot_ppi_coexpression_by_types,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
boxplot_ppi_coexpression_by_types
```

Violin plot plotting the zeroes (corresponding to the unsignificant correlations):

```{r}
violinplot_ppi_coexpression_by_types_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_pair_type_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("PPI", "Distance > 3")]
violinplot_ppi_coexpression_by_types = coexpression_pairs_filt %>%
  select(-distances_ppi, -distances_nci, -nci, -random_not_ppi, -random_far_not_nci) %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  mutate(category = replace(category, category == "ppi", "PPI")) %>%
  mutate(category = replace(category, category == "random_far_not_ppi", "Distance > 3")) %>%
  mutate(category = factor(category, levels = c("PPI", "Distance > 3"))) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      y = abs(correlation),
      fill = category,
      colour = category) +
  geom_violin(alpha = 0.8, 
              scale = "width", 
              adjust = 0.5) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        y = "Absolute co-expression weight",
        color = "",
        fill = ""); violinplot_ppi_coexpression_by_types
ggsave(
  violinplot_ppi_coexpression_by_types_file,
  plot = violinplot_ppi_coexpression_by_types,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
```

Violin plot showing only the weights of significant correlations (removing the unsignificant ones):

```{r}
violinplot_ppi_coexpression_by_types_nozero_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_pair_type_nozero_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("PPI", "Distance > 3")]
violinplot_ppi_coexpression_by_types_nozero = coexpression_pairs_filt %>%
  select(-distances_ppi, -distances_nci, -nci, -random_not_ppi, -random_far_not_nci) %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  filter(abs(correlation) > 0 ) %>% 
  mutate(category = replace(category, category == "ppi", "PPI")) %>%
  mutate(category = replace(category, category == "random_far_not_ppi", "Distance > 3")) %>%
  mutate(category = factor(category, levels = c("PPI", "Distance > 3"))) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      #y = abs(correlation),
      y = correlation,
      fill = category,
      colour = category) +
  geom_violin(alpha = 0.8, 
              scale = "width", 
              adjust = 0.5) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        #y = "Absolute co-expression weight",
        y = "Co-expression weight",
        color = "",
        fill = ""); violinplot_ppi_coexpression_by_types_nozero
ggsave(
  violinplot_ppi_coexpression_by_types_nozero_file,
  plot = violinplot_ppi_coexpression_by_types_nozero,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
```

Violin plot showing all co-expression weights (without setting unsignificant co-expression links to 0):

```{r}
violinplot_ppi_coexpression_by_types_nofilt_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_pair_type_%s_%s.png", dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("PPI", "Distance > 3")]
violinplot_ppi_coexpression_by_types_nofilt = coexpression_pairs %>%
  select(-distances_ppi, -distances_nci, -nci, -random_not_ppi, -random_far_not_nci) %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  mutate(category = replace(category, category == "ppi", "PPI")) %>%
  mutate(category = replace(category, category == "random_far_not_ppi", "Distance > 3")) %>%
  mutate(category = factor(category, levels = c("PPI", "Distance > 3"))) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      #y = abs(correlation),
      y = correlation,
      fill = category,
      colour = category) +
  geom_violin(
    alpha = 0.8
    #scale = "width",
    #adjust = 0.5
  ) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Number of samples",
        #y = "Absolute co-expression weight",
        y = "Co-expression weight",
        color = "",
        fill = ""); violinplot_ppi_coexpression_by_types_nofilt
ggsave(
  violinplot_ppi_coexpression_by_types_nofilt_file,
  plot = violinplot_ppi_coexpression_by_types_nofilt,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
```


### Calculate differences between different sample sizes of the same protein pair group

Use Friedmanâ€™s test to statistically evaluate differences between a set of protein pairs at different sample sizes.
Use Dunn's test to identify the type of difference between sample sizes if the difference exists.

#### By PPI network distance

Check differences when setting insignificant links to 0:

```{r}
# Select data to compare
ppi_coexpression_across_sizes_by_distances_df = coexpression_pairs_filt %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

friedman_results_across_sizes_by_distance_df = data.frame(matrix(ncol=length(c("distance", "friedman.statistic", "friedman.pvalue")),nrow=0, dimnames=list(NULL, c("distance", "friedman.statistic", "friedman.pvalue"))))
dunn_results_across_sizes_by_distance_df = data.frame(matrix(ncol=length(c("distance", "Comparison", "Z", "P.unadj", "P.adj")),nrow=0, dimnames=list(NULL, c("distance", "Comparison", "Z", "P.unadj", "P.adj"))))
for(distance_selected in sort(unique(ppi_coexpression_across_sizes_by_distances_df$distances_ppi))){
  # Filter co-expression by distance
  ppi_coexpression_by_selected_distance_df = ppi_coexpression_across_sizes_by_distances_df %>% filter(distances_ppi == distance_selected)
  # Run Friedman test by distance groups
  friedman_by_distances = friedman.test(y=abs(ppi_coexpression_by_selected_distance_df$correlation), groups=ppi_coexpression_by_selected_distance_df$size, blocks=ppi_coexpression_by_selected_distance_df$edge)
  # Save result
  friedman_results_across_sizes_by_distance_df = rbind(friedman_results_across_sizes_by_distance_df, data.frame(distance=distance_selected, friedman.statistic=friedman_by_distances$statistic[[1]], friedman.pvalue=friedman_by_distances$p.value))
  # Run Dunn test by distance groups
  dunn_by_distances = ppi_coexpression_by_selected_distance_df %>% 
    mutate(size = factor(size, levels = sizes_selected)) %>% 
    FSA::dunnTest(abs(correlation) ~ size, .)
  # Save result
  dunn_results_across_sizes_by_distance_df= rbind(dunn_results_across_sizes_by_distance_df, cbind(data.frame(distance=distance_selected), dunn_by_distances$res))
  rm(ppi_coexpression_by_selected_distance_df)
}
rm(ppi_coexpression_across_sizes_by_distances_df)

# Merge and save results
test_results_across_sizes_by_distance_df = friedman_results_across_sizes_by_distance_df %>% full_join(dunn_results_across_sizes_by_distance_df, by=c("distance")) %>% rename("comparison"="Comparison", "dunn.z"="Z", "dunn.p.unadj"="P.unadj", "dunn.p.adj"="P.adj")
test_results_by_distance_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_tests_across_sizes_by_distance_pval_%s_%s_%s.txt", as.character(threshold_selected), dataset, type_dataset)
test_results_across_sizes_by_distance_df %>% fwrite(test_results_by_distance_file)
```

The results are the following:

```{r}
head(test_results_across_sizes_by_distance_df)
```

```{r}
test_results_across_sizes_by_distance_df %>% select(distance, friedman.statistic, friedman.pvalue) %>% unique() %>% arrange(distance)
```

```{r}
test_results_across_sizes_by_distance_df %>% 
  select(distance, comparison, dunn.z, dunn.p.adj) %>% 
  filter(comparison %in% c("100 - 20", "20 - 200", "20 - 300", "100 - 200", "100 - 300")) %>% 
  unique() %>% 
  arrange(distance)
```

Check differences when keeping all links:

```{r}
# Select data to compare
ppi_coexpression_across_sizes_by_distances_df = coexpression_pairs %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

friedman_results_across_sizes_by_distance_df = data.frame(matrix(ncol=length(c("distance", "friedman.statistic", "friedman.pvalue")),nrow=0, dimnames=list(NULL, c("distance", "friedman.statistic", "friedman.pvalue"))))
dunn_results_across_sizes_by_distance_df = data.frame(matrix(ncol=length(c("distance", "Comparison", "Z", "P.unadj", "P.adj")),nrow=0, dimnames=list(NULL, c("distance", "Comparison", "Z", "P.unadj", "P.adj"))))
for(distance_selected in sort(unique(ppi_coexpression_across_sizes_by_distances_df$distances_ppi))){
  # Filter co-expression by distance
  ppi_coexpression_by_selected_distance_df = ppi_coexpression_across_sizes_by_distances_df %>% filter(distances_ppi == distance_selected)
  # Run Friedman test by distance groups
  friedman_by_distances = friedman.test(y=abs(ppi_coexpression_by_selected_distance_df$correlation), groups=ppi_coexpression_by_selected_distance_df$size, blocks=ppi_coexpression_by_selected_distance_df$edge)
  # Save result
  friedman_results_across_sizes_by_distance_df = rbind(friedman_results_across_sizes_by_distance_df, data.frame(distance=distance_selected, friedman.statistic=friedman_by_distances$statistic[[1]], friedman.pvalue=friedman_by_distances$p.value))
  # Run Dunn test by distance groups
  dunn_by_distances = ppi_coexpression_by_selected_distance_df %>% 
    mutate(size = factor(size, levels = sizes_selected)) %>% 
    FSA::dunnTest(abs(correlation) ~ size, .)
  # Save result
  dunn_results_across_sizes_by_distance_df= rbind(dunn_results_across_sizes_by_distance_df, cbind(data.frame(distance=distance_selected), dunn_by_distances$res))
  rm(ppi_coexpression_by_selected_distance_df)
}
rm(ppi_coexpression_across_sizes_by_distances_df)

# Merge and save results
test_results_across_sizes_by_distance_df = friedman_results_across_sizes_by_distance_df %>% full_join(dunn_results_across_sizes_by_distance_df, by=c("distance")) %>% rename("comparison"="Comparison", "dunn.z"="Z", "dunn.p.unadj"="P.unadj", "dunn.p.adj"="P.adj")
test_results_by_distance_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_tests_across_sizes_by_distance_%s_%s.txt", dataset, type_dataset)
test_results_across_sizes_by_distance_df %>% fwrite(test_results_by_distance_file)
```

The results are the following:

```{r}
head(test_results_across_sizes_by_distance_df)
```

```{r}
test_results_across_sizes_by_distance_df %>% select(distance, friedman.statistic, friedman.pvalue) %>% unique() %>% arrange(distance)
```

```{r}
test_results_across_sizes_by_distance_df %>% 
  select(distance, comparison, dunn.z, dunn.p.adj) %>% 
  filter(comparison %in% c("100 - 20", "20 - 200", "20 - 300", "100 - 200", "100 - 300")) %>% 
  unique() %>% 
  arrange(distance)
```

#### By protein/gene pair sets

Check differences when setting insignificant links to 0:

```{r}
# Select data to compare
ppi_coexpression_by_types_df = coexpression_pairs_filt %>%
  select(-distances_ppi, -distances_nci, -nci, -random_far_not_nci) %>%
  #pivot_longer(c("ppi", "random_not_ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

friedman_results_by_type_df = data.frame(matrix(ncol=length(c("category", "friedman.statistic", "friedman.pvalue")),nrow=0, dimnames=list(NULL, c("category", "friedman.statistic", "friedman.pvalue"))))
dunn_results_by_type_df = data.frame(matrix(ncol=length(c("category", "Comparison", "Z", "P.unadj", "P.adj")),nrow=0, dimnames=list(NULL, c("category", "Comparison", "Z", "P.unadj", "P.adj"))))
for(category_selected in sort(unique(ppi_coexpression_by_types_df$category))){
  # Filter co-expression by type
  ppi_coexpression_by_selected_type_df = ppi_coexpression_by_types_df %>% filter(category == category_selected)
  # Run Friedman test by protein/gene pair groups
  friedman_by_type = friedman.test(y=abs(ppi_coexpression_by_selected_type_df$correlation), groups=ppi_coexpression_by_selected_type_df$size, blocks=ppi_coexpression_by_selected_type_df$edge)
  # Save result
  friedman_results_by_type_df = rbind(friedman_results_by_type_df, data.frame(category=category_selected, friedman.statistic=friedman_by_type$statistic[[1]], friedman.pvalue=friedman_by_type$p.value))
  # Run Dunn test by protein/gene pair groups
  dunn_by_type = ppi_coexpression_by_selected_type_df %>% 
    mutate(size = factor(size, levels = sizes_selected)) %>% 
    FSA::dunnTest(abs(correlation) ~ size, .)
  # Save result
  dunn_results_by_type_df= rbind(dunn_results_by_type_df, cbind(data.frame(category=category_selected), dunn_by_type$res))
  rm(ppi_coexpression_by_selected_type_df)
}
rm(ppi_coexpression_by_types_df)

# Merge and save results
test_results_by_type_df = friedman_results_by_type_df %>% full_join(dunn_results_by_type_df, by=c("category")) %>% rename("comparison"="Comparison", "dunn.z"="Z", "dunn.p.unadj"="P.unadj", "dunn.p.adj"="P.adj")
test_results_by_type_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_tests_across_sizes_by_pair_type_pval_%s_%s_%s.txt", as.character(threshold_selected), dataset, type_dataset)
test_results_by_type_df %>% fwrite(test_results_by_type_file)
```

The results are the following:

```{r}
head(test_results_by_type_df)
```

Check differences when keeping all links:

```{r}
# Select data to compare
ppi_coexpression_by_types_df = coexpression_pairs %>%
  select(-distances_ppi, -distances_nci, -nci, -random_far_not_nci) %>%
  #pivot_longer(c("ppi", "random_not_ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

friedman_results_by_type_df = data.frame(matrix(ncol=length(c("category", "friedman.statistic", "friedman.pvalue")),nrow=0, dimnames=list(NULL, c("category", "friedman.statistic", "friedman.pvalue"))))
dunn_results_by_type_df = data.frame(matrix(ncol=length(c("category", "Comparison", "Z", "P.unadj", "P.adj")),nrow=0, dimnames=list(NULL, c("category", "Comparison", "Z", "P.unadj", "P.adj"))))
for(category_selected in sort(unique(ppi_coexpression_by_types_df$category))){
  # Filter co-expression by type
  ppi_coexpression_by_selected_type_df = ppi_coexpression_by_types_df %>% filter(category == category_selected)
  # Run Friedman test by protein/gene pair groups
  friedman_by_type = friedman.test(y=abs(ppi_coexpression_by_selected_type_df$correlation), groups=ppi_coexpression_by_selected_type_df$size, blocks=ppi_coexpression_by_selected_type_df$edge)
  # Save result
  friedman_results_by_type_df = rbind(friedman_results_by_type_df, data.frame(category=category_selected, friedman.statistic=friedman_by_type$statistic[[1]], friedman.pvalue=friedman_by_type$p.value))
  # Run Dunn test by protein/gene pair groups
  dunn_by_type = ppi_coexpression_by_selected_type_df %>% 
    mutate(size = factor(size, levels = sizes_selected)) %>% 
    FSA::dunnTest(abs(correlation) ~ size, .)
  # Save result
  dunn_results_by_type_df= rbind(dunn_results_by_type_df, cbind(data.frame(category=category_selected), dunn_by_type$res))
  rm(ppi_coexpression_by_selected_type_df)
}
rm(ppi_coexpression_by_types_df)

# Merge and save results
test_results_by_type_df = friedman_results_by_type_df %>% full_join(dunn_results_by_type_df, by=c("category")) %>% rename("comparison"="Comparison", "dunn.z"="Z", "dunn.p.unadj"="P.unadj", "dunn.p.adj"="P.adj")
test_results_by_type_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_tests_across_sizes_by_pair_type_%s_%s.txt", dataset, type_dataset)
test_results_by_type_df %>% fwrite(test_results_by_type_file)
```

```{r}
test_results_by_type_df
```

### Calculate differences between groups of protein/gene pairs of the same sample size

#### By PPI network distance

Check differences when setting insignificant links to 0:

```{r}
# Select data to compare
ppi_coexpression_across_distances_by_sizes_df = coexpression_pairs_filt %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

kw_results_across_distances_by_sizes_df = data.frame(matrix(ncol=length(c("size", "kw.statistic", "kw.pvalue")),nrow=0, dimnames=list(NULL, c("size", "kw.statistic", "kw.pvalue"))))
dunn_results_across_distances_by_sizes_df = data.frame(matrix(ncol=length(c("size", "Comparison", "Z", "P.unadj", "P.adj")),nrow=0, dimnames=list(NULL, c("size", "Comparison", "Z", "P.unadj", "P.adj"))))
for(size_selected in sort(unique(ppi_coexpression_across_distances_by_sizes_df$size))){
  # Filter co-expression by size
  ppi_coexpression_by_selected_size_df = ppi_coexpression_across_distances_by_sizes_df %>% filter(size == size_selected)
  # Run KW test across distance groups
  kw_across_distances = ppi_coexpression_by_selected_size_df %>% 
    mutate(distances_ppi = factor(distances_ppi, levels = c(1,2,3,4,5,6))) %>% 
    kruskal.test(abs(correlation) ~ distances_ppi, .)
  # Save result
  kw_results_across_distances_by_sizes_df = rbind(kw_results_across_distances_by_sizes_df, data.frame(size=size_selected, kw.statistic=kw_across_distances$statistic[[1]], kw.pvalue=kw_across_distances$p.value))
  # Run Dunn test across distance groups
  dunn_across_distances = ppi_coexpression_by_selected_size_df %>% 
    mutate(distances_ppi = factor(distances_ppi, levels = c(1,2,3,4,5,6))) %>% 
    FSA::dunnTest(abs(correlation) ~ distances_ppi, .)
  # Save result
  dunn_results_across_distances_by_sizes_df= rbind(dunn_results_across_distances_by_sizes_df, cbind(data.frame(size=size_selected), dunn_across_distances$res))
  rm(ppi_coexpression_by_selected_size_df)
}
rm(ppi_coexpression_across_distances_by_sizes_df)

# Merge and save results
test_results_across_distances_by_sizes_df = kw_results_across_distances_by_sizes_df %>% full_join(dunn_results_across_distances_by_sizes_df, by=c("size")) %>% rename("comparison"="Comparison", "dunn.z"="Z", "dunn.p.unadj"="P.unadj", "dunn.p.adj"="P.adj")
test_results_across_distances_by_sizes_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_statistic_tests_across_distance_by_sizes_pval_%s_%s_%s.txt", as.character(threshold_selected), dataset, type_dataset)
test_results_across_distances_by_sizes_df %>% fwrite(test_results_across_distances_by_sizes_file)
```

The results are the following:

```{r}
head(test_results_across_distances_by_sizes_df)
```

Here we find that the absolute co-expression weights from protein pairs that are close in the protein-protein network interactome are, on median, higher.

```{r}
test_results_across_distances_by_sizes_df %>% select(size, kw.statistic, kw.pvalue) %>% unique() %>% arrange(size)
```

```{r}
test_results_across_distances_by_sizes_df %>% 
  select(size, comparison, dunn.z, dunn.p.adj) %>% 
  filter(comparison %in% c("1 - 6", "1 - 4", "1 - 2")) %>% 
  unique() %>% 
  arrange(comparison, size)
```

Check differences when keeping all links (TABLE S2):

```{r}
# Function to Compute Pairwise Differences with Comparison Column
compute_pairwise_differences <- function(data, group_col, value_col) {

  unique_groups <- sort(unique(data[[group_col]]))
  
  # Generate all pairwise combinations of groups
  group_pairs <- combn(unique_groups, 2)
  
  results <- data.frame(
    comparison = character(0),
    difference = numeric(0)
  )
  
  for (i in seq_len(ncol(group_pairs))) {
    g1 <- group_pairs[1, i]
    g2 <- group_pairs[2, i]

    mean_abs_g1 <- mean(abs(data[data[[group_col]] == g1, ][[value_col]]))
    mean_abs_g2 <- mean(abs(data[data[[group_col]] == g2, ][[value_col]]))
    mean_diff <- round((mean_abs_g1 - mean_abs_g2), 3)

    results <- rbind(results, data.frame(
      comparison = paste(g1, "-", g2),
      mean_diff = mean_diff
    ))
  }
  
  return(results)
}

# Select protein pairs from different distances and transform to long format
ppi_coexpression_across_distances_by_sizes_df <- coexpression_pairs %>%
  dplyr::select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  dplyr::filter(!(is.na(distances_ppi))) %>%
  tidyr::pivot_longer(
    as.character(sizes_selected),
    names_to = "size",
    values_to = "correlation",
    names_transform = list(size = as.numeric),
    values_transform = list(correlation = as.numeric)
  ) %>%
  dplyr::filter(!(is.na(correlation))) %>%
  tidyr::unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

# Calculate mean, SD and difference across distances
mean_sd_corr_across_distances_by_sizes_df <- ppi_coexpression_across_distances_by_sizes_df %>%
  dplyr::group_by(size, distances_ppi) %>%
  dplyr::summarize(
    mean_abs_corr = round(mean(abs(correlation)), 3),
    sd_abs_corr = round(sd(abs(correlation)), 3)
  ) %>%
  dplyr::ungroup()

# Calculate statistical tests between protein pairs from different distances for
# each sample size
kw_results_across_distances_by_sizes_df <- data.frame()
mean_diff_across_distances_by_sizes_df <- data.frame()
dunn_results_across_distances_by_sizes_df <- data.frame()

for(size_selected in sort(unique(ppi_coexpression_across_distances_by_sizes_df$size))) {

  # Filter co-expression by size
  ppi_coexpression_by_selected_size_df <- ppi_coexpression_across_distances_by_sizes_df %>%
    dplyr::filter(size == size_selected)

  # Run KW test across distance groups
  kw_across_distances <- ppi_coexpression_by_selected_size_df %>% 
    dplyr::mutate(distances_ppi = factor(distances_ppi, levels = c(1,2,3,4,5,6))) %>% 
    kruskal.test(abs(correlation) ~ distances_ppi, .)

  # Save result
  kw_results_across_distances_by_sizes_df <- rbind(
    kw_results_across_distances_by_sizes_df,
    data.frame(
      size = size_selected,
      kw.statistic = kw_across_distances$statistic[[1]],
      kw.pvalue = ifelse(kw_across_distances$p.value == 0, "< 2.2e-16", kw_across_distances$p.value)
    )
  )

  # Pairwise differences of means across distances
  mean_diff_across_distances <- compute_pairwise_differences(
    data = ppi_coexpression_by_selected_size_df,
    group_col = "distances_ppi",
    value_col = "correlation"
  )

  # Save result
  mean_diff_across_distances_by_sizes_df <- rbind(
    mean_diff_across_distances_by_sizes_df,
    cbind(
      data.frame(size = size_selected),
      mean_diff_across_distances
    )
  )

  # Run Dunn test across distance groups
  dunn_across_distances <- ppi_coexpression_by_selected_size_df %>% 
    dplyr::mutate(distances_ppi = factor(distances_ppi, levels = c(1,2,3,4,5,6))) %>% 
    FSA::dunnTest(abs(correlation) ~ distances_ppi, .)

  # Save result
  dunn_results_across_distances_by_sizes_df <- rbind(
    dunn_results_across_distances_by_sizes_df,
    cbind(
      data.frame(size = size_selected),
      dunn_across_distances$res
    )
  )

  rm(ppi_coexpression_by_selected_size_df)
}

rm(ppi_coexpression_across_distances_by_sizes_df)

# Merge and save results
test_results_across_distances_by_sizes_df <- kw_results_across_distances_by_sizes_df %>%
  dplyr::full_join(
    dunn_results_across_distances_by_sizes_df %>%
      dplyr::full_join(
        mean_diff_across_distances_by_sizes_df,
        by = c("size" = "size", "Comparison" = "comparison")
      ),
    by = c("size")
  ) %>%
  dplyr::rename(
    "comparison" = "Comparison",
    "dunn.z" = "Z",
    "dunn.p.unadj" = "P.unadj",
    "dunn.p.adj" = "P.adj"
  )

test_results_across_distances_by_sizes_file <- sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_statistic_tests_across_distance_by_sizes_%s_%s.txt", dataset, type_dataset)
test_results_across_distances_by_sizes_df %>%
  data.table::fwrite(test_results_across_distances_by_sizes_file)

test_results_across_distances_by_sizes_df
```

The results are the following:

```{r}
test_results_across_distances_by_sizes_df %>% select(size, kw.statistic, kw.pvalue) %>% unique() %>% arrange(size)
```

```{r}
test_results_across_distances_by_sizes_df %>% 
  select(size, comparison, dunn.z, dunn.p.adj) %>% 
  filter(comparison %in% c("1 - 6", "1 - 4", "1 - 2")) %>% 
  unique() %>% 
  arrange(comparison, size)
```

#### By protein/gene pair sets

Check differences when setting insignificant links to 0:

```{r}
# Select data to compare
ppi_coexpression_across_types_by_sizes_df = coexpression_pairs_filt %>%
  select(-distances_ppi, -distances_nci, -nci, -random_far_not_nci) %>%
  #pivot_longer(c("ppi", "random_not_ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

kw_results_across_types_by_sizes_df = data.frame(matrix(ncol=length(c("size", "kw.statistic", "kw.pvalue")),nrow=0, dimnames=list(NULL, c("size", "kw.statistic", "kw.pvalue"))))
wilcox_results_across_types_by_sizes_df = data.frame(matrix(ncol=length(c("size", "w.statistic", "w.pvalue")),nrow=0, dimnames=list(NULL, c("size", "w.statistic", "w.pvalue"))))
for(size_selected in sort(unique(ppi_coexpression_across_types_by_sizes_df$size))){
  # Filter co-expression by size
  ppi_coexpression_by_selected_size_df = ppi_coexpression_across_types_by_sizes_df %>% 
    filter(size == size_selected) %>% 
    mutate(category = factor(category, levels = c("ppi", "random_far_not_ppi"))) 
  # Run KW test across types
  kw_across_types = ppi_coexpression_by_selected_size_df %>% 
    kruskal.test(abs(correlation) ~ category, .)
  # Save result
  kw_results_across_types_by_sizes_df = rbind(kw_results_across_types_by_sizes_df, data.frame(size=size_selected, kw.statistic=kw_across_types$statistic[[1]], kw.pvalue=kw_across_types$p.value))
  # # Run Dunn test across types
  # dunn_across_types = ppi_coexpression_by_selected_size_df %>% 
  #   mutate(category = factor(category, levels = c("ppi", "random_far_not_ppi"))) %>% 
  #   FSA::dunnTest(abs(correlation) ~ category, .) # ERROR: Error in Psort[1, i] : incorrect number of dimensions
  # Run Wilcoxon test instead of Dunn, because there are only two groups
  wilcox_across_types = wilcox.test(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "ppi"))$correlation), abs((ppi_coexpression_by_selected_size_df %>% filter(category == "random_far_not_ppi"))$correlation), alternative = "greater", correct = FALSE, exact = FALSE)
  # Save result
  wilcox_results_across_types_by_sizes_df= rbind(wilcox_results_across_types_by_sizes_df, data.frame(size=size_selected, w.statistic=as.numeric(wilcox_across_types$statistic), w.pvalue=wilcox_across_types$p.value))
  rm(ppi_coexpression_by_selected_size_df)
}
rm(ppi_coexpression_across_types_by_sizes_df)

# Merge and save results
test_results_across_types_by_sizes_df = kw_results_across_types_by_sizes_df %>% full_join(wilcox_results_across_types_by_sizes_df, by=c("size"))
test_results_across_types_by_sizes_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_statistic_tests_across_pair_type_by_sizes_pval_%s_%s_%s.txt", as.character(threshold_selected), dataset, type_dataset)
test_results_across_types_by_sizes_df %>% fwrite(test_results_across_types_by_sizes_file)
```

The results are the following:

```{r}
test_results_across_types_by_sizes_df
```

Check differences when keeping all links (TABLE S1):

```{r}
# Select PPIs and far protein pairs and transform the data to long format
ppi_coexpression_across_types_by_sizes_df <- coexpression_pairs %>%
  dplyr::select(-distances_ppi, -distances_nci, -nci, -random_far_not_nci) %>%
  tidyr::pivot_longer(
    c("ppi", "random_far_not_ppi"),
    names_to = "category",
    values_to = "category_bool"
  ) %>%
  dplyr::filter(category_bool == TRUE) %>%
  dplyr::select(-category_bool) %>%
  tidyr::pivot_longer(
    as.character(sizes_selected),
    names_to = "size",
    values_to = "correlation",
    names_transform = list(size = as.numeric),
    values_transform = list(correlation = as.numeric)
  ) %>%
  dplyr::filter(!(is.na(correlation))) %>%
  tidyr::unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

# Calculate statistical tests between PPIs and far protein pairs for each sample
# size
wilcox_results_across_types_by_sizes_df <- data.frame()
for(size_selected in sort(unique(ppi_coexpression_across_types_by_sizes_df$size))) {

  # Filter co-expression by size
  ppi_coexpression_by_selected_size_df <- ppi_coexpression_across_types_by_sizes_df %>% 
    dplyr::filter(size == size_selected) %>% 
    dplyr::mutate(category = factor(category, levels = c("ppi", "random_far_not_ppi")))

  # Run Wilcoxon test instead of Dunn, because there are only two groups
  wilcox_across_types <- wilcox.test(
    abs((ppi_coexpression_by_selected_size_df %>% filter(category == "ppi"))$correlation),
    abs((ppi_coexpression_by_selected_size_df %>% filter(category == "random_far_not_ppi"))$correlation),
    alternative = "greater",
    correct = FALSE,
    exact = FALSE
  )

  # Calculate mean and SD
  mean_abs_corr_ppi <- mean(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "ppi"))$correlation))
  mean_abs_corr_far <- mean(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "random_far_not_ppi"))$correlation))
  diff_mean <- mean_abs_corr_ppi - mean_abs_corr_far
  sd_abs_corr_ppi <- sd(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "ppi"))$correlation))
  sd_abs_corr_far <- sd(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "random_far_not_ppi"))$correlation))
  diff_sd <- sd_abs_corr_ppi - sd_abs_corr_far

  # Save result
  wilcox_results_across_types_by_sizes_df <- rbind(
    wilcox_results_across_types_by_sizes_df,
    data.frame(
      size = size_selected,
      mean_abs_corr_ppi = round(mean_abs_corr_ppi, 3),
      mean_abs_corr_far = round(mean_abs_corr_far, 3),
      diff_mean = round(diff_mean, 3),
      sd_abs_corr_ppi = round(sd_abs_corr_ppi, 3),
      sd_abs_corr_far = round(sd_abs_corr_far, 3),
      diff_sd = round(diff_sd, 3),
      w.statistic = as.numeric(wilcox_across_types$statistic),
      w.pvalue = ifelse(wilcox_across_types$p.value == 0, "< 2.2e-16", wilcox_across_types$p.value)
    )
  )

  rm(ppi_coexpression_by_selected_size_df)
}

rm(ppi_coexpression_across_types_by_sizes_df)

# Save results
wilcox_results_across_types_by_sizes_file <- sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_wilcox_across_pair_type_by_sizes_%s_%s.txt", dataset, type_dataset)
wilcox_results_across_types_by_sizes_df %>%
  data.table::fwrite(wilcox_results_across_types_by_sizes_file)

wilcox_results_across_types_by_sizes_df
```

The results are the following:

```{r}
test_results_across_types_by_sizes_df
```


### Calculate performance predicting PPIs using gene co-expression

Define functions to get performance metrics:

```{r}
#'  get_auc
#'  Get AUC from ROC calculation and organize it.
#'  @param rocs ROC result.
#'
get_auc = function(rocs, base = "xxx", measure = 'wto'){
  o = rocs$ci %>% 
    as.vector() %>% 
    t %>%
    as.data.frame()  %>%
    mutate(base = base, 
           measure = measure) 
  names(o)[1:3] = c("CI_l", "AUC", "CI_h")
  
  return(o)
}

#'  performances
#'  Calculate performances
#'  @param data Score/Value list.
#'  @param label Label list.
#'
performances = function(data, label){
  # Calculate ROC
  roc <- ROCR::performance(data,"tpr","fpr")
  roc = data.frame(x = as.numeric(roc@x.values[[1]]),
                   y = as.numeric(roc@y.values[[1]]),
                   cutoff = as.numeric(roc@alpha.values[[1]]),
                   type = "ROC")
  
  # Calculate AUC
  auc <- ROCR::performance(data,"auc")
  auc = auc@y.values
  
  # Calculate precision
  prec_rec <- ROCR::performance(data, "prec", "rec")
  prec_rec = data.frame(x = as.numeric(prec_rec@x.values[[1]]),
                        y = as.numeric(prec_rec@y.values[[1]]),
                        cutoff = as.numeric(prec_rec@alpha.values[[1]]),
                        type = "ROCPR")

  # Calculate AUCPR
  aucpr <- ROCR::performance(data, "aucpr")
  aucpr = aucpr@y.values

  # Merge results
  out =  bind_rows(roc, prec_rec) %>%
    mutate(base = label) %>%
    mutate(AUC = auc[[1]]) %>%
    mutate(AUCPR = aucpr[[1]])

  return(out)
}
```

#### Results setting unsignificant links to 0

Calculate the AUC of predicting PPIs using gene co-expression across sample sizes:

```{r}
# Select correlations to analyze
ppi_vs_random_far_df = coexpression_pairs_filt %>%
  select(-distances_ppi, -distances_nci, -nci, -random_not_ppi, -random_far_not_nci) %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  mutate(correlation=abs(correlation))

sum_rocs_df = data.frame(matrix(ncol=length(c("size", "CI_l", "AUC", "CI_h", "base", "measure")),nrow=0, dimnames=list(NULL, c("size", "CI_l", "AUC", "CI_h", "base", "measure"))))
perf_df = data.frame(matrix(ncol=length(c("size", "x", "y", "cutoff", "type", "base", "AUC", "AUCPR")),nrow=0, dimnames=list(NULL, c("size", "x", "y", "cutoff", "type", "base", "AUC", "AUCPR"))))
for (size_selected in sort(as.numeric(unique(ppi_vs_random_far_df$size)))){
  # Filter correlations by size
  data_filtered_by_size = ppi_vs_random_far_df %>% filter(size == size_selected)
  
  # Calculate ROC
  roc_result = pROC::roc(data_filtered_by_size, category, correlation, ci = T, direction="auto")
  
  # Get AUC from sum of ROC calculations
  sum_rocs_result = roc_result %>%
    get_auc(., base = "ppi", 
            measure = "pearson")
  sum_rocs_df = rbind(sum_rocs_df, cbind(data.frame(size=size_selected), sum_rocs_result))

  # Create prediction instance
  pred_instance = ROCR::prediction(predictions = data_filtered_by_size$correlation,
                                  labels = data_filtered_by_size$category,
                                  label.ordering = c("random_far_not_ppi", "ppi")) # Negative and positive class label
  
  # Calculate performances
  perf_result = pred_instance  %>%
    performances(., label = "PPI")
  perf_df = rbind(perf_df, cbind(data.frame(size=size_selected), perf_result))
  
  #rm(data_filtered_by_size)
  #rm(pred_instance)
}

# Save results
sum_rocs_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_ppi_vs_random_far_pairs_sum_rocs_pval_%s_%s_%s.txt", as.character(threshold_selected), dataset, type_dataset)
sum_rocs_df %>% fwrite(sum_rocs_file) 
perf_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_ppi_vs_random_far_pairs_performance_pval_%s_%s_%s.txt", as.character(threshold_selected), dataset, type_dataset)
perf_df %>% fwrite(perf_file) 
```

Compare the results of the two ways of calculating the AUC:

```{r}
sum_rocs_df %>% select(size, AUC)
perf_df %>% select("size", "AUC", "AUCPR") %>% unique()
```

Plot the AUC:

```{r}
perf_sum = perf_df %>%
  select(size, AUC,AUCPR) %>%
  unique() %>% 
  pivot_longer(cols = c("AUC", "AUCPR"))
print(perf_sum)

perf_plot = perf_sum %>% 
  mutate(size = factor(size, 
                       levels = sizes_selected, 
                       labels = sizes_selected)) %>% 
  ggplot() +
  aes(x = size, 
      #fill = base, 
      #colour = base, 
      weight = value) +
  geom_bar(col="#DBA9CA",
         fill="#DBA9CA") +
  #scale_fill_manual(values = colors_defined) +
  #scale_color_manual(values = colors_defined) +
  theme_linedraw() +
  facet_grid(vars(name))+ 
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold", hjust = 0.5), 
        axis.text = element_text(size = 13), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        strip.text = element_text(face = "bold", hjust = 0.5),
        text = element_text(family = "Helvetica")) +
  labs(y = NULL, #"AUC / AUCPR", 
       x = "Number of samples"
       )
perf_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
ggsave(
  perf_plot_file,
  plot = perf_plot,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
 perf_plot
```

```{r}
perf_auc = perf_df %>%
  select(size, AUC) %>%
  unique()
print(perf_auc)

auc_plot = perf_auc %>% 
  mutate(size = factor(size, 
                       levels = sizes_selected, 
                       labels = sizes_selected)) %>% 
  ggplot() +
  aes(x = size, 
      #fill = base, 
      #colour = base, 
      weight = AUC) +
  geom_bar(col="#DBA9CA",
         fill="#DBA9CA") +
  #scale_fill_manual(values = colors_defined) +
  #scale_color_manual(values = colors_defined) +
  coord_cartesian(ylim=c(0.5,NA)) +
  theme_linedraw() +
  #facet_grid(vars(name))+ 
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs(y = "AUROC",
       x = "Number of samples"
       )

auc_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_AUC_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
ggsave(
  auc_plot_file,
  plot = auc_plot,
  dpi = 1200,
  #width = 9300,
  height = 6000,
  units = c("px")
)
 auc_plot
```

#### Results keeping the value of all the links

Calculate the AUC of predicting PPIs using gene co-expression across sample sizes:

```{r}
# Select correlations to analyze
ppi_vs_random_far_df = coexpression_pairs %>%
  select(-distances_ppi, -distances_nci, -nci, -random_not_ppi, -random_far_not_nci) %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.numeric), values_transform = list(correlation = as.numeric)) %>%
  filter(!(is.na(correlation))) %>%
  mutate(correlation=abs(correlation))

sum_rocs_df = data.frame(matrix(ncol=length(c("size", "CI_l", "AUC", "CI_h", "base", "measure")),nrow=0, dimnames=list(NULL, c("size", "CI_l", "AUC", "CI_h", "base", "measure"))))
perf_df = data.frame(matrix(ncol=length(c("size", "x", "y", "cutoff", "type", "base", "AUC", "AUCPR")),nrow=0, dimnames=list(NULL, c("size", "x", "y", "cutoff", "type", "base", "AUC", "AUCPR"))))
for (size_selected in sort(as.numeric(unique(ppi_vs_random_far_df$size)))){
  # Filter correlations by size
  data_filtered_by_size = ppi_vs_random_far_df %>% filter(size == size_selected)
  
  # Calculate ROC
  roc_result = pROC::roc(data_filtered_by_size, category, correlation, ci = T, direction="auto")
  
  # Get AUC from sum of ROC calculations
  sum_rocs_result = roc_result %>%
    get_auc(., base = "ppi", 
            measure = "pearson")
  sum_rocs_df = rbind(sum_rocs_df, cbind(data.frame(size=size_selected), sum_rocs_result))

  # Create prediction instance
  pred_instance = ROCR::prediction(predictions = data_filtered_by_size$correlation,
                                  labels = data_filtered_by_size$category,
                                  label.ordering = c("random_far_not_ppi", "ppi")) # Negative and positive class label
  
  # Calculate performances
  perf_result = pred_instance  %>%
    performances(., label = "PPI")
  perf_df = rbind(perf_df, cbind(data.frame(size=size_selected), perf_result))
  
  #rm(data_filtered_by_size)
  #rm(pred_instance)
}

# Save results
sum_rocs_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_ppi_vs_random_far_pairs_sum_rocs_%s_%s.txt", dataset, type_dataset)
sum_rocs_df %>% fwrite(sum_rocs_file) 
perf_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_ppi_vs_random_far_pairs_performance_%s_%s.txt", dataset, type_dataset)
perf_df %>% fwrite(perf_file) 
```

Compare the results of the two ways of calculating the AUC:

```{r}
sum_rocs_df %>% select(size, AUC)
perf_df %>% select("size", "AUC", "AUCPR") %>% unique()
```

Plot the AUC and AUCPR barplot:

```{r}
perf_sum = perf_df %>%
  select(size, AUC,AUCPR) %>%
  unique() %>% 
  pivot_longer(cols = c("AUC", "AUCPR"))
print(perf_sum)

perf_plot = perf_sum %>% 
  mutate(size = factor(size, 
                       levels = sizes_selected, 
                       labels = sizes_selected)) %>% 
  ggplot() +
  aes(x = size, 
      #fill = base, 
      #colour = base, 
      weight = value) +
  geom_bar(col="#DBA9CA",
         fill="#DBA9CA") +
  #scale_fill_manual(values = colors_defined) +
  #scale_color_manual(values = colors_defined) +
  theme_linedraw() +
  facet_grid(vars(name))+ 
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold", hjust = 0.5), 
        axis.text = element_text(size = 13), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        strip.text = element_text(face = "bold", hjust = 0.5),
        text = element_text(family = "Helvetica")) +
  labs(y = NULL, #"AUC / AUCPR", 
       x = "Number of samples"
       )
perf_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_%s_%s.png", dataset, type_dataset)
ggsave(
  perf_plot_file,
  plot = perf_plot,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
 perf_plot
```

Plot the AUC barplot:

```{r}
perf_auc = perf_df %>%
  select(size, AUC) %>%
  unique()
print(perf_auc)

auc_plot = perf_auc %>% 
  mutate(size = factor(size, 
                       levels = sizes_selected, 
                       labels = sizes_selected)) %>% 
  ggplot() +
  aes(x = size, 
      #fill = base, 
      #colour = base, 
      weight = AUC) +
  geom_bar(col="#DBA9CA",
         fill="#DBA9CA") +
  #scale_fill_manual(values = colors_defined) +
  #scale_color_manual(values = colors_defined) +
  coord_cartesian(ylim=c(0.5,NA)) +
  theme_linedraw() +
  #facet_grid(vars(name))+ 
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs(y = "AUROC",
       x = "Number of samples"
       )

auc_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_AUC_%s_%s.png", dataset, type_dataset)
ggsave(
  auc_plot_file,
  plot = auc_plot,
  dpi = 1200,
  #width = 9300,
  height = 6000,
  units = c("px")
)
 auc_plot
```

Plot the ROC curve:

```{r}
my_orange = RColorBrewer::brewer.pal(n = (length(unique(perf_df$size)) + 2), "Oranges")[3:(length(unique(perf_df$size))+2)] # I get 8 and exclude the 2 first colors, which are more light 
roc_curve_plot = perf_df %>% 
  filter(type == "ROC") %>% 
  mutate(size = factor(size, levels = as.character(sort(as.numeric(unique(perf_df$size)))))) %>%
  ggplot(aes(x=x, y=y)) +
  geom_line(aes(col=size)) +
  scale_color_manual(values=my_orange) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "FPR",
        y = "TPR",
        color = "Num. samples")

roc_curve_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_ROC_%s_%s.png", dataset, type_dataset)
ggsave(
  roc_curve_plot_file,
  plot = roc_curve_plot,
  dpi = 1200,
  #width = 9300,
  height = 6000,
  units = c("px")
)
roc_curve_plot
```

Plot the ROCPR curve:

```{r}
my_orange = RColorBrewer::brewer.pal(n = (length(unique(perf_df$size)) + 2), "Oranges")[3:(length(unique(perf_df$size))+2)] # I get 8 and exclude the 2 first colors, which are more light 
rocpr_curve_plot = perf_df %>% 
  filter(type == "ROCPR") %>% 
  mutate(size = factor(size, levels = as.character(sort(as.numeric(unique(perf_df$size)))))) %>%
  ggplot(aes(x=x, y=y)) +
  geom_line(aes(col=size)) +
  scale_color_manual(values=my_orange) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Recall",
        y = "Precision",
        color = "Num. samples")

rocpr_curve_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_ROCPR_%s_%s.png", dataset, type_dataset)
ggsave(
  rocpr_curve_plot_file,
  plot = rocpr_curve_plot,
  dpi = 1200,
  #width = 9300,
  height = 6000,
  units = c("px")
)
rocpr_curve_plot
```

Plot the Specificity vs. Sensitivity curve:

```{r}
my_orange = RColorBrewer::brewer.pal(n = (length(unique(perf_df$size)) + 2), "Oranges")[3:(length(unique(perf_df$size))+2)] # I get 8 and exclude the 2 first colors, which are more light 
sensspeccurve_curve_plot = perf_df %>% 
  filter(type == "ROC") %>% 
  mutate(size = factor(size, levels = as.character(sort(as.numeric(unique(perf_df$size)))))) %>%
  ggplot(aes(x=(1-x), y=y)) +
  geom_line(aes(col=size)) +
  scale_color_manual(values=my_orange) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 15, face="bold"), 
        axis.title = element_text(size = 14, face="bold"), 
        axis.text = element_text(size = 13),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 13), 
        legend.title=element_text(size=14, face="bold"),
        text = element_text(family = "Helvetica")) +
  labs (x = "Specificity",
        y = "Sensitivity",
        color = "Num. samples")

sensspeccurve_curve_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_sensspeccurve_%s_%s.png", dataset, type_dataset)
ggsave(
  sensspeccurve_curve_plot_file,
  plot = sensspeccurve_curve_plot,
  dpi = 1200,
  #width = 9300,
  height = 6000,
  units = c("px")
)
sensspeccurve_curve_plot
```

Plot the balanced accuracy vs. cut-off:

```{r}
metric2name = list(baccuracy="Balanced Accuracy", TPR="Sensitivity", TNR="Specificity")

perf_vs_cutoff_df = perf_df %>%
  filter(type == "ROC") %>%
  select(-type, -base) %>%
  rename(c(FPR = x, TPR = y)) %>%
  mutate(cutoff = replace(cutoff, cutoff == Inf, 1)) %>%
  mutate(TNR = 1 - FPR) %>%
  mutate(baccuracy = (TPR + TNR) / 2) %>%
  mutate(size = factor(size, levels = as.character(sort(as.numeric(unique(perf_df$size)))))) %>% 
  pivot_longer(!c(size, cutoff), names_to = "metric", values_to = "result")

for(metric_selected in c("baccuracy", "TPR", "TNR")){
  perf_vs_cutoff_plot = perf_vs_cutoff_df %>%
    filter(metric == metric_selected) %>%
    ggplot(aes(x=cutoff, y=result)) +
    geom_line(aes(col=size)) +
    scale_color_manual(values=my_orange) +
    theme_linedraw() +
    theme(aspect.ratio=1,
          plot.title =  element_text(size = 15, face="bold"), 
          axis.title = element_text(size = 14, face="bold"), 
          axis.text = element_text(size = 13),
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 13), 
          legend.title=element_text(size=14, face="bold"),
          text = element_text(family = "Helvetica")) +
    labs (x = "Cut-off",
          y = metric2name[[metric_selected]],
          color = "Num. samples")
  perf_vs_cutoff_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_%s_vs_cutoff_%s_%s.png", metric_selected, dataset, type_dataset)
  ggsave(
    perf_vs_cutoff_plot_file,
    plot = perf_vs_cutoff_plot,
    dpi = 1200,
    #width = 9300,
    height = 6000,
    units = c("px")
  )
  print(perf_vs_cutoff_plot)
}
```

Calculate optimal threshold based on Sensitivity and Specificity:

```{r}
cutoff_df = data.frame(matrix(ncol=length(c("size", "cutoff")),nrow=0, dimnames=list(NULL, c("size", "cutoff"))))
for(size_selected in unique(ppi_vs_random_far_df$size)){
  ppi_vs_random_far_size_df = ppi_vs_random_far_df %>% filter(size == size_selected)
  cp <- cutpointr::cutpointr(ppi_vs_random_far_size_df, correlation, category, 
                  method = cutpointr::maximize_metric, metric = cutpointr::sum_sens_spec)
  cutoff_df = rbind(cutoff_df, data.frame(size=size_selected, cutoff=cp$optimal_cutpoint))
  #summary(cp)
  #plot(cp)
}
cutoff_df
```

```{r}
perf_by_selected_cutoff_df = data.frame(matrix(ncol=length(c("size", "metric", "result", "cutoff")),nrow=0, dimnames=list(NULL, c("size", "metric", "result", "cutoff"))))
cutoffs = c(0.40, 0.50)
for(cutoff_selected in cutoffs){
  perf_by_cutoff_df = ppi_vs_random_far_df %>%
    mutate(prediction = if_else(correlation >= cutoff_selected, "ppi", "random_far_not_ppi")) %>%
    mutate(result = if_else(((category == "ppi") & (prediction == "ppi")), "TP", if_else(((category == "random_far_not_ppi") & (prediction == "random_far_not_ppi")), "TN", ifelse(((category == "ppi") & (prediction == "random_far_not_ppi")), "FP", "FN"))))
  perf_sum_by_cutoff_df = perf_by_cutoff_df %>%
    group_by(size, result) %>%
    summarize(n_result = n()) %>%
    ungroup() %>% 
    pivot_wider(names_from = result, values_from = n_result) %>%
    as.data.frame()
  perf_sum_by_cutoff_df$TPR = perf_sum_by_cutoff_df$TP / (perf_sum_by_cutoff_df$TP + perf_sum_by_cutoff_df$FN) # TPR = Sensitivity: Proportion of predicted positives identified correctly
  perf_sum_by_cutoff_df$TNR = perf_sum_by_cutoff_df$TN / (perf_sum_by_cutoff_df$TN + perf_sum_by_cutoff_df$FP) # TNR = Specificity: Proportion of predicted negatives identified correctly
  perf_sum_by_cutoff_df$FPR = perf_sum_by_cutoff_df$FP / (perf_sum_by_cutoff_df$TN + perf_sum_by_cutoff_df$FP)
  perf_sum_by_cutoff_df$baccuracy = (perf_sum_by_cutoff_df$TPR + perf_sum_by_cutoff_df$TNR) / 2 # Accuracy: Proportion of predictions (positives or negatives) that the model got correctly
  perf_sum_by_cutoff_df$precision = perf_sum_by_cutoff_df$TP / (perf_sum_by_cutoff_df$TP + perf_sum_by_cutoff_df$FP) # Precision: Proportion of positives (predicted or not) identified correctly
  perf_sum_by_cutoff_df = perf_sum_by_cutoff_df %>% pivot_longer(-size, names_to = "metric", values_to = "result")
  perf_by_selected_cutoff_df = rbind(perf_by_selected_cutoff_df, cbind(perf_sum_by_cutoff_df, data.frame(cutoff=cutoff_selected)))
  
  perf_sum_by_cutoff_plot = perf_sum_by_cutoff_df %>%
    filter(metric %in% c("TPR", "TNR", "baccuracy", "precision")) %>%
    ggplot(aes(x=size, y=result)) +
    geom_line(aes(col=metric)) +
    scale_x_continuous(breaks = unique(perf_sum_by_cutoff_df$size)) +
    theme_linedraw() +
    theme(aspect.ratio=1,
          plot.title =  element_text(size = 15, face="bold"), 
          axis.title = element_text(size = 14, face="bold"), 
          axis.text = element_text(size = 13),
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 13), 
          legend.title=element_text(size=14, face="bold"),
          text = element_text(family = "Helvetica")) +
    labs (x = "Num. samples",
          y = "Metric value",
          color = "Metric")
  perf_sum_by_cutoff_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_by_cutoff_%s_%s_%s.png", cutoff_selected, dataset, type_dataset)
  ggsave(
    perf_sum_by_cutoff_plot_file,
    plot = perf_sum_by_cutoff_plot,
    dpi = 1200,
    #width = 9300,
    height = 6000,
    units = c("px")
  )
  print(perf_sum_by_cutoff_plot)
}
```



