---
title: "scipher_preprocessing"
author: "Joaquim Aguirre-Plans & Deisy Morselli Gysi"
date: "1/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Pre-process Scipher data (version v8) before creating gene co-expression networks.

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
library(igraph)
require(magrittr)
library(tidyr)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```

## Read files

```{r}
# Define working directories
scipher_data_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/Dec2021'

# Define input files
RNA_metadata_file = paste(scipher_data_dir, '00_data/RNA_metadata.csv', sep='/')
```

We read a file containing the gene expression and metadata together, which was created in a previous analysis:

```{r}
# Read RNA metadata processed and select the clean dataset of genes
RNA_metadata = fread(RNA_metadata_file)
genes_dataset = unique(toupper(names(RNA_metadata)[53:length(names(RNA_metadata))]))
print(length(genes_dataset))
```

```{r}
RNA_metadata[1:5,1:60]
```

The `RNA_metadata` dataframe contains `r ncol(RNA_metadata)` features (`r length(genes_dataset)` genes) and `r nrow(RNA_metadata)` samples.

> Important to note: There are many samples that have the same Scipher_id, meaning that they come from the same patient. They only differ by the visit (which can be either 0m, 3m or 6m) and the "batch" (which can be either "seq_fac1" or "seq_fac2").

We combine the Scipher_id, the visit and the batch, so that we can have a unique ID for each sample:

```{r}
scipherid_visit_batch = paste(RNA_metadata$Scipher_id, RNA_metadata$visit, RNA_metadata$batch, sep=".")
RNA_metadata = cbind(data.frame(Scipher_id.visit.batch=scipherid_visit_batch), RNA_metadata) %>% relocate(Scipher_id.visit.batch, Scipher_id, visit, batch, .before = age)
RNA_metadata[1:5,1:5]
```

The new `RNA_metadata` dataframe contains `r ncol(RNA_metadata)` features (`r length(genes_dataset)` genes) and `r nrow(RNA_metadata)` samples.


## Analyze the metadata

We analyze the response of the patients at different time points:

```{r}
# Read file with metadata
metadata_raw_file = paste(scipher_data_dir, '00_data/RA-netk042021-10-25.xlsx', sep='/')
metadata = readxl::read_xlsx(metadata_raw_file)

metadata = metadata %>% 
  rename(CRP_metadata = CRP)

# Select type of response of patients
metadata_response = 
  metadata %>%
  select(Scipher_id, visit, 
         eular_binary, acr50) %>%
  pivot_wider(names_from = visit, 
              values_from = c(eular_binary, acr50))

head(metadata_response)
```

When combining the metadata with the RNAseq we have, in total, `r nrow(RNA_metadata)` samples from `r length(unique(RNA_metadata$Scipher_id))` individuals. Out of those, **`r nrow(RNA_metadata[RNA_metadata$visit == "0m", ])`** samples from `r length(unique(RNA_metadata[RNA_metadata$visit == "0m", ]$Scipher_id))` individuals are at baseline.

Let's focus on the samples that we have in the baseline (visit = 0m). 

```{r, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
# We filter gene expression by visit = 0m (at baseline)
RNA_metadata_0m = RNA_metadata %>% filter(visit == "0m")
RNA_metadata_0m = dplyr::inner_join(metadata_response, RNA_metadata_0m)
```

This is the response (in eular) of patients after 3 months:

```{r, echo=FALSE, results='asis'}
RNA_metadata_0m %>% 
  group_by(eular_binary_3m) %>% 
  summarise(`#` = n()) %>%
  ungroup() %>%
  mutate(`%` = round((`#`/n()), 2))
```

This is the response (in eular) of patients after 6 months:

```{r, echo=FALSE, results='asis'}
RNA_metadata_0m %>% 
  group_by(eular_binary_6m) %>% 
  summarise(`#` = n()) %>%
  ungroup() %>%
  mutate(`%` = round((`#`/n()), 2))
```

This is a summary of the response (in eular) after both 3 and 6 months:

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
RNA_metadata_0m %>% 
  group_by(eular_binary_3m, eular_binary_6m) %>%
  summarise(`#` = n()) %>%
  ungroup() %>%
  mutate(`%` = round((`#`/n()), 2))
```


## Sample processing

In this section, we will divide the samples in groups depending on the type of analysis that we want to perform.

* `scipher.complete.dataset`: All samples, no matter the visit or batch.
* `scipher.complete.responder`: All samples at any visit where the patient responds to the treatment.
* `scipher.complete.nonresponder`: All samples at any visit where the patient does not respond to the treatment.
* `scipher.sample.per.patient.all.visits`: One sample per patient at different visits.
* `scipher.sample.per.patient.baseline`: One sample per patient at baseline.
* `scipher.sample.per.responder.baseline`: One sample per responder patient at baseline.
* `scipher.sample.per.nonresponder.baseline`: One sample per non-responder patient at baseline.

First, we will merge samples repeated in the same patient so that we can create a dataset with a unique sample per patient and timepoint

### Combine samples from same patients and visit

We combine the Scipher_id and visit, so that we can have a unique ID for each sample:

```{r}
scipherid_visit = paste(RNA_metadata$Scipher_id, RNA_metadata$visit, sep=".")
RNA_metadata_unique_patients = cbind(data.frame(Scipher_id.visit=scipherid_visit), (RNA_metadata %>% select(-Scipher_id.visit.batch))) %>% relocate(Scipher_id.visit, Scipher_id, visit, batch, .before = age)
RNA_metadata_unique_patients[1:10,1:5]
```

We calculate the median expression of samples from same patients and visits:

```{r}
RNA_metadata_unique_patients = RNA_metadata_unique_patients %>% select(-batch)
RNA_metadata_unique_patients = RNA_metadata_unique_patients %>% group_by_at(colnames(RNA_metadata_unique_patients)[1:52]) %>% summarise_all(median) %>% as.data.frame()
RNA_metadata_unique_patients_file = paste(scipher_data_dir, "00_data/RNA_metadata_unique_patients.csv", sep='/')
fwrite(RNA_metadata_unique_patients, RNA_metadata_unique_patients_file)
RNA_metadata_unique_patients[1:10,1:5]
```

The new `RNA_metadata_unique_patients` dataframe contains `r ncol(RNA_metadata_unique_patients)` features (`r length(genes_dataset)` genes) and `r nrow(RNA_metadata_unique_patients)` samples.


### Make groups of samples

Now, we will make different groups of samples depending on the analyses that we are planning to perform:

```{r}
# Select all samples
scipher_complete_dataset = RNA_metadata$Scipher_id.visit.batch

# Responders are selected when they respond in eular and do not have any non-response
responder_euler_complete_dataset_df = metadata_response %>% inner_join(RNA_metadata, by="Scipher_id") %>% filter(( (eular_binary_3m == "responder") & (eular_binary_6m == "responder")) | ((eular_binary_3m == "responder") & (eular_binary_6m == "")) | ((eular_binary_3m == "") & (eular_binary_6m == "responder")) ) # Response in eular and not any non-response
scipher_complete_responder = responder_euler_complete_dataset_df$Scipher_id.visit.batch

# Nonresponders are selected when they don't respond in eular and do not have any response
nonresponder_euler_complete_dataset_df = metadata_response %>% inner_join(RNA_metadata, by="Scipher_id") %>% filter(( (eular_binary_3m == "nonresponder") & (eular_binary_6m == "nonresponder")) | ((eular_binary_3m == "nonresponder") & (eular_binary_6m == "")) | ((eular_binary_3m == "") & (eular_binary_6m == "nonresponder")) )
#nonresponder_df = sample_per_patient_baseline_df %>% filter( (!(eular_binary_3m == "responder")) & (!(eular_binary_6m == "responder")) & (!(acr50_3m == "responder")) & (!(acr50_6m == "responder")) & ( (eular_binary_3m == "nonresponder") | (eular_binary_6m == "nonresponder") | (acr50_3m == "nonresponder") | (acr50_6m == "nonresponder") ) ) # Non-response in eular and not any response
scipher_complete_nonresponder = nonresponder_euler_complete_dataset_df$Scipher_id.visit.batch

# Select all samples of unique patients at different visits
sample_per_patient_all_visits = RNA_metadata_unique_patients$Scipher_id.visit

# Select all samples of unique patients at baseline
sample_per_patient_baseline = (RNA_metadata_unique_patients %>% filter(visit == "0m"))$Scipher_id.visit

# Responders are selected when they respond in eular and do not have any non-response
responder_euler_unique_patients_baseline_df = metadata_response %>% inner_join((RNA_metadata_unique_patients %>% filter(visit == "0m")), by="Scipher_id") %>% filter(( (eular_binary_3m == "responder") & (eular_binary_6m == "responder")) | ((eular_binary_3m == "responder") & (eular_binary_6m == "")) | ((eular_binary_3m == "") & (eular_binary_6m == "responder")) ) # Response in eular and not any non-response
sample_per_responder_baseline = responder_euler_unique_patients_baseline_df$Scipher_id.visit

# Nonresponders are selected when they don't respond in eular and do not have any response
nonresponder_euler_unique_patients_baseline_df = metadata_response %>% inner_join((RNA_metadata_unique_patients %>% filter(visit == "0m")), by="Scipher_id") %>% filter(( (eular_binary_3m == "nonresponder") & (eular_binary_6m == "nonresponder")) | ((eular_binary_3m == "nonresponder") & (eular_binary_6m == "")) | ((eular_binary_3m == "") & (eular_binary_6m == "nonresponder")) )
sample_per_nonresponder_baseline = nonresponder_euler_unique_patients_baseline_df$Scipher_id.visit

# Output table with different groups of samples
group_to_samples_df = data.frame(group="scipher.complete.dataset", sample_id=scipher_complete_dataset)
group_to_samples_df = rbind(group_to_samples_df, data.frame(group="scipher.complete.responder", sample_id=scipher_complete_responder))
group_to_samples_df = rbind(group_to_samples_df, data.frame(group="scipher.complete.nonresponder", sample_id=scipher_complete_nonresponder))
group_to_samples_df = rbind(group_to_samples_df, data.frame(group="scipher.sample.per.patient.all.visits", sample_id=sample_per_patient_all_visits))
group_to_samples_df = rbind(group_to_samples_df, data.frame(group="scipher.sample.per.patient.baseline", sample_id=sample_per_patient_baseline))
group_to_samples_df = rbind(group_to_samples_df, data.frame(group="scipher.sample.per.responder.baseline", sample_id=sample_per_responder_baseline))
group_to_samples_df = rbind(group_to_samples_df, data.frame(group="scipher.sample.per.nonresponder.baseline", sample_id=sample_per_nonresponder_baseline))
group_to_samples_file = paste(scipher_data_dir, "00_data/scipher_group_to_samples.csv", sep='/')
fwrite(group_to_samples_df, group_to_samples_file)
```

Each group has the following number of samples:

* `scipher.complete.dataset` (all samples, no matter the visit or batch): `r length(scipher_complete_dataset)`.
* `scipher.complete.responder` (all samples at any visit where the patient responds to the treatment): `r length(scipher_complete_responder)`.
* `scipher.complete.nonresponder` (all samples at any visit where the patient does not respond to the treatment): `r length(scipher_complete_nonresponder)`.
* `scipher.sample.per.patient.all.visits` (one sample per patient at different visits): `r length(sample_per_patient_all_visits)`.
* `scipher.sample.per.patient.baseline` (one sample per patient at baseline): `r length(sample_per_patient_baseline)`.
* `scipher.sample.per.responder.baseline` (one sample per responder patient at baseline): `r length(sample_per_responder_baseline)`.
* `scipher.sample.per.nonresponder.baseline` (one sample per non-responder patient at baseline): `r length(sample_per_nonresponder_baseline)`.


## Gene pre-processing

### Filter genes without classification

We filter the genes in the matrix keeping only the ones that have been properly annotated and classified in the HGNC database:

```{r}
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_category_file = paste(gene_annotations_dir, 'hgnc_category_set_2022-01-01.txt', sep='/')
gene_category_df = fread(gene_category_file)
gene_category_df$HGNC_Symbol = toupper(gene_category_df$HGNC_Symbol)
head(gene_category_df)
```

```{r}
genes_with_category = genes_dataset[genes_dataset %in% gene_category_df$HGNC_Symbol]
genes_without_category = genes_dataset[!(genes_dataset %in% gene_category_df$HGNC_Symbol)]
genes_dataset_df = gene_category_df %>% filter(HGNC_Symbol %in% genes_with_category)
head(genes_dataset_df)
```

```{r}
table(genes_dataset_df$HGNC_category)
round(prop.table(table(genes_dataset_df$HGNC_category))*100,2)
```

From the initial `r length(genes_dataset)` genes, there are `r length(genes_with_category)` with category and `r length(genes_without_category)` without category.


### Filter genes with low counts

We filter the genes that have low expression in all samples, keeping the ones that meet these rules:

* Sum of the expression of all samples above 0.
* Mean of the expression of all samples above 5.
* Median of the expression of all samples above 5.
* Standard deviation of all samples above 0.
* More than half of the genes with counts above 0.

```{r}
# Merge data from multiple batches with data from unique patients
RNA_metadata_merged = rbind((RNA_metadata %>% select(-batch)), (RNA_metadata_unique_patients %>% rename("Scipher_id.visit.batch" = "Scipher_id.visit"))) 

# Get only the gene expression
RNA = RNA_metadata_merged %>% select(genes_dataset_df$HGNC_Symbol)
genes = colnames(RNA)
sample_ids = RNA_metadata_merged$Scipher_id.visit.batch

# Transpose the matrix
RNA = t(as.matrix(RNA)) %>% as.data.frame()
colnames(RNA) = sample_ids
RNA = cbind(data.frame(Gene=genes), RNA)
RNA[1:5,1:5]
```

```{r}
rnaseq_filt_dir = paste(scipher_data_dir, '00_data/rnaseq_filtered_files_by_group', sep='/')
genes_filt_dir = paste(scipher_data_dir, '00_data/genes_from_rnaseq_filtered_files_by_group', sep='/')
dir.create(rnaseq_filt_dir, showWarnings = FALSE)
dir.create(genes_filt_dir, showWarnings = FALSE)

cols = c("group", "n_genes_raw", "n_genes_processed", "sum",  "mean", "median", "sd", "sum_zero_counts")
statistics_low_counts_df <- setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)

for (group_selected in unique(group_to_samples_df$group)){
  
  # Filter RNA matrix by samples of the group
  samples_group = (group_to_samples_df %>% filter(group == group_selected))$sample_id
  RNA_group = RNA %>% select(c("Gene", all_of(samples_group)))

  # Calculate sum, mean, median and standard deviation
  info_df = data.frame(Gene=RNA_group$Gene, sum=rowSums(RNA_group[,-1]))
  info_df$sd = apply(RNA_group[,-1], 1, sd)
  info_df$mean = rowMeans(RNA_group[,-1])
  info_df$median = apply(RNA_group[,-1], 1, median)
  info_df$zero_sum_counts = as.numeric(apply(RNA_group[,-1], 1, function(i) sum(i == 0) )) # Number of times that the counts are equal to 0
  
  # We keep the genes following all criteria
  info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 5) & (info_df$median > 5) & (info_df$sd > 0) & (info_df$zero_sum_counts < dim(RNA_group[, -1])[2]/2) ,]
  
  # Write rnaseq file
  RNA_group_filt = RNA_group[RNA_group$Gene %in% info_filt_df$Gene,]
  RNA_group_filt_file = paste(rnaseq_filt_dir, paste('scipher_rnaseq_', group_selected, ".csv", sep=''), sep='/')
  fwrite(RNA_group_filt, RNA_group_filt_file)

  # Write gene info file
  genes_dataset_filt_file = paste(genes_filt_dir, paste('scipher_rnaseq_gene_info_', group_selected, ".csv", sep=''), sep='/')
  genes_dataset_filt_df = data.frame(genes_dataset_df)
  genes_dataset_filt_df$enough_counts = ifelse(genes_dataset_filt_df$HGNC_Symbol %in% info_filt_df$Gene, TRUE, FALSE)
  fwrite(genes_dataset_filt_df, genes_dataset_filt_file)
  
  # Count statistics about the filter
  statistics_low_counts_df = rbind(statistics_low_counts_df, data.frame(group=group_selected, n_genes_raw=length(RNA_group$Gene), n_genes_processed=length(RNA_group_filt$Gene), sum=length(info_df[(info_df$sum > 0),]$Gene), mean=length(info_df[(info_df$mean > 5),]$Gene), median=length(info_df[(info_df$median > 5),]$Gene), sd=length(info_df[(info_df$sd > 0),]$Gene), sum_zero_counts=length(info_df[(info_df$sum_zero_counts < dim(RNA_group[, -1])[2]/2),]$Gene)))
}

statistics_low_counts_file = paste(scipher_data_dir, "00_data/scipher_rnaseq_low_counts_statistics.csv", sep='/')
fwrite(statistics_low_counts_df, statistics_low_counts_file)
```

These are the different number of genes after filtering by the criteria mentioned above:

```{r}
statistics_low_counts_df
```


## Prepare disease genes information dataset

Read the disease-gene associations dataset:

```{r}
gene_associations_dir = '/work/ccnr/j.aguirreplans/data/gene_associations/data/out'
disease_genes_file = paste(gene_associations_dir, 'GDA_Filtered_04042022.csv', sep='/')
GDA = fread(disease_genes_file) %>% unique()
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$NewName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes)

Prepare the disease-gene associations for the Scipher dataset: 

```{r}
disease_genes_info_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/disease_genes/disease_genes_info_2022_scipher.csv"

GDA[is.na(GDA)]<-0
GDA$NewName = tolower(GDA$NewName)
GDA$DescriptorName = tolower(GDA$DescriptorName)
GDA$MESHID = toupper(GDA$MESHID)
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
GDA %<>% 
  filter(Strong > 0 | 
           Weak > 0 |
           Incompatible > 0) %>%
  mutate(DiseaseName = NewName) %>%
  filter(DescriptorName %ni% c("behavioral disciplines and activities")) %>%
  mutate(MESHID_remove = stringr::str_detect(TreeNumber, pattern = "C22|C25|F04")) %>%
  mutate(MESHID_remove = ifelse(TreeNumber == "F01", TRUE,MESHID_remove)) %>%
  filter(!(MESHID_remove)) %>%
  select(DiseaseName, HGNC_Symbol, DescriptorName) %>% 
  unique() %>% 
  mutate(gene_in_dataset = ifelse(HGNC_Symbol %in% RNA$Gene, TRUE, FALSE))

# Make columns without special characters  
GDA$DiseaseName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DiseaseName))))
GDA$DescriptorName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DescriptorName))))

# Write output file
GDA %>% fwrite(disease_genes_info_file)

head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$DiseaseName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes), and `r nrow((GDA %>% filter(gene_in_dataset == TRUE)))` disease-gene associations in the Scipher dataset (`r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$DiseaseName))` diseases, `r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$HGNC_Symbol))` genes).


## Prepare essential genes dataset

Prepare the essential genes of the Scipher dataset:

```{r}
# Read the essential genes file
essential_genes_dir = '/work/ccnr/j.aguirreplans/Databases/OGEE'
essential_genes_original_file = paste(essential_genes_dir, 'OGEE_esential_genes_20190416.txt', sep='/')
# Filter by human genes
essential_genes_df = fread(essential_genes_original_file, header=TRUE) %>% rename("taxID"="#taxID")
head(essential_genes_df)
```

```{r}
# Read the gene annotations dictionary
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_dict_file = paste(gene_annotations_dir, 'multiple_hgnc_dict_set_2022-01-01.txt', sep='/')
gene_dict_df = fread(gene_dict_file) %>% unique()
gene_dict_df = gene_dict_df[!((gene_dict_df$HGNC_Symbol=="")|(gene_dict_df$alias=="")),]
gene_dict_df$HGNC_Symbol = toupper(gene_dict_df$HGNC_Symbol)
gene_dict_df$alias = toupper(gene_dict_df$alias)
gene_dict_df %<>% unique()
head(gene_dict_df)
```

```{r}
# Map ENSEMBL IDs to HGNC symbols
essential_genes_df = essential_genes_df %>% filter((taxID == 9606) & (essential=="E"))
essential_genes_df = essential_genes_df %>% inner_join(gene_dict_df, by=c("locus"="alias")) %>% select(taxID, sciName, HGNC_Symbol) %>% unique()
essential_genes_df$gene_in_dataset = ifelse(essential_genes_df$HGNC_Symbol %in% RNA$Gene, TRUE, FALSE)
head(essential_genes_df)
essential_genes_scipher_file = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/essential_genes/OGEE_essential_genes_scipher.csv'
essential_genes_df %>% fwrite(essential_genes_scipher_file)
```

There are `r nrow(essential_genes_df)` essential genes, `r nrow((essential_genes_df %>% filter(gene_in_dataset == TRUE)))` in the Scipher dataset.

