---
title: "Pre-process TCGA"
author: "Joaquim Aguirre-Plans"
date: "3/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

Pre-process TCGA data before creating gene co-expression networks.

```{r, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
require(magrittr)
library(tidyr)
library(igraph)
set.seed(1510)
`%ni%` <- Negate(`%in%`)
```

## Define variables

```{r define_files}
# Define type of counts used
type_counts = "reads" # tpm, reads
if(!(type_counts %in% c("tpm", "reads"))){
  stop("Unknown type of data")
}

# Define working directories
gdc_dir = '/work/ccnr/j.aguirreplans/Databases/TCGA/2022-11-18-Dataset/TCGA'
output_dir = paste(gdc_dir, 'out', type_counts, sep='/')
metadata_file = paste(gdc_dir, 'raw/metadata/metadata.txt', sep='/')

# Define the input file
tcga_processed_file = paste(output_dir, 'TCGA_genes_and_samples_processed.csv', sep='/')
```

## Read files

```{r}
RNAseq = fread(tcga_processed_file)
```

The TCGA dataframe contains `r as.integer(ncol(RNAseq)-1)` columns (case IDs) and `r nrow(RNAseq)` rows (genes).

```{r}
head(RNAseq[1:10,1:5])
```

```{r}
metadata_df = fread(metadata_file)
metadata_df %>% select(participant_sample_type, participant, sample_type_simplified, project_id, project_disease_type) %>% unique() %>% group_by(project_id, project_disease_type, sample_type_simplified) %>% summarize(n_samples=n_distinct(participant_sample_type), n_subjects=n_distinct(participant))
```

## Prepare groups of samples

We have the following groups of samples:

```{r}
# Include tumor and normal sample groups
sample_groups_df = metadata_df %>% select(participant_sample_type, project_id, sample_type_simplified, gender) %>% unique() %>% rename("sample_group_id" = "project_id", "sample_group_type" = "sample_type_simplified", "sex"="gender")

# Include normal samples matched to tissues
sample_groups_df = rbind(sample_groups_df, (metadata_df %>% select(participant_sample_type, tissue_simplified, sample_type_simplified, gender) %>% filter(sample_type_simplified == "normal") %>% select(-sample_type_simplified) %>% unique() %>% rename("sample_group_id" = "tissue_simplified", "sex"="gender") %>% mutate(sample_group_type = "tissue")))

# Include subtype samples
sample_groups_df = rbind(sample_groups_df, (metadata_df %>% select(participant_sample_type, cancer_subtype, sample_type_simplified, gender) %>% filter(sample_type_simplified == "tumor")  %>% filter(!(cancer_subtype == "")) %>% select(-sample_type_simplified) %>% unique() %>% rename("sample_group_id" = "cancer_subtype", "sex"="gender") %>% mutate(sample_group_type = "subtype"))) %>% arrange(sample_group_type, sample_group_id, participant_sample_type)

# Save file
sample_groups_file = paste(gdc_dir, paste('raw/metadata/sample_groups.txt', sep=""), sep='/')
fwrite(sample_groups_df, sample_groups_file, sep='\t')

head(sample_groups_df)
tail(sample_groups_df)
```

- Tumor types: `r length(unique((sample_groups_df %>% filter(sample_group_type == "tumor"))$sample_group_id))` types and `r length(unique((sample_groups_df %>% filter(sample_group_type == "tumor"))$participant_sample_type))` samples.
- Tumor subtypes: `r length(unique((sample_groups_df %>% filter(sample_group_type == "subtype"))$sample_group_id))` types and `r length(unique((sample_groups_df %>% filter(sample_group_type == "subtype"))$participant_sample_type))` samples.
- Normal tissue types matched to tumors: `r length(unique((sample_groups_df %>% filter(sample_group_type == "normal"))$sample_group_id))` types and `r length(unique((sample_groups_df %>% filter(sample_group_type == "normal"))$participant_sample_type))` samples.
- Normal tissue types matched to tissues: `r length(unique((sample_groups_df %>% filter(sample_group_type == "tissue"))$sample_group_id))` types and `r length(unique((sample_groups_df %>% filter(sample_group_type == "tissue"))$participant_sample_type))` samples.


## Filter genes without classification

We keep genes that have a classification in HGNC:

```{r}
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_category_file = paste(gene_annotations_dir, 'hgnc_category_set_2022-01-01.txt', sep='/')
gene_category_df = fread(gene_category_file)
gene_category_df$HGNC_Symbol = toupper(gene_category_df$HGNC_Symbol)
head(gene_category_df)
```

We use this dataframe to map gene category, and we only keep genes that are in this dataframe:

```{r}
dim(RNAseq)
genes_with_category = RNAseq$Gene[RNAseq$Gene %in% gene_category_df$HGNC_Symbol]
genes_dataset_df = gene_category_df %>% filter(HGNC_Symbol %in% genes_with_category)
RNAseq %<>% filter(Gene %in% genes_with_category)
dim(RNAseq)
```

The TCGA dataframe contains `r as.integer(ncol(RNAseq)-1)` case IDs and `r nrow(RNAseq)` genes with category.

It contains the following number of genes for each category:

```{r}
table(genes_dataset_df$HGNC_category)
round(prop.table(table(genes_dataset_df$HGNC_category))*100,2)
```

## Filter genes with empty counts in half of the samples or more

We only keep the genes that have empty counts in less than half of the samples.

```{r}
for(sample_group_type_selected in c("tumor", "normal")){
  # We separate samples depending on sample type
  sample_group_type_dir = paste(output_dir, sample_group_type_selected, sep='/')
  filter_zeros_dir = paste(sample_group_type_dir, 'filter_genes_with_zeros', sep='/')
  dir.create(sample_group_type_dir, showWarnings = FALSE)
  dir.create(filter_zeros_dir, showWarnings = FALSE)
  samples_selected_by_type = unique((sample_groups_df %>% filter(sample_group_type == sample_group_type_selected))$participant_sample_type)
  RNAseq_by_type = RNAseq %>% select(c("Gene", all_of(samples_selected_by_type)))
  # We calculate the number of times that a gene has 0 counts
  info_df = data.frame(Gene=RNAseq_by_type$Gene, sum=rowSums(RNAseq_by_type[,-1]))
  info_df$sum_zero_counts = as.numeric(apply(RNAseq_by_type[,-1], 1, function(i) sum(i == 0) )) # Number of times that the counts are equal to 0
  # We keep genes that have empty counts in less than half of the samples
  info_filt_df = info_df[(info_df$sum_zero_counts < dim(RNAseq_by_type[, -1])[2]/2) ,]
  print(paste("Sample type:", sample_group_type_selected, "Number of genes with empty counts in less than half of samples", length(info_filt_df$Gene)))
  # We filter the gene expression
  RNAseq_filt_zeros = RNAseq_by_type %>% filter(Gene %in% info_filt_df$Gene)
  # We write final file
  tcga_final_file = paste(filter_zeros_dir, paste('tcga_rnaseq_TCGA.csv', sep=""), sep='/')
  fwrite(RNAseq_filt_zeros, tcga_final_file, sep='\t')
  print(paste("Sample type:", sample_group_type_selected, "Number of columns (samples):", as.integer(ncol(RNAseq_filt_zeros)-1), "Number of rows (genes):", nrow(RNAseq_filt_zeros)))
  # Write the information about the selected genes of the dataset in a separate file
  genes_dataset_df$enough_counts = ifelse(genes_dataset_df$HGNC_Symbol %in% info_filt_df$Gene, TRUE, FALSE)
  genes_dataset_file = paste(filter_zeros_dir, 'rnaseq_gene_info_TCGA.csv', sep='/')
  fwrite(genes_dataset_df, genes_dataset_file)
  # Check the percentage of genes in each category after filtering by counts
  genes_dataset_filtered_df = genes_dataset_df %>% filter(HGNC_Symbol %in% info_filt_df$Gene) %>% unique
  table(genes_dataset_filtered_df$HGNC_category)
  round(prop.table(table(genes_dataset_filtered_df$HGNC_category))*100,2)
}
rm(RNAseq_by_type)
rm(RNAseq_filt_zeros)
```


## Filter genes with low counts

We filter the genes that have low expression in all samples, keeping the ones that meet these rules:

In the case of TPM:

* Sum of the expression of all samples above 0.
* Mean of the expression of all samples above 1.
* Median of the expression of all samples above 1.
* Standard deviation of all samples above 0.
* Less than half of the samples with empty counts (0s).

In the case of reads:

* Sum of the expression of all samples above 0.
* Mean of the expression of all samples above 5.
* Median of the expression of all samples above 5.
* Standard deviation of all samples above 0.
* Less than half of the samples with empty counts (0s).

```{r}
cols = c("sample_group_id", "sex", "sample_group_type", "n_genes_raw", "n_genes_processed", "sum",  "mean", "median", "sd", "sum_zero_counts")
statistics_low_counts_df <- setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)
for(sample_group_type_selected in c("tumor", "normal")){
  # We separate samples depending on sample type
  sample_group_type_dir = paste(output_dir, sample_group_type_selected, sep='/')
  filter_low_counts_dir = paste(sample_group_type_dir, 'filter_genes_low_counts', sep='/')
  dir.create(sample_group_type_dir, showWarnings = FALSE)
  dir.create(filter_low_counts_dir, showWarnings = FALSE)
  samples_selected_by_type = unique((sample_groups_df %>% filter(sample_group_type == sample_group_type_selected))$participant_sample_type)
  RNAseq_by_type = RNAseq %>% select(c("Gene", all_of(samples_selected_by_type)))
  # We calculate the number of times that a gene has 0 counts
  info_df = data.frame(Gene=RNAseq_by_type$Gene, sum=rowSums(RNAseq_by_type[,-1]))
  info_df$sd = apply(RNAseq_by_type[,-1], 1, sd)
  info_df$mean = rowMeans(RNAseq_by_type[,-1])
  info_df$median = apply(RNAseq_by_type[,-1], 1, median)
  info_df$sum_zero_counts = as.numeric(apply(RNAseq_by_type[,-1], 1, function(i) sum(i == 0) )) # Number of times that the counts are equal to 0
  # We keep the genes following all criteria
  if(type_counts == "tpm"){
    info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 1) & (info_df$median > 1) & (info_df$sd > 0) & (info_df$sum_zero_counts < dim(RNAseq_by_type[, -1])[2]/2) ,]
  } else if(type_counts == "reads"){
    info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 5) & (info_df$median > 5) & (info_df$sd > 0) & (info_df$sum_zero_counts < dim(RNAseq_by_type[, -1])[2]/2) ,]
  } else {
    stop("Unknown type of data")
  }
  # We filter the gene expression
  RNAseq_filt_low_counts = RNAseq_by_type %>% filter(Gene %in% info_filt_df$Gene)
  # We write final file
  tcga_final_file = paste(filter_low_counts_dir, 'tcga_rnaseq_TCGA.csv', sep='/')
  fwrite(RNAseq_filt_low_counts, tcga_final_file, sep='\t')
  print(paste("Sample type:", sample_group_type_selected, "Number of columns (samples):", as.integer(ncol(RNAseq_filt_low_counts)-1), "Number of rows (genes):", nrow(RNAseq_filt_low_counts)))
  # Write the information about the selected genes of the dataset in a separate file
  genes_dataset_df$enough_counts = ifelse(genes_dataset_df$HGNC_Symbol %in% info_filt_df$Gene, TRUE, FALSE)
  genes_dataset_file = paste(filter_low_counts_dir, 'rnaseq_gene_info_TCGA.csv', sep='/')
  fwrite(genes_dataset_df, genes_dataset_file)
  # Check the percentage of genes in each category after filtering by counts
  genes_dataset_filtered_df = genes_dataset_df %>% filter(HGNC_Symbol %in% info_filt_df$Gene) %>% unique
  table(genes_dataset_filtered_df$HGNC_category)
  round(prop.table(table(genes_dataset_filtered_df$HGNC_category))*100,2)
  # Count statistics about the filter
  statistics_low_counts_df = rbind(statistics_low_counts_df, data.frame(sample_group_id="TCGA", sample_group_type=sample_group_type_selected, n_genes_raw=length(RNAseq_by_type$Gene), n_genes_processed=length(RNAseq_filt_low_counts$Gene), sum=length(info_df[(info_df$sum > 0),]$Gene), mean=length(info_df[(info_df$mean > 1),]$Gene), median=length(info_df[(info_df$median > 1),]$Gene), sd=length(info_df[(info_df$sd > 0),]$Gene), sum_zero_counts=length(info_df[(info_df$sum_zero_counts < dim(RNAseq_by_type[, -1])[2]/2),]$Gene)))
}
rm(RNAseq_by_type)
rm(RNAseq_filt_low_counts)
```

```{r}
statistics_low_counts_df
```


## Filtering the genes with low counts by type of sample and sex

We repeat the same procedure but specific for the different types of samples:

```{r}
cols = c("sample_group_id", "sex", "sample_group_type", "n_genes_raw", "n_genes_processed", "sum",  "mean", "median", "sd", "sum_zero_counts")
statistics_low_counts_df <- setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)
sex_types = c("male","female", "")
for(sample_group_type_selected in unique(sample_groups_df$sample_group_type)){
  sample_group_type_dir = paste(output_dir, sample_group_type_selected, sep='/')
  filter_low_counts_dir = paste(sample_group_type_dir, 'filter_genes_low_counts', sep='/')
  dir.create(sample_group_type_dir, showWarnings = FALSE)
  dir.create(filter_low_counts_dir, showWarnings = FALSE)
  sample_group_ids = unique((sample_groups_df %>% filter(sample_group_type == sample_group_type_selected))$sample_group_id)
  for (sample_group_id_selected in sample_group_ids){
    rnaseq_filt_dir = paste(filter_low_counts_dir, 'rnaseq_filtered_files_by_sample_group', sep='/')
    genes_filt_dir = paste(filter_low_counts_dir, 'genes_filtered_files_by_sample_group', sep='/')
    dir.create(rnaseq_filt_dir, showWarnings = FALSE)
    dir.create(genes_filt_dir, showWarnings = FALSE)

    for (sex_selected in sex_types){
      
      # Get the samples of the selected tissue/sex
      if (sex_selected == ""){
        samples_selected = unique((sample_groups_df %>% filter((sample_group_type == sample_group_type_selected) & (sample_group_id == sample_group_id_selected)))$participant_sample_type)
        rnaseq_group_file = paste(rnaseq_filt_dir, paste("tcga_rnaseq_", sample_group_id_selected, ".csv", sep=''), sep='/')
        genes_group_file = paste(genes_filt_dir, paste("rnaseq_gene_info_", sample_group_id_selected, ".csv", sep=''), sep='/')
      } else {
        samples_selected = unique((sample_groups_df %>% filter((sample_group_type == sample_group_type_selected) & (sample_group_id == sample_group_id_selected) & (sex == sex_selected)))$participant_sample_type)
        rnaseq_group_file = paste(rnaseq_filt_dir, paste("tcga_rnaseq_", sample_group_id_selected, "_", sex_selected, ".csv", sep=''), sep='/')
        genes_group_file = paste(genes_filt_dir, paste("rnaseq_gene_info_", sample_group_id_selected, "_", sex_selected, ".csv", sep=''), sep='/')
      }

      if(length(samples_selected) >= 10){
        # Select expression from selected samples
        RNAseq_group = RNAseq %>% select(c("Gene", all_of(samples_selected)))
        # Calculate metrics of expression
        info_df = data.frame(Gene=RNAseq_group$Gene, sum=rowSums(RNAseq_group[,-1]))
        info_df$sd = apply(RNAseq_group[,-1], 1, sd)
        info_df$mean = rowMeans(RNAseq_group[,-1])
        info_df$median = apply(RNAseq_group[,-1], 1, median)
        info_df$sum_zero_counts = as.numeric(apply(RNAseq_group[,-1], 1, function(i) sum(i == 0) )) # Number of times that the counts are equal to 0
        # Filter low expressed genes
        if(type_counts == "tpm"){
          info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 1) & (info_df$median > 1) & (info_df$sd > 0) & (info_df$sum_zero_counts < dim(RNAseq_group[, -1])[2]/2) ,]
        } else if(type_counts == "reads"){
          info_filt_df = info_df[(info_df$sum > 0) & (info_df$mean > 5) & (info_df$median > 5) & (info_df$sd > 0) & (info_df$sum_zero_counts < dim(RNAseq_group[, -1])[2]/2) ,]
        } else {
          stop("Unknown type of data")
        }
        RNAseq_group_filt = RNAseq_group[RNAseq_group$Gene %in% info_filt_df$Gene,]
        fwrite(RNAseq_group_filt, rnaseq_group_file)
        # Count statistics about the filter
        statistics_low_counts_df = rbind(statistics_low_counts_df, data.frame(sample_group_id=sample_group_id_selected, sex=sex_selected, n_genes_raw=length(RNAseq_group$Gene), n_genes_processed=length(RNAseq_group_filt$Gene), sum=length(info_df[(info_df$sum > 0),]$Gene), mean=length(info_df[(info_df$mean > 1),]$Gene), median=length(info_df[(info_df$median > 1),]$Gene), sd=length(info_df[(info_df$sd > 0),]$Gene), sum_zero_counts=length(info_df[(info_df$sum_zero_counts < dim(RNAseq_group[, -1])[2]/2),]$Gene)))
        # Write gene info file
        genes_dataset_filt_group_df = data.frame(genes_dataset_df)
        genes_dataset_filt_group_df$enough_counts = ifelse(genes_dataset_filt_group_df$HGNC_Symbol %in% info_filt_df$Gene, TRUE, FALSE)
        fwrite(genes_dataset_filt_group_df, genes_group_file)
      }
    }
  }
}
statistics_low_counts_file = paste(output_dir, "tcga_rnaseq_group_low_counts_statistics.csv", sep='/')
fwrite(statistics_low_counts_df, statistics_low_counts_file)
```

These are the different number of genes after filtering by the criteria mentioned above:

```{r}
head(statistics_low_counts_df)
```


## Prepare disease genes information dataset

Read the disease-gene associations dataset:

```{r}
gene_associations_dir = '/work/ccnr/j.aguirreplans/data/gene_associations/data/out'
disease_genes_file = paste(gene_associations_dir, 'GDA_Filtered_04042022.csv', sep='/')
GDA = fread(disease_genes_file) %>% unique()
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$NewName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes)

Prepare the disease-gene associations for the TCGA dataset: 

```{r}
disease_genes_info_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/disease_genes/disease_genes_info_2022_tcga.csv"

GDA[is.na(GDA)]<-0
GDA$NewName = tolower(GDA$NewName)
GDA$DescriptorName = tolower(GDA$DescriptorName)
GDA$MESHID = toupper(GDA$MESHID)
GDA$HGNC_Symbol = toupper(GDA$HGNC_Symbol)
GDA %<>% 
  filter(Strong > 0 | 
           Weak > 0 |
           Incompatible > 0) %>%
  mutate(DiseaseName = NewName) %>%
  filter(DescriptorName %ni% c("behavioral disciplines and activities")) %>%
  mutate(MESHID_remove = stringr::str_detect(TreeNumber, pattern = "C22|C25|F04")) %>%
  mutate(MESHID_remove = ifelse(TreeNumber == "F01", TRUE,MESHID_remove)) %>%
  filter(!(MESHID_remove)) %>%
  select(DiseaseName, HGNC_Symbol, DescriptorName) %>% 
  unique() %>% 
  mutate(gene_in_dataset = ifelse(HGNC_Symbol %in% RNAseq$Gene, TRUE, FALSE))

# Make columns without special characters  
GDA$DiseaseName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DiseaseName))))
GDA$DescriptorName.no.sp.char <- gsub(' ', '.', gsub(', ', '.', gsub('-', '.', gsub('[\\(\\)]', '', GDA$DescriptorName))))

# Write output file
GDA %>% fwrite(disease_genes_info_file)

head(GDA)
```

There are `r nrow(GDA)` disease-gene associations (`r length(unique(GDA$DiseaseName))` diseases, `r length(unique(GDA$HGNC_Symbol))` genes), and `r nrow((GDA %>% filter(gene_in_dataset == TRUE)))` disease-gene associations in the TCGA dataset (`r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$DiseaseName))` diseases, `r length(unique((GDA %>% filter(gene_in_dataset == TRUE))$HGNC_Symbol))` genes).

## Prepare essential genes dataset

Prepare the essential genes of the TCGA dataset:

```{r}
# Read the essential genes file
essential_genes_dir = '/work/ccnr/j.aguirreplans/Databases/OGEE'
essential_genes_original_file = paste(essential_genes_dir, 'OGEE_esential_genes_20190416.txt', sep='/')
# Filter by human genes
essential_genes_df = fread(essential_genes_original_file, header=TRUE) %>% rename("taxID"="#taxID")
head(essential_genes_df)
```

```{r}
# Read the gene annotations dictionary
gene_annotations_dir = '/work/ccnr/j.aguirreplans/data/gene_annotations/data/out'
gene_dict_file = paste(gene_annotations_dir, 'multiple_hgnc_dict_set_2022-01-01.txt', sep='/')
gene_dict_df = fread(gene_dict_file) %>% unique()
gene_dict_df = gene_dict_df[!((gene_dict_df$HGNC_Symbol=="")|(gene_dict_df$alias=="")),]
gene_dict_df$HGNC_Symbol = toupper(gene_dict_df$HGNC_Symbol)
gene_dict_df$alias = toupper(gene_dict_df$alias)
gene_dict_df %<>% unique()
head(gene_dict_df)
```

```{r}
# Map ENSEMBL IDs to HGNC symbols
essential_genes_df = essential_genes_df %>% filter((taxID == 9606) & (essential=="E"))
essential_genes_df = essential_genes_df %>% inner_join(gene_dict_df, by=c("locus"="alias")) %>% select(taxID, sciName, HGNC_Symbol) %>% unique()
essential_genes_df$gene_in_dataset = ifelse(essential_genes_df$HGNC_Symbol %in% RNAseq$Gene, TRUE, FALSE)
head(essential_genes_df)
essential_genes_tcga_file = '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/essential_genes/OGEE_essential_genes_tcga.csv'
essential_genes_df %>% fwrite(essential_genes_tcga_file)
```

There are `r nrow(essential_genes_df)` essential genes, `r nrow((essential_genes_df %>% filter(gene_in_dataset == TRUE)))` in the TCGA dataset.
