---
title: "Create Figures"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Description

Create figures for the manuscript, extended abstract and poster.

```{r}
#packrat::init("/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeR")
```

```{r}
library(data.table)
library(dplyr)
library(tidyr)
library(igraph)
library(ggplot2)
library(ggalluvial)
library(gghalves)
library(ggnewscale)
require(ggrepel)
library(grid)
library(gridExtra)
library(gtable)
require(magrittr)
library(patchwork)
set.seed(1510)
options(bitmapType='cairo')
`%ni%` <- Negate(`%in%`)
```

### Load data

Un-normalized data:

```{r}
input_dir <- '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data'
plots_dir <- '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots'
tables_dir <- '/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables'
font_selected <- "Helvetica"
numbers_file <- paste(input_dir, 'dataset_numbers_complete_graph.txt', sep='/')
numbers_df <- fread(numbers_file)
numbers_df$dataset <- tolower(numbers_df$dataset)
meta_file <- paste(input_dir, "metadata_summary.txt", sep="/")
meta_df <- fread(meta_file)
sd_genes_file <- paste(input_dir, "variation_by_gene.txt", sep="/") # from => calculate_rnaseq_datasets_variation.Rmd
sd_genes_df <- fread(sd_genes_file)
my_theme <- theme_linedraw() +
  theme(aspect.ratio = 1, 
        plot.title =  element_text(size = 20, face = "bold"), 
        axis.title = element_text(size = 17, face = "bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        legend.position = "none")

name2color <- c(
  "TCGA: Lung cancer" = "#56B4E9",
  "tcga:tcga-luad" = "#56B4E9",
  "tcga:tcga-luad-tumor" = "#56B4E9", # #D55E00
  "TCGA: Breast cancer" = "#0072B2",
  "TCGA: Breast c." = "#0072B2",
  "tcga:tcga-brca_female" = "#0072B2",
  "tcga:tcga-brca-tumor_female" = "#0072B2",
  "breast neoplasms"="#0072B2", 
  "gse193677" = "#E032EC",
  "GSE193677" = "#E032EC",
  "UC inflamed" = "#E032EC",
  "GTEx: Whole blood" = "#65F3BF",
  "gtex:whole.blood" = "#65F3BF",
  "GTEx: Lung" = "#24D157",
  "gtex:lung" = "#24D157",
  "GTEx: Breast" = "#00BA37", #44AA99
  "gtex:breast.mammary.tissue" = "#00BA37",
  "gtex:breast.mammary.tissue_female" = "#00BA37",
  "R. arthritis" = "#D55E00", #E69F00
  "scipher:scipher.sample.per.patient.baseline" = "#D55E00",
  "arthritis rheumatoid"="#D55E00", 
  "asthma"="#d9f365", 
  "thyroid neoplasms"="#0072B2",
  "tcga" = "#0072B2",
  "TCGA" = "#0072B2",
  "gtex" = "#00BA37",
  "GTEx" = "#00BA37",
  "scipher" = "#D55E00",
  #"Analytical model"="#e3f705",
  "Analytical model"="#cc68f7",
  "Power law"="red", #cc68f7
  "Logarithm"="#09b863",
  "Exponential decay"="orange", #09b863
  "Constant discovery"="grey",
  "\u2265 0.2" = "#F8766D",
  "\u2265 0.4" = "#7CAE00",
  "\u2265 0.6" = "#00BFC4",
  "\u2265 0.8" = "#C77CFF",
  "Very Weak" = "#ff7b00",
  "Weak" = "#F8766D",
  "Moderate" = "#7CAE00",
  "Strong" = "#00BFC4",
  "Very Strong" = "#C77CFF",
  "Convergence point" = "red",
  "Model prediction" = "red",
  "P-value < 0.05" = "red",
  "common" = "#619CFF",
  "disease-gene" = "#cc68f7",
  "disease-specific" = "#F8766D",
  "normal-specific" = "#00BA38",
  "undefined" = "#808080",
  "different" = "#C77CFF",
  "all" = "#CD9600",
  "common (constant)" = "#619CFF",
  "common (change)" = "#b5d1ff",
  "disease-specific (constant)" = "#F8766D",
  "disease-specific (change)" = "#ffcbc7",
  "normal-specific (constant)" = "#00BA38",
  "normal-specific (change)" = "#a3d1b1",
  "undefined (constant)" = "#808080",
  "undefined (change)" = "#c2c2c2"
  )

dataset2name <- c(
  "gtex:breast.mammary.tissue_female" = "GTEx: Breast",
  "gtex:whole.blood" = "GTEx: Whole blood",
  "gtex:lung" = "GTEx: Lung",
  "tcga:tcga-brca_female" = "TCGA: Breast cancer",
  "tcga:tcga-brca-tumor_female" = "TCGA: Breast cancer",
  "tcga:tcga-luad" = "TCGA: Lung cancer",
  "scipher:scipher.sample.per.patient.baseline" = "R. arthritis",
  "gse193677:rectum_uc_inflamed" = "UC inflamed",
  "weak-moderate-strong-very strong" = "\u2265 0.2",
  "moderate-strong-very strong" = "\u2265 0.4",
  "strong-very strong" = "\u2265 0.6",
  "very strong" = "\u2265 0.8"
)

input_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data/example_pearson_pval_0.05'
topology_results_file = paste(input_dir, 'topology_results_pearson_pval_0.05.txt', sep='/')
results_selected_df <- data.table::fread(topology_results_file)
topology_results_by_size_file = paste(input_dir, 'topology_results_mean_pearson_pval_0.05.txt', sep='/')
topology_results_selected_by_size_df <- data.table::fread(topology_results_by_size_file)
predictions_file = paste(input_dir, 'predictions_pearson_pval_0.05.txt', sep='/')
predicted_results_df <- data.table::fread(predictions_file)
analytical_results_file = paste(input_dir, 'analytical_model_results_pearson_pval_0.05.txt', sep='/')
topology_results_selected_analytical_df <- data.table::fread(analytical_results_file)
analytical_regression_results_file = paste(input_dir, 'analytical_model_regression_results_pearson_pval_0.05.txt', sep='/')
stretched_exponential_regression_df <- data.table::fread(analytical_regression_results_file)
theoretical_sample_size_file = paste(input_dir, 'theoretical_sample_size_for_correlations_of_datasets.txt', sep='/') # from => calculate_convergence_correlation_types.Rmd
sample_size_correlation_df <- data.table::fread(theoretical_sample_size_file)
```

Normalized data:

```{r}
topology_type_normalization = "divide.L"
input_dir = '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data/example_pearson_pval_0.05'
topology_results_file = paste(input_dir, '/', 'topology_results_norm_', topology_type_normalization, '_pearson_pval_0.05.txt', sep='')
results_selected_norm_df <- data.table::fread(topology_results_file)
topology_results_by_size_file = paste(input_dir, '/', 'topology_results_mean_norm_', topology_type_normalization, '_pearson_pval_0.05.txt', sep='')
topology_results_selected_by_size_norm_df <- data.table::fread(topology_results_by_size_file)
predictions_file = paste(input_dir, '/', 'predictions_', topology_type_normalization, '_pearson_pval_0.05.txt', sep='')
predicted_results_norm_df <- data.table::fread(predictions_file)
analytical_results_file = paste(input_dir, '/', 'analytical_model_results_', topology_type_normalization, '_pearson_pval_0.05.txt', sep='')
topology_results_selected_analytical_norm_df <- data.table::fread(analytical_results_file)
```

Number of networks and maximum sample size:

```{r}
nrow((results_selected_df %>% 
        dplyr::select(method, type_dataset, size, rep) %>% 
        unique()))
max(((results_selected_df %>% 
        dplyr::select(method, type_dataset, size, rep) %>% 
        unique()))$size)
```

Number of networks and maximum sample size without counting TCGA full dataset:

```{r}
nrow((results_selected_df %>% 
        dplyr::select(method, type_dataset, size, rep) %>% 
        filter(!(type_dataset == "tcga:tcga")) %>% 
        unique()))
max(((results_selected_df %>% 
        dplyr::select(method, type_dataset, size, rep) %>% 
        filter(!(type_dataset == "tcga:tcga")) %>% 
        unique()))$size)
```

Read and process analytical model data:

```{r}
analytical_summary_file = paste(input_dir, 'analytical_model_summary_pearson_pval_0.05.txt', sep='/')
analytical_model_summary_df <- data.table::fread(analytical_summary_file)
analytical_model_summary_df <- analytical_model_summary_df %>% 
  inner_join((results_selected_norm_df %>%
                dplyr::select("type_dataset", "subclassification")), 
             by = c("type_dataset")) %>%
  unique()

analytical_model_summary_df$L_vs_total <- abs((analytical_model_summary_df$unnorm_L - analytical_model_summary_df$total_num_edges)) / analytical_model_summary_df$total_num_edges
analytical_model_summary_df$dataset_name <- analytical_model_summary_df$type_dataset

analytical_model_summary_df <- analytical_model_summary_df %>%
  tidyr::separate(col = "dataset_name", 
           into = c("dataset_name", "sex"), 
           sep = "_", 
           remove = TRUE) %>%
  tidyr::separate(col = "dataset_name", 
           into = c("dataset", NULL), 
           sep = ":", 
           remove = FALSE) %>%
  mutate(sex = if_else(is.na(sex), "", sex))

analytical_model_summary_df$dataset_name <- ifelse(
  !(analytical_model_summary_df$subclassification == ""),
  paste(
    analytical_model_summary_df$dataset_name, 
    analytical_model_summary_df$subclassification, 
    sep = "-"
  ),
  analytical_model_summary_df$dataset_name
)

analytical_model_summary_df$dataset_name <- ifelse(
  !(analytical_model_summary_df$sex == ""),
  paste(
    analytical_model_summary_df$dataset_name,
    analytical_model_summary_df$sex,
    sep = "_"
  ),
  analytical_model_summary_df$dataset_name
)
```

### Make plots

#### Curve plot

```{r}
datasets_selected <- c("gtex:breast.mammary.tissue_female", 
                      "tcga:tcga-brca_female")

results_curve_lessdatasets_df <- results_selected_norm_df %>%
  dplyr::filter(type_dataset %in% datasets_selected) %>%
  dplyr::select(size, rep, unnorm, type_dataset) %>%
  dplyr::mutate(type_dataset = dplyr::recode(type_dataset, !!!dataset2name)) %>%
  inner_join( (name2color %>% 
                 as.data.frame() %>% 
                 dplyr::rename("rgb_col"=".") %>% 
                 tibble::rownames_to_column("type_dataset")), 
              by=c("type_dataset") ) %>%
  dplyr::mutate(geom_text_repel_label = case_when(
    type_dataset == "GTEx: Breast" & size == 140 & rep == 1 ~ "GTEx: Breast",
    type_dataset == "TCGA: Breast cancer" & size == 600 & rep == 1 ~ "TCGA: Breast cancer",
    TRUE ~ ""
  ))

# Plot with labels
curve_plot_lessdatasets <- results_curve_lessdatasets_df %>%
  ggplot(aes(x=size, y=unnorm, label = geom_text_repel_label)) +
  geom_point(aes(col=rgb_col), alpha=0.5, size=3) +
  scale_colour_identity() +
  theme_linedraw() +
  xlab("Num. samples") +
  ylab("Num. significant correlations") +
  my_theme +
  geom_label_repel(
    aes(col = rgb_col),
    family = font_selected,
    fill = "white",
    max.overlaps = Inf,
    label.size = 0.75, 
    size = 5,
    alpha = 0.9, 
    box.padding = unit(0.75, "cm"),
    label.padding = 0.25, 
    fontface = "bold",
    na.rm = TRUE,
    seed = 1234
  )

plot_file = paste(plots_dir, "curve_pearson_pval_0.05_lessdatasets.png", sep="/")
ggsave(
  plot_file,
  plot=curve_plot_lessdatasets,
  dpi = 1200,
  #width = 10000,
  height = 6000,
  units = c("px")
)
print(curve_plot_lessdatasets)
```

#### Calculate power law, exponential decay and constant discovery

```{r calculate_pl_ed_cd}
datasets_selected <- c(
  "gse193677:rectum_uc_inflamed",
  "gtex:breast.mammary.tissue_female",
  "gtex:lung",
  "scipher:scipher.sample.per.patient.baseline",
  "tcga:tcga-luad",
  "tcga:tcga-brca_female"
)

cols = c("size", 
         "S", 
         "type_dataset", 
         "S_lag1", 
         "Fn", 
         "model",
         "model_result",
         "slope", 
         "intercept", 
         "adj.r.squared")

scaling_relation_comparison_df = data.frame(matrix(ncol = length(cols),
                                                   nrow = 0,
                                                   dimnames = list(NULL, cols)))

for (dataset_selected in datasets_selected){
  scaling_relation_comp = topology_results_selected_by_size_df %>% 
    filter(type_dataset == dataset_selected) %>%
    select(size, mean, type_dataset) %>%
    unique() %>%
    rename("S" = "mean") %>%
    arrange(size) %>%
    mutate(S_lag1 = if_else(size == min(size), 0, lag(S))) %>%
    mutate(Fn = ((S - S_lag1) / S_lag1)) %>%
    filter((!(is.infinite(Fn))) & (Fn > 0))

  # Calculate power law
  pl_summary = summary(lm(log(scaling_relation_comp$Fn)~log(scaling_relation_comp$size)))
  pl_pred = exp(coef(pl_summary)[1]) * scaling_relation_comp$size**(coef(pl_summary)[2])

  # Calculate exponential decay
  ed_summary = summary(lm(log(scaling_relation_comp$Fn)~scaling_relation_comp$size))
  ed_pred = exp(coef(ed_summary)[1]) * exp(coef(ed_summary)[2] * scaling_relation_comp$size)

  # Calculate constant discovery
  constant_discovery = rep(mean(scaling_relation_comp$Fn), length(scaling_relation_comp$size))
  #constant_discovery = rep(mean((scaling_relation_comp %>% dplyr::filter(size >= 120))$Fn), length(scaling_relation_comp$size))
  SS_res <- sum((scaling_relation_comp$Fn - constant_discovery)^2)
  mean_y <- mean(scaling_relation_comp$Fn)
  SS_tot <- sum((scaling_relation_comp$Fn - mean_y)^2)
  R_squared <- 1 - (SS_res / SS_tot)

  scaling_relation_comparison_df = rbind(scaling_relation_comparison_df, 
                                           cbind(scaling_relation_comp, 
                                                 data.frame(model = "Power law",
                                                            model_result = pl_pred),
                                                 data.frame(slope = coef(pl_summary)[2], 
                                                            intercept = coef(pl_summary)[1], 
                                                            adj.r.squared = pl_summary$adj.r.squared)))
  scaling_relation_comparison_df = rbind(scaling_relation_comparison_df,
                                           cbind(scaling_relation_comp, 
                                                 data.frame(model = "Exponential decay",
                                                            model_result = ed_pred),
                                                 data.frame(slope = coef(ed_summary)[2], 
                                                            intercept = coef(ed_summary)[1], 
                                                            adj.r.squared = ed_summary$adj.r.squared)))
  scaling_relation_comparison_df = rbind(scaling_relation_comparison_df,
                                           cbind(scaling_relation_comp, 
                                                 data.frame(model = "Constant discovery",
                                                            model_result = constant_discovery),
                                                 data.frame(slope = 0, 
                                                            intercept = constant_discovery[1], 
                                                            adj.r.squared = R_squared)))
}

# Calculate relative error
scaling_relation_comparison_df$relative.error <- abs((scaling_relation_comparison_df$Fn - scaling_relation_comparison_df$model_result)) / scaling_relation_comparison_df$Fn

# Calculate relative error mean
scaling_relation_comparison_df <- scaling_relation_comparison_df %>%
  dplyr::group_by(type_dataset, model) %>%
  dplyr::mutate(relative.error.mean = mean(relative.error)) %>%
  dplyr::ungroup()
```

#### Power law plot

```{r}
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "tcga:tcga-brca_female")

# Process results
power_law_lessdatasets_df <- scaling_relation_comparison_df %>%
  dplyr::filter(type_dataset %in% datasets_selected) %>%
  dplyr::filter(model == "Power law") %>%
  dplyr::mutate_at(c('adj.r.squared'), ~sprintf("%.2f",.)) %>%
  dplyr::mutate_at(c('adj.r.squared'), as.character) %>%
  dplyr::mutate(type_dataset = dplyr::recode(type_dataset, !!!dataset2name)) %>%
  dplyr::inner_join(
    (name2color %>%
       as.data.frame() %>%
       dplyr::rename("rgb_col" = ".") %>%
       tibble::rownames_to_column("type_dataset")),
    by = c("type_dataset") ) %>%
  dplyr::mutate(
    geom_text_repel_label = dplyr::case_when(
      type_dataset == "GTEx: Breast" & size == 140 ~ "GTEx: Breast",
      type_dataset == "TCGA: Breast cancer" & size == 400 ~ "TCGA: Breast cancer",
      TRUE ~ ""
    )
  )

power_law_lessdatasets_df$r2labels <- ifelse(
  power_law_lessdatasets_df$type_dataset == "GTEx: Breast", 
  paste(
    "GTEx Breast: ",
    unique((power_law_lessdatasets_df %>% dplyr::filter(type_dataset == "GTEx: Breast"))$adj.r.squared),
    sep = ""
  ), 
  ifelse(
    power_law_lessdatasets_df$type_dataset == "TCGA: Breast cancer",
    paste(
      "TCGA Breast c.: ",
      unique((power_law_lessdatasets_df %>% dplyr::filter(type_dataset == "TCGA: Breast cancer"))$adj.r.squared),
      sep = ""),
    ""
  )
)

# Plot with labels
scaling_relation_plot_lessdatasets <- power_law_lessdatasets_df %>%
  ggplot(aes(x = log(size), y = log(Fn), col = r2labels)) +
  geom_point(alpha = 0.9, size = 3) +
  geom_abline(aes(intercept = intercept, slope = slope, col=r2labels), size = 1, key_glyph = "smooth") +
  theme_linedraw() +
  xlab("log(n)") +
  ylab(bquote(bold(log((S[n]-S[n-1])/S[n-1])))) +
  scale_color_manual(guide = "legend", values=as.vector(name2color[levels(factor(power_law_lessdatasets_df$type_dataset))])) +
  guides(col = guide_legend(title=bquote(bold(.('Power law') ~ R^2)))) +
  theme(aspect.ratio = 1, 
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        legend.position = c(0.64, 0.86),
        legend.text = element_text(size = 16), 
        legend.title = element_text(size = 16, face = "bold")
        )

plot_file = paste(plots_dir, 
                  "scaling_relation_pearson_pval_0.05_lessdatasets.png", 
                  sep="/")
ggsave(
  plot_file,
  plot = scaling_relation_plot_lessdatasets,
  dpi = 1200,
  #width = 9500,
  height = 6000,
  units = c("px")
)
print(scaling_relation_plot_lessdatasets)
```

#### Comparison of power law with exponential decay

```{r}
# Process results
scaling_relation_comparison_proc_df <- scaling_relation_comparison_df %>%
  dplyr::filter(type_dataset == "tcga:tcga-brca_female") %>%
  dplyr::mutate_at(c('adj.r.squared'), ~sprintf("%.2f",.)) %>%
  dplyr::mutate_at(c('adj.r.squared'), as.character) %>%
  dplyr::mutate(type_dataset = dplyr::recode(type_dataset, !!!dataset2name)) %>%
  dplyr::inner_join(
    (name2color %>%
       as.data.frame() %>%
       dplyr::rename("rgb_col" = ".") %>%
       tibble::rownames_to_column("type_dataset")),
    by = c("type_dataset") )

# Create r2 labels
scaling_relation_comparison_proc_df <- scaling_relation_comparison_proc_df %>%
  dplyr::mutate(
    r2labels = dplyr::case_when(
      model == "Power law" ~ paste("Power law: ", unique((scaling_relation_comparison_proc_df %>% dplyr::filter(model == "Power law"))$adj.r.squared), sep = ""),
      model == "Exponential decay" ~ paste("Exp. decay: ", unique((scaling_relation_comparison_proc_df %>% dplyr::filter(model == "Exponential decay"))$adj.r.squared), sep = ""),
      model == "Constant discovery" ~ paste("Constant: ", unique((scaling_relation_comparison_proc_df %>% dplyr::filter(model == "Constant discovery"))$adj.r.squared), sep = ""),
      TRUE ~ ""
    )
  ) %>%
  dplyr::mutate(model = factor(model, levels = sort(unique(.data[["model"]]), decreasing = TRUE))) %>%
  dplyr::mutate(r2labels = factor(r2labels, levels = sort(unique(.data[["r2labels"]]), decreasing = TRUE)))

# Plot with labels
scaling_relation_plot_comparison <- scaling_relation_comparison_proc_df %>%
  ggplot(aes(x = size, y = Fn)) +
  geom_point(alpha = 0.9, size = 3, col = "#0072B2") +
  geom_line(
    aes(
      x = size,
      y = model_result,
      col = r2labels
    ),
    size = 1
  ) +
  theme_linedraw() +
  xlab("n") +
  ylab(bquote(bold((S[n]-S[n-1])/S[n-1]))) +
  scale_color_manual(guide = 'legend', values = as.vector(name2color[levels(factor(scaling_relation_comparison_proc_df$model))])) +
  guides(col = guide_legend(title = bquote(bold(R^2)))) +
  theme(aspect.ratio=1, 
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        legend.position = c(0.64, 0.85),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        legend.background = element_rect(fill = 'transparent', color = "transparent"))

plot_file = paste(plots_dir, "/comparison_pl_ed_pearson_pval_0.05_", dataset_selected, ".png", sep="")
ggsave(
  plot_file,
  plot = scaling_relation_plot_comparison,
  dpi = 1200,
  #width = 9500,
  height = 6000,
  units = c("px")
)
print(scaling_relation_plot_comparison)
```

#### Summary table

```{r}
# Define data sets
datasets_selected <- c(
  "gse193677:rectum_uc_inflamed",
  "gtex:breast.mammary.tissue_female",
  "gtex:lung",
  "scipher:scipher.sample.per.patient.baseline",
  "tcga:tcga-luad",
  "tcga:tcga-brca_female"
)

# Define model
model_selected <- "Stretched exponential (by linear fit)"

# Get R2 information from scaling relationship
pl_model_summary_df <- scaling_relation_comparison_df %>%
  dplyr::filter((type_dataset %in% datasets_selected) & (model == "Power law")) %>%
  dplyr::select(type_dataset, adj.r.squared, relative.error.mean) %>%
  unique() %>% 
  dplyr::arrange(factor(type_dataset, levels = datasets_selected)) %>%
  dplyr::mutate(type_dataset = dplyr::recode(type_dataset, !!!dataset2name)) %>%
  dplyr::rename("R**2" = "adj.r.squared", "epsilon" = "relative.error.mean") %>%
  dplyr::mutate_if(is.numeric, ~sprintf("%.2f",.))
print(pl_model_summary_df)

# Get R2 and error for exponential decay model
ed_model_summary_df <- scaling_relation_comparison_df %>% 
  dplyr::filter((type_dataset %in% datasets_selected) & (model == "Exponential decay")) %>%
  dplyr::select(type_dataset, adj.r.squared, relative.error.mean) %>%
  unique() %>% 
  dplyr::arrange(factor(type_dataset, levels = datasets_selected)) %>%
  dplyr::mutate(type_dataset = dplyr::recode(type_dataset, !!!dataset2name)) %>%
  dplyr::rename("R**2" = "adj.r.squared", "epsilon" = "relative.error.mean") %>%
  dplyr::mutate_if(is.numeric, ~sprintf("%.2f",.))
print(ed_model_summary_df)

# Get model parameters
analytical_model_summary_df %>% 
  dplyr::select(type_dataset, model, a, b, L, L_vs_total, adj.r.squared, relative.error.mean) %>% 
  dplyr::filter((type_dataset %in% datasets_selected) & (model == model_selected)) %>% unique() %>%
  dplyr::arrange(factor(type_dataset, levels = datasets_selected, datasets_selected))

# Get R2 and error for power law
analytical_model_summary_proc_df <- analytical_model_summary_df %>% 
  dplyr::select(type_dataset, model, a, adj.r.squared, relative.error.mean) %>% 
  dplyr::filter((type_dataset %in% datasets_selected) & (model == model_selected)) %>%
  dplyr::select(-model) %>%
  dplyr::arrange(factor(type_dataset, levels = datasets_selected)) %>%
  dplyr::mutate(type_dataset = dplyr::recode(type_dataset, !!!dataset2name)) %>%
  dplyr::rename("alpha"="a", "R**2" = "adj.r.squared", "epsilon" = "relative.error.mean") %>%
  unique() %>% 
  dplyr::mutate_if(is.numeric, ~sprintf("%.2f",.))
print(analytical_model_summary_proc_df)

# Bind power law with logarithm results
figure_summary_table <- cbind(
  pl_model_summary_df,
  (ed_model_summary_df %>% dplyr::select(-type_dataset)),
  (analytical_model_summary_proc_df %>% dplyr::select(-type_dataset))
)

# Assign first row as row names
row.names(figure_summary_table) <- figure_summary_table$type_dataset
figure_summary_table$type_dataset <- NULL

# Customize and print table
name2colorlight <- c(
  "UC inflamed" = "#f8b6fc",
  "TCGA: Breast cancer" = "#2bb3ff",
  "TCGA: Lung cancer" = "#c0e5fa",
  "R. arthritis" = "#ffd6b5",
  "GTEx: Lung" = "#78ffa0",
  "GTEx: Breast" = "#47d671"
  )

colors_original_figure <- as.vector(name2color[unique(rownames(figure_summary_table))])
colors_original_figure <- c(tail(colors_original_figure, 1), head(colors_original_figure, -1)) # I don't know why, the color in fg_params starts from 2nd position, so I have to reorder the vector of colors in order to match the other colors
colors_light_figure <- as.vector(name2colorlight[unique(rownames(figure_summary_table))])

core_figure <- gridExtra::tableGrob(
  figure_summary_table,
  theme = ttheme_minimal(
    colhead = list(
      bg_params = list(fill = NA, col = "black", lwd = 1),
      fg_params = list(parse = TRUE, fontface = "bold", fontsize = 18)
    ),
    rowhead = list(
      bg_params = list(fill = NA, col = NA),
      fg_params = list(col = "black", fontface = "plain", fontsize = 18)
    ), 
    core = list(
      bg_params = list(fill = colors_light_figure, col = "black", lwd = 1),
      fg_params = list(fontsize = 18)
    )
  )
)

header <- gridExtra::tableGrob(
  figure_summary_table[1, 1:3],
  rows = NULL,
  cols = c("Power law", "Exponential decay", "Analytical model"),
  theme = ttheme_minimal(
    colhead = list(
      bg_params = list(fill = NA, col = "black", lwd = 1),
      fg_params = list(parse = TRUE, fontface = "plain", fontsize = 18)
    )
  )
)

figure_summary_table_plot <- gtable_combine(header[1,], core_figure, along = 2)

# figure_summary_table_plot$widths[2:length(figure_summary_table_plot$widths)] <- rep(
#   max(figure_summary_table_plot$widths[2:length(figure_summary_table_plot$widths)]),
#   length(figure_summary_table_plot$widths[2:length(figure_summary_table_plot$widths)])
# ) # make column widths equal

# figure_summary_table_plot$widths[3:length(figure_summary_table_plot$widths)] <- rep(
#   max(figure_summary_table_plot$widths[3:length(figure_summary_table_plot$widths)]),
#   length(figure_summary_table_plot$widths[3:length(figure_summary_table_plot$widths)])
# ) # make columns 3 to end widths equal
# figure_summary_table_plot$widths[2] <- rep(
#   max(figure_summary_table_plot$widths[2]) * 0.7,
#   length(figure_summary_table_plot$widths[2])
# ) # make column 2 smaller

# Reduce widths
current_widths <- figure_summary_table_plot$widths[2:length(figure_summary_table_plot$widths)]
figure_summary_table_plot$widths[2:length(figure_summary_table_plot$widths)] <- unit(rep(2.5, length(current_widths)), "cm")

grid.newpage()
#figure_summary_table_plot$layout[1:4 , c("l", "r")] <- list(c(2, 7), c(6, 8)) # For 8 columns (including b and L)
#figure_summary_table_plot$layout[1:4 , c("l", "r")] <- list(c(2, 5), c(4, 6))
#figure_summary_table_plot$layout[1:6 , c("l", "r")] <- list(c(2, 3, 6), c(2, 5, 7)) # For 6 columns
figure_summary_table_plot$layout[1:6 , c("l", "r")] <- list(c(2, 4, 6), c(3, 5, 8)) # Power law spans 2 "l" to 3 "r", Exponential decay spans 4 "l" to 5 "r", and Analytical model spans 6 "l" to 8 "r"
grid.draw(figure_summary_table_plot)
```

#### Prediction plot

All datasets:

```{r}
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "gtex:lung", 
                      "gtex:whole.blood", 
                      "tcga:tcga-brca_female", 
                      "tcga:tcga-luad", 
                      "scipher:scipher.sample.per.patient.baseline")

#model_selected = "Stretched exponential (by optimization)"
model_selected = "Stretched exponential (by linear fit)"

# Process data
prediction_plot_df = results_selected_norm_df %>% 
  dplyr::right_join((predicted_results_norm_df %>% dplyr::select(type_dataset, model, model_result, size) %>% mutate_at(c('model_result'), as.numeric) %>% unique()), by=c("type_dataset", "size")) %>%
  dplyr::filter((type_dataset %in% datasets_selected) & (model == model_selected)) %>%
  dplyr::select(size, rep, type_dataset, num_edges, model_result) %>%
  dplyr::mutate(type_dataset = dplyr::recode(type_dataset, !!!dataset2name)) %>%
  inner_join( (name2color %>% as.data.frame() %>% dplyr::rename("rgb_col"=".") %>% tibble::rownames_to_column("type_dataset")), by=c("type_dataset") ) %>%
  mutate(geom_text_repel_label=dplyr::if_else(type_dataset == "R. arthritis" & size == 240 & rep == 1, "R. arthritis", dplyr::if_else(type_dataset == "GTEx: Whole blood" & size == 700 & rep == 1, "GTEx: Whole blood", dplyr::if_else(type_dataset == "GTEx: Lung" & size == 560 & rep == 1, "GTEx: Lung", dplyr::if_else(type_dataset == "GTEx: Breast" & size == 140 & rep == 1, "GTEx: Breast", dplyr::if_else(type_dataset == "TCGA: Lung cancer" & size == 460 & rep == 1, "TCGA: Lung cancer", dplyr::if_else(type_dataset == "TCGA: Breast cancer" & size == 800 & rep == 1, "TCGA: Breast cancer", "")))))))

# Plot with labels
prediction_plot = prediction_plot_df %>%
  ggplot(aes(x=log10(size), y=num_edges, col=rgb_col)) +
  geom_point(alpha=0.5, size=3) +
  #geom_line(data=(predicted_results_norm_df %>% filter((type_dataset %in% datasets_selected) & (model == model_selected)) %>% unique()), aes(x=size, y=model_result, col=type_dataset), size=1) +
  geom_line(aes(x=log10(size), y=model_result, col=rgb_col), size=1) +
  #scale_x_continuous(trans = scales::log10_trans()) +
  #scale_color_manual(values = as.vector(name2color[levels(factor(prediction_plot_df$type_dataset))])) + # Plot using the dictionary
  scale_colour_identity() +
  theme_linedraw() +
  xlab("Num. samples (Log10)") +
  ylab("Frac. significant correlations") +
  my_theme +
  geom_label_repel(
    aes(col = rgb_col, label = geom_text_repel_label),
    family = font_selected,
    fill = "white",
    max.overlaps = Inf,
    label.size = 0.75, 
    size = 5,
    alpha = 0.9, 
    box.padding = unit(0.75, "cm"),
    label.padding = 0.25, 
    fontface = "bold",
    na.rm = TRUE,
    seed = 1234
  )

plot_file = paste(plots_dir, "prediction_pearson_pval_0.05.png", sep="/")
print(prediction_plot)
ggsave(
  plot_file,
  plot=prediction_plot,
  dpi = 1200,
  #width = 9500,
  #height = 6000,
  units = c("px")
)
```

Less datasets:

```{r}
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "tcga:tcga-brca_female")
model_selected = "Stretched exponential (by linear fit)"

# Process data
prediction_plot_lessdatasets_df = results_selected_norm_df %>% 
  dplyr::right_join((predicted_results_norm_df %>% 
                       dplyr::select(type_dataset, model, model_result, size) %>% 
                       mutate_at(c('model_result'), as.numeric) %>% 
                       unique()), by=c("type_dataset", "size")) %>%
  dplyr::filter((type_dataset %in% datasets_selected) & (model == model_selected)) %>%
  dplyr::select(size, rep, type_dataset, num_edges, model_result) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "gtex:breast.mammary.tissue_female", "GTEx: Breast")) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "tcga:tcga-brca_female", "TCGA: Breast c.")) %>%
  inner_join( (name2color %>% as.data.frame() %>% dplyr::rename("rgb_col"=".") %>% tibble::rownames_to_column("type_dataset")), by=c("type_dataset") ) %>%
  mutate(geom_text_repel_label=dplyr::if_else(type_dataset == "R. arthritis" & size == 240 & rep == 1, "R. arthritis", dplyr::if_else(type_dataset == "GTEx: Whole blood" & size == 700 & rep == 1, "GTEx: Whole blood", dplyr::if_else(type_dataset == "GTEx: Lung" & size == 560 & rep == 1, "GTEx: Lung", dplyr::if_else(type_dataset == "GTEx: Breast" & size == 140 & rep == 1, "GTEx: Breast", dplyr::if_else(type_dataset == "TCGA: Lung cancer" & size == 460 & rep == 1, "TCGA: Lung cancer", dplyr::if_else(type_dataset == "TCGA: Breast c." & size == 1000 & rep == 1, "TCGA: Breast c.", "")))))))

# Plot with labels
prediction_plot_lessdatasets = prediction_plot_lessdatasets_df %>%
  ggplot(aes(x=log10(size), y=num_edges, col=rgb_col)) +
  geom_point(alpha=0.5, size=3) +
  geom_line(aes(x=log10(size), y=model_result, col=rgb_col), size=1) +
  scale_colour_identity(labels = unique(prediction_plot_lessdatasets_df$type_dataset), breaks = unique(prediction_plot_lessdatasets_df$rgb_col), guide = "legend") +
  theme_linedraw() +
  xlab("Num. samples (Log10)") +
  ylab("Frac. significant correlations") +
  guides(col=guide_legend(title = NULL)) +
  theme(aspect.ratio=1, 
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.position = c(0.28, 0.92),
        legend.text = element_text(size = 16), 
        legend.title = element_text(size = 16, face="bold"),
        #legend.position="none"),
        legend.background = element_rect(fill = 'transparent', color = "transparent"),
        text = element_text(family = font_selected)) #+

plot_file = paste(plots_dir, "prediction_pearson_pval_0.05_lessdatasets.png", sep="/")
print(prediction_plot_lessdatasets)
ggsave(
  plot_file,
  plot=prediction_plot_lessdatasets,
  dpi = 1200,
  #width = 9500,
  #height = 6000,
  units = c("px")
)
```

Number of samples for a given % of significant edges:

```{r}
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "gtex:lung", 
                      "gtex:whole.blood", 
                      "tcga:tcga-brca_female", 
                      "tcga:tcga-luad", 
                      "scipher:scipher.sample.per.patient.baseline")

model_selected = "Stretched exponential (by linear fit)"
predicted_results_norm_df %>% dplyr::select(type_dataset, model, model_result, size) %>%
  mutate_at(c('model_result'), as.numeric) %>%
  filter((type_dataset %in% datasets_selected) & (model == model_selected)) %>%
  group_by(type_dataset) %>% 
  filter((abs(model_result - 0.25) == min(abs(model_result - 0.25))) | (abs(model_result - 0.5) == min(abs(model_result - 0.5))) | (abs(model_result - 0.75) == min(abs(model_result - 0.75))))
```

#### Comparison of models: stretched exponential vs. logarithm

```{r}
models_selected = c("Stretched exponential (by linear fit)", "Logarithmic", "Mean")
#dataset_selected = "gtex:whole.blood"
dataset_selected = "tcga:tcga-brca_female"
comparison_palette = name2color[c(dataset_selected, "Analytical model", "Logarithm")]

# Join predicted results with mean
predicted_results_norm_and_mean_df = rbind((topology_results_selected_by_size_norm_df %>% dplyr::select(type_dataset, size, mean_norm) %>% unique() %>% rename("model_result" = "mean_norm") %>% mutate(model="Mean")), (predicted_results_norm_df %>% dplyr::select(type_dataset, size, model_result, model) %>% unique() %>% mutate_at(c('model_result'), as.numeric))) %>% 
  filter(model %in% models_selected) %>%
  mutate(model = replace(model, model == "Stretched exponential (by linear fit)", "Analytical model")) %>%
  mutate(model = replace(model, model == "Logarithmic", "Logarithm"))
predicted_results_norm_and_mean_df$model <- factor(predicted_results_norm_and_mean_df$model, levels = c("Mean", "Analytical model", "Logarithm"))

# Process data
comparison_log_df = results_selected_norm_df %>% 
  dplyr::select(type_dataset, size, rep, num_edges) %>%
  right_join(predicted_results_norm_and_mean_df, by=c("type_dataset", "size")) %>%
  filter(type_dataset == dataset_selected) %>%
  filter(size <= max((results_selected_norm_df %>% filter(type_dataset == dataset_selected))$size)) %>%
  filter(size >= min((results_selected_norm_df %>% filter(type_dataset == dataset_selected))$size)) %>%
  inner_join( (name2color %>% as.data.frame() %>% dplyr::rename("rgb_col"=".") %>% tibble::rownames_to_column("model") %>% mutate(model = replace(model, model == dataset_selected, "Mean"))), by=c("model") ) %>%
  mutate(geom_text_repel_label=dplyr::if_else(model == "Logarithm" & size == 280 & rep == 1, "Logarithm", dplyr::if_else(model == "Analytical model" & size == 760 & rep == 1, "Analytical model", dplyr::if_else(model == "Mean" & size == 560 & rep == 1, "Mean", ""))))

# # Plot with legend
# comparison_plot = comparison_log_df %>%
#   ggplot(aes(x=size, y=num_edges, col=model)) +
#   geom_point(alpha=0.1, size=3, col=comparison_palette[1]) +
#   geom_line(aes(x=size, y=model_result, col=model), size=1) +
#   #scale_x_continuous(trans = scales::log10_trans()) +
#   scale_color_manual(values = comparison_palette) +
#   theme_linedraw() +
#   xlab("Num. samples") +
#   ylab("Frac. significant correlations") +
#   #guides(col=guide_legend(title="Model")) +
#   theme(aspect.ratio=1,
#         plot.title =  element_text(size = 20, face="bold"),
#         axis.title = element_text(size = 17, face="bold"),
#         axis.text = element_text(size = 16),
#         legend.text = element_text(size = 16), 
#         #legend.title=element_text(size=16, face="bold"), 
#         legend.title=element_blank(), 
#         legend.position="bottom")

# Plot with labels
comparison_plot = comparison_log_df %>%
  ggplot(aes(x=size, y=num_edges)) +
  geom_point(alpha=0.1, size=3, col=comparison_palette[[dataset_selected]]) +
  geom_line(aes(x=size, y=model_result, col=rgb_col), size=1) +
  scale_colour_identity() +
  theme_linedraw() +
  xlab("Num. samples") +
  ylab("Frac. significant correlations") +
  theme(aspect.ratio=1, 
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        legend.position="none") +
  geom_label_repel(
    aes(
      x = size,
      y = model_result,
      col = rgb_col,
      label = geom_text_repel_label
    ),
    family = font_selected,
    fill = "white",
    max.overlaps = Inf,
    label.size = 0.75, 
    size = 5,
    alpha = 0.9, 
    box.padding = unit(0.75, "cm"),
    label.padding = 0.25, 
    fontface = "bold",
    na.rm = TRUE,
    seed = 1234
  )

plot_file = paste(plots_dir, "/comparison_models_pearson_pval_0.05_", dataset_selected, ".png", sep="")
print(comparison_plot)
ggsave(
  plot_file,
  plot=comparison_plot,
  dpi = 1200,
  #width = 6600,
  #height = 6700,
  units = c("px")
)

```

#### Demonstration of the model goodness-of-fit for 1 model

```{r}
models_selected = c("Stretched exponential (by linear fit)")
#dataset_selected = "gtex:whole.blood"
dataset_selected = "tcga:tcga-brca_female"
comparison_palette = name2color[c(dataset_selected, "Analytical model")]

# Join predicted results with mean
predicted_results_norm_and_mean_df <- 
  rbind((topology_results_selected_by_size_norm_df %>% 
           dplyr::select(type_dataset, size, mean_norm) %>% 
           unique() %>% 
           rename("model_result" = "mean_norm") %>% 
           mutate(model="Mean")), 
        (predicted_results_norm_df %>% 
           dplyr::select(type_dataset, size, model_result, model) %>% 
           unique() 
         %>% mutate_at(c('model_result'), as.numeric))) %>%
  filter(model %in% models_selected) %>%
  mutate(model = replace(model, 
                         model == "Stretched exponential (by linear fit)", 
                         "Analytical model"))
predicted_results_norm_and_mean_df$model <- factor(predicted_results_norm_and_mean_df$model, levels = c("Mean", "Analytical model", "Logarithm"))

# Process data
goodnessfit_df = results_selected_norm_df %>% 
  dplyr::select(type_dataset, size, rep, num_edges) %>%
  right_join(predicted_results_norm_and_mean_df, by=c("type_dataset", "size")) %>%
  filter(type_dataset == dataset_selected) %>%
  filter(size <= max((results_selected_norm_df %>% filter(type_dataset == dataset_selected))$size)) %>%
  filter(size >= min((results_selected_norm_df %>% filter(type_dataset == dataset_selected))$size)) %>%
  inner_join( (name2color %>% as.data.frame() %>% dplyr::rename("rgb_col"=".") %>% tibble::rownames_to_column("model") %>% mutate(model = replace(model, model == dataset_selected, "Mean"))), by=c("model") ) %>%
  mutate(geom_text_repel_label=dplyr::if_else(model == "Logarithm" & size == 280 & rep == 1, "Logarithm", dplyr::if_else(model == "Analytical model" & size == 760 & rep == 1, "Analytical model", dplyr::if_else(model == "Mean" & size == 560 & rep == 1, "Mean", ""))))

# Plot with labels
goodnessfit_plot = goodnessfit_df %>%
  ggplot(aes(x=size, y=num_edges)) +
  geom_point(alpha=0.6, size=3, col=comparison_palette[[dataset_selected]]) +
  geom_line(aes(x=size, y=model_result, col=rgb_col), size=1) +
  scale_colour_identity() +
  theme_linedraw() +
  xlab("Num. samples") +
  ylab("Frac. significant correlations") +
  theme(aspect.ratio=1, 
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        legend.position="none") +
  geom_label_repel(
    aes(
      x = size,
      y = model_result,
      col = rgb_col,
      label = geom_text_repel_label
    ),
    family = font_selected,
    fill = "white",
    max.overlaps = Inf,
    label.size = 0.75, 
    size = 5,
    alpha = 1,
    box.padding = unit(0.75, "cm"),
    label.padding = 0.25, 
    fontface = "bold",
    na.rm = TRUE,
    seed = 1234
  )

plot_file = paste(plots_dir, "/goodnessfit_model_pearson_pval_0.05_", dataset_selected, ".png", sep="")
print(goodnessfit_plot)
ggsave(
  plot_file,
  plot=goodnessfit_plot,
  dpi = 1200,
  #width = 6600,
  #height = 6700,
  units = c("px")
)

```

Demonstration of the model goodness-of-fit for 2 models:

```{r}
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "tcga:tcga-brca_female")
models_selected = c("Stretched exponential (by linear fit)")

# Join predicted results with mean
#predicted_results_norm_and_mean_df <- 
predicted_results_mean_df <- 
  #rbind((topology_results_selected_by_size_norm_df %>% 
  rbind((topology_results_selected_by_size_df %>% 
           #dplyr::select("type_dataset", "size", "mean_norm") %>% 
           dplyr::select("type_dataset", "size", "mean") %>% 
           unique() %>% 
           #rename("model_result" = "mean_norm") %>% 
           rename("model_result" = "mean") %>% 
           mutate(model="Mean")), 
        #(predicted_results_norm_df %>% 
        (predicted_results_df %>% 
           dplyr::select(type_dataset, size, model_result, model) %>% 
           unique() 
         %>% mutate_at(c('model_result'), as.numeric))) %>%
  filter(type_dataset %in% datasets_selected) %>%
  filter(model %in% models_selected) %>%
  inner_join((analytical_model_summary_df %>% 
                dplyr::select("type_dataset", "model", "adj.r.squared")),
             by = c("type_dataset", "model")) %>%
  mutate_at(c('adj.r.squared'), ~sprintf("%.2f",.)) %>%
  mutate_at(c('adj.r.squared'), as.character)

# Process data
#goodnessfit_lessdatasets_df = results_selected_norm_df %>% 
goodnessfit_lessdatasets_df = results_selected_df %>% 
  dplyr::select(type_dataset, size, rep, num_edges) %>%
  #right_join(predicted_results_norm_and_mean_df, by=c("type_dataset", "size")) %>%
  right_join(predicted_results_mean_df, by=c("type_dataset", "size")) %>%
  filter(type_dataset %in% datasets_selected) %>%
  #filter(size <= max((results_selected_norm_df %>% 
  filter(size <= max((results_selected_df %>% 
                        filter(type_dataset %in% datasets_selected))$size)) %>%
  #filter(size >= min((results_selected_norm_df %>% 
  filter(size >= min((results_selected_df %>% 
                        filter(type_dataset %in% datasets_selected))$size)) %>%
  filter(!(is.na(num_edges))) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "gtex:breast.mammary.tissue_female", "GTEx: Breast")) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "tcga:tcga-brca_female", "TCGA: Breast cancer")) %>%
  inner_join( (name2color %>% as.data.frame() %>% dplyr::rename("rgb_col"=".") %>% tibble::rownames_to_column("type_dataset")), by=c("type_dataset") ) %>%
  mutate(geom_text_repel_label=dplyr::if_else(type_dataset == "R. arthritis" & size == 240 & rep == 1, "R. arthritis", dplyr::if_else(type_dataset == "GTEx: Whole blood" & size == 700 & rep == 1, "GTEx: Whole blood", dplyr::if_else(type_dataset == "GTEx: Lung" & size == 560 & rep == 1, "GTEx: Lung", dplyr::if_else(type_dataset == "GTEx: Breast" & size == 140 & rep == 1, "GTEx: Breast", dplyr::if_else(type_dataset == "TCGA: Lung cancer" & size == 460 & rep == 1, "TCGA: Lung cancer", dplyr::if_else(type_dataset == "TCGA: Breast cancer" & size == 1000 & rep == 1, "TCGA: Breast cancer", "")))))))

goodnessfit_lessdatasets_df$r2labels = 
  ifelse(goodnessfit_lessdatasets_df$type_dataset == "GTEx: Breast", 
         paste("GTEx Breast: ", 
               unique((goodnessfit_lessdatasets_df %>% 
                         filter(type_dataset == "GTEx: Breast"))$adj.r.squared), 
               sep=""), 
         ifelse(goodnessfit_lessdatasets_df$type_dataset == "TCGA: Breast cancer", 
                paste("TCGA Breast c.: ", 
                      unique((goodnessfit_lessdatasets_df %>% 
                                filter(type_dataset == "TCGA: Breast cancer"))$adj.r.squared), 
                      sep=""), 
                ""))

# Plot with labels
goodnessfit_plot_lessdatasets = goodnessfit_lessdatasets_df %>%
  ggplot(aes(x=size, y=num_edges, col=r2labels)) +
  geom_point(alpha=0.6, size=3) +
  geom_line(aes(x=size, y=model_result, col=r2labels), size=1) +
  #scale_colour_identity() +
  scale_color_manual(guide = 'legend', values=as.vector(name2color[levels(factor(goodnessfit_lessdatasets_df$type_dataset))])) +
  theme_linedraw() +
  xlab("Num. samples") +
  ylab("Num. significant correlations") +
  guides(col=guide_legend(title=bquote(bold(.('Analytical model') ~ R^2)))) +
  theme(aspect.ratio=1, 
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        #legend.position="none") +
        legend.position = c(0.64, 0.15),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        #legend.background = element_rect(linewidth=0.2, linetype="solid", colour ="black")
        legend.background = element_rect(fill = "transparent", color = "transparent")
        )
  # geom_label_repel(aes(x=size, y=model_result, col=rgb_col, label=geom_text_repel_label),
  #                  fill="white",
  #                  box.padding = 0.5, max.overlaps = Inf,
  #                  label.size = 0.75, 
  #                  size = 5,
  #                  alpha = 1, 
  #                  label.padding=0.25, 
  #                  fontface = "bold",
  #                  na.rm=TRUE,
  #                  seed = 1234)

plot_file = paste(plots_dir, "/goodnessfit_2models_pearson_pval_0.05_", dataset_selected, ".png", sep="")
print(goodnessfit_plot_lessdatasets)
ggsave(
  plot_file,
  plot = goodnessfit_plot_lessdatasets,
  dpi = 1200,
  #width = 6600,
  #height = 6700,
  units = c("px")
)

```

#### a vs. fraction of significant correlations plot

Define functions:

```{r}
#'  calculate_predictions_using_stretched_exponential_model_optimized
#'  Formula to calculate exponential decay of significant interactions from a list of sample sizes.
#'  This formula requires the use of the L parameter
#'  @param x List of sample sizes.
#'  @param a Slope coefficient.
#'  @param b Intercept coefficient.
#'  
calculate_predictions_using_stretched_exponential_model_optimized = function(x, L, a, b){
  y = L * exp((b * x ** (-a + 1)) / (-a + 1))
  #y = exp(log(L) - exp(b+a*log(x)))
  return(y)
}

#'  calculate_prediction_from_analytical_model
#'  Function to calculate predictions using a specific analytical model
#'  @param model Name of the model used.
#'  @param x_list List of values in the x axis (e.g. size)
#'  @param a Parameter a.
#'  @param b Parameter b.
#'  @param L Parameter L.
#'  
calculate_prediction_from_analytical_model = function(model, x_list, a, b, L){
  if(model == "Logarithmic"){
    prediction_result = (log(x_list) *a + b)
  } else if((model == "Stretched exponential (by optimization)") | (model == "Stretched exponential (by linear fit)")){
    prediction_result = calculate_predictions_using_stretched_exponential_model_optimized(x=x_list, L=L, a=a, b=b)
  } else if(model == "Stretched exponential (without L)"){
    prediction_result = calculate_predictions_using_stretched_exponential_model_without_L(x=x_list, a=a, b=b)
  }
  return(prediction_result)
}
```

Calculate predictions of the models:

```{r}
#datasets_selected = c("gtex:whole.blood", "gtex:artery.tibial", "tcga:tcga-coad", "tcga:tcga-brca")
#datasets_selected = c("gtex:whole.blood", "gtex:artery.tibial", "tcga:tcga-coad", "tcga:tcga-brca", "scipher:scipher.complete.dataset")
#datasets_selected = c("gtex:breast.mammary.tissue", "gtex:thyroid", "tcga:tcga-brca", "tcga:tcga-thca", "scipher:scipher.complete.dataset")
#datasets_selected = c("gtex:breast.mammary.tissue", "gtex:thyroid", "gtex:whole.blood", "tcga:tcga-brca", "tcga:tcga-thca", "scipher:scipher.complete.dataset")
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "gtex:lung", 
                      "gtex:whole.blood", 
                      "tcga:tcga-brca_female", 
                      "tcga:tcga-luad", 
                      "scipher:scipher.sample.per.patient.baseline")
#model_selected = "Stretched exponential (by optimization)"
model_selected = "Stretched exponential (by linear fit)"
sample_size_vs_a_df = analytical_model_summary_df %>% 
  inner_join(sample_size_correlation_df, 
             by = c("type_dataset" = "dataset")) %>% 
  #filter((type_dataset %in% datasets_selected) & (model == model_selected)) 
  filter(model == model_selected)

# Calculate number of edges for each critical correlation using model
sample_size_vs_a_df$num_edges_from_statistical_corrected = 
  calculate_prediction_from_analytical_model(model = model_selected, 
                                             x_list = sample_size_vs_a_df$sample_size_statistical_corrected, 
                                             a = sample_size_vs_a_df$a, 
                                             b = sample_size_vs_a_df$b, 
                                             L = sample_size_vs_a_df$L)
sample_size_vs_a_df$num_edges_from_statistical_corrected = 
  sample_size_vs_a_df$num_edges_from_statistical_corrected * sample_size_vs_a_df$max_value_in_dataset
sample_size_vs_a_df$num_edges_from_statistical_corrected_norm = 
  sample_size_vs_a_df$num_edges_from_statistical_corrected / sample_size_vs_a_df$total_num_edges

# sample_size_vs_a_df = sample_size_vs_a_df %>% 
#   tidyr::tidyr::separate(type_dataset, 
#                   into = c("dataset", NA), 
#                   sep = ":", 
#                   remove = FALSE) %>%
#   tidyr::tidyr::separate(type_dataset, 
#                   into = c("dataset_name", "sex"), 
#                   sep = "_", 
#                   remove = TRUE) %>%
#   mutate(subclassification = 
#            if_else(dataset == "tcga", 
#                   "tumor", 
#                   "")) %>%
#   mutate(dataset_name = 
#            if_else(dataset == "tcga", 
#                    paste(dataset_name, 
#                          subclassification, 
#                          sep = "-"), 
#                    dataset_name)) %>%
#   mutate(dataset_name = 
#            if_else(!(is.na(sex)), 
#                    paste(dataset_name, 
#                          sex, 
#                          sep = "_"), 
#                    dataset_name))

sample_size_vs_a_df %>% head(10)
```

Save the predictions:

```{r}
type_correlation_selected = "weak"

sample_size_vs_a_df %>%
  filter(type_correlation == type_correlation_selected) %>%
  arrange(desc(a))

a_vs_fraction_corr_file = paste(tables_dir, 
                                "a_vs_fraction_sig_correlations_pearson_pval_0.05.txt", 
                                sep = "/")

sample_size_vs_a_df %>% 
  arrange(desc(a)) %>% 
  fwrite(a_vs_fraction_corr_file)
```

Plot alpha vs. fraction of correlations above 0.2:

```{r}
# Process data
datasets_selected <- c("scipher:scipher.sample.per.patient.baseline",
                       "gse193677:rectum-disease_inflamed_vs_control_cd",
                       "gse193677:rectum-disease_inflamed_vs_control_uc",
                       "gse193677:rectum-disease_inflamed_vs_control_control",
                       "gse193677:rectum-inflamed_vs_noninflamed_inflamed",
                       "gse193677:rectum-inflamed_vs_noninflamed_noninflamed",
                       "gtex:breast.mammary.tissue_female",
                       "gtex:lung",
                       "gtex:skin.sun.exposed.lower.leg",
                       "gtex:thyroid",
                       "gtex:whole.blood",
                       "tcga:tcga-brca-tumor_female",
                       "tcga:tcga-kirc-tumor",
                       "tcga:tcga-kirp-tumor",
                       "tcga:tcga-luad-tumor",
                       "tcga:tcga-lusc-tumor",
                       "tcga:tcga-thca-tumor",
                       "tcga:breast-tissue_female", 
                       "tcga:kidney-tissue",
                       "tcga:lung-tissue",
                       "tcga:brca.luma-subtype",
                       "tcga:brca.lumb-subtype")

sample_size_vs_a_plot_df = sample_size_vs_a_df %>% 
  filter(type_correlation == type_correlation_selected) %>% 
  filter(dataset_name %in% datasets_selected) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "scipher:scipher.sample.per.patient.baseline", 
                   "R. arthritis")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "gtex:whole.blood",
                   "GTEx: Whole blood")) %>%
  #mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:thyroid", "GTEx: Thyroid")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "gtex:lung", 
                   "GTEx: Lung")) %>%
  #mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:artery.tibial", "GTEx: Artery tibial")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "gtex:breast.mammary.tissue_female", 
                   "GTEx: Breast")) %>%
  #mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-coad", "TCGA: Colon cancer")) %>%
  #mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-thca", "TCGA: Thyroid cancer")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "tcga:tcga-luad-tumor", 
                   "TCGA: Lung cancer")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "tcga:tcga-brca-tumor_female", 
                   "TCGA: Breast cancer")) %>%
  left_join( (name2color %>% 
                as.data.frame() %>% 
                dplyr::rename("rgb_col" = ".") %>% 
                tibble::rownames_to_column("dataset_name")), 
             by = c("dataset_name") ) %>%
  mutate(rgb_col = replace(rgb_col, is.na(rgb_col), "#000000")) %>% # Non-selected datasets with color black
  mutate(geom_text_repel_label=dplyr::if_else(dataset_name == "R. arthritis", "R. arthritis", 
                                              dplyr::if_else(dataset_name == "GTEx: Whole blood", "GTEx: Whole blood", 
                                                             dplyr::if_else(dataset_name == "GTEx: Lung", "GTEx: Lung", 
                                                                            dplyr::if_else(dataset_name == "GTEx: Breast", "GTEx: Breast", 
                                                                                           dplyr::if_else(dataset_name == "TCGA: Lung cancer", "TCGA: Lung cancer", 
                                                                                                          dplyr::if_else(dataset_name == "TCGA: Breast cancer", "TCGA: Breast cancer", "")))))))

# Calculate correlation
ss_vs_a_cor <- cor(sample_size_vs_a_plot_df$a, sample_size_vs_a_plot_df$num_edges_from_statistical_corrected_norm)

# Calculate regression
lm_summary <- summary(lm(sample_size_vs_a_plot_df$num_edges_from_statistical_corrected_norm ~ sample_size_vs_a_plot_df$a))
slope <- coef(lm_summary)[2]
intercept <- coef(lm_summary)[1]
adj.r.squared <- lm_summary$adj.r.squared
lm_cor_result <- paste("R =", 
                       round(adj.r.squared, 3), 
                       "\ncorr. =", 
                       round(ss_vs_a_cor, 3))


# Make scatter plot
a_vs_fraction_sig_correlations_plot <- sample_size_vs_a_plot_df %>%
    ggplot(aes(x=a, y=num_edges_from_statistical_corrected_norm)) + 
    geom_point(aes(col=rgb_col), show.legend = FALSE) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, label = geom_text_repel_label),
                     fill="white",
                     family = font_selected,
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     max.overlaps = Inf,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 5555) +
    new_scale_color() +
    geom_abline(aes(slope = slope,
                    intercept = intercept,
                    col = lm_cor_result),
              linetype = "dashed", 
              show.legend = FALSE) +
  scale_color_manual(values = c("gray50")) +
  annotate("text", x = 2.07, y = 0.77, label = lm_cor_result, size = 5) +
  xlab(expression(alpha)) +
  ylab("Frac. correlations above 0.2") +
  theme_linedraw() +
  theme(aspect.ratio = 1, 
        plot.title = element_text(size = 20, face = "bold"), 
        axis.title = element_text(size = 17, face = "bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        legend.position="none")
        #legend.position = c(0.70, 0.25),
        #legend.text = element_text(size = 14), 
        #legend.title=element_blank())

print(a_vs_fraction_sig_correlations_plot)
plot_file <- paste(plots_dir,
                   "a_vs_fraction_sig_correlations_pearson_pval_0.05.png",
                   sep="/")
ggsave(
  plot_file,
  plot=a_vs_fraction_sig_correlations_plot,
  dpi = 1200,
  #width = 6400,
  #height = 6000,
  units = c("px")
)
```

Less datasets:

```{r}
# Process data
datasets_selected <- c("scipher:scipher.sample.per.patient.baseline",
                       "gse193677:rectum-disease_inflamed_vs_control_cd",
                       "gse193677:rectum-disease_inflamed_vs_control_uc",
                       "gse193677:rectum-disease_inflamed_vs_control_control",
                       "gse193677:rectum-inflamed_vs_noninflamed_inflamed",
                       "gse193677:rectum-inflamed_vs_noninflamed_noninflamed",
                       "gtex:breast.mammary.tissue_female",
                       "gtex:lung",
                       "gtex:skin.sun.exposed.lower.leg",
                       "gtex:thyroid",
                       "gtex:whole.blood",
                       "tcga:tcga-brca-tumor_female",
                       "tcga:tcga-kirc-tumor",
                       "tcga:tcga-kirp-tumor",
                       "tcga:tcga-luad-tumor",
                       "tcga:tcga-lusc-tumor",
                       "tcga:tcga-thca-tumor",
                       "tcga:breast-tissue_female", 
                       "tcga:kidney-tissue",
                       "tcga:lung-tissue",
                       "tcga:brca.luma-subtype",
                       "tcga:brca.lumb-subtype")

sample_size_vs_a_plot_lessdatasets_df = sample_size_vs_a_df %>% 
  filter(type_correlation == type_correlation_selected) %>% 
  filter(dataset_name %in% datasets_selected) %>%
  mutate(dataset = replace(dataset, dataset == "scipher", "R. arthritis"),
         dataset = replace(dataset, dataset == "gse193677", "GSE193677"),
         dataset = replace(dataset, dataset == "gtex", "GTEx"),
         dataset = replace(dataset, dataset == "tcga", "TCGA")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "gtex:breast.mammary.tissue_female", 
                   "GTEx: Breast")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "tcga:tcga-brca-tumor_female", 
                   "TCGA: Breast cancer")) %>%
  left_join( (name2color %>% 
                as.data.frame() %>% 
                dplyr::rename("rgb_col" = ".") %>% 
                tibble::rownames_to_column("dataset")# %>% 
                #filter(dataset_name %in% c("TCGA: Breast cancer", "GTEx: Breast"))
              ), 
             by = c("dataset") ) %>%
  mutate(rgb_col = replace(rgb_col, is.na(rgb_col), "#000000")) %>% # Non-selected datasets with color black
  mutate(geom_text_repel_label=dplyr::if_else(dataset_name == "R. arthritis", "R. arthritis", 
                                              dplyr::if_else(dataset_name == "GTEx: Whole blood", "GTEx: Whole blood", 
                                                             dplyr::if_else(dataset_name == "GTEx: Lung", "GTEx: Lung", 
                                                                            dplyr::if_else(dataset_name == "GTEx: Breast", "GTEx: Breast", 
                                                                                           dplyr::if_else(dataset_name == "TCGA: Lung cancer", "TCGA: Lung cancer", 
                                                                                                          dplyr::if_else(dataset_name == "TCGA: Breast cancer", "TCGA: Breast cancer", "")))))))

# Calculate correlation
a_vs_fr_cor <- cor(sample_size_vs_a_plot_lessdatasets_df$a, sample_size_vs_a_plot_lessdatasets_df$num_edges_from_statistical_corrected_norm)

# Calculate regression
lm_summary_avsfr <- summary(lm(sample_size_vs_a_plot_lessdatasets_df$num_edges_from_statistical_corrected_norm ~ sample_size_vs_a_plot_lessdatasets_df$a))
slope_avsfr <- coef(lm_summary_avsfr)[2]
intercept_avsfr <- coef(lm_summary_avsfr)[1]
adj.r.squared_avsfr <- lm_summary_avsfr$adj.r.squared
lm_cor_result_avsfr <- paste("R =", 
                       round(adj.r.squared_avsfr, 3), 
                       "\ncorr. =", 
                       round(a_vs_fr_cor, 3))


# Make scatter plot
a_vs_fraction_sig_correlations_lessdatasets_plot <- sample_size_vs_a_plot_lessdatasets_df %>%
    ggplot(aes(x=a, y=num_edges_from_statistical_corrected_norm)) + 
    geom_point(aes(col=rgb_col), alpha = 0.6, size = 3, show.legend = TRUE) +
    #scale_colour_identity() +
    scale_colour_identity(labels = unique(sample_size_vs_a_plot_lessdatasets_df$dataset), breaks = unique(sample_size_vs_a_plot_lessdatasets_df$rgb_col), guide = "legend") +
    geom_label_repel(aes(col = rgb_col, label = geom_text_repel_label),
                     fill="white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     max.overlaps = Inf,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding = 0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     show.legend = FALSE,
                     max.iter = 1e5,
                     seed = 5555) +
    geom_abline(aes(slope = slope_avsfr,
                    intercept = intercept_avsfr),
                linetype = "dashed", 
                col = "gray50",
                show.legend = FALSE) +
  annotate("text", x = 2.00, y = 0.45, label = lm_cor_result_avsfr, size = 5) +
  xlab(expression(alpha)) +
  ylab("Frac. correlations above 0.2") +
  theme_linedraw() +
  theme(aspect.ratio = 1, 
        plot.title = element_text(size = 20, face = "bold"), 
        axis.title = element_text(size = 17, face = "bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        #legend.position="none")
        legend.position = c(0.20, 0.84),
        legend.background = element_rect(fill = 'transparent', color = "transparent"),
        legend.text = element_text(size = 16), 
        legend.title=element_blank())

print(a_vs_fraction_sig_correlations_lessdatasets_plot)
plot_file <- paste(plots_dir,
                   "a_vs_fraction_sig_correlations_pearson_pval_0.05_lessdatasets.png",
                   sep="/")
ggsave(
  plot_file,
  plot=a_vs_fraction_sig_correlations_lessdatasets_plot,
  dpi = 1200,
  #width = 6400,
  #height = 6000,
  units = c("px")
)
```

#### Variation plot

Create plots showing the variation of each gene across samples:

```{r}
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "gtex:lung", 
                      "gtex:whole.blood", 
                      "tcga:tcga-brca-tumor_female", 
                      "tcga:tcga-luad-tumor", 
                      "scipher:scipher.sample.per.patient.baseline")
parameters = c("sd", "mean", "cv")
parameter2label <- list("sd"="SD", "mean"="mean", "cv"="CV")
type_counts_selected = "reads"
distribution_plots = list()
sd_genes_filtered = sd_genes_df %>%
  filter(dataset_name %in% datasets_selected) %>%
  filter(type_counts == type_counts_selected) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "scipher.sample.per.patient.baseline", "R. arthritis")) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "Whole.Blood", "GTEx: Whole blood")) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "Lung", "GTEx: Lung")) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "Breast.Mammary.Tissue", "GTEx: Breast")) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "TCGA-LUAD", "TCGA: Lung cancer")) %>%
  mutate(type_dataset = replace(type_dataset, type_dataset == "TCGA-BRCA", "TCGA: Breast cancer")) %>%
  inner_join( (name2color %>% as.data.frame() %>% dplyr::rename("rgb_col"=".") %>% tibble::rownames_to_column("type_dataset")), by=c("type_dataset") )

print(unique(sd_genes_filtered$dataset_name)) # test to check if datasets are selected correctly

for (x in 1:length(parameters)){
  parameter = parameters[x]
  
  # # Plot with legend
  # distribution_plot = sd_genes_filtered %>%
  #   ggplot(aes(.data[[parameter]], color=type_dataset)) +
  #   geom_freqpoly(size=1.5) +
  #   scale_color_manual(values = c("#D55E00", "#E69F00", "#44AA99", "#65F3BF", "#0072B2", "#56B4E9")) +
  #   scale_x_log10() + 
  #   scale_y_log10() + 
  #   labs(x=paste("Gene expression ", parameter2label[[parameter]], " (Log)", sep=""), y = "Number of genes (Log)") +
  #   theme_linedraw() + 
  #   guides(col=guide_legend(title="Dataset")) +
  #   theme(aspect.ratio=1, plot.title =  element_text(size = 20, face="bold"), axis.title = element_text(size = 17, face="bold"), axis.text = element_text(size = 16), legend.text = element_text(size = 16), legend.title=element_text(size=16, face="bold"))

  # Plot without legend
  distribution_plot = sd_genes_filtered %>%
    ggplot(aes(.data[[parameter]], color=rgb_col)) +
    geom_freqpoly(size=1.5) +
    scale_colour_identity() +
    scale_x_log10() + 
    scale_y_log10() + 
    labs(x=paste("Gene expression ", parameter2label[[parameter]], " (Log)", sep=""), y = "Number of genes (Log)") +
    theme_linedraw() + 
    guides(col=guide_legend(title="Dataset")) +
    theme(aspect.ratio=1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")
  
  # Save distribution plot without changes in the theme, because we will tune it afterwards with patchwork
  distribution_plots[[x]] = distribution_plot
  
  plot_file = paste(plots_dir, "/distribution_", parameter, "_genes.png", sep="")
  ggsave(
    plot_file,
    plot=distribution_plot,
    dpi = 1200,
    #width = 10000,
    #height = 6000,
    units = c("px")
  )
  print(distribution_plot)
}
```

Less datasets:

```{r}
datasets_selected = c("gtex:breast.mammary.tissue_female", 
                      "tcga:tcga-brca-tumor_female")
parameters = c("sd", "mean", "cv")
parameter2label <- list("sd"="SD", "mean"="mean", "cv"="CV")
type_counts_selected = "reads"
distribution_plots_lessdatasets = list()
sd_genes_filtered = sd_genes_df %>%
  filter(dataset_name %in% datasets_selected) %>%
  filter(type_counts == type_counts_selected) %>%
  mutate(type_dataset = replace(type_dataset, 
                                type_dataset == "Breast.Mammary.Tissue", 
                                "GTEx: Breast")) %>%
  mutate(type_dataset = replace(type_dataset, 
                                type_dataset == "TCGA-BRCA", 
                                "TCGA: Breast cancer")) %>%
  inner_join( (name2color %>%
                 as.data.frame() %>%
                 dplyr::rename("rgb_col"=".") %>%
                 tibble::rownames_to_column("type_dataset")), 
              by=c("type_dataset") )

print(unique(sd_genes_filtered$dataset_name)) # test to check if datasets are selected correctly

for (x in 1:length(parameters)){
  parameter = parameters[x]

  # Plot without legend
  distribution_plot_lessdatasets = sd_genes_filtered %>%
    ggplot(aes(.data[[parameter]], color=rgb_col)) +
    geom_freqpoly(size=1.5) +
    scale_colour_identity() +
    scale_x_log10() + 
    scale_y_log10() + 
    labs(x=paste("Gene expression ", 
                 parameter2label[[parameter]], 
                 " (Log)", sep=""), 
         y = "Number of genes (Log)") +
    theme_linedraw() + 
    guides(col=guide_legend(title="Dataset")) +
    theme(aspect.ratio=1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")
  
  # Save distribution plot without changes in the theme, because we will tune it 
  # afterwards with patchwork
  distribution_plots_lessdatasets[[parameter]] = distribution_plot_lessdatasets
  
  plot_file = paste(plots_dir, 
                    "/distribution_", 
                    parameter, 
                    "_genes_lessdatasets.png", 
                    sep="")
  ggsave(
    plot_file,
    plot=distribution_plot_lessdatasets,
    dpi = 1200,
    #width = 10000,
    #height = 6000,
    units = c("px")
  )
  print(distribution_plot_lessdatasets)
}
```

Show numbers:

```{r}
a_vs_fraction_corr_file = 
  paste(tables_dir, 
        "a_vs_fraction_sig_correlations_pearson_pval_0.05.txt", 
        sep='/')
a_vs_fraction_corr_df <- data.table::fread(a_vs_fraction_corr_file)
head(a_vs_fraction_corr_df)
```

```{r}
sd_cut <- 10
mean_cut <- 10
cv_cut <- 50

sd_genes_df %>% 
  filter(type_counts == type_counts_selected) %>% 
  group_by(dataset_name, subclassification) %>% 
  summarize(mean_sd = mean(sd), 
            median_sd = median(sd), 
            mean_mean = mean(mean), 
            median_mean = median(mean), 
            mean_cv = mean(cv), 
            median_cv = median(cv)) %>%
  head(10)

# Calculate number of genes above certain SD, mean and CV cut-offs
sd_genes_summary_df = sd_genes_df %>% 
  filter(type_counts == type_counts_selected) %>% 
  group_by(dataset, 
           type_dataset, 
           dataset_name, 
           type_counts, 
           sex, 
           subclassification) %>% 
  summarize(num_genes_above_sd_cut = length(sd[sd >= sd_cut]), 
            num_genes_above_mean_cut = length(mean[mean >= mean_cut]), 
            num_genes_above_cv_cut = length(cv[cv >= cv_cut]), 
            num_genes = n()) %>% 
  ungroup() %>% 
  mutate(fraction_genes_above_sd_cut = num_genes_above_sd_cut / num_genes, 
         fraction_genes_above_mean_cut = num_genes_above_mean_cut / num_genes, 
         fraction_genes_above_cv_cut = num_genes_above_cv_cut / num_genes)

sd_genes_summary_gse <- sd_genes_summary_df %>%
  filter(dataset == "gse193677") %>%
  dplyr::select(-type_dataset, dataset_name) %>%
  slice(rep(1, 5)) %>%
  mutate(
    type_dataset = (a_vs_fraction_corr_df %>% filter(dataset == "gse193677") %>% pull(type_dataset) %>% unique()),
    dataset_name = (a_vs_fraction_corr_df %>% filter(dataset == "gse193677") %>% pull(dataset_name) %>% unique())
  )

sd_genes_summary_df <- sd_genes_summary_df %>%
  filter(!(dataset == "gse193677")) %>%
  bind_rows(sd_genes_summary_gse)

sd_genes_vs_a_df <- sd_genes_summary_df %>%
  inner_join((a_vs_fraction_corr_df %>% select(-type_dataset, -sex, -subclassification)), 
             by = c("dataset_name", "dataset")) %>% 
  select(!c("type_correlation", 
            "correlation", 
            "sample_size_statistical", 
            "sample_size_statistical_corrected", 
            "num_edges_from_statistical_corrected", 
            "num_edges_from_statistical_corrected_norm")) %>% 
  unique() %>%
  as.data.frame()

head(sd_genes_vs_a_df)
```

Plot different parameters (sd, mean, cv) vs. alpha:

```{r}
parameters <- c("sd", "mean", "cv")
parameter2label <- list("sd" = "SD", "mean" = "mean", "cv" = "CV")
datasets_selected <- c("scipher:scipher.sample.per.patient.baseline",
                       "gse193677:rectum-disease_inflamed_vs_control_cd",
                       "gse193677:rectum-disease_inflamed_vs_control_uc",
                       "gse193677:rectum-disease_inflamed_vs_control_control",
                       "gse193677:rectum-inflamed_vs_noninflamed_inflamed",
                       "gse193677:rectum-inflamed_vs_noninflamed_noninflamed",
                       "gtex:breast.mammary.tissue_female",
                       "gtex:lung",
                       "gtex:skin.sun.exposed.lower.leg",
                       "gtex:thyroid",
                       "gtex:whole.blood",
                       "tcga:tcga-brca-tumor_female",
                       "tcga:tcga-kirc-tumor",
                       "tcga:tcga-kirp-tumor",
                       "tcga:tcga-luad-tumor",
                       "tcga:tcga-lusc-tumor",
                       "tcga:tcga-thca-tumor",
                       "tcga:breast-tissue_female", 
                       "tcga:kidney-tissue",
                       "tcga:lung-tissue",
                       "tcga:brca.luma-subtype",
                       "tcga:brca.lumb-subtype")

# Prepare table for plots
sd_genes_vs_a_filtered <- sd_genes_vs_a_df %>%
  dplyr::filter(type_counts == type_counts_selected) %>%
  dplyr::filter(dataset_name %in% datasets_selected) %>%
  dplyr::mutate(dataset_name = dplyr::recode(dataset_name, !!!dataset2name)) %>%
  dplyr::left_join(
    (name2color %>%
       as.data.frame() %>%
       dplyr::rename("rgb_col"=".") %>%
       tibble::rownames_to_column("dataset")),
    by = c("dataset")
  ) %>%
  dplyr::mutate(rgb_col = replace(rgb_col, is.na(rgb_col), "#000000")) %>% # Non-selected datasets with color black
  dplyr::mutate(geom_text_repel_label = dplyr::if_else(
    dataset_name %in% c("R. arthritis", "GTEx: Whole blood", "GTEx: Lung", "GTEx: Breast", "TCGA: Lung cancer", "TCGA: Breast cancer"),
    dataset_name,
    ""
  )) %>% 
  as.data.frame()
scatter_plots = list()

print(unique(sd_genes_vs_a_filtered$dataset_name)) # test to check if datasets are selected correctly

# Make plot for each type of parameter (SD, mean, CV)
for (x in 1:length(parameters)){
  parameter = parameters[x]
  if (parameter == "cv"){
    cut <- cv_cut
    x_legend <- 1.55
    y_legend <- 0.45
  } else if (parameter == "sd"){
    cut <- sd_cut
    x_legend <- 1.585
    y_legend <- 0.938
  } else if (parameter == "mean"){
    cut <- mean_cut
    x_legend <- 1.585
    y_legend <- 0.968
  }

  # Calculate correlation
  ss_vs_a_cor <- cor(sd_genes_vs_a_filtered$a, 
                     sd_genes_vs_a_filtered[[paste("fraction_genes_above_",
                                                   parameter,
                                                   "_cut",
                                                   sep = "")]])
  
  # Calculate regression
  lm_summary <- summary(lm(sd_genes_vs_a_filtered[[paste("fraction_genes_above_",
                                                   parameter,
                                                   "_cut",
                                                   sep = "")]] ~ 
                             sd_genes_vs_a_filtered$a))
  slope <- coef(lm_summary)[2]
  intercept <- coef(lm_summary)[1]
  adj.r.squared <- lm_summary$adj.r.squared
  lm_cor_result <- paste("R =", 
                         round(adj.r.squared, 3), 
                         "\ncorr. =", 
                         round(ss_vs_a_cor, 3))

  param_vs_a_plot = sd_genes_vs_a_filtered %>%
    ggplot(aes(x = a,
               y = .data[[paste("fraction_genes_above_", 
                                parameter, 
                                "_cut", 
                                sep = "")]])) + 
    geom_point(aes(col=rgb_col), show.legend = FALSE) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, 
                         label = geom_text_repel_label),
                     fill = "white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"), 
                     max.overlaps = Inf,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    new_scale_color() +
    geom_abline(aes(slope = slope,
                    intercept = intercept,
                    col = lm_cor_result),
              linetype = "dashed", 
              show.legend = FALSE) +
    scale_color_manual(values = c("gray50")) +
    annotate("text", x = x_legend, y = y_legend, label = lm_cor_result, size = 5) +
    xlab(expression(alpha)) +
    ylab(paste("Frac. genes with ", 
               parameter2label[[parameter]], 
               " above ", 
               cut, 
               sep = "")) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")
          #legend.position = c(0.70, 0.25),
          #legend.text = element_text(size = 14), 
          #legend.title=element_blank())

  plot_file = paste(plots_dir, "/a_vs_fraction_genes_above_", parameter, "_cut.png", sep="")
  ggsave(
    plot_file,
    dpi = 1200,
    width = 6000,
    #height = 6000,
    units = c("px")
  )
  print(param_vs_a_plot)
  scatter_plots[[x]] = param_vs_a_plot
}
```

With less datasets:

```{r}
parameters <- c("sd", "mean", "cv")
parameter2label <- list("sd" = "SD", "mean" = "mean", "cv" = "CV")
datasets_selected <- c("scipher:scipher.sample.per.patient.baseline",
                       "gse193677:rectum-disease_inflamed_vs_control_cd",
                       "gse193677:rectum-disease_inflamed_vs_control_uc",
                       "gse193677:rectum-disease_inflamed_vs_control_control",
                       "gse193677:rectum-inflamed_vs_noninflamed_inflamed",
                       "gse193677:rectum-inflamed_vs_noninflamed_noninflamed",
                       "gtex:breast.mammary.tissue_female",
                       "gtex:lung",
                       "gtex:skin.sun.exposed.lower.leg",
                       "gtex:thyroid",
                       "gtex:whole.blood",
                       "tcga:tcga-brca-tumor_female",
                       "tcga:tcga-kirc-tumor",
                       "tcga:tcga-kirp-tumor",
                       "tcga:tcga-luad-tumor",
                       "tcga:tcga-lusc-tumor",
                       "tcga:tcga-thca-tumor",
                       "tcga:breast-tissue_female", 
                       "tcga:kidney-tissue",
                       "tcga:lung-tissue",
                       "tcga:brca.luma-subtype",
                       "tcga:brca.lumb-subtype")

# Prepare table for plots
sd_genes_vs_a_filtered <- sd_genes_vs_a_df %>%
  dplyr::filter(dataset_name %in% datasets_selected) %>%
  dplyr::mutate(dataset_name = dplyr::recode(dataset_name, !!!dataset2name)) %>%
  dplyr::left_join(
    (name2color %>%
       as.data.frame() %>%
       dplyr::rename("rgb_col" = ".") %>%
       tibble::rownames_to_column("dataset")),
    by = c("dataset")
  ) %>%
  dplyr::mutate(rgb_col = replace(rgb_col, is.na(rgb_col), "#000000")) %>% # Non-selected datasets with color black
  dplyr::mutate(geom_text_repel_label = dplyr::if_else(
    dataset_name %in% c("GTEx: Breast", "TCGA: Breast cancer"),
    dataset_name,
    ""
  )) %>% 
  as.data.frame()

scatter_plots_lessdatasets = list()
lm_summary_avspar = list()
lm_cor_result_avspar = list()

for (x in 1:length(parameters)){
  parameter = parameters[x]
  if (parameter == "cv"){
    cut <- cv_cut
    x_legend <- 2
    y_legend <- 0.38
  } else if (parameter == "sd"){
    cut <- sd_cut
    x_legend <- 1.585
    y_legend <- 0.938
  } else if (parameter == "mean"){
    cut <- mean_cut
    x_legend <- 1.585
    y_legend <- 0.968
  }

  # Calculate correlation
  a_vs_par_cor <- cor(sd_genes_vs_a_filtered$a, 
                     sd_genes_vs_a_filtered[[paste("fraction_genes_above_",
                                                   parameter,
                                                   "_cut",
                                                   sep = "")]])
  
  # Calculate regression
  lm_summary_avspar[[parameter]] <- summary(lm(sd_genes_vs_a_filtered[[paste("fraction_genes_above_",
                                                   parameter,
                                                   "_cut",
                                                   sep = "")]] ~ 
                             sd_genes_vs_a_filtered$a))
  lm_cor_result_avspar[[parameter]] <- paste("R =", 
                         round(lm_summary_avspar[[parameter]]$adj.r.squared, 3), 
                         "\ncorr. =", 
                         round(a_vs_par_cor, 3))
  
  scatter_plots_lessdatasets[[parameter]] <- sd_genes_vs_a_filtered %>%
    ggplot(aes(x = a, 
               y = .data[[paste("fraction_genes_above_", 
                                parameter, 
                                "_cut", 
                                sep = "")]], 
               col = rgb_col)) + 
    geom_point(alpha = 0.6, size = 3) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, label = geom_text_repel_label),
                     fill="white",
                     family = font_selected,
                     max.iter = 1e6,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     box.padding = unit(0.75, "cm"),
                     label.padding = 0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    geom_abline(aes(slope = coef(lm_summary_avspar[[parameter]])[2],
                  intercept = coef(lm_summary_avspar[[parameter]])[1]),
              col = "gray50",
              linetype = "dashed", 
              show.legend = FALSE) +
    annotate("text", x = x_legend, y = y_legend, label = lm_cor_result_avspar[[parameter]], size = 5) +
    xlab(expression(alpha)) +
    ylab(paste("Frac. genes with ", 
               parameter2label[[parameter]], 
               " above ", 
               cut, 
               sep = "")) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position = "none")

  plot_file = paste(plots_dir, 
                    "/a_vs_fraction_genes_above_", 
                    parameter, 
                    "_cut_lessdatasets.png", 
                    sep = "")
  ggsave(
    plot_file,
    plot = scatter_plots_lessdatasets[[parameter]],
    dpi = 1200,
    width = 6000,
    #height = 6000,
    units = c("px")
  )
  print(scatter_plots_lessdatasets[[parameter]])
}
```

Make figures for Supplementary Information:

```{r}
layout <- "
ABC
DFG
"

((distribution_plots_lessdatasets[["mean"]] + labs(tag = 'A')) +
  (distribution_plots_lessdatasets[["sd"]] + labs(tag = 'B')) +
  (distribution_plots_lessdatasets[["cv"]] + labs(tag = 'C')) +
  ((scatter_plots_lessdatasets[["mean"]] + 
    geom_abline(aes(slope = coef(lm_summary_avspar[["mean"]])[2],
                  intercept = coef(lm_summary_avspar[["mean"]])[1]),
              col = "gray50",
              linetype = "dashed", 
              show.legend = FALSE)) + labs(tag = 'D')) +
  ((scatter_plots_lessdatasets[["sd"]] +
    geom_abline(aes(slope = coef(lm_summary_avspar[["sd"]])[2],
                  intercept = coef(lm_summary_avspar[["sd"]])[1]),
              col = "gray50",
              linetype = "dashed", 
              show.legend = FALSE)) + labs(tag = 'E')) +
  (scatter_plots_lessdatasets[["cv"]] + labs(tag = 'F'))) +
  plot_layout(design = layout) &
  theme(plot.tag = element_text(face = 'bold', size = 17),
        plot.tag.position  = c(.1, 1.02),
        plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))

plot_file = paste(plots_dir, "/cv_results_SI.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  width = 17500,
  #height = 12500,
  height = 12000,
  units = c("px")
)

```

#### CV vs. sample size

We calculate the sample size necessary to reveal 50% of the significant correlations:

```{r}
calculate_optimal_N <- function(L, a, b, s, N_guess=c(10000)) {
  optimize_N = function(par){
    N = par[1]
    f = s - L * exp((b * N ** (-a + 1)) / (-a + 1))
    return((abs(f)))
  }
  res = optim(par=N_guess, fn=optimize_N, method="Brent", lower=3, upper=50000)
  return(res$par)
}
```

```{r}
model_selected = "Stretched exponential (by linear fit)"
cols = c("dataset_name", "s", "a", "N_opt")
fraction_links <- 0.5

N_opt_df = data.frame(matrix(ncol=length(cols),nrow=0, dimnames=list(NULL, cols)))
for (dataset_selected in unique(analytical_model_summary_df$dataset_name)){
  analytical_params_selected = analytical_model_summary_df %>%
    filter((dataset_name == dataset_selected) & (model == model_selected))
  N_opt <- calculate_optimal_N(L = analytical_params_selected$L, 
                               a = analytical_params_selected$a,
                               b = analytical_params_selected$b, 
                               s = fraction_links * analytical_params_selected$L)
                               # Why multiply 0.5 x L? Because the normalized result (e.g., 0.5) is the result of:
                               # 0.5 = (s * max_value_in_dataset / max_value_in_dataset) / L = s / L
                               # --> s = 0.5 * L
  N_opt_df = rbind(N_opt_df, 
                   data.frame(dataset_name = dataset_selected, 
                              s = fraction_links, 
                              a = analytical_params_selected$a, 
                              N_opt = N_opt))
}

N_opt_file = paste(tables_dir,
                   "/optimal_sample_size_",
                   fraction_links,
                   "links",
                   sep = "")

N_opt_df %>% 
  fwrite(N_opt_file)

N_opt_df
```

Then, we match the optimal sample size to obtain 50% of significant correlations with the CV of the dataset.

```{r}
sd_cut <- 10
mean_cut <- 10
cv_cut <- 50

datasets_selected <- c("scipher:scipher.sample.per.patient.baseline",
                       "gse193677:rectum-disease_inflamed_vs_control_cd",
                       "gse193677:rectum-disease_inflamed_vs_control_uc",
                       "gse193677:rectum-disease_inflamed_vs_control_control",
                       "gse193677:rectum-inflamed_vs_noninflamed_inflamed",
                       "gse193677:rectum-inflamed_vs_noninflamed_noninflamed",
                       "gtex:breast.mammary.tissue_female",
                       "gtex:lung",
                       "gtex:skin.sun.exposed.lower.leg",
                       "gtex:thyroid",
                       "gtex:whole.blood",
                       "tcga:tcga-brca-tumor_female",
                       "tcga:tcga-kirc-tumor",
                       "tcga:tcga-kirp-tumor",
                       "tcga:tcga-luad-tumor",
                       "tcga:tcga-lusc-tumor",
                       "tcga:tcga-thca-tumor",
                       "tcga:breast-tissue_female", 
                       "tcga:kidney-tissue",
                       "tcga:lung-tissue",
                       "tcga:brca.luma-subtype",
                       "tcga:brca.lumb-subtype")

sd_genes_vs_ss_df <- sd_genes_summary_df %>% 
  inner_join(N_opt_df, by=c("dataset_name")) %>%
  dplyr::filter(dataset_name %in% datasets_selected) %>%
  unique() %>%
  as.data.frame()

sd_genes_vs_ss_file = paste(tables_dir,
                            "/sd_vs_optimal_sample_size_",
                            fraction_links,
                            "links",
                            sep = "")

sd_genes_vs_ss_df %>% 
  fwrite(sd_genes_vs_ss_file)

# Modify dataset names for the plots
sd_genes_vs_ss_df = sd_genes_vs_ss_df %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "tcga:tcga-brca-tumor_female", 
                   "TCGA: Breast cancer")) %>%
  mutate(dataset_name = 
           replace(dataset_name, 
                   dataset_name == "gtex:breast.mammary.tissue_female", 
                   "GTEx: Breast")) %>%
  # mutate(dataset_name = 
  #          replace(dataset_name, 
  #                  dataset_name == "scipher:scipher.sample.per.patient.baseline", 
  #                  "R. arthritis")) %>%
  # mutate(dataset_name = 
  #          replace(dataset_name, 
  #                  dataset_name == "gtex:whole.blood",
  #                  "GTEx: Whole blood")) %>%
  # #mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:thyroid", "GTEx: Thyroid")) %>%
  # mutate(dataset_name = 
  #          replace(dataset_name, 
  #                  dataset_name == "gtex:lung", 
  #                  "GTEx: Lung")) %>%
  # #mutate(dataset_name = replace(dataset_name, dataset_name == "gtex:artery.tibial", "GTEx: Artery tibial")) %>%
  # #mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-coad", "TCGA: Colon cancer")) %>%
  # #mutate(dataset_name = replace(dataset_name, dataset_name == "tcga:tcga-thca", "TCGA: Thyroid cancer")) %>%
  # mutate(dataset_name = 
  #          replace(dataset_name, 
  #                  dataset_name == "tcga:tcga-luad-tumor", 
  #                  "TCGA: Lung cancer")) %>%
  left_join( (name2color %>% 
                as.data.frame() %>% 
                dplyr::rename("rgb_col"=".") %>% 
                tibble::rownames_to_column("dataset") #%>% 
                #filter(dataset_name %in% c("TCGA: Breast cancer", "GTEx: Breast"))
              ), 
             by=c("dataset") ) %>%
  mutate(rgb_col = replace(rgb_col, is.na(rgb_col), "#000000")) %>% # Non-selected datasets with color black
  mutate(geom_text_repel_label=dplyr::if_else(dataset_name == "R. arthritis", "R. arthritis", 
                                              dplyr::if_else(dataset_name == "GTEx: Whole blood", "GTEx: Whole blood", 
                                                             dplyr::if_else(dataset_name == "GTEx: Lung", "GTEx: Lung", 
                                                                            dplyr::if_else(dataset_name == "GTEx: Breast", "GTEx: Breast", 
                                                                                           dplyr::if_else(dataset_name == "TCGA: Lung cancer", "TCGA: Lung cancer", 
                                                                                                          dplyr::if_else(dataset_name == "TCGA: Breast cancer", "TCGA: Breast cancer", "")))))))

sd_genes_vs_ss_df
```

We plot the results of log10 alpha vs log10 sample size:

```{r}
# Calculate correlation
ss_vs_a_cor <- cor(log10(sd_genes_vs_ss_df$N_opt),
                   log10(sd_genes_vs_ss_df$a))

# Calculate regression
lm_summary_ssvsa <- summary(lm(log10(sd_genes_vs_ss_df$a) ~ 
                           log10(sd_genes_vs_ss_df$N_opt)))
slope_ssvsa <- coef(lm_summary_ssvsa)[2]
intercept_ssvsa <- coef(lm_summary_ssvsa)[1]
adj.r.squared_ssvsa <- lm_summary_ssvsa$adj.r.squared
lm_cor_result_ssvsa <- paste("R =", 
                       round(adj.r.squared_ssvsa, 3), 
                       "\ncorr. =", 
                       round(ss_vs_a_cor, 3))

a_vs_ss_plot <- sd_genes_vs_ss_df %>%
    ggplot(aes(x = log10(N_opt), 
               y = log10(a), 
               col = rgb_col)) + 
    geom_point(alpha = 0.6, size = 3) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, 
                         label = geom_text_repel_label),
                     fill = "white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     max.iter = 1e5,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding = 0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    new_scale_color() +
    geom_abline(aes(slope = slope_ssvsa,
                    intercept = intercept_ssvsa,
                    col = lm_cor_result_ssvsa),
              linetype = "dashed", 
              show.legend = FALSE) +
    scale_color_manual(values = c("gray50")) +
    #annotate("text", x = 4500, y = 2.05, label = lm_cor_result_ssvsa, size = 5) +
    annotate("text", x = 3.3, y = 0.32, label = lm_cor_result_ssvsa, size = 5) +
    xlab("log10(n) to get 50% sig. links") +
    ylab(expression(bold(log10(alpha)))) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")

plot_file = paste(plots_dir, "/ss_vs_a.png", sep="")
ggsave(
  plot_file,
  plot = a_vs_ss_plot,
  dpi = 1200,
  width = 8000,
  #height = 6000,
  units = c("px")
)
print(a_vs_ss_plot)
```

We plot the results of alpha vs sample size (lin-lin):

```{r}
# Calculate correlation
ss_vs_a_cor <- cor(sd_genes_vs_ss_df$N_opt,
                   sd_genes_vs_ss_df$a)

# Calculate regression
lm_summary_ssvsa_linlin <- summary(lm(sd_genes_vs_ss_df$a ~ 
                           sd_genes_vs_ss_df$N_opt))
slope_ssvsa_linlin <- coef(lm_summary_ssvsa_linlin)[2]
intercept_ssvsa_linlin <- coef(lm_summary_ssvsa_linlin)[1]
adj.r.squared_ssvsa <- lm_summary_ssvsa_linlin$adj.r.squared
lm_cor_result_ssvsa <- paste("R =", 
                       round(adj.r.squared_ssvsa, 3), 
                       "\ncorr. =", 
                       round(ss_vs_a_cor, 3))

a_vs_ss_lin_lin_plot <- sd_genes_vs_ss_df %>%
    ggplot(aes(x = N_opt, 
               y = a, 
               col = rgb_col)) + 
    geom_point(alpha = 0.6, size = 3) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, 
                         label = geom_text_repel_label),
                     fill = "white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     max.iter = 1e5,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    new_scale_color() +
    geom_abline(aes(slope = slope_ssvsa_linlin,
                    intercept = intercept_ssvsa_linlin,
                    col = lm_cor_result_ssvsa),
              linetype = "dashed", 
              show.legend = FALSE) +
    scale_color_manual(values = c("gray50")) +
    #annotate("text", x = 4500, y = 2.05, label = lm_cor_result_ssvsa, size = 5) +
    annotate("text", x = 4000, y = 2, label = lm_cor_result_ssvsa, size = 5) +
    xlab("n to get 50% sig. links") +
    ylab(expression(bold(alpha))) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")
print(a_vs_ss_lin_lin_plot)
```

We plot the results of log10 alpha vs sample size:

```{r}
# Calculate correlation
ss_vs_a_cor <- cor(sd_genes_vs_ss_df$N_opt,
                   log10(sd_genes_vs_ss_df$a))

# Calculate regression
lm_summary_ssvsa_loglin <- summary(lm(log10(sd_genes_vs_ss_df$a) ~ 
                           sd_genes_vs_ss_df$N_opt))
slope_ssvsa_loglin <- coef(lm_summary_ssvsa_loglin)[2]
intercept_ssvsa_loglin <- coef(lm_summary_ssvsa_loglin)[1]
adj.r.squared_ssvsa <- lm_summary_ssvsa_loglin$adj.r.squared
lm_cor_result_ssvsa <- paste("R =", 
                       round(adj.r.squared_ssvsa, 3), 
                       "\ncorr. =", 
                       round(ss_vs_a_cor, 3))

a_vs_ss_log_lin_plot <- sd_genes_vs_ss_df %>%
    ggplot(aes(x = N_opt, 
               y = log10(a), 
               col = rgb_col)) + 
    geom_point(alpha = 0.6, size = 3) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, 
                         label = geom_text_repel_label),
                     fill = "white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     max.overlaps = Inf,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    new_scale_color() +
    geom_abline(aes(slope = slope_ssvsa_loglin,
                    intercept = intercept_ssvsa_loglin,
                    col = lm_cor_result_ssvsa),
              linetype = "dashed", 
              show.legend = FALSE) +
    scale_color_manual(values = c("gray50")) +
    #annotate("text", x = 4500, y = 2.05, label = lm_cor_result_ssvsa, size = 5) +
    annotate("text", x = 4000, y = 0.3, label = lm_cor_result_ssvsa, size = 5) +
    xlab("n to get 50% sig. links") +
    ylab(expression(bold(log10(alpha)))) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")
print(a_vs_ss_log_lin_plot)
```


We plot the results of alpha vs log10 sample size:

```{r}
# Calculate correlation
ss_vs_a_cor <- cor(log10(sd_genes_vs_ss_df$N_opt),
                   sd_genes_vs_ss_df$a)

# Calculate regression
lm_summary_ssvsa_linlog <- summary(lm(sd_genes_vs_ss_df$a ~ 
                           log10(sd_genes_vs_ss_df$N_opt)))
slope_ssvsa_linlog <- coef(lm_summary_ssvsa_linlog)[2]
intercept_ssvsa_linlog <- coef(lm_summary_ssvsa_linlog)[1]
adj.r.squared_ssvsa <- lm_summary_ssvsa_linlog$adj.r.squared
lm_cor_result_ssvsa <- paste("R =", 
                       round(adj.r.squared_ssvsa, 3), 
                       "\ncorr. =", 
                       round(ss_vs_a_cor, 3))

a_vs_ss_lin_log_plot <- sd_genes_vs_ss_df %>%
    ggplot(aes(x = log10(N_opt), 
               y = a, 
               col = rgb_col)) + 
    geom_point(alpha = 0.6, size = 3) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, 
                         label = geom_text_repel_label),
                     fill = "white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     max.iter = 1e5,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    new_scale_color() +
    geom_abline(aes(slope = slope_ssvsa_linlog,
                    intercept = intercept_ssvsa_linlog,
                    col = lm_cor_result_ssvsa),
              linetype = "dashed", 
              show.legend = FALSE) +
    scale_color_manual(values = c("gray50")) +
    #annotate("text", x = 4500, y = 2.05, label = lm_cor_result_ssvsa, size = 5) +
    annotate("text", x = 3.3, y = 2, label = lm_cor_result_ssvsa, size = 5) +
    xlab("log10(n) to get 50% sig. links") +
    ylab(expression(bold(alpha))) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")
plot_file = paste(plots_dir, "/ss_vs_a_lin_log.png", sep="")
ggsave(
  plot_file,
  plot = a_vs_ss_lin_log_plot,
  dpi = 1200,
  width = 8000,
  #height = 6000,
  units = c("px")
)
print(a_vs_ss_lin_log_plot)
```

We plot all results of alpha vs. sample size together:

```{r}

layout <- "
AB
CD
"

((a_vs_ss_lin_lin_plot +
    labs(tag = 'A') +
    theme(plot.tag.position = c(.07, 1.04),
          plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
  ) +
    (a_vs_ss_lin_log_plot +
       labs(tag = 'B') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (a_vs_ss_log_lin_plot +
       labs(tag = 'C') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (a_vs_ss_plot +
       labs(tag = 'D') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     )) +
  plot_layout(design = layout) &
  theme(plot.tag = element_text(face = 'bold', size = 17))

plot_file = paste(plots_dir, "/ss_vs_a_all_combinations.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  width = 12000,
  height = 11000,
  units = c("px")
)

```

We plot the results of CV vs sample size:

```{r}
# Calculate correlation
cv_vs_ss_cor <- cor(sd_genes_vs_ss_df$N_opt, 
                   sd_genes_vs_ss_df$fraction_genes_above_cv_cut)

# Calculate regression
lm_summary_cvvsss <- summary(lm(sd_genes_vs_ss_df$fraction_genes_above_cv_cut ~ 
                           sd_genes_vs_ss_df$N_opt))
slope_cvvsss <- coef(lm_summary_cvvsss)[2]
intercept_cvvsss <- coef(lm_summary_cvvsss)[1]
adj.r.squared_cvvsss <- lm_summary_cvvsss$adj.r.squared
lm_cor_result_cvvsss <- paste("R =", 
                       round(adj.r.squared_cvvsss, 3), 
                       "\ncorr. =", 
                       round(cv_vs_ss_cor, 3))


cv_vs_ss_plot <- sd_genes_vs_ss_df %>%
    ggplot(aes(x = N_opt, 
               y = fraction_genes_above_cv_cut, 
               col = rgb_col)) + 
    geom_point(alpha = 0.6, size = 3) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, 
                         label = geom_text_repel_label),
                     fill = "white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     max.overlaps = Inf,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    new_scale_color() +
    geom_abline(aes(slope = slope_cvvsss,
                    intercept = intercept_cvvsss,
                    col = lm_cor_result_cvvsss),
              linetype = "dashed", 
              show.legend = FALSE) +
    scale_color_manual(values = c("gray50")) +
    annotate("text", x = 4500, y = 0.41, label = lm_cor_result_cvvsss, size = 5) +
    xlab("Num. samples for 50% sig. links") +
    ylab(paste("Frac. genes with CV above ", 
               cv_cut, 
               sep = "")) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")

plot_file = paste(plots_dir, "/ss_vs_fraction_genes_above_cv_cut.png", sep="")
ggsave(
  plot_file,
  plot = cv_vs_ss_plot,
  dpi = 1200,
  width = 8000,
  #height = 6000,
  units = c("px")
)
print(cv_vs_ss_plot)
```

We plot the results of SD vs sample size:

```{r}
# Calculate correlation
sd_vs_ss_cor <- cor(sd_genes_vs_ss_df$N_opt, 
                   sd_genes_vs_ss_df$fraction_genes_above_sd_cut)

# Calculate regression
lm_summary_sdvsss <- summary(lm(sd_genes_vs_ss_df$fraction_genes_above_sd_cut ~ 
                           sd_genes_vs_ss_df$N_opt))
slope_sdvsss <- coef(lm_summary_sdvsss)[2]
intercept_sdvsss <- coef(lm_summary_sdvsss)[1]
adj.r.squared_sdvsss <- lm_summary_sdvsss$adj.r.squared
lm_cor_result_sdvsss <- paste("R =", 
                       round(adj.r.squared_sdvsss, 3), 
                       "\ncorr. =", 
                       round(sd_vs_ss_cor, 3))


sd_vs_ss_plot <- sd_genes_vs_ss_df %>%
    ggplot(aes(x = N_opt, 
               y = fraction_genes_above_sd_cut, 
               col = rgb_col)) + 
    geom_point(alpha = 0.6, size = 3) +
    scale_colour_identity() +
    geom_label_repel(aes(col = rgb_col, 
                         label = geom_text_repel_label),
                     fill = "white",
                     family = font_selected,
                     box.padding = unit(0.75, "cm"),
                     #max.overlaps = Inf,
                     max.iter = 1e5,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.6, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm = TRUE,
                     seed = 1234) +
    new_scale_color() +
    geom_abline(aes(slope = slope_sdvsss,
                    intercept = intercept_sdvsss,
                    col = lm_cor_result_sdvsss),
              linetype = "dashed", 
              show.legend = FALSE) +
    scale_color_manual(values = c("gray50")) +
    annotate("text", x = 3500, y = 0.93, label = lm_cor_result_sdvsss, size = 5) +
    xlab("Num. samples for 50% sig. links") +
    ylab(paste("Frac. genes with SD above ", 
               sd_cut, 
               sep = "")) +
    theme_linedraw() +
    theme(aspect.ratio = 1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none")

plot_file = paste(plots_dir, "/ss_vs_fraction_genes_above_sd_cut.png", sep="")
ggsave(
  plot_file,
  plot = sd_vs_ss_plot,
  dpi = 1200,
  width = 8000,
  #height = 6000,
  units = c("px")
)
print(sd_vs_ss_plot)
```

#### Figure 2: Panel containing power law demonstration plots

Create a panel with all the results:

```{r}
# # Use patchwork to concatenate plots
# (((curve_plot + labs(tag = 'A')) |
#   (scaling_relation_plot + labs(tag = 'B')) |
#   (comparison_plot + labs(tag = 'C'))) /
#   (wrap_elements(figure_summary_table_plot) + labs(tag='D')) / 
#   ((prediction_plot + labs(tag = 'E')) |
#   (a_vs_fraction_sig_correlations_plot + labs(tag = 'F')) |
#   (scatter_plots[[3]] + labs(tag='G')))) & 
#   theme(element_text(face = 'bold', size = 17)) # bold tags

layout <- "
ABC
DDD
EFG
"

((goodnessfit_plot_lessdatasets +
    labs(tag = 'A') +
    theme(plot.tag.position = c(.09, 1.04),
          plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
  ) +
    (scaling_relation_plot_comparison +
       labs(tag = 'B') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (prediction_plot_lessdatasets +
       labs(tag = 'C') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (wrap_elements(figure_summary_table_plot) +
       labs(tag='D') +
       theme(plot.tag.position = c(.20, .87),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (a_vs_fraction_sig_correlations_lessdatasets_plot +
       labs(tag = 'E') +
       theme(plot.tag.position = c(.09, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (scatter_plots_lessdatasets[["sd"]] +
      geom_abline(
        aes(
          slope = coef(lm_summary_avspar[["sd"]])[2],
          intercept = coef(lm_summary_avspar[["sd"]])[1]
        ),
        col = "gray50",
        linetype = "dashed",
        show.legend = FALSE
      ) +
    labs(tag = 'F') +
    theme(plot.tag.position = c(.04, 1.04),
          plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (a_vs_ss_lin_log_plot +
       labs(tag='G')
     ) +
    theme(plot.tag.position  = c(.04, 1.04),
          plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
  ) +
  plot_layout(design = layout) &
  theme(plot.tag = element_text(face = 'bold', size = 17))
  
plot_file = paste(plots_dir, "/modelling_and_cv_results.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  width = 20500,
  #height = 12500,
  height = 18000,
  units = c("px")
)

# (((curve_plot +
#       labs(tag = 'A') +
#       theme(legend.position="none")) | 
#     (comparison_plot +
#       labs(tag = 'D') +
#       guides(col=guide_legend(title="Model", title.position="top")) +
#       theme(legend.position = c(0.7, 0.24), legend.text = element_text(size = 11), legend.title=element_text(size=13, face="bold"), legend.background = element_rect(size=0.2, linetype="solid", colour ="black"))) |
#     (a_vs_fraction_sig_correlations_plot +
#      labs(tag = 'F') +
#      theme(legend.position="none"))) /
#   ((scaling_relation_plot +
#       labs(tag = 'B') +
#       theme(legend.position="none")) | 
#     (prediction_plot +
#       labs(tag = 'E') +
#       #guides(col=guide_legend(title="Dataset", title.position="top", nrow=3,byrow=TRUE)) +
#       #theme(legend.position = 'bottom', legend.text = element_text(size = 11), legend.title=element_text(size=13, face="bold"))) |
#       theme(legend.position = 'none')) |
#      (distribution_plots[[3]] +
#      labs(tag = 'G') +
#      theme(legend.position="none"))) /
#    ((wrap_elements(figure_summary_table_plot) + labs(tag='C')) | 
#       (scatter_plots[[3]] + labs(tag='H')))) & 
#   #plot_annotation(tag_levels = 'A') & # Include letters 
#   theme(element_text(face = 'bold', size = 17)) # that are bold
```


#### P-value variation plot

```{r}
dataset <- "gtex"
type_dataset <- "Whole.Blood"
sizes_selected <- c(20, 60, 100, 200, 300)
num_links_to_select <- 1E5
coexpression_dir <- sprintf("/scratch/j.aguirreplans/Scipher/SampleSizeProject/networks_%s/reads/%s/consensus", dataset, type_dataset)
```

Read the co-expression networks and extract the co-expression p-values.

```{r}
pvalue_variation_file <- sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/pval_variation_analysis_%s_%s.txt", dataset, type_dataset)
pvalue_df <- data.frame()
if (!(file.exists(pvalue_variation_file))) {
  
  # Fetch the files containing the gene co-expression networks of different
  # sample sizes
  all_coexpression_files <- list.files(coexpression_dir)
  coexpression_files_selected <- data.frame(matrix(ncol = length(c("size", "file")), nrow = 0, dimnames = list(NULL, c("size", "file"))))

  for (coexpression_file in all_coexpression_files){
    file_split <- strsplit(
      gsub(".net", "", coexpression_file),
      split = "_"
    )[[1]]
    size <- as.numeric(file_split[(length(file_split)-1)])
    if (size %in% sizes_selected){
      coexpression_files_selected <- rbind(
        coexpression_files_selected,
        data.frame(
          size = size,
          file = paste(coexpression_dir, coexpression_file, sep = "/")
        )
      )
    }
  }  

  # Read the co-expression networks and extract the co-expression p-values
  sizes_selected <- sort(unique(coexpression_files_selected$size))
  for (i in seq_len(length(sizes_selected))) {
    size_selected <- sizes_selected[i]

    coexpression_file <- (coexpression_files_selected %>% filter(size == size_selected))$file
    coexpression_df <- data.table::fread(coexpression_file) %>%
      dplyr::mutate(strength = dplyr::case_when(
        CN >= 0.8 ~ "Very Strong",
        CN < 0.8 & CN >= 0.6 ~ "Strong",
        CN < 0.6 & CN >= 0.4 ~ "Moderate",
        CN < 0.4 & CN >= 0.2 ~ "Weak",
        TRUE ~ "Very Weak"
      )) %>%
      dplyr::select("Node.1", "Node.2", "strength", "pval.fisher") %>%
      dplyr::rename(!!paste("pval", size_selected, sep = ".") := "pval.fisher")

    if (i == 1) {
      pvalue_df <- as.data.frame(coexpression_df)
    } else {
      pvalue_df <- pvalue_df %>%
        dplyr::inner_join(
          coexpression_df,
          by = c("Node.1", "Node.2", "strength")
        )
    }
    rm(coexpression_df)
  }

  # Select 1M links from each correlation strength
  pvalue_df <- pvalue_df %>%
    dplyr::group_by(strength) %>%
    dplyr::slice_sample(n = num_links_to_select, replace = FALSE) %>%
    dplyr::ungroup() %>%
    as.data.frame()

  pvalue_df %>% data.table::fwrite(pvalue_variation_file, sep = "\t")
} else {
  pvalue_df <- data.table::fread(pvalue_variation_file, sep = "\t")
}
```

Create violin plot showing distributions of p-values across sample sizes and correlation strengths:

```{r}
violinplot_pvalue_variation_file <-  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/violin_pvalue_variation_pearson_%s_%s.png", dataset, type_dataset)

violinplot_pvalue_variation <- pvalue_df %>%
  # Transform data frame to long format
  tidyr::pivot_longer(
    !(c("Node.1", "Node.2", "strength")),
    names_prefix = "pval.",
    names_to = "size",
    values_to = "pval",
    names_transform = list(size = as.character),
    values_transform = list(pval = as.numeric)
  ) %>%
  # Transform p-values lower than 1E-20 to 1E-20 so that there are no p-values
  # equal to 0 and the visualization improves
  dplyr::mutate(
    pval = ifelse(pval <= 1E-20, 1E-20, pval),
    strength = factor(
      strength,
      levels = c("Very Strong", "Strong", "Moderate", "Weak", "Very Weak")
    )
  ) %>%
  ggplot(aes(
    x = reorder(size, as.numeric(size)),
    y = -log(pval),
    fill = strength,
    color = strength
  )) +
  geom_violin(
    alpha = 0.3, 
    scale = "width", 
    adjust = 0.5
  ) +
  geom_hline(
    aes(
      yintercept = -log(0.05),
      linetype = "P-value = 0.05"
    ),
    color = "red",
    show.legend = TRUE
  ) +
  scale_fill_manual(values = name2color) +
  scale_color_manual(values = name2color) +
  scale_linetype_manual(
    values = "dashed"
  ) +
  labs(
    x = "Num. samples",
    y = "-Log( P-value )",
    color = "Correlation strength",
    fill = "Correlation strength",
    linetype = ""
  ) +
  guides(
    color = guide_legend(
      override.aes = list(color = NA) # Remove the red horizontal line
    )
  ) +
  theme_linedraw() +
  theme(
    aspect.ratio = 1,
    plot.title =  element_text(size = 20, face = "bold"), 
    axis.title = element_text(size = 17, face = "bold"), 
    axis.text = element_text(size = 16),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    legend.text = element_text(size = 16), 
    legend.title = element_text(size = 16, face = "bold"),
    text = element_text(family = font_selected)
  )

rm(pvalue_df)
print(violinplot_pvalue_variation)

ggsave(
  violinplot_pvalue_variation_file,
  plot = violinplot_pvalue_variation,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
```


#### Levels of correlations plot

Define function to calculate critical sample size for correlation:

```{r}
calculate_sample_size_for_pearson_correlation_using_t_distribution_with_optimization <- function(r, total_num_edges, alpha=0.05, N_guess=c(10000)){
  optimize_N = function(par){
    N = par[1]
    t_guess = qt(alpha, df=N-2, lower.tail = F)
    t = r * sqrt((N-2)) / sqrt((1-r**2))
    return((abs(t-t_guess)))
  }
  res = optim(par=N_guess, fn=optimize_N, method="Brent", lower=3, upper=50000)
  return(res$par)
}

calculate_sample_size_for_pearson_correlation_using_z_distribution_with_power <- function(r, total_num_edges, alpha=0.05, power=0.8){
  alpha_corrected = alpha / total_num_edges # Correct alpha by multiple error
  Za = qnorm(alpha, lower.tail=F)
  Za_corrected = qnorm(alpha_corrected, lower.tail=F)
  Zb = qnorm(power, lower.tail=F)
  N = ((Za + Zb) / (0.5 * log((1+r)/(1-r))))**2 + 3
  N_corrected = ((Za_corrected + Zb) / (0.5 * log((1+r)/(1-r))))**2 + 3
  return(list(N=N, N_corrected=N_corrected))
}

calculate_sample_size_for_pearson_correlation_using_z_distribution <- function(r, alpha=0.05){
  Za = qnorm(alpha, lower.tail=F)
  N = r / (2*Za - log((1+r)/(1-r))) + 1
  return(N)
}

calculate_sample_size_for_pearson_correlation_using_bonnet_and_wright <- function(r, w, alpha=0.05){
  Za = qnorm(alpha, lower.tail=F)
  N0 = round(4 * (1-r**2)**2 * (Za/w)**2)
  L1 = 0.5 * (log(1+r) - log(1-r)) - Za/sqrt(N0-3)
  L2 = 0.5 * (log(1+r) - log(1-r)) + Za/sqrt(N0-3)
  w0 = ((exp(2*L2) - 1) / (exp(2*L2) + 1)) - ((exp(2*L1) - 1) / (exp(2*L1) + 1))
  N = ((N0 - 3) * (w0 / w)**2) + 3
  return(N)
}

calculate_predictions_using_stretched_exponential_model_optimized <- function(x, L, a, b){
  y = L * exp((b*x**(-a+1))/(-a+1))
  #y = exp(log(L) - exp(b+a*log(x)))
  return(y)
}
```

Calculate sample size for different levels of correlation and different data sets:

```{r}
types_correlation <- c("weak", "moderate", "strong", "very strong")
type_correlation_df <- data.frame(
  type_correlation = types_correlation,
  lower_val = c(0.2,0.4,0.6,0.8),
  upper_val = c(0.4,0.6,0.8,NA)
)

cols <- c("dataset", "type_correlation", "correlation", "sample_size_statistical", "sample_size_statistical_corrected", "sample_size_z")
sample_size_correlation_df <- data.frame(matrix(
  ncol = length(cols),
  nrow = 0,
  dimnames = list(NULL, cols)
))

for (dataset in numbers_df$dataset) {
  total_num_edges <- unique((numbers_df %>% dplyr::filter(dataset == !!dataset))$total_num_edges)
  for (type_correlation in types_correlation) {
    r <- (type_correlation_df %>% filter(type_correlation == !!type_correlation))$lower_val
  
    N <- calculate_sample_size_for_pearson_correlation_using_t_distribution_with_optimization(
      r = r,
      total_num_edges = total_num_edges,
      alpha = 0.05,
      N_guess = c(10000)
    )
  
    N_corrected <- calculate_sample_size_for_pearson_correlation_using_t_distribution_with_optimization(
      r = r,
      total_num_edges = total_num_edges,
      alpha = 0.05 / total_num_edges,
      N_guess = c(10000)
    )
  
    N_z <- calculate_sample_size_for_pearson_correlation_using_bonnet_and_wright(
      r = r,
      w = 0.8,
      alpha = 0.05 / total_num_edges
    )
  
    sample_size_correlation_df <- rbind(
      sample_size_correlation_df,
      data.frame(
        dataset = dataset,
        type_correlation = type_correlation,
        correlation = r,
        sample_size_statistical = N,
        sample_size_statistical_corrected = N_corrected,
        sample_size_z = N_z
      )
    )
  }
}

print(sample_size_correlation_df)
```

Calculate predictions from analytical model for GTEx whole blood data set at predicted optimal sample sizes:

```{r}
dataset_selected <- "gtex:whole.blood"
model_selected <- "Stretched exponential (by linear fit)"

# Get analytical model parameters for GTEx whole blood
results_analytical_gtex_wb <- topology_results_selected_analytical_norm_df %>% 
  dplyr::filter(
    (model == model_selected) & (type_dataset == dataset_selected)
  ) %>%
  dplyr::select(model, type_dataset, a, b, L, max_value_in_dataset) %>%
  unique()

# Get optimal sample size for different types of correlations and GTEx whole
# blood data set
sample_size_correlation_gtex_wb <- sample_size_correlation_df %>%
  dplyr::filter((dataset == dataset_selected) & (!(type_correlation == "weak")))

# Calculate fraction of edges for optimal sample sizes calculated before
predictions_convergence_gtex_wb_norm <- calculate_predictions_using_stretched_exponential_model_optimized(
  x = sample_size_correlation_gtex_wb$sample_size_statistical_corrected,
  L = results_analytical_gtex_wb$L,
  a = results_analytical_gtex_wb$a,
  b = results_analytical_gtex_wb$b
)

# Create table to store predictions
predictions_convergence_gtex_wb_df <- data.frame(
  size = round(sample_size_correlation_gtex_wb$sample_size_statistical_corrected),
  num_edges = predictions_convergence_gtex_wb_norm * results_analytical_gtex_wb$max_value_in_dataset,
  num_edges_norm = predictions_convergence_gtex_wb_norm
)

# Include normalized num. of edges
predictions_convergence_gtex_wb_df$num_edges_norm <- (
  predictions_convergence_gtex_wb_df$num_edges / results_analytical_gtex_wb$max_value_in_dataset
) / results_analytical_gtex_wb$L

predictions_convergence_gtex_wb_df

# Select prediction results for GTEx whole blood
predicted_results_wb_df <- predicted_results_df %>%
  dplyr::filter(
    (model == model_selected) & (type_dataset == dataset_selected)
  ) %>%
  dplyr::mutate_at(c('model_result'), as.numeric)

# Normalize model predictions
predicted_results_wb_df$model_result_norm <- (
  predicted_results_wb_df$model_result / results_analytical_gtex_wb$max_value_in_dataset
) / results_analytical_gtex_wb$L
```


```{r}
dataset_selected <- "Whole.Blood"
all_topology_results_file <- '/home/j.aguirreplans/Projects/Scipher/SampleSize/scripts/SampleSizeShiny/data/analysis_topology.csv'
threshold_selected <- 0.05
method_selected <- "pearson"

# Read information about number of edges
all_results_df <- data.table::fread(all_topology_results_file)

# Select number of edges for GTEx whole blood
results_gtex_wb <- all_results_df %>%
  dplyr::filter(
    (type_dataset == dataset_selected) &
      (method == method_selected) &
      (threshold == threshold_selected) &
      (type_correlation %in% c("weak-moderate-strong-very strong", "moderate-strong-very strong", "strong-very strong", "very strong"))
  ) %>%
  dplyr::select("method", "type_dataset", "size", "rep", "type_correlation", "num_edges")

# Normalize number of edges
results_gtex_wb$num_edges_norm <- (
  results_gtex_wb$num_edges / results_analytical_gtex_wb$max_value_in_dataset
) / results_analytical_gtex_wb$L

# Merge info about num. of edges with model predictions and convergence points
correlation_levels_plot_df <- results_gtex_wb %>% 
  dplyr::mutate(type_correlation = dplyr::recode(type_correlation, !!!dataset2name)) %>%
  dplyr::full_join(
    (predicted_results_wb_df %>%
       dplyr::filter(
         size <= max((all_results_df %>% dplyr::filter((type_dataset == dataset_selected) & (threshold == threshold_selected)))$size)
        ) %>%
       dplyr::mutate(type_correlation = "Model prediction")
    ),
    by = c("type_dataset", "size", "type_correlation", "num_edges_norm" = "model_result_norm")
  ) %>%
  dplyr::full_join(
    (predictions_convergence_gtex_wb_df %>%
       dplyr::mutate(type_correlation = "Convergence point")
    ),
    by = c("size", "num_edges", "num_edges_norm", "type_correlation")
  ) %>%
  dplyr::left_join(
    (name2color %>% as.data.frame() %>%
       dplyr::rename("rgb_col" = ".") %>%
       tibble::rownames_to_column("type_correlation")),
    by = c("type_correlation")
  )

correlation_levels_plot <- correlation_levels_plot_df %>%
  ggplot(aes(x = size, y = num_edges_norm)) +
  # Plot num. of edges
  geom_point(
    data = .%>% dplyr::filter(!(type_correlation %in% c("Convergence point", "Model prediction"))),
    aes(
      col = type_correlation,
      size = num_edges_norm
    ),
    alpha = 0.5
  ) +
  scale_color_manual(
    name = "Correlation strength",
    values = as.vector(name2color[levels(factor((correlation_levels_plot_df %>% dplyr::filter(!(type_correlation == "Model prediction")))$type_correlation))])
  ) +
  new_scale_color() +
  # Plot convergence points
  geom_point(
    data = .%>% dplyr::filter(type_correlation == "Convergence point"),
    aes(
      col = factor(type_correlation)
    ),
    alpha = 1,
    size = 3
  ) +
  scale_color_manual(name = NULL, values = c("red")) +
  new_scale_color() +
  # Plot model prediction
  geom_line(
    data = .%>% dplyr::filter(type_correlation == "Model prediction"),
    aes(col = type_correlation),
    linewidth = 1
  ) +
  scale_color_manual(name = NULL, values = c("red")) +
  theme_linedraw() +
  xlab("Num. samples") +
  ylab("Frac. significant correlations") +
  guides(
    col = guide_legend(
      override.aes = list(alpha = 1, size = 3)
    ),
    col = guide_legend(title=NULL),
    size = guide_legend(title="Frac. sig. correlations")
  ) +
  theme(
    aspect.ratio = 1,
    plot.title = element_text(size = 20, face="bold"), 
    axis.title = element_text(size = 17, face="bold"), 
    axis.text = element_text(size = 16),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    legend.text = element_text(size = 16), 
    legend.title = element_text(size=16, face = "bold"),
    text = element_text(family = font_selected)
  )

plot_file <- paste(
  plots_dir,
  "correlation_levels_pearson_pval_0.05_data_gtexwholeblood.png",
  sep = "/"
)
ggsave(
  plot_file,
  plot = correlation_levels_plot,
  dpi = 1200,
  width = 9000,
  height = 6000,
  units = c("px")
)
print(correlation_levels_plot)
```


#### Plot with 5 specific links

Read data: 

```{r}
input_data_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/analysis_specific_edges_pearson_gtex_Whole.Blood.txt"
input_df <- data.table::fread(input_data_file)
input_df$link = paste(input_df$Node.1, input_df$Node.2, sep="---")
input_df = input_df %>% mutate(pval.adj = replace(pval.adj, pval.adj < 0.000001, 0.000001))
head(input_df)
```

Calculate mean and SD of correlation among replicates:

```{r}
input_mean_df = input_df %>% 
  group_by(link, method, size) %>%
  summarize(mean_score = mean(abs(score)), sd_score = sd(abs(score)), min_score=min(abs(score)), max_score=max(abs(score)), mean_pval = mean(pval.adj), sd_pval = sd(pval.adj), min_pval=min(pval.adj), max_pval=max(pval.adj)) %>%
  ungroup()
head(input_mean_df)
```

Select the links of different correlation levels:

```{r}
set.seed(2106)
max_size_links_df = input_mean_df %>% 
  filter(size == max(size)) %>% 
  mutate(type_correlation="Very Weak") %>% 
  mutate(type_correlation = replace(type_correlation, abs(mean_score) >= 0.2, "Weak")) %>%
  mutate(type_correlation = replace(type_correlation, abs(mean_score) >= 0.4, "Moderate")) %>%
  mutate(type_correlation = replace(type_correlation, abs(mean_score) >= 0.6, "Strong")) %>%
  mutate(type_correlation = replace(type_correlation, abs(mean_score) >= 0.8, "Very Strong"))
num_links_selected = 1
specific_links_df = max_size_links_df %>% select(link, type_correlation, mean_score) %>% filter(type_correlation == "Very Weak") %>% sample_n(size=num_links_selected, replace = FALSE)
specific_links_df = rbind(specific_links_df, max_size_links_df %>% select(link, type_correlation, mean_score) %>% filter(type_correlation == "Weak") %>% filter((abs(mean_score) > 0.25) & (abs(mean_score) < 0.35)) %>% sample_n(size=num_links_selected, replace = FALSE))
specific_links_df = rbind(specific_links_df, max_size_links_df %>% select(link, type_correlation, mean_score) %>% filter(type_correlation == "Moderate") %>% filter((abs(mean_score) > 0.45) & (abs(mean_score) < 0.55)) %>% sample_n(size=num_links_selected, replace = FALSE))
specific_links_df = rbind(specific_links_df, max_size_links_df %>% select(link, type_correlation, mean_score) %>% filter(type_correlation == "Strong") %>% filter((abs(mean_score) > 0.65) & (abs(mean_score) < 0.75)) %>% sample_n(size=num_links_selected, replace = FALSE))
specific_links_df = rbind(specific_links_df, max_size_links_df %>% select(link, type_correlation, mean_score) %>% filter(type_correlation == "Very Strong") %>% sample_n(size=num_links_selected, replace = FALSE))
head(specific_links_df)
```

Make the plot:

```{r}
specific_links_plot_df = input_mean_df %>% 
  inner_join((specific_links_df %>% select(link, type_correlation)), by=c("link")) %>% 
  mutate(max_pval_discrete=if_else(max_pval >= 0.05, "P. adj. \u2265 0.05", "P. adj. < 0.05")) %>% 
  mutate(max_pval_size=if_else(max_pval >= 0.05, "P. adj. \u2265 0.05", if_else(max_pval >= 0.01, "P. adj. < 0.05", if_else(max_pval >= 0.001, "P. adj. < 0.01", "P. adj. < 0.001")))) %>%
  mutate(type_correlation = factor(type_correlation, levels = c("Very Strong", "Strong", "Moderate", "Weak", "Very Weak")))
# name2size <- c(
#   "P. adj. \u2265 0.05" = 0.1,
#   "P. adj. < 0.05" = 1,
#   "P. adj. < 0.01" = 1.5,
#   "P. adj. < 0.001" = 2
#   )

specific_links_plot = specific_links_plot_df %>%
  ggplot(aes(x=size, y=abs(mean_score))) +
  geom_line(aes(col=type_correlation)) +
  #geom_point(data=(input_df %>% filter(link %in% specific_links)), aes(x = size, y = abs(score), col=link), alpha=0.2) +
  geom_ribbon(aes(ymin = abs(min_score), ymax = abs(max_score), fill=type_correlation), alpha=0.3) +
  #geom_point(aes(size=max_pval_size), col="red") +
  #scale_color_manual(name = "Correlation strength", values = c("#ff7b00", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF")) +
  scale_color_manual(name = "Correlation strength", values = as.vector(name2color[levels(factor(specific_links_plot_df$type_correlation))])) + # Plot using the dictionary
  #scale_fill_manual(name = "Correlation strength", values = c("#ff7b00", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF")) +
  scale_fill_manual(name = "Correlation strength", values = as.vector(name2color[levels(factor(specific_links_plot_df$type_correlation))])) + # Plot using the dictionary
  new_scale_color() +
  geom_point(data = ~filter(.x, max_pval < 0.05), aes(x=size, y=abs(mean_score), col=max_pval_discrete)) +
  scale_color_manual(name = NULL, values = c("red")) +
  #geom_errorbar(aes(ymin=abs(mean_score)-sd_score, ymax=abs(mean_score)+sd_score), width=.2, position=position_dodge(0.05)) +
  theme_linedraw() +
  xlab("Num. samples") +
  ylab("Pearson correlation (abs)") +
  #scale_size_manual(values = as.vector(name2size[levels(factor(specific_links_plot_df$max_pval_size))])) +
  #scale_color_manual(values = c("#C77CFF", "#7CAE00", "#FF7B00", "#F8766D", "#00BFC4")) +
  #scale_fill_manual(values = c("#C77CFF", "#7CAE00", "#FF7B00", "#F8766D", "#00BFC4")) +
  #guides(col=guide_legend(title="Correlation strength"), fill=guide_legend(title="Correlation strength")) +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = font_selected))

plot_file = paste(plots_dir, "specific_links_gtex_Whole.Blood.png", sep="/")
ggsave(
  plot_file,
  plot = specific_links_plot,
  dpi = 1200,
  #width = 9000,
  #height = 6000,
  units = c("px")
)
print(specific_links_plot)
```

#### Standard deviation of co-expression value plot

Read data:

```{r}
sd_table_file = paste(tables_dir, "analysis_sd_coexpression_weight_gtex_Whole.Blood.txt", sep="/")
sd_table_df <- data.table::fread(sd_table_file)
```

Plot:

```{r}
# Define palette using the scale_fill_brewer Oranges palette and selecting manually the colors using RColorBrewer
# https://stackoverflow.com/questions/29466239/how-to-set-the-color-range-of-scale-colour-brewer-in-ggplot2-palette-selecte 
my_orange = RColorBrewer::brewer.pal(n = (length(unique(sd_table_df$ss)) + 2), "Oranges")[3:(length(unique(sd_table_df$ss))+2)] # I get 8 and exclude the 2 first colors, which are more light 

sd_plot_file = paste(plots_dir, "analysis_sd_coexpression_weight_gtex_Whole.Blood.png", sep="/")
sd_plot = sd_table_df %>%
  mutate(ss = factor(ss, levels = as.character(sort(as.numeric(unique(sd_table_df$ss)))))) %>%
  ggplot(aes(x = ss, y = sd_correlation)) +
  #geom_half_violin(side = "r", color="orangered4", fill="orangered1") +
  geom_half_violin(aes(fill=ss), color="transparent", side = "r") +
  #scale_fill_brewer(palette="Oranges") +
  scale_fill_manual(values=my_orange) +
  theme_linedraw() +
  theme(aspect.ratio=1, 
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(family = font_selected),
        legend.position="none") +
  labs (x = "Num. samples",
        y = "Pearson correlation SD",
        color = "",
        fill = "")
ggsave(
  sd_plot_file,
  plot = sd_plot,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
print(sd_plot)
rm(sd_table_df)
```

Patchwork:

```{r}
# # Use patchwork to concatenate plots
# (correlation_levels_plot /
#   specific_links_plot /
#   sd_plot) + 
#   plot_annotation(tag_levels = 'A') & # Include letters 
#     theme(plot.tag = element_text(face = 'bold', size = 17))
# 
# plot_file = paste(plots_dir, "/sd_coexpression_results.png", sep="")
# ggsave(
#   plot_file,
#   dpi = 1200,
#   width = 10000,
#   height = 17000,
#   units = c("px")
# )
```

```{r}
layout <- "
AB
CD
"

((violinplot_pvalue_variation +
    labs(tag = 'A') +
    theme(plot.tag.position = c(.07, 1.04),
          plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
  ) +
    (correlation_levels_plot +
       labs(tag = 'B') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (specific_links_plot +
       labs(tag = 'C') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     ) +
    (sd_plot +
       labs(tag = 'D') +
       theme(plot.tag.position = c(.07, 1.04),
             plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))
     )) +
  plot_layout(design = layout) &
  theme(plot.tag = element_text(face = 'bold', size = 17))

plot_file = paste(plots_dir, "/sd_coexpression_results.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  width = 16000,
  height = 11000,
  units = c("px")
)
```


Supplementary figure with the analysis of SD for different datasets:

```{r}
datasets_analysis = c("tcga_TCGA-BRCA_female", "tcga_TCGA-LUAD")
datasets_analysis_titles = c("TCGA: Breast cancer", "TCGA: Lung cancer")
sd_plots = list()
for (x in 1:length(datasets_analysis)){
  dataset_selected = datasets_analysis[x]
  sd_table_file = paste(tables_dir, "/analysis_sd_coexpression_weight_", dataset_selected, ".txt", sep="")
  sd_table_df <- data.table::fread(sd_table_file)
  sd_plot_file = paste(plots_dir, "/analysis_sd_coexpression_weight_", dataset_selected, ".png", sep="")
  my_orange = RColorBrewer::brewer.pal(n = (length(unique(sd_table_df$ss)) + 2), "Oranges")[3:(length(unique(sd_table_df$ss))+2)]
  sd_plot = sd_table_df %>%
    mutate(ss = factor(ss, levels = as.character(sort(as.numeric(unique(sd_table_df$ss)))))) %>%
    ggplot(aes(x = ss, y = sd_correlation)) +
    #geom_half_violin(side = "r", color="orangered4", fill="orangered1") +
    geom_half_violin(aes(fill=ss), color="transparent", side = "r") +
    #scale_fill_brewer(palette="Oranges") +
    scale_fill_manual(values=my_orange) +
    theme_linedraw() +
    theme(aspect.ratio=1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none") +
    labs (x = "Num. samples",
          y = "Pearson correlation SD",
          title = datasets_analysis_titles[[x]],
          color = "",
          fill = "")
  sd_plots[[x]] = sd_plot
  ggsave(
    sd_plot_file,
    plot = sd_plot,
    dpi = 1200,
    width = 6500,
    height = 6500,
    units = c("px")
  )
  print(sd_plot)
  rm(sd_table_df)
}
```

Same plots but for small sample sizes:

```{r}
datasets_analysis = c("tcga_TCGA-BRCA_female", "tcga_TCGA-LUAD")
datasets_analysis_titles = c("TCGA: Breast cancer", "TCGA: Lung cancer")
sd_plots_small_sizes = list()
for (x in 1:length(datasets_analysis)){
  dataset_selected = datasets_analysis[x]
  sd_table_file = paste(tables_dir, "/analysis_sd_coexpression_weight_with_small_sizes_", dataset_selected, ".txt", sep="")
  sd_table_df <- data.table::fread(sd_table_file)
  sd_plot_file = paste(plots_dir, "/analysis_sd_coexpression_weight_with_small_sizes_", dataset_selected, ".png", sep="")
  my_orange = RColorBrewer::brewer.pal(n = (length(unique(sd_table_df$ss)) + 2), "Oranges")[3:(length(unique(sd_table_df$ss))+2)]
  sd_plot_small_sizes = sd_table_df %>%
    mutate(ss = factor(ss, levels = as.character(sort(as.numeric(unique(sd_table_df$ss)))))) %>%
    ggplot(aes(x = ss, y = sd_correlation)) +
    #geom_half_violin(side = "r", color="orangered4", fill="orangered1") +
    geom_half_violin(aes(fill=ss), color="transparent", side = "r") +
    #scale_fill_brewer(palette="Oranges") +
    scale_fill_manual(values=my_orange) +
    theme_linedraw() +
    theme(aspect.ratio=1, 
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16), 
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          text = element_text(family = font_selected),
          legend.position="none") +
    labs (x = "Num. samples",
          y = "Pearson correlation SD",
          title = datasets_analysis_titles[[x]],
          color = "",
          fill = "")
  sd_plots_small_sizes[[x]] = sd_plot_small_sizes
  ggsave(
    sd_plot_file,
    plot = sd_plot_small_sizes,
    dpi = 1200,
    width = 6500,
    height = 6500,
    units = c("px")
  )
  print(sd_plot_small_sizes)
  rm(sd_table_df)
}
```

Patchwork for the supplementary figure:

```{r}
layout <- "
AB
"
patch_breast = ((sd_plots_small_sizes[[1]] + labs(tag = 'A', title = NULL)) +
  (sd_plots[[1]] + labs(tag = 'B', title = NULL))) +
  plot_annotation(title = 'TCGA: Breast cancer', theme=theme(plot.title =  element_text(size = 20, face="bold"))) &
  theme(plot.tag = element_text(face = 'bold', size = 17),
        plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))

layout <- "
CD
"
patch_lung = ((sd_plots_small_sizes[[2]] + labs(tag = 'C', title = NULL)) +
  (sd_plots[[2]] + labs(tag = 'D', title = NULL))) +
  plot_annotation(title = 'TCGA: Lung cancer', theme=theme(plot.title =  element_text(size = 20, face="bold"))) &
  theme(plot.tag = element_text(face = 'bold', size = 17),
        plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm"))

(wrap_elements(patch_breast) / wrap_elements(patch_lung))

plot_file = paste(plots_dir, "/analysis_sd_coexpression_weight_SI.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  width = 12000,
  height = 14000,
  units = c("px")
)
```

#### Differential co-expression analysis plots

```{r}
name_D = "tcga.brca.female"
name_N = "tcga.breast.female"
diffcoexp_plots_dir = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/plots_differential_coexpression/TCGA-BRCA_female___TCGA-Breast_female"
diffcoexp_tables_dir = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/tables_differential_coexpression/TCGA-BRCA_female___TCGA-Breast_female"
name2color_df = name2color %>% as.data.frame() %>% dplyr::rename("rgb_col"=".") %>% tibble::rownames_to_column("name")
```

Read file:

```{r}
change_of_category_file = paste(diffcoexp_tables_dir, paste("codina_", name_D, "_", name_N, "_changes.txt", sep=""), sep="/")
change_of_category_df <- data.table::fread(change_of_category_file)
```

Make plot:

```{r}
sizes_list = sort(unique(change_of_category_df$size.prev))
if (max(sizes_list) <= 160) {
  sizes_to_plot = c(20, 40, 60, 80, 100, 120, 140, 160)
} else if ((max(sizes_list) > 160) & (max(sizes_list) <= 300)) {
  sizes_to_plot = c(20, 60, 100, 140, 180, 220, 260, 300)
} else if ((max(sizes_list) > 300) & (max(sizes_list) <= 440)) {
  sizes_to_plot = c(20, 80, 140, 200, 260, 320, 380, 440)
} else {
  sizes_to_plot = seq(20, max(sizes_list), 100)
}

# Create last column of plot (size 100)
change_of_category_size100_df = change_of_category_df %>%
  filter(size.curr == 100) %>%
  dplyr::select(Node, Phi_name.curr, size.curr) %>%
  dplyr::rename("Phi_name.prev"="Phi_name.curr", "size.prev"="size.curr") %>%
  mutate(Phi_name.curr = NA) %>%
  mutate(size.curr = NA) %>%
  mutate(transition_type = NA) %>%
  mutate(transition_name = NA) %>%
  mutate(transition_name_simple = Phi_name.prev) %>%
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "common", "common (constant)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "disease-specific", "disease-specific (constant)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "normal-specific", "normal-specific (constant)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "undefined", "undefined (constant)")) %>% 
  left_join(name2color_df, by=c("Phi_name.prev"="name"))

# Count transitions / merge with colors / merge with last column
change_of_category_col_df = change_of_category_df %>%
  mutate(transition_name_simple = transition_name) %>%
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "common / common", "common (constant)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple %in% c("common / disease-specific", "common / normal-specific", "common / undefined"), "common (change)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "disease-specific / disease-specific", "disease-specific (constant)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple %in% c("disease-specific / common", "disease-specific / normal-specific", "disease-specific / undefined"), "disease-specific (change)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "normal-specific / normal-specific", "normal-specific (constant)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple %in% c("normal-specific / common", "normal-specific / disease-specific", "normal-specific / undefined"), "normal-specific (change)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple == "undefined / undefined", "undefined (constant)")) %>% 
  mutate(transition_name_simple = replace(transition_name_simple, transition_name_simple %in% c("undefined / disease-specific", "undefined / normal-specific", "undefined / common"), "undefined (change)")) %>% 
  left_join(name2color_df, by=c("transition_name_simple"="name")) %>%
  full_join(change_of_category_size100_df)

# Define factors
change_of_category_col_df$transition_name_simple <- factor(change_of_category_col_df$transition_name_simple, levels=c("common (constant)", "common (change)", "disease-specific (constant)",  "disease-specific (change)", "normal-specific (constant)", "normal-specific (change)", "undefined (constant)", "undefined (change)"))
change_of_category_col_df$size.prev = as.character(change_of_category_col_df$size.prev)
change_of_category_col_df$size.prev <- factor(change_of_category_col_df$size.prev, levels=as.character(sort(unique(as.integer(change_of_category_col_df$size.prev)))))

# Plot
plot_flow = change_of_category_col_df %>% 
  filter(size.prev %in% sizes_to_plot) %>%
  ggplot(aes(x = size.prev, stratum = transition_name_simple, alluvium = Node, 
             fill = transition_name_simple, label = transition_name_simple)) +
  geom_flow(stat = "alluvium", lode.guidance = "frontback") +
  #geom_stratum() +
  geom_stratum(aes(col=transition_name_simple), size=0) +
  scale_fill_manual(name="Gene category", values=setNames(change_of_category_col_df$rgb_col, change_of_category_col_df$transition_name_simple)) + # If I want to plot only specific elements in the legend, use the argument breaks with a list of the elements that I want to show
  scale_color_manual(name="Gene category", values=setNames(change_of_category_col_df$rgb_col, change_of_category_col_df$transition_name_simple)) +
  xlab("Num. samples") +
  ylab("Num. disease genes") +
  guides(fill=guide_legend(title="Gene category")) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = font_selected))
plot_file = paste(diffcoexp_plots_dir, paste("codina_", name_D, "_", name_N, "_change_disease_genes_improved.png", sep=""), sep="/")
ggsave(
  plot_file,
  plot = plot_flow,
  dpi = 1200,
  width = 10000,
  height = 6000,
  units = c("px")
)
print(plot_flow)
```

Calculate stable genes by category:

```{r}
# Calculate total of stable genes
total_stable_change_genes_df = change_of_category_df %>%
  group_by(size.curr) %>% 
  mutate(n_total = n()) %>%
  ungroup() %>%
  select(Node, Phi_name.curr, size.curr, transition_type, n_total) %>% 
  group_by(size.curr, transition_type, n_total) %>% 
  summarize(n_genes = n()) %>% 
  ungroup() %>%
  mutate(frac_genes = n_genes / n_total)
total_stable_change_genes_df$Phi_name.curr = "all"

# Calculate stable genes by category
categories_stable_change_genes_df = change_of_category_df %>%
  group_by(size.curr) %>% 
  mutate(n_total = n()) %>%
  ungroup() %>%
  filter(transition_type == "stable") %>%
  select(Node, Phi_name.curr, size.curr, transition_type, n_total) %>% 
  group_by(Phi_name.curr, size.curr, transition_type, n_total) %>% 
  summarize(n_genes = n()) %>% 
  ungroup() %>%
  mutate(frac_genes = n_genes / n_total)

# Merge two tables
stable_genes_df = rbind(total_stable_change_genes_df, categories_stable_change_genes_df) %>%
  filter(transition_type == "stable") %>%
  left_join(name2color_df, by=c("Phi_name.curr"="name"))
```

Make plot:

```{r}
plot_stable_genes = stable_genes_df %>%
  ggplot(aes(x = size.curr, y = frac_genes, col = Phi_name.curr)) + 
  geom_line(aes(size = Phi_name.curr)) +
  xlab("Num. samples") +
  ylab("Fraction of stable disease genes") +
  scale_size_manual(values = c("all" = 1, "common" = 0.5, "disease-specific" = 0.5, "normal-specific" = 0.5, "undefined" = 0.5, "different" = 0.5)) +
  scale_color_manual(values=setNames(stable_genes_df$rgb_col, stable_genes_df$Phi_name.curr)) +
  guides(col=guide_legend(title="Gene category"), size="none") +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16),
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = font_selected))

plot_file = paste(diffcoexp_plots_dir, paste("codina_", name_D, "_", name_N, "_stable_genes.png", sep=""), sep="/")
ggsave(
  plot_file,
  plot = plot_stable_genes,
  dpi = 1200,
  width = 9000,
  height = 6000,
  units = c("px")
)
print(plot_stable_genes)
```

Calculate ranking of widely known genes:

```{r}
ranking_nodes_file = paste(diffcoexp_tables_dir, paste("codina_", name_D, "_", name_N, "_nodes_ranked.txt", sep=""), sep="/")
ranking_nodes_df= fread(ranking_nodes_file)
nodes_to_follow_file = "/home/j.aguirreplans/Projects/Scipher/SampleSize/data/nodes_to_follow/breast_cancer.txt"
nodes_to_follow <- data.table::fread(nodes_to_follow_file, header = F)$V1
nodes_to_follow = c("ARNT2", "EGR3", "NR3C2", "ZNF652")
```

Make plot:

```{r}
nodes_to_follow_ranking_plot = ranking_nodes_df %>%
  filter(Node %in% nodes_to_follow) %>%
  mutate(label=if_else(size==max(size), Node, "")) %>%
  ggplot(aes(x = size, y = rank)) + 
    geom_point(aes(group = Node, col = Phi_name)) +
    geom_line(aes(group = Node, col = Phi_name)) +
    xlab("Num. samples") +
    ylab("Ranking") +
    scale_color_manual(values=as.vector(name2color_df[name2color_df$name %in% levels(factor(unique((ranking_nodes_df %>% filter(Node %in% nodes_to_follow))$Phi_name))),]$rgb_col)) +
    scale_y_continuous(trans = "reverse") +
    guides(col=guide_legend(title="Gene category"), size="none") +
    theme_linedraw() +
    theme(aspect.ratio=1,
          plot.title =  element_text(size = 20, face="bold"), 
          axis.title = element_text(size = 17, face="bold"), 
          axis.text = element_text(size = 16),
          panel.grid.major=element_blank(), 
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 16), 
          legend.title=element_text(size=16, face="bold"),
          text = element_text(family = font_selected)) +
    geom_label_repel(aes(col=Phi_name, label = label),
                     fill="white",
                     family = font_selected,
                     box.padding = 0.2,
                     max.overlaps = Inf,
                     label.size = 0.75, 
                     size = 5,
                     alpha = 0.9, 
                     label.padding=0.25, 
                     fontface = "bold",
                     na.rm=TRUE,
                     show.legend=F,
                     seed = 1234)

plot_file = paste(diffcoexp_plots_dir, "nodes_to_follow_ranking.png", sep="/")
ggsave(
  plot_file,
  plot=nodes_to_follow_ranking_plot,
  dpi = 1200,
  #width = 9500,
  height = 6000,
  units = c("px")
)
print(nodes_to_follow_ranking_plot)
```

Use patchwork to plot everything together:

```{r}

(plot_flow /
  nodes_to_follow_ranking_plot /
  plot_stable_genes) + 
  plot_annotation(tag_levels = 'A') & # Include letters 
  theme(plot.tag = element_text(face = 'bold', size = 17)
        ,plot.tag.position  = c(.05, 1.04)
        ,plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm")
        ) # that are bold


plot_file = paste(diffcoexp_plots_dir, "/codina_stability_results.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  width = 11000,
  #height = 12000,
  height = 17500,
  units = c("px")
)
```


#### Protein-protein interaction plots

Load PPI + co-expression dataframe:

```{r}
dataset = "gtex"
type_dataset = "Whole.Blood"
sample_group_type = "tumor" # if dataset = "tcga"
type_counts = "reads" # tpm, reads
sizes_selected = c(20, 100, 200, 300, 400, 500)
sizes_selected_comparison_plot = c(20, 100, 200, 300)
threshold_selected = 0.05
coexpression_pairs_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_analysis_pairs_%s_%s.txt", dataset, type_dataset)
coexpression_pairs <- data.table::fread(coexpression_pairs_file)
perf_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_ppi_vs_random_far_pairs_performance_%s_%s.txt", dataset, type_dataset)
perf_df <- data.table::fread(perf_file)

colors_defined = c("PPI" = "#DBA9CA",
                   "PPI & iNCI" = "#FAF69E",
                   "Random" = "#B8BECC",
                   "Random pairs in PPI" = "#B8BECC",
                   "Random (distance > 3)" = "#543F4D",
                   "Distance > 3" = "#543F4D",
                   "Random pairs in PPI (dist. > 3)" = "#543F4D",
                   "Random pairs in PPI & iNCI (dist. > 3)" = "#545334",
                   "1" = "#DBA9CA",
                   "2" = "#a8829b",
                   "3" = "#87677c",
                   "4" = "#5e4857",
                   "5" = "#40303b",
                   "6" = "#1c151a")
```

Violin plot of PPI vs. pairs > 3 distance across sample size (without removing unsignificant weights)

```{r}
violinplot_ppi_coexpression_by_types_nofilt_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_pair_type_%s_%s.png", dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("PPI", "Distance > 3")]
violinplot_ppi_coexpression_by_types_nofilt = coexpression_pairs %>%
  select(-distances_ppi, -distances_nci, -nci, -random_not_ppi, -random_far_not_nci) %>%
  pivot_longer(c("ppi", "random_far_not_ppi"), names_to = "category", values_to = "category_bool") %>%
  filter(category_bool == TRUE) %>%
  select(-category_bool) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  mutate(category = replace(category, category == "ppi", "PPI")) %>%
  mutate(category = replace(category, category == "random_far_not_ppi", "Distance > 3")) %>%
  mutate(category = factor(category, levels = c("PPI", "Distance > 3"))) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      #y = abs(correlation),
      y = correlation,
      fill = category,
      colour = category) +
  geom_violin(
    alpha = 0.8
    #scale = "width",
    #adjust = 0.5
  ) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = font_selected)) +
  labs (x = "Number of samples",
        #y = "Absolute co-expression weight",
        y = "Co-expression weight",
        color = "",
        fill = ""); violinplot_ppi_coexpression_by_types_nofilt
ggsave(
  violinplot_ppi_coexpression_by_types_nofilt_file,
  plot = violinplot_ppi_coexpression_by_types_nofilt,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
```

Table S1: Mean and SD values of the different groups across sample size and statistical comparison using Wilcoxon test:

```{r}
# Select PPIs and far protein pairs and transform the data to long format
ppi_coexpression_across_types_by_sizes_df <- coexpression_pairs %>%
  dplyr::select(-distances_ppi, -distances_nci, -nci, -random_far_not_nci) %>%
  tidyr::pivot_longer(
    c("ppi", "random_far_not_ppi"),
    names_to = "category",
    values_to = "category_bool"
  ) %>%
  dplyr::filter(category_bool == TRUE) %>%
  dplyr::select(-category_bool) %>%
  tidyr::pivot_longer(
    as.character(sizes_selected),
    names_to = "size",
    values_to = "correlation",
    names_transform = list(size = as.numeric),
    values_transform = list(correlation = as.numeric)
  ) %>%
  dplyr::filter(!(is.na(correlation))) %>%
  tidyr::unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

# Calculate statistical tests between PPIs and far protein pairs for each sample
# size
wilcox_results_across_types_by_sizes_df <- data.frame()
for(size_selected in sort(unique(ppi_coexpression_across_types_by_sizes_df$size))) {

  # Filter co-expression by size
  ppi_coexpression_by_selected_size_df <- ppi_coexpression_across_types_by_sizes_df %>% 
    dplyr::filter(size == size_selected) %>% 
    dplyr::mutate(category = factor(category, levels = c("ppi", "random_far_not_ppi")))

  # Run Wilcoxon test instead of Dunn, because there are only two groups
  wilcox_across_types <- wilcox.test(
    abs((ppi_coexpression_by_selected_size_df %>% filter(category == "ppi"))$correlation),
    abs((ppi_coexpression_by_selected_size_df %>% filter(category == "random_far_not_ppi"))$correlation),
    alternative = "greater",
    correct = FALSE,
    exact = FALSE
  )

  # Calculate mean and SD
  mean_abs_corr_ppi <- mean(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "ppi"))$correlation))
  mean_abs_corr_far <- mean(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "random_far_not_ppi"))$correlation))
  diff_mean <- mean_abs_corr_ppi - mean_abs_corr_far
  sd_abs_corr_ppi <- sd(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "ppi"))$correlation))
  sd_abs_corr_far <- sd(abs((ppi_coexpression_by_selected_size_df %>% filter(category == "random_far_not_ppi"))$correlation))
  diff_sd <- sd_abs_corr_ppi - sd_abs_corr_far

  # Save result
  wilcox_results_across_types_by_sizes_df <- rbind(
    wilcox_results_across_types_by_sizes_df,
    data.frame(
      size = size_selected,
      mean_abs_corr_ppi = round(mean_abs_corr_ppi, 3),
      mean_abs_corr_far = round(mean_abs_corr_far, 3),
      diff_mean = round(diff_mean, 3),
      sd_abs_corr_ppi = round(sd_abs_corr_ppi, 3),
      sd_abs_corr_far = round(sd_abs_corr_far, 3),
      diff_sd = round(diff_sd, 3),
      w.statistic = as.numeric(wilcox_across_types$statistic),
      w.pvalue = ifelse(wilcox_across_types$p.value == 0, "< 2.2e-16", wilcox_across_types$p.value)
    )
  )

  rm(ppi_coexpression_by_selected_size_df)
}

rm(ppi_coexpression_across_types_by_sizes_df)

# Save results
wilcox_results_across_types_by_sizes_file <- sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_wilcox_across_pair_type_by_sizes_%s_%s.txt", dataset, type_dataset)
wilcox_results_across_types_by_sizes_df %>%
  data.table::fwrite(wilcox_results_across_types_by_sizes_file)

wilcox_results_across_types_by_sizes_df
```

Line plot of mean and variance of co-expression of PPI vs. pairs > 3 distance across sample size:

```{r}
lineplot_ppi_coexpression_by_types_nofilt_file <-  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_lineplot_by_pair_type_%s_%s.png", dataset, type_dataset)

colors_defined_filt <- colors_defined[names(colors_defined) %in% c("PPI", "Distance > 3")]

lineplot_ppi_coexpression_by_types_nofilt <- wilcox_results_across_types_by_sizes_df %>%
  dplyr::select(-diff_mean, -diff_sd, -w.statistic, -w.pvalue) %>%
  tidyr::pivot_longer(
    -size,
    names_to = c("metric", "category"),
    names_pattern = "(.*)_abs_corr_(.*)",
    values_to = "value"
  ) %>%
  tidyr::pivot_wider(
    names_from = metric,
    values_from = value,
    names_glue = "{.name}_corr"
  ) %>%
  dplyr::mutate(category = replace(category, category == "ppi", "PPI")) %>%
  dplyr::mutate(category = replace(category, category == "far", "Distance > 3")) %>%
  dplyr::mutate(category = factor(category, levels = c("PPI", "Distance > 3"))) %>%
  ggplot(aes(x = reorder(size, as.numeric(size)), group = category)) +
  geom_line(aes(y = mean_corr, color = category), linewidth = 1) +
  geom_point(aes(y = mean_corr, color = category)) +
  geom_errorbar(aes(ymin = mean_corr - sd_corr, ymax = mean_corr + sd_corr, color = category), width = .2, position = position_dodge(0.05)) +
  #geom_ribbon(aes(y = mean_corr, ymin = mean_corr - sd_corr, ymax = mean_corr + sd_corr, fill = category), alpha = 0.2) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = font_selected)) +
  labs (x = "Number of samples",
        #y = "Absolute co-expression weight",
        y = "Mean abs. co-expression weight",
        color = "",
        fill = ""); lineplot_ppi_coexpression_by_types_nofilt
ggsave(
  lineplot_ppi_coexpression_by_types_nofilt_file,
  plot = lineplot_ppi_coexpression_by_types_nofilt,
  dpi = 1200,
  width = 8400,
  #height = 6000,
  units = c("px")
)
```

Violin plot of PPI distances across sample size (without removing unsignificant correlations):

```{r}
violinplot_ppi_coexpression_by_distances_nofilt_file = sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_violinplot_by_ppi_distance_%s_%s.png", dataset, type_dataset)
colors_defined_filt = colors_defined[names(colors_defined) %in% c("1", "2", "3", "4", "5", "6")]
violinplot_ppi_coexpression_by_distances_nofilt = coexpression_pairs %>%
  select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  filter(!(is.na(distances_ppi))) %>%
  pivot_longer(as.character(sizes_selected), names_to = "size", values_to = "correlation", names_transform = list(size = as.character), values_transform = list(correlation = as.numeric)) %>%
  filter(size %in% sizes_selected_comparison_plot) %>%
  filter(!(is.na(correlation))) %>%
  mutate(across(c(distances_ppi), as.character)) %>%
  ggplot() +
  aes(x = reorder(size, as.numeric(size)),
      #y = abs(correlation),
      y = correlation,
      fill = distances_ppi,
      colour = distances_ppi) +
  geom_violin(
    alpha = 0.8
    #scale = "width",
    #adjust = 0.5
  ) +
  scale_fill_manual(values = colors_defined_filt) +
  scale_color_manual(values = colors_defined_filt) +
  theme_linedraw() +
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = font_selected)) +
  labs (x = "Number of samples",
        #y = "Absolute co-expression weight",
        y = "Co-expression weight",
        color = "PPI distance",
        fill = "PPI distance")
ggsave(
  violinplot_ppi_coexpression_by_distances_nofilt_file,
  plot = violinplot_ppi_coexpression_by_distances_nofilt,
  dpi = 1200,
  #width = 9300,
  #height = 6000,
  units = c("px")
)
violinplot_ppi_coexpression_by_distances_nofilt
```

Table S2: Mean and SD values of protein pairs from different distances across sample size and statistical comparison using KW and Dunn tests:

```{r}
# Function to Compute Pairwise Differences with Comparison Column
compute_pairwise_differences <- function(data, group_col, value_col) {

  unique_groups <- sort(unique(data[[group_col]]))
  
  # Generate all pairwise combinations of groups
  group_pairs <- combn(unique_groups, 2)
  
  results <- data.frame(
    comparison = character(0),
    difference = numeric(0)
  )
  
  for (i in seq_len(ncol(group_pairs))) {
    g1 <- group_pairs[1, i]
    g2 <- group_pairs[2, i]

    mean_abs_g1 <- mean(abs(data[data[[group_col]] == g1, ][[value_col]]))
    mean_abs_g2 <- mean(abs(data[data[[group_col]] == g2, ][[value_col]]))
    mean_diff <- round((mean_abs_g1 - mean_abs_g2), 3)

    results <- rbind(results, data.frame(
      comparison = paste(g1, "-", g2),
      mean_diff = mean_diff
    ))
  }
  
  return(results)
}

# Select protein pairs from different distances and transform to long format
ppi_coexpression_across_distances_by_sizes_df <- coexpression_pairs %>%
  dplyr::select(-ppi, -nci, -distances_nci, -random_not_ppi, -random_far_not_ppi, -random_far_not_nci) %>%
  dplyr::filter(!(is.na(distances_ppi))) %>%
  tidyr::pivot_longer(
    as.character(sizes_selected),
    names_to = "size",
    values_to = "correlation",
    names_transform = list(size = as.numeric),
    values_transform = list(correlation = as.numeric)
  ) %>%
  dplyr::filter(!(is.na(correlation))) %>%
  tidyr::unite(edge, c("Node.1", "Node.2"), sep = "_", remove = FALSE)

# Calculate mean, SD and difference across distances
mean_sd_corr_across_distances_by_sizes_df <- ppi_coexpression_across_distances_by_sizes_df %>%
  dplyr::group_by(size, distances_ppi) %>%
  dplyr::summarize(
    mean_abs_corr = round(mean(abs(correlation)), 3),
    sd_abs_corr = round(sd(abs(correlation)), 3)
  ) %>%
  dplyr::ungroup()

# Calculate statistical tests between protein pairs from different distances for
# each sample size
kw_results_across_distances_by_sizes_df <- data.frame()
mean_diff_across_distances_by_sizes_df <- data.frame()
dunn_results_across_distances_by_sizes_df <- data.frame()

for(size_selected in sort(unique(ppi_coexpression_across_distances_by_sizes_df$size))) {

  # Filter co-expression by size
  ppi_coexpression_by_selected_size_df <- ppi_coexpression_across_distances_by_sizes_df %>%
    dplyr::filter(size == size_selected)

  # Run KW test across distance groups
  kw_across_distances <- ppi_coexpression_by_selected_size_df %>% 
    dplyr::mutate(distances_ppi = factor(distances_ppi, levels = c(1,2,3,4,5,6))) %>% 
    kruskal.test(abs(correlation) ~ distances_ppi, .)

  # Save result
  kw_results_across_distances_by_sizes_df <- rbind(
    kw_results_across_distances_by_sizes_df,
    data.frame(
      size = size_selected,
      kw.statistic = kw_across_distances$statistic[[1]],
      kw.pvalue = ifelse(kw_across_distances$p.value == 0, "< 2.2e-16", kw_across_distances$p.value)
    )
  )

  # Pairwise differences of means across distances
  mean_diff_across_distances <- compute_pairwise_differences(
    data = ppi_coexpression_by_selected_size_df,
    group_col = "distances_ppi",
    value_col = "correlation"
  )

  # Save result
  mean_diff_across_distances_by_sizes_df <- rbind(
    mean_diff_across_distances_by_sizes_df,
    cbind(
      data.frame(size = size_selected),
      mean_diff_across_distances
    )
  )

  # Run Dunn test across distance groups
  dunn_across_distances <- ppi_coexpression_by_selected_size_df %>% 
    dplyr::mutate(distances_ppi = factor(distances_ppi, levels = c(1,2,3,4,5,6))) %>% 
    FSA::dunnTest(abs(correlation) ~ distances_ppi, .)

  # Save result
  dunn_results_across_distances_by_sizes_df <- rbind(
    dunn_results_across_distances_by_sizes_df,
    cbind(
      data.frame(size = size_selected),
      dunn_across_distances$res
    )
  )

  rm(ppi_coexpression_by_selected_size_df)
}

rm(ppi_coexpression_across_distances_by_sizes_df)

# Merge and save results
test_results_across_distances_by_sizes_df <- kw_results_across_distances_by_sizes_df %>%
  dplyr::full_join(
    dunn_results_across_distances_by_sizes_df %>%
      dplyr::full_join(
        mean_diff_across_distances_by_sizes_df,
        by = c("size" = "size", "Comparison" = "comparison")
      ),
    by = c("size")
  ) %>%
  dplyr::rename(
    "comparison" = "Comparison",
    "dunn.z" = "Z",
    "dunn.p.unadj" = "P.unadj",
    "dunn.p.adj" = "P.adj"
  ) %>%
  dplyr::arrange(size, comparison)

test_results_across_distances_by_sizes_file <- sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/tables/ppi_coexpression_analysis_statistic_tests_across_distance_by_sizes_%s_%s.txt", dataset, type_dataset)
test_results_across_distances_by_sizes_df %>%
  data.table::fwrite(test_results_across_distances_by_sizes_file)

test_results_across_distances_by_sizes_df
```

AUC of PPI prediction:

```{r}
perf_auc = perf_df %>%
  select(size, AUC) %>%
  unique()
print(perf_auc)

auc_plot = perf_auc %>% 
  mutate(size = factor(size, 
                       levels = sizes_selected, 
                       labels = sizes_selected)) %>% 
  ggplot() +
  aes(x = size, 
      #fill = base, 
      #colour = base, 
      weight = AUC) +
  geom_bar(col="#DBA9CA",
         fill="#DBA9CA") +
  #scale_fill_manual(values = colors_defined) +
  #scale_color_manual(values = colors_defined) +
  coord_cartesian(ylim=c(0.5,NA)) +
  theme_linedraw() +
  #facet_grid(vars(name))+ 
  theme(aspect.ratio=1,
        plot.title =  element_text(size = 20, face="bold"), 
        axis.title = element_text(size = 17, face="bold"), 
        axis.text = element_text(size = 16), 
        panel.grid.major=element_blank(), 
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 16), 
        legend.title=element_text(size=16, face="bold"),
        text = element_text(family = font_selected)) +
  labs(y = "AUROC",
       x = "Number of samples"
       )

auc_plot_file =  sprintf("/home/j.aguirreplans/Projects/Scipher/SampleSize/data/out/plots/ppi_coexpression_perf_AUC_pval_%s_%s_%s.png", as.character(threshold_selected), dataset, type_dataset)
ggsave(
  auc_plot_file,
  plot = auc_plot,
  dpi = 1200,
  #width = 9300,
  height = 6000,
  units = c("px")
)
 auc_plot
```

Patchwork (filtered by p-value):

```{r}
# Use patchwork to concatenate plots
((violinplot_ppi_coexpression_by_types_nofilt + theme(legend.position="bottom")) |
  (violinplot_ppi_coexpression_by_distances_nofilt + theme(legend.position="bottom")) |
  auc_plot) + 
  plot_annotation(tag_levels = 'A') &
    theme(plot.tag = element_text(face = 'bold', size = 17),
          plot.tag.position  = c(.01, 1.06),
          plot.margin = margin(0.6, 0.6, 0.6, 0.6, "cm")
          )

plot_file = paste(plots_dir, "/ppi_coexpression_results.png", sep="")
ggsave(
  plot_file,
  dpi = 1200,
  width = 20000,
  height = 9000,
  units = c("px")
)
```


